## code to prepare `DATASET` dataset goes here

gene_blacklist = system.file("extdata", "gene_blacklist_with_IG.tsv", package = "GAMBLR") %>%
  read_tsv()

usethis::use_data(gene_blacklist, overwrite = TRUE)

grch37_all_gene_coordinates = system.file("extdata", "grch37_gene_coordinates.tsv", package = "GAMBLR") %>%
  read_tsv() %>% dplyr::filter(grepl("PATCH",chromosome))

usethis::use_data(grch37_all_gene_coordinates, overwrite = TRUE)


hg38_oncogene = system.file("extdata", "oncogene_regions.hg38.tsv", package = "GAMBLR") %>%
  read_tsv(col_types="ciici")

usethis::use_data(hg38_oncogene, overwrite = TRUE)

grch37_oncogene = system.file("extdata", "oncogene_regions.grch37.tsv", package = "GAMBLR") %>%
  read_tsv(col_types="ciici")

usethis::use_data(grch37_oncogene, overwrite = TRUE)

hg38_partners = system.file("extdata","superenhancer_regions.hg38.tsv",package="GAMBLR") %>%
  read_tsv(col_types="ciici")

usethis::use_data(hg38_partners, overwrite = TRUE)

grch37_partners = system.file("extdata", "superenhancer_regions.grch37.tsv", package = "GAMBLR") %>%
  read_tsv(col_types="ciici")


grch37_ashm_regions = system.file("extdata", "somatic_hypermutation_locations_GRCh37.txt", package = "GAMBLR") %>%
  read_tsv() %>% mutate(name=paste0(gene,"-",region))

usethis::use_data(grch37_ashm_regions, overwrite = TRUE)

hg38_ashm_regions = system.file("extdata", "somatic_hypermutation_locations_GRCh38.txt", package = "GAMBLR") %>%
  read_tsv()

usethis::use_data(hg38_ashm_regions, overwrite = TRUE)

lymphoma_genes = system.file("extdata","lymphoma_genes.tsv",package="GAMBLR") %>%
  read_tsv(col_types="clllll")

usethis::use_data(lymphoma_genes, overwrite = TRUE)

#THis needs to be syncronized with the list above (it currently is out of sync!)
grch37_lymphoma_genes_bed = system.file("extdata","lymphoma_genes.grch37.bed",package="GAMBLR") %>%
  read_tsv()

usethis::use_data(grch37_lymphoma_genes_bed, overwrite = TRUE)

hg38_lymphoma_genes_bed = system.file("extdata","lymphoma_genes.hg38.bed",package="GAMBLR") %>%
  read_tsv()

usethis::use_data(hg38_lymphoma_genes_bed, overwrite = TRUE)


wright_genes_with_weights = system.file("extdata","WrightGenesWithWeights.txt",package="GAMBLR") %>%
  read.table(sep="\t",header=1) %>% rename(Ensembl_ID=EnsemblGeneID,Hugo_Symbol=GeneName)

usethis::use_data(wright_genes_with_weights, overwrite = TRUE)

dhitsig_genes_with_weights = system.file("extdata","DHITsigGenesWithWeights.txt",package="GAMBLR") %>%
  read.table(sep="\t",header=1) %>% rename(Ensembl_ID=ensembl_gene_id,Hugo_Symbol=GeneName)
usethis::use_data(dhitsig_genes_with_weights, overwrite = TRUE)

target_regions_hg38 = system.file("extdata","target_regions_hg38.txt",package="GAMBLR") %>%
  read.table(sep="\t",header=1)
usethis::use_data(target_regions_hg38, overwrite = TRUE)

target_regions_grch37 = system.file("extdata","target_regions_grch37.txt",package="GAMBLR") %>%
  read.table(sep="\t",header=1)
usethis::use_data(target_regions_grch37, overwrite = TRUE)

hotspot_regions_grch37 = system.file("extdata","hotspot_regions.grch37.tsv",package="GAMBLR") %>%
  read.table(sep="\t",header=1) %>%
  column_to_rownames("gene")
usethis::use_data(hotspot_regions_grch37, overwrite = TRUE)

hotspot_regions_hg38 = system.file("extdata","hotspot_regions.hg38.tsv",package="GAMBLR") %>%
  read.table(sep="\t",header=1) %>%
  column_to_rownames("gene")
usethis::use_data(hotspot_regions_hg38, overwrite = TRUE)

chromosome_arms_grch37 = system.file("extdata","chromosome_arms_grch37.tsv",package="GAMBLR") %>%
  read.table(sep="\t",header=1)
usethis::use_data(chromosome_arms_grch37, overwrite = TRUE)

chromosome_arms_hg38 = system.file("extdata","chromosome_arms_hg38.tsv",package="GAMBLR") %>%
  read.table(sep="\t",header=1)
usethis::use_data(chromosome_arms_hg38, overwrite = TRUE)

lymphgen_entrez = system.file("extdata","lymphgen_genes_entrez.txt",package="GAMBLR") %>%
  read_tsv()

entrez_map = system.file("extdata","hugo2entrez.tsv",package="GAMBLR") %>%
  read_tsv()

lymphgen_anno = left_join(lymphgen_entrez,entrez_map) %>% dplyr::rename("Hugo_Symbol"="Approved symbol") %>% dplyr::select(1:3)


library("biomaRt")

ensembl = useMart(biomart="ENSEMBL_MART_ENSEMBL", host="https://grch37.ensembl.org", path="/biomart/martservice" ,dataset="hsapiens_gene_ensembl")
#need to get entrezgene_id, hgnc_symbol using ensembl_gene_id
gene_detail = getBM(attributes=c( 'ensembl_gene_id','entrezgene_id','hgnc_symbol'),
      filters = 'ensembl_gene_id',
      values = lymphoma_genes$ensembl_gene_id,
      mart = ensembl,useCache = FALSE)

#fill in missing entrezgene_id

gene_detail[gene_detail$ensembl_gene_id =="ENSG00000036448","entrezgene_id"] = 9172

gene_detail[gene_detail$ensembl_gene_id =="ENSG00000065526","entrezgene_id"] = 23013

gene_detail[gene_detail$ensembl_gene_id =="ENSG00000102096","entrezgene_id"] = 11040

gene_detail[gene_detail$ensembl_gene_id =="ENSG00000172578","entrezgene_id"] = 89857

gene_detail[gene_detail$ensembl_gene_id =="ENSG00000205542","entrezgene_id"] = 7114

lymphoma_genes = left_join(lymphoma_genes,gene_detail)
lymphoma_genes$LymphGen=FALSE
lymphoma_genes[lymphoma_genes$entrezgene_id %in% lymphgen_anno$NCBI_Gene_ID,"LymphGen"] = TRUE

#load the Reddy gene list
reddy_genes = system.file("extdata","reddy_genes.tsv",package="GAMBLR") %>%
  read_tsv() %>% dplyr::rename("hgnc_symbol"="Approved symbol")

usethis::use_data(reddy_genes, overwrite = TRUE)

lymphoma_genes$Reddy = FALSE
lymphoma_genes[lymphoma_genes$hgnc_symbol %in% reddy_genes$hgnc_symbol,"Reddy"]=TRUE

usethis::use_data(lymphoma_genes, overwrite = TRUE)

reddy_only = reddy_genes[which(!reddy_genes$hgnc_symbol %in% lymphoma_genes$hgnc_symbol),"hgnc_symbol"]

ensembl = useMart("ensembl",dataset="hsapiens_gene_ensembl")

reddy_detail = getBM(attributes=c( 'ensembl_gene_id','entrezgene_id','hgnc_symbol'),
                    filters = 'hgnc_symbol',
                    values = reddy_only$hgnc_symbol,
                    mart = ensembl,useCache = FALSE)

#update any based on mapping to Ensembl ID
lymphoma_genes[lymphoma_genes$ensembl_gene_id %in% reddy_detail$ensembl_gene_id,"Reddy"]=TRUE

#reddy_only = dplyr::filter(reddy_detail,!ensembl_gene_id %in% lymphoma_genes$ensembl_gene_id ) %>%
#  group_by(ensembl_gene_id,hgnc_symbol) %>% slice_head() %>% ungroup() %>% dplyr::select(-entrezgene_id) %>% dplyr::rename("Gene"="hgnc_symbol") %>%
#  mutate(Reddy=TRUE)
#lymphoma_genes_comprehensive = bind_rows(reddy_only,lymphoma_genes) %>% dplyr::select(ensembl_gene_id,Gene,Reddy,LymphGen,Chapuy)



chapuy_genes = system.file("extdata","chapuy_genes.tsv",package="GAMBLR") %>%
  read_tsv() %>% dplyr::rename("hgnc_symbol"="gene")

lymphoma_genes$Chapuy = FALSE
lymphoma_genes[lymphoma_genes$hgnc_symbol %in% chapuy_genes$hgnc_symbol,"Chapuy"]=TRUE

#lymphoma_genes_comprehensive[lymphoma_genes_comprehensive$Gene %in% chapuy_genes$hgnc_symbol,"Chapuy"]=TRUE
#lymphoma_genes_comprehensive[!lymphoma_genes_comprehensive$Gene %in% chapuy_genes$hgnc_symbol,"Chapuy"]=FALSE


lymphoma_genes_comprehensive = read_tsv("inst/extdata/lymphoma_genes_comprehensive.tsv")

lacy = read_tsv("inst/extdata/lacy_genes.tsv") %>% dplyr::filter(`Included in statistical analysis`=='Yes',Feature!="Amplification")
lacy_genes = pull(lacy,Gene)
lymphoma_genes$Lacy = FALSE
lymphoma_genes[lymphoma_genes$hgnc_symbol %in% lacy_genes,"Lacy"]=TRUE

lymphoma_genes_comprehensive$Lacy = FALSE
lymphoma_genes_comprehensive[lymphoma_genes_comprehensive$Gene %in% lacy_genes,"Lacy"]=TRUE

lacy_ashm = dplyr::filter(lacy,!is.na(`Annotation as aSHM`)) %>% pull(Gene)
lymphoma_genes_comprehensive$aSHM = FALSE
lymphoma_genes_comprehensive[lymphoma_genes_comprehensive$Gene %in% lacy_ashm,"aSHM"]=TRUE
lymphoma_genes_comprehensive[lymphoma_genes_comprehensive$Gene %in% grch37_ashm_regions$gene,"aSHM"]=TRUE
lymphoma_genes_comprehensive = mutate(lymphoma_genes_comprehensive,aSHM=ifelse(grepl("HIST",Gene),TRUE,aSHM))
usethis::use_data(lymphoma_genes_comprehensive, overwrite = TRUE)



#ensembl = useMart(biomart="ENSEMBL_MART_ENSEMBL", host="https://grch37.ensembl.org", path="/biomart/martservice" ,dataset="hsapiens_gene_ensembl")
#need to get entrezgene_id, hgnc_symbol using ensembl_gene_id
#gene_detail = getBM(attributes=c( 'ensembl_gene_id','hgnc_symbol','chromosome_name'),
#                    filters = 'hgnc_symbol',
#                    values = chapuy_only$hgnc_symbol,
#                    mart = ensembl,useCache = FALSE) %>% dplyr::filter(chromosome_name %in% c(c(1:22),"X","Y")) %>%
#  dplyr::select(-chromosome_name) %>% dplyr::rename("Gene"="hgnc_symbol") %>%
#  mutate(Chapuy=TRUE) %>% dplyr:: filter(!Gene %in% lymphoma_genes_comprehensive$Gene)

#lymphoma_genes_comprehensive = bind_rows(gene_detail,lymphoma_genes_comprehensive) %>%
#  mutate(Reddy=ifelse(is.na(Reddy),FALSE,Reddy))

#lymphoma_genes_comprehensive[lymphoma_genes_comprehensive$Gene %in% lymphgen_anno$Hugo_Symbol,"LymphGen"] = TRUE
#lymphoma_genes_comprehensive[!lymphoma_genes_comprehensive$Gene %in% lymphgen_anno$Hugo_Symbol,"LymphGen"] = FALSE

#lymphoma_genes_comprehensive %>% dplyr::filter(Chapuy ==FALSE, Reddy==FALSE, LymphGen == FALSE)

lymphoma_genes = dplyr::select(lymphoma_genes,-entrezgene_id) %>% group_by(Gene,ensembl_gene_id) %>% slice_head() %>% ungroup()

#DLBCL_curated_genes = lymphoma_genes %>% dplyr::filter(DLBCL==TRUE) %>% pull(Gene)

#lymphoma_genes_comprehensive$curated = FALSE
#lymphoma_genes_comprehensive[lymphoma_genes_comprehensive$Gene %in% DLBCL_curated_genes,]$curated = TRUE
#lymphoma_genes_comprehensive = dplyr::filter(lymphoma_genes_comprehensive,Reddy==TRUE | Chapuy == TRUE | LymphGen == TRUE | curated == TRUE)
#lymphoma_genes_comprehensive$other_support = ""

# Example SSM data from Grande et. al, 2019
grande_maf = system.file("extdata", "blood8871418-suppl2-ssm.csv", package = "GAMBLR") %>%
  read_tsv(.) %>%
  as.data.frame
grande_maf = grande_maf[!colnames(grande_maf) %in% c("tumor_biospecimen_id","normal_biospecimen_id")]

usethis::use_data(grande_maf, overwrite = TRUE)

# Features of Chapuy classifier with cluster weights
chapuy_features <- list()

chapuy_features$feature_weights <- system.file(
  "extdata",
  "chapuy_weights.tsv",
  package="GAMBLR") %>%
  read_tsv

chapuy_features$ssm_features <- chapuy_features$feature_weights$Feature[
    !str_detect(
        chapuy_features$feature_weights$Feature,
        "SV|AMP|DEL"
    )
]

chapuy_features$cnv_features <- chapuy_features$feature_weights$Feature[
    str_detect(
        chapuy_features$feature_weights$Feature,
        "AMP|DEL"
    )
]

chapuy_features$cnv_features_cytoband <-
chapuy_features$cnv_features[!grepl("p:|q:", chapuy_features$cnv_features)] %>%
as.data.frame %>%
separate(
  .,
  `.`,
  sep = ":",
  into = c("cytoband", "CNV")
)

chapuy_features$cnv_features_arm <-
chapuy_features$cnv_features[grepl("p:|q:", chapuy_features$cnv_features)] %>%
as.data.frame %>%
separate(
  .,
  `.`,
  sep = ":",
  into = c("arm", "CNV")
)

chapuy_features$sv_features <-
chapuy_features$feature_weights$Feature[
    str_detect(
        chapuy_features$feature_weights$Feature,
        "SV"
    )
] %>%
gsub(
    "SV:",
    "",
    .
) %>%
gsub(
    '[/][A-Z0-9]*',
    "",
    .
)

usethis::use_data(chapuy_features, overwrite = TRUE)


# Lacy classifier

# RF model
RFmodel_Lacy <- system.file(
  "extdata",
  "Lacy_rf_model.rds",
  package="GAMBLR") %>%
  readRDS

# Features
lacy_features <- list()

lacy_features$all <- system.file(
  "extdata",
  "lacy_weights.tsv",
  package="GAMBLR") %>%
  read_tsv

lacy_features$cnv <-
lacy_features$all$Feature[
    str_detect(lacy_features$all$Feature, "amp|del")
  ] %>%
  as.data.frame %>%
  dplyr::rename(
    "Gene" = "."
  ) %>%
  dplyr::mutate(
    Feature = Gene,
    CNV = ifelse(
      grepl(
        "amp",
        Gene
      ),
      "AMP",
      "DEL"
    ),
    Dual = ifelse(
      grepl(
        "_OR_",
        Gene
      ),
      "TRUE",
      "FALSE"
    ),
    Gene = gsub(
    '[_][A-Za-z]*',
    "",
    Gene)
  )

lacy_features$shm <-
  lacy_features$all$Feature[
    str_detect(lacy_features$all$Feature, "_S")
  ] %>%
  as.data.frame %>%
  mutate(
    Gene = gsub(
      "_S",
      "",
      `.`
    )
  ) %>%
  `names<-`(c(
    "Feature",
    "Gene"
  )) %>%
  mutate(
    genome_build = "grch37"
  )

lacy_features$grch37_shm <-
lacy_features$shm %>%
    left_join(
        .,
        grch37_gene_coordinates,
        by=c("Gene"="gene_name")
    ) %>%
    dplyr::select(chromosome, start, end, everything()) %>%
    arrange(chromosome, start, end) %>%
    dplyr::mutate(
      feature_start = start - 2000,
      feature_end = end,
    )

lacy_features$hg38_shm <-
hg38_gene_coordinates %>%
    dplyr::filter(
        ensembl_gene_id %in% c(
            lacy_features$grch37_shm$ensembl_gene_id,
            "ENSG00000278677",
            "ENSG00000273802",
            "ENSG00000286522",
            "ENSG00000273983")
    ) %>%
    dplyr::mutate(
        Feature = lacy_features$grch37_shm$Feature,
        genome_build = "hg38",
        Gene = gene_name,
        feature_start = start-2000,
        feature_end = end,
    ) %>%
    select(
        colnames(lacy_features$grch37_shm)
    )

lacy_features$hotspots <-
  lacy_features$all$Feature[
    str_detect(lacy_features$all$Feature, "_noncan")
  ] %>%
  gsub(
    "_noncan",
    "",
    .
  )

lacy_features$ssm <-
  lacy_features$all$Feature[
    str_detect(
      lacy_features$all$Feature,
      "_noncan|_S|amp|del",
      negate = TRUE)
  ] %>%
  gsub(
    '[_][0-9]*',
    "",
    .
  )

lacy_features$ssm <-
c(
  lacy_features$ssm,
  lacy_features$cnv %>%
      dplyr::filter(
        Dual == "TRUE"
      ) %>%
      pull(
        Gene
      )
) %>% sort

usethis::use_data(RFmodel_Lacy, overwrite = TRUE)
usethis::use_data(lacy_features, overwrite = TRUE)
