<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>CN &amp; SV Tutorial • GAMBLR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="CN &amp; SV Tutorial">
<meta property="og:description" content="GAMBLR">
<meta property="og:image" content="https://github.com/morinlab/GAMBLR/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-12345678-0"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-12345678-0');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9500</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-light fa-memo-circle-info"></span>
     
    How To
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="https://morinlab.github.io/GAMBLR/#installation" class="external-link">Install</a>
    </li>
    <li>
      <a href="https://morinlab.github.io/GAMBLR/#running-gamblr-on-your-own-computer" class="external-link">Running GAMBLR On Your Own Computer</a>
    </li>
    <li>
      <a href="https://morinlab.github.io/GAMBLR/#contributing" class="external-link">Contribute</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-light fa-person-chalkboard"></span>
     
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/ssm_tutorial.html">SSM</a>
    </li>
    <li>
      <a href="../articles/cn_sv_tutorial.html">CN &amp; SV</a>
    </li>
    <li>
      <a href="../articles/fancy_plots.html">Fancy Plots</a>
    </li>
    <li>
      <a href="../articles/utilities_tutorial.html">Utilities</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="../reference/index.html">
    <span class="fa fa-file-code-o"></span>
     
    Functions &amp; Bundled Data
  </a>
</li>
<li>
  <a href="https://github.com/morinlab/GAMBLR">
    <span class="fa fa-github fa-lg"></span>
     
    github
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>CN &amp; SV Tutorial</h1>
            
      
      
      <div class="hidden name"><code>cn_sv_tutorial.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="copy-number-structural-variants">Copy Number &amp; Structural Variants<a class="anchor" aria-label="anchor" href="#copy-number-structural-variants"></a>
</h2>
<p>This vignette demonstrates how to return Copy Number (CN) and Structural Variant (SV) data in a variety of different ways. The method of choice might be different for any given application or planned downstream analysis, as well as what information you, as a researcher, have at hand. This vignette also highlights how to visualize CN data, using established GAMBLR functions.</p>
<div class="section level3">
<h3 id="lets-get-started">Let’s get started<a class="anchor" aria-label="anchor" href="#lets-get-started"></a>
</h3>
<p>A good way to start is to retrieve metadata for all GAMBL samples. In this tutorial, we are only interested in samples with seq type <strong>genome</strong>. To get metadata for all such samples, we call <code>get_gambl_metadata</code>. Let us also subset the main metadata based on a few criteria.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#get metadata for all samples</span></span>
<span><span class="va">all_metadata</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span>seq_type_filter <span class="op">=</span> <span class="st">"genome"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#subset metadata to only include DLBCL samples</span></span>
<span><span class="va">dlbcl_metadata</span> <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">all_metadata</span>, <span class="va">pathology</span> <span class="op">==</span> <span class="st">"DLBCL"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#subset metadata to only include DLBCL and FL samples</span></span>
<span><span class="va">dlbcl_fl_metadata</span> <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">all_metadata</span>, <span class="va">pathology</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DLBCL"</span>, <span class="st">"FL"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>While we are at it, let us also define a few variables that we will use in the subsequent steps.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#define a sample</span></span>
<span><span class="va">this_sample_id</span> <span class="op">=</span> <span class="st">"HTMCP-01-06-00422-01A-01D"</span></span>
<span><span class="va">these_sample_list</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"00-15201_tumorA"</span>, <span class="st">"00-15201_tumorB"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#return regions for an example gene</span></span>
<span><span class="va">myc_region</span> <span class="op">=</span> <span class="fu"><a href="../reference/gene_to_region.html">gene_to_region</a></span><span class="op">(</span>gene_symbol <span class="op">=</span> <span class="st">"MYC"</span>, genome_build <span class="op">=</span> <span class="st">"grch37"</span>, return_as <span class="op">=</span> <span class="st">"region"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#print myc_region</span></span>
<span><span class="va">myc_region</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "8:128747680-128753674"</span></span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#subset first 15 rows from the lymphoma genes bed file</span></span>
<span><span class="va">some_lymphoma_genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TNFRSF14"</span>,<span class="st">"NOL9"</span>,<span class="st">"KLHL21"</span>,</span>
<span>                        <span class="st">"PIK3CD"</span>,<span class="st">"MTOR"</span>,<span class="st">"TNFRSF1B"</span>,</span>
<span>                        <span class="st">"SPEN"</span>,<span class="st">"ID3"</span>,<span class="st">"ARID1A"</span>,</span>
<span>                        <span class="st">"ZC3H12A"</span>,<span class="st">"RRAGC"</span>,<span class="st">"JAK1"</span>,</span>
<span>                        <span class="st">"BCL10"</span>,<span class="st">"NRAS"</span>,<span class="st">"CD58"</span><span class="op">)</span> </span></code></pre></div>
</div>
<div class="section level3">
<h3 id="cn">CN<a class="anchor" aria-label="anchor" href="#cn"></a>
</h3>
<div class="section level4">
<h4 id="cn-segments">CN Segments<a class="anchor" aria-label="anchor" href="#cn-segments"></a>
</h4>
<p>If we are interested in CN segments in specific genomic regions, the recommended approach is to call <code>get_cn_segments</code>. This function takes a variety of parameters that will allow the user to easily define what regions they are interested in. For example, if this function is being called using the <code>region</code> parameter, the function expects a region defined in the following format; <strong>“chr:start-end”</strong>. It is also possible to individually specify these coordinates with <code>chr</code>, <code>qstart</code> and <code>qend</code>. Another parameter to be aware of is <code>streamlined</code>, similarly to other GAMBLR function, this parameter controls the number of columns present in the returned data frame. If set to <code>FALSE</code>, full MAF-like format is returned and with this parameter set to <code>TRUE</code> only basic columns are kept. In addition, the parameter <code>remove_chr_prefix</code> is available for dealing with chr prefixes of the returned data frame. In the following example we are demonstrating how to use these different parameters in various combinations.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#using the region parameter</span></span>
<span><span class="va">my_segments_grch37</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_cn_segments.html">get_cn_segments</a></span><span class="op">(</span>region <span class="op">=</span> <span class="st">"chr8:128,723,128-128,774,067"</span>, </span>
<span>                                     projection <span class="op">=</span> <span class="st">"grch37"</span>, </span>
<span>                                     this_seq_type <span class="op">=</span> <span class="st">"genome"</span>, </span>
<span>                                     streamlined <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                                     from_flatfile <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#get cn segments for specified region, using chromosome, qstart and qend parameters.</span></span>
<span><span class="va">my_segments_hg38</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_cn_segments.html">get_cn_segments</a></span><span class="op">(</span>chromosome <span class="op">=</span> <span class="st">"8"</span>, </span>
<span>                                   qstart <span class="op">=</span> <span class="fl">128723128</span>, </span>
<span>                                   qend <span class="op">=</span> <span class="fl">128774067</span>, </span>
<span>                                   projection <span class="op">=</span> <span class="st">"hg38"</span>, </span>
<span>                                   this_seq_type <span class="op">=</span> <span class="st">"genome"</span>, </span>
<span>                                   with_chr_prefix <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                   streamlined <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                   from_flatfile <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#get cn segments, using a previously defined region (from gene_to_region) and filtering out all copy number states equal to 2</span></span>
<span><span class="va">myc_cns</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_cn_segments.html">get_cn_segments</a></span><span class="op">(</span>region <span class="op">=</span> <span class="va">myc_region</span>, </span>
<span>                          projection <span class="op">=</span> <span class="st">"grch37"</span>,</span>
<span>                          this_seq_type <span class="op">=</span> <span class="st">"genome"</span>, </span>
<span>                          streamlined <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                          from_flatfile <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                          with_chr_prefix <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">CN</span> <span class="op">!=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="cn-states">CN States<a class="anchor" aria-label="anchor" href="#cn-states"></a>
</h4>
<p>If you are interested in returning the copy number states for a specific region, one can call <em>get_cn_states</em>. This function takes a region (or regions) in the centralized region-format (chr:start-end) as well as region names (e.g gene symbols) to subset the returned cn states on. The function returns a data frame with absolute copy number states for the queried regions. In this example, we are returning copy number states for MYC (previously defined region) across all available samples. We are then filtering out all samples that show the expected copy number state (i.e 2) and are visualizing the counts of each copy number state for the remaining samples.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#get cn states for a specific region</span></span>
<span><span class="va">single_gene_cn</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_cn_states.html">get_cn_states</a></span><span class="op">(</span>regions_list <span class="op">=</span> <span class="va">myc_region</span>,</span>
<span>                               region_names <span class="op">=</span> <span class="st">"MYC"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#filter out all copy number states that are equal to 2, i.e keep all samples that have copy number mutations in MYC and count the number of occurrences for each copy number state.</span></span>
<span><span class="va">myc_cn_mutated</span> <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">single_gene_cn</span>, <span class="va">MYC</span> <span class="op">!=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">group_by</span><span class="op">(</span><span class="va">MYC</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">summarize</span><span class="op">(</span>count <span class="op">=</span> <span class="fu">n</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span><span class="st">"cn_state"</span> <span class="op">=</span> <span class="st">"MYC"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">mutate_at</span><span class="op">(</span><span class="fu">vars</span><span class="op">(</span><span class="va">cn_state</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">factor</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#visualize cn states for this gene with a simple barplot</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">myc_cn_mutated</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">cn_state</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_bar</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>y <span class="op">=</span> <span class="va">count</span>, fill <span class="op">=</span> <span class="va">cn_state</span>, label <span class="op">=</span> <span class="va">count</span><span class="op">)</span>, position <span class="op">=</span> <span class="st">"dodge"</span>, stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_text</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">cn_state</span>, y <span class="op">=</span> <span class="va">count</span>, label <span class="op">=</span> <span class="va">count</span><span class="op">)</span>, colour <span class="op">=</span> <span class="st">"#000000"</span>, size <span class="op">=</span> <span class="fl">5</span>, position <span class="op">=</span> <span class="fu">position_dodge</span><span class="op">(</span>width<span class="op">=</span><span class="fl">0.9</span><span class="op">)</span>, vjust<span class="op">=</span><span class="op">-</span><span class="fl">0.25</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_fill_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="../reference/get_gambl_colours.html">get_gambl_colours</a></span><span class="op">(</span><span class="st">"copy_number"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>title <span class="op">=</span> <span class="st">"CN States for MYC"</span>, x <span class="op">=</span> <span class="st">"Coppy Number State"</span>, y <span class="op">=</span> <span class="st">"Counts"</span>, fill <span class="op">=</span> <span class="st">""</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/theme_Morons.html">theme_Morons</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<p><img src="cn_sv_tutorial_files/figure-html/get_cn_states-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="sample-cn-segments">Sample CN Segments<a class="anchor" aria-label="anchor" href="#sample-cn-segments"></a>
</h4>
<p>If you are interested in returning all CN segments for one sample (or multiple samples), <code>get_sample_cn_segments</code> is available for this purpose. This functions takes an optional parameter called <code>this_sample_id</code> for query of a singular sample. If CN segments for multiple samples are to be returned, first toggle the Boolean parameter <code>multiple_samples</code> to TRUE and provide a list of sample IDs in the <code>samples_list</code> parameter. Other convenient parameters are <code>with_chr_prefis</code>, if set to <strong>TRUE</strong>, chromosome prefixes (chr) will be added to the returned data frame. For returning minimal output (rather than full details)m, set <code>streamlined</code> to <strong>TRUE</strong>. In this example we will first call the described function with default parameters for one sample, the next example we will demonstrate the multiple-samples-return functionality.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#get cn segments for one sample.</span></span>
<span><span class="va">one_sample_cn_segs</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_sample_cn_segments.html">get_sample_cn_segments</a></span><span class="op">(</span>this_sample_id <span class="op">=</span> <span class="va">this_sample_id</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#return chr-prefixed cn segments for multiple samples.</span></span>
<span><span class="va">multiple_samples_cn_seqs</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_sample_cn_segments.html">get_sample_cn_segments</a></span><span class="op">(</span>multiple_samples <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                                  sample_list <span class="op">=</span> <span class="va">these_sample_list</span>,</span>
<span>                                                  projection <span class="op">=</span> <span class="st">"hg38"</span>, </span>
<span>                                                  with_chr_prefix <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="assign-cn-to-ssm">Assign CN to SSM<a class="anchor" aria-label="anchor" href="#assign-cn-to-ssm"></a>
</h4>
<p>For annotating mutations with CN information, we have <code>assign_cn_to_ssm</code> available. This function provides great flexibility for annotating mutations from a variety of sources. For example, the user can either provide an already loaded MAF with mutations to be annotated with the parameter <code>maf_df</code> or a path to a local MAF file with <code>maf_file</code>. If neither of these parameters are used, the function will run <code>get_ssm_by_sample</code> with a sample ID specified in <code>this_sample</code> to return sample-specific SSMs.Similarly, one can also use a local file path for associated CN segments. This is achieved with calling the function with the <code>seg_file</code> parameter. This function also has a variety of convenience parameters for improved flexibility. For example, the return can be subset to specific genes with <code>genes</code> parameter. We can also restrict the return to only show coding mutations (<code>coding_only = TRUE</code>). In addition, include silent mutations can also be toggled with the parameter <code>include_silent</code> (Boolean). The returned data frame shares the same columns as your standard MAF but with two extra columns (<strong>CN</strong> and <strong>log.ratio</strong>). In this example we will call this function with a selection of different parameters to thoroughly demonstrate each different scenario (parameter combinations).</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#annotate a MAF wigh compy number information.</span></span>
<span><span class="va">annotated_ssm</span> <span class="op">=</span> <span class="fu"><a href="../reference/assign_cn_to_ssm.html">assign_cn_to_ssm</a></span><span class="op">(</span>this_sample_id <span class="op">=</span> <span class="va">this_sample_id</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#use same sample for annotating SSM with CN restricted to a specific set of genes.</span></span>
<span><span class="va">lymph_genes</span> <span class="op">=</span> <span class="va">lymphoma_genes_comprehensive</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">pull</span><span class="op">(</span><span class="va">Gene</span><span class="op">)</span></span>
<span></span>
<span><span class="va">annotated_ssm_lymphgenes</span> <span class="op">=</span> <span class="fu"><a href="../reference/assign_cn_to_ssm.html">assign_cn_to_ssm</a></span><span class="op">(</span>this_sample_id <span class="op">=</span> <span class="va">this_sample_id</span>,</span>
<span>                                            genes <span class="op">=</span> <span class="va">lymph_genes</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#additional example, using more parameters!</span></span>
<span><span class="va">annotated_ssm_extended</span> <span class="op">=</span> <span class="fu"><a href="../reference/assign_cn_to_ssm.html">assign_cn_to_ssm</a></span><span class="op">(</span>this_sample_id <span class="op">=</span> <span class="va">this_sample_id</span>,</span>
<span>                                          coding_only <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                          include_silent <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="sv">SV<a class="anchor" aria-label="anchor" href="#sv"></a>
</h3>
<div class="section level4">
<h4 id="combined-svs">Combined SVs<a class="anchor" aria-label="anchor" href="#combined-svs"></a>
</h4>
<p>IF you want to retrieve combined SV calls (manta and GRIDSS-derived) for one or more samples, you should look into <code>get_combined_sv</code>. This function does exactly this. Similarly to other GAMBLR function, we have a collection of familiar parameters available. The required parameters are <code>sample_ids</code>, a character vector of tumour sample IDs you wish to retrieve SVs for. Another useful way to subset your returned SVs based on VAF score is with the <code>min_vaf</code> parameter, this parameter hgolds the minimum tumour VAF for a SV to be returned. Recommended: 0. (default: 0). The bedpe files used as input to this function were pre-filtered for a minimum VAF of 0.05, and SVs affecting common translocation regions (BCL2, BCL6, MYC, CCND1) were whitelisted (e.g. no VAF filter applied). Therefore if you wish to post-filter the SVs we recommend doing so carefully after loading this data frame. Further, the input bedpe file is annotated with oncogenes and superenhancers from naive and germinal centre B-cells. You can subset to events affecting certain loci using the <code>oncogenes</code> argument.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#example with minimal parameter usage.</span></span>
<span><span class="va">combined_svs</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_combined_sv.html">get_combined_sv</a></span><span class="op">(</span>these_sample_ids <span class="op">=</span> <span class="va">this_sample_id</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#extended example, using additional parameters that are available. </span></span>
<span><span class="va">combined_sv_extended</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_combined_sv.html">get_combined_sv</a></span><span class="op">(</span>these_sample_ids <span class="op">=</span> <span class="va">these_sample_list</span>,</span>
<span>                                       projection <span class="op">=</span> <span class="st">"hg38"</span>,</span>
<span>                                       oncogenes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MYC"</span>, <span class="st">"BCL2"</span>, <span class="st">"BCL6"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="manta-svs">Manta SVs<a class="anchor" aria-label="anchor" href="#manta-svs"></a>
</h4>
<p>If you are interested in just retrieving manta SV call this is of course also possible. With <code>get_manta_sv</code> you can retrieve SVs from this specific call-set and filter with available function parameters. To call this function with the minimum parameters, no parameters are required. If the function is called in this way, all SVs are returned with default filtering parameters. If you have a sample ID you are interested in, you simply call this function with the <code>sample_id</code> parameter. Let’s go over the filtering options available for this function. Similarly to <code>get_combined_sv</code>, the following parameters are also available for this function; <code>min_vaf</code>, and <code>from_flatfile</code>. In addition, we also have; <code>min_score</code> describing the lowest Manta somatic score for a SV to be returned. <code>pass</code>, If set to TRUE, include SVs that are annotated with PASS in FILTER column. <code>pair_status</code> To restrict results (if desired) to matched or unmatched results (default is to return all). <code>chromosome</code> The chromosome you are restricting to. <code>qstart</code> Query start coordinate of the range you are restricting to. <code>qend</code> Query end coordinate of the range you are restricting to. <code>region</code> Formatted like chrX:1234-5678 instead of specifying chromosome, start and end separately. In the following example, we will look at a few different examples exploring some of the different parameters.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#get all SVs.</span></span>
<span><span class="va">all_svs</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_manta_sv.html">get_manta_sv</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#peek dimensions of returned SV calls</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">all_svs</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 413880     21</span></span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#extended example, return all SVs with a minimum vaf score of 0.1, minimum manta somatic score of 40, for MYC region, with chromosome prefixes. </span></span>
<span><span class="va">manta_svs</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_manta_sv.html">get_manta_sv</a></span><span class="op">(</span>min_vaf <span class="op">=</span> <span class="fl">0.1</span>, </span>
<span>                         min_score <span class="op">=</span> <span class="fl">40</span>, </span>
<span>                         pass <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                         region <span class="op">=</span> <span class="st">"8:128723128-128774067"</span>, </span>
<span>                         verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="sv-calls-to-ucsc-genome-browser-custom-track">SV Calls to UCSC Genome Browser Custom Track<a class="anchor" aria-label="anchor" href="#sv-calls-to-ucsc-genome-browser-custom-track"></a>
</h4>
<p>If you want to convert your SV calls into a UCSC-ready custom track, for interrogation on the genome browser, you can call <code>sv_to_customtrack</code>. This parameter only takes four parameters, <code>sv_bedpe</code>, <code>output_file</code>, <code>is_annotated</code>, and <code>sv_name</code>. Were <code>sv_bedpe</code> is a bedpe formatted data frame of SVs. <code>output_file</code> is a bed file with UCSC custom header. <code>is_annotated</code> is a Boolean aprameter, set to TRUE if input SV bedpe is annotated, default is TRUE. <code>sv_name</code> is SV name. Default is set to “all” = include all subtypes of SVs.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#convert SV calls to UCSC custom track file.</span></span>
<span><span class="co">#sv_to_custom_track(sv_bedpe = manta_svs, output_file = "../manta_sv_custom_track.bed", is_annotated = FALSE)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h3>
<div class="section level4">
<h4 id="variant-counts">Variant Counts<a class="anchor" aria-label="anchor" href="#variant-counts"></a>
</h4>
<p>Interested in how many variants are present in a given sample? A quick way to overview such information would be to call <code>fancy_svbar</code> on a given sample ID. This function returns a bar plot visualizing SVs or SSMs. A variety of sub-setting options are also available. Such as chromosome selection and variant types to be included in plot.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#build plot</span></span>
<span><span class="fu"><a href="../reference/fancy_v_count.html">fancy_v_count</a></span><span class="op">(</span>this_sample_id <span class="op">=</span> <span class="va">this_sample_id</span>,</span>
<span>              ssm <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>              plot_title <span class="op">=</span> <span class="st">"SV Count Example"</span>,</span>
<span>              chr_select <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span><span class="op">)</span>,</span>
<span>              coding_only <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="cn_sv_tutorial_files/figure-html/svbar-1.png" width="1093.75"></p>
</div>
<div class="section level4">
<h4 id="variant-counts-per-chromosome-">Variant Counts Per Chromosome.<a class="anchor" aria-label="anchor" href="#variant-counts-per-chromosome-"></a>
</h4>
<p>Sometimes it can be useful to overview variant distributions across the complete genomic landscape. To do so, one could call <code>fancy_v_chrcount</code> with sample ID. Optional arguments for sub-setting your data are available (see parameter descriptions for more information). In the following example we are plotting SV (ssm = FALSE) counts for autosomes. One can also call <code>add_qc_metric</code> to add a second y-axis annotating mean corrected coverage for the selected sample.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#build plot</span></span>
<span><span class="fu"><a href="../reference/fancy_v_chrcount.html">fancy_v_chrcount</a></span><span class="op">(</span>this_sample_id <span class="op">=</span> <span class="st">"HTMCP-01-06-00422-01A-01D"</span>,</span>
<span>                 plot_subtitle <span class="op">=</span>  <span class="st">"grch37"</span>,</span>
<span>                 plot_title <span class="op">=</span> <span class="st">"SV Genomic Distribution Example"</span>,</span>
<span>                 ssm <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                 projection <span class="op">=</span> <span class="st">"grch37"</span>,</span>
<span>                 min_vaf <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                 chr_select <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                 coding_only <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                 from_flatfile <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                 use_augmented_maf <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="cn_sv_tutorial_files/figure-html/svdist-1.png" width="1093.75"></p>
</div>
<div class="section level4">
<h4 id="variant-size-densities-">Variant Size Densities.<a class="anchor" aria-label="anchor" href="#variant-size-densities-"></a>
</h4>
<p>Another useful plotting function for visualizing SV size distributions is with <code>fancy_sv_sizedens</code>. In this example, we are showing SVs (larger 50bp and with a with minimum vaf of 0) size distributions for a specific sample and selected chromosomes.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#build plot</span></span>
<span><span class="fu"><a href="../reference/fancy_sv_sizedens.html">fancy_sv_sizedens</a></span><span class="op">(</span>this_sample_id <span class="op">=</span> <span class="va">this_sample_id</span>,</span>
<span>                  vaf_cutoff <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                  size_cutoff <span class="op">=</span> <span class="fl">50</span>,</span>
<span>                  chr_select <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                  plot_title <span class="op">=</span> <span class="st">"SV (&gt;50bp) Size Distribution Example"</span>,</span>
<span>                  plot_subtitle <span class="op">=</span> <span class="st">"grch37"</span>,</span>
<span>                  projection <span class="op">=</span> <span class="st">"grch37"</span><span class="op">)</span></span></code></pre></div>
<p><img src="cn_sv_tutorial_files/figure-html/sv_size-1.png" width="1093.75"></p>
</div>
</div>
<div class="section level3">
<h3 id="copy-number-states-summaries-">Copy Number States Summaries.<a class="anchor" aria-label="anchor" href="#copy-number-states-summaries-"></a>
</h3>
<p>Another way to visualize copy number information is with the the <code>fancy_cnvbar</code>. This function takes one required parameter (sample ID), calls <code>get_cn_segments</code> and <code>assign_cn_ssm</code> to retrieve data for plotting. The plot highlights the number of CN states for any given sample (left y-axis), together with n bases included in each given CN state (right y-axis).</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#build plot</span></span>
<span><span class="fu"><a href="../reference/fancy_cnbar.html">fancy_cnbar</a></span><span class="op">(</span>this_sample_id <span class="op">=</span> <span class="va">this_sample_id</span>,</span>
<span>            plot_title <span class="op">=</span> <span class="st">"CNV Barplot Example"</span>,</span>
<span>            plot_subtitle <span class="op">=</span> <span class="st">"grch37"</span>,</span>
<span>            chr_select <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span><span class="op">)</span>,</span>
<span>            include_cn2 <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="cn_sv_tutorial_files/figure-html/cnbar-1.png" width="1093.75"></p>
</div>
<div class="section level3">
<h3 id="copy-number-segments-plot">Copy Number Segments Plot<a class="anchor" aria-label="anchor" href="#copy-number-segments-plot"></a>
</h3>
<p>Another way to visualize CN segments is with the <code>focal_cn_plot</code>. This function visualizes all CN segments for a defined region, colours the returned segments based on lymphgen information. In addition, this function takes either a specified region (chr:start-end format). If no region is supplied, the user can give the function a gene symbol with <code>gene</code>. If so, the function will internally retrieve the region for the specified gene.Sample IDs are specified along the y-axis and the genomic postilion iis visualized along the x-axis.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#get metadata</span></span>
<span><span class="va">this_metadata</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#get myc region</span></span>
<span><span class="va">myc_region</span> <span class="op">=</span> <span class="fu"><a href="../reference/gene_to_region.html">gene_to_region</a></span><span class="op">(</span>gene_symbol <span class="op">=</span> <span class="st">"MYC"</span>, return_as <span class="op">=</span> <span class="st">"region"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#build plot</span></span>
<span><span class="fu"><a href="../reference/focal_cn_plot.html">focal_cn_plot</a></span><span class="op">(</span>these_samples_metadata <span class="op">=</span> <span class="va">this_metadata</span>, region <span class="op">=</span> <span class="va">myc_region</span>, type <span class="op">=</span> <span class="st">"loss"</span>, crop_distance <span class="op">=</span> <span class="fl">100000000</span><span class="op">)</span></span></code></pre></div>
<p><img src="cn_sv_tutorial_files/figure-html/cnfocal-1.png" width="1093.75"></p>
</div>
<div class="section level3">
<h3 id="sample-level-ideograms-">Sample-level Ideograms.<a class="anchor" aria-label="anchor" href="#sample-level-ideograms-"></a>
</h3>
<p>A great way to overview SVs and Copy Number Segments on sample-level is to call <code>fancy_ideogram</code>. This function creates an ideogram based on the chromosomes defined in the <code>chr_select</code> parameter. Copy Number states are plotted along the x-axis (selected contigs plotted vertically on the y-axis). The plot annotates Copy Number States from 0 to 6+. It is also possible to highlight genomic regions fo interest, such as genes, hotspots, difficult to map regions, etc. For convenience, one can specify a list of genes under the <code>gene_annotation</code> parameter. The function will then implement <code>gene_to_region</code> function to convert gene names to bed-style coordinates.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fancy_ideogram.html">fancy_ideogram</a></span><span class="op">(</span>this_sample_id <span class="op">=</span> <span class="va">this_sample_id</span>, </span>
<span>               include_ssm <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>               ssm_count <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>               gene_annotation <span class="op">=</span> <span class="st">"MYC"</span>,</span>
<span>               plot_title <span class="op">=</span> <span class="st">"Sample-level Ideogram Example"</span>,</span>
<span>               plot_subtitle <span class="op">=</span> <span class="st">"grch37"</span>,</span>
<span>               coding_only <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>               from_flatfile <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>               use_augmented_maf <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="cn_sv_tutorial_files/figure-html/fancy_ideogram-1.png" width="2100"></p>
</div>
<div class="section level3">
<h3 id="multi-sample-ideograms">Multi-sample Ideograms<a class="anchor" aria-label="anchor" href="#multi-sample-ideograms"></a>
</h3>
<p>It’s also possible to plot multiple samples (2, 3 or 4) in a similar manner. This can be used to infer inheritance patterns, hotspots, etc. across multiple samples or the same sample over different time points. For this purpose <code>fancy_ideogram_multisamp</code> was developed. This function is called with similar parameters as <code>fancy_ideogram</code> with the addition that <code>this_sample</code> has been replaced with <code>these_sample_ids</code>. The function automatically detects the number of samples provided and sets the plotting parameters accordingly.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#load two samples</span></span>
<span><span class="va">two_samples</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"00-15201_tumorA"</span>, <span class="st">"00-15201_tumorB"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#build plot</span></span>
<span><span class="fu"><a href="../reference/fancy_multisamp_ideogram.html">fancy_multisamp_ideogram</a></span><span class="op">(</span>these_sample_ids <span class="op">=</span> <span class="va">two_samples</span>,</span>
<span>                         plot_title <span class="op">=</span> <span class="st">"Multi-sample Ideograms Example"</span>,</span>
<span>                         plot_sub <span class="op">=</span> <span class="st">"grch37"</span>,</span>
<span>                         chr_anno_dist <span class="op">=</span> <span class="fl">2.5</span>, </span>
<span>                         include_cn2 <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                         chr_select <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                         coding_only <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                         from_flatfile <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                         use_augmented_maf <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="cn_sv_tutorial_files/figure-html/fancy_ideogram_multisamp-1.png" width="2100"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Ryan Morin, Kostia Dreval, Adam Mattsson.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
