<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>GAMBLR - working with GAMBL data painlessly in R â€¢ GAMBLR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="GAMBLR - working with GAMBL data painlessly in R">
<meta property="og:description" content="GAMBLR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/examples.html">GAMBLR - working with GAMBL data painlessly in R</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><link href="examples_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet">
<script src="examples_files/anchor-sections-1.0/anchor-sections.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>GAMBLR - working with GAMBL data painlessly in R</h1>
            
      
      
      <div class="hidden name"><code>examples.Rmd</code></div>

    </div>

    
    
<p>Load the GAMBLR package and other packages required by these examples.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">GAMBLR</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va"><a href="http://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va"><a href="https://github.com/PoisonAlien/maftools">maftools</a></span><span class="op">)</span></code></pre></div>
<p>Get the metadata, automagically joining both tables and dropping normals and RNA-seq samples.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">my_meta</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span>

<span class="co">#reduce to some of the more useful metadata fields</span>

<span class="va">my_meta</span> <span class="op">=</span> <span class="va">my_meta</span> <span class="op">%&gt;%</span> <span class="fu">select</span><span class="op">(</span><span class="va">sample_id</span>,<span class="va">biopsy_id</span>,<span class="va">myc_ba</span>,<span class="va">cohort</span>,<span class="va">pathology</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">my_meta</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt;                   sample_id            biopsy_id myc_ba       cohort pathology</span>
<span class="co">#&gt; 1 BLGSP-71-06-00001-01A-11D BLGSP-71-06-00001-01    POS BL_Pediatric        BL</span>
<span class="co">#&gt; 2 BLGSP-71-06-00004-01A-11D BLGSP-71-06-00004-01    POS BL_Pediatric        BL</span>
<span class="co">#&gt; 3 BLGSP-71-06-00013-01B-11D BLGSP-71-06-00013-01    POS BL_Pediatric        BL</span>
<span class="co">#&gt; 4 BLGSP-71-06-00080-01A-11D BLGSP-71-06-00080-01    POS BL_Pediatric        BL</span>
<span class="co">#&gt; 5 BLGSP-71-06-00019-01A-11D BLGSP-71-06-00019-01    POS BL_Pediatric        BL</span>
<span class="co">#&gt; 6 BLGSP-71-06-00008-01A-11D BLGSP-71-06-00008-01    NEG BL_Pediatric        BL</span></code></pre></div>
<p>Obtain all SV breakpoints (called by Manta) within a specific region of the genome</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">myc_locus_sv</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_manta_sv.html">get_manta_sv</a></span><span class="op">(</span>region<span class="op">=</span><span class="st">"8:128723128-128774067"</span>,pass<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span> 

<span class="co">#we can override default that requires SV to meet the Manta "Pass" filtering criterion</span>

<span class="va">myc_locus_sv</span> <span class="op">=</span> <span class="va">myc_locus_sv</span> <span class="op">%&gt;%</span> <span class="fu">select</span><span class="op">(</span><span class="va">tumour_sample_id</span>,<span class="va">CHROM_A</span>,<span class="va">START_A</span>,<span class="va">CHROM_B</span>,<span class="va">START_B</span>,<span class="va">STRAND_A</span>,<span class="va">STRAND_B</span>,<span class="va">VAF_tumour</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">myc_locus_sv</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt;   tumour_sample_id CHROM_A   START_A CHROM_B   START_B STRAND_A STRAND_B</span>
<span class="co">#&gt; 1  00-15201_tumorB       8 128727469       8 128727743        +        +</span>
<span class="co">#&gt; 2  01-12047_tumorC       8 128748814      14 106994498        -        +</span>
<span class="co">#&gt; 3  01-14774_tumorA       8 128727469       8 128727743        +        +</span>
<span class="co">#&gt; 4  01-23117_tumorB       8 128749370      14 106323836        -        +</span>
<span class="co">#&gt; 5  01-23117_tumorB       8 128749348      14 106323928        +        -</span>
<span class="co">#&gt; 6  02-18356_tumorB       8 128727463       8 128727752        +        +</span>
<span class="co">#&gt;   VAF_tumour</span>
<span class="co">#&gt; 1      0.151</span>
<span class="co">#&gt; 2      0.429</span>
<span class="co">#&gt; 3      0.163</span>
<span class="co">#&gt; 4      0.345</span>
<span class="co">#&gt; 5      0.154</span>
<span class="co">#&gt; 6      0.169</span></code></pre></div>
<p>Get all CN segments that overlap with a gene of interest and annotate them using the metadata.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">my_segments</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_cn_segments.html">get_cn_segments</a></span><span class="op">(</span>chromosome<span class="op">=</span><span class="st">"4"</span>,qstart<span class="op">=</span><span class="fl">83274467</span>,qend<span class="op">=</span><span class="fl">83295149</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">my_segments</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt;             ID chrom    start       end LOH_flag log.ratio</span>
<span class="co">#&gt; 1  1064-01-1TD  chr4    11870 190789536        0         0</span>
<span class="co">#&gt; 2 1182-01-03TD  chr4    69567 190785390        0         1</span>
<span class="co">#&gt; 3 1227-02-02TD  chr4 62724209 190790246        0         0</span>
<span class="co">#&gt; 4 1242-02-01BD  chr4    70761 190790246        0         0</span>
<span class="co">#&gt; 5 1296-01-03TD  chr4    98530 190762427        0         0</span>
<span class="co">#&gt; 6 1318-02-04BD  chr4    81303 190784332        0         0</span>

<span class="va">deleted_segments</span> <span class="op">=</span> <span class="va">my_segments</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">log.ratio</span><span class="op">&lt;</span><span class="fl">0</span><span class="op">)</span> <span class="co">#use some lower value if you want to be more stringent</span>

<span class="va">annotated_segments</span> <span class="op">=</span> <span class="fu">left_join</span><span class="op">(</span><span class="va">deleted_segments</span>,<span class="va">my_meta</span>,by<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ID"</span> <span class="op">=</span> <span class="st">"sample_id"</span><span class="op">)</span><span class="op">)</span>

<span class="va">annotated_segments</span> <span class="op">%&gt;%</span> <span class="fu">pull</span><span class="op">(</span><span class="va">cohort</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; .</span>
<span class="co">#&gt;   CLL_GenomeCanada        DLBCL_ctDNA     DLBCL_Gascoyne DLBCL_GenomeCanada </span>
<span class="co">#&gt;                  4                  2                  5                  9 </span>
<span class="co">#&gt;        DLBCL_HTMCP         DLBCL_ICGC  DLBCL_LSARP_Trios        DLBCL_Marra </span>
<span class="co">#&gt;                  3                  5                 53                 10 </span>
<span class="co">#&gt;    FL_GenomeCanada            FL_ICGC          FL_Kridel    MALY_Other_ICGC </span>
<span class="co">#&gt;                  1                  4                  6                  3 </span>
<span class="co">#&gt;        MM_mmsanger </span>
<span class="co">#&gt;                  4</span></code></pre></div>
<p>Retrieve all somatic mutations in a given gene and visualize them using MAFtools.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">all_ssms</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_ssm_by_gene.html">get_ssm_by_gene</a></span><span class="op">(</span>gene_symbol<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CCND3"</span><span class="op">)</span>,coding_only <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="va">all_ssms</span> <span class="op">=</span> <span class="va">all_ssms</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span>

<span class="co">#make a MAFtools object and plot a lollipop plot</span>

<span class="va">maf_obj</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/maftools/man/read.maf.html">read.maf</a></span><span class="op">(</span><span class="va">all_ssms</span><span class="op">)</span>
<span class="co">#&gt; -Validating</span>
<span class="co">#&gt; -Silent variants: 2 </span>
<span class="co">#&gt; -Summarizing</span>
<span class="co">#&gt; -Processing clinical data</span>
<span class="co">#&gt; --Missing clinical data</span>
<span class="co">#&gt; -Finished in 0.244s elapsed (2.948s cpu)</span>
<span class="fu"><a href="https://rdrr.io/pkg/maftools/man/lollipopPlot.html">lollipopPlot</a></span><span class="op">(</span><span class="va">maf_obj</span>,gene<span class="op">=</span><span class="st">"CCND3"</span><span class="op">)</span>
<span class="co">#&gt; Assuming protein change information are stored under column HGVSp_Short. Use argument AACol to override if necessary.</span>
<span class="co">#&gt; 4 transcripts available. Use arguments refSeqID or proteinID to manually specify tx name.</span>
<span class="co">#&gt;     HGNC    refseq.ID   protein.ID aa.length</span>
<span class="co">#&gt; 1: CCND3 NM_001136125 NP_001129597       220</span>
<span class="co">#&gt; 2: CCND3    NM_001760    NP_001751       292</span>
<span class="co">#&gt; 3: CCND3 NM_001136017 NP_001129489       211</span>
<span class="co">#&gt; 4: CCND3 NM_001136126 NP_001129598        96</span>
<span class="co">#&gt; Using longer transcript NM_001760 for now.</span></code></pre></div>
<p><img src="examples_files/figure-html/ssms_and_maftools-1.png" width="700"></p>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Ryan Morin.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
