<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>GAMBLR - working with GAMBL data painlessly in R • GAMBLR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="GAMBLR - working with GAMBL data painlessly in R">
<meta property="og:description" content="GAMBLR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9500</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/examples.html">GAMBLR - working with GAMBL data painlessly in R</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><link href="examples_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet">
<script src="examples_files/anchor-sections-1.0/anchor-sections.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>GAMBLR - working with GAMBL data painlessly in R</h1>
            
      
      
      <div class="hidden name"><code>examples.Rmd</code></div>

    </div>

    
    
<p>If you haven’t already, clone the github repo: <code>git clone git@github.com:morinlab/GAMBLR.git</code> Run this vignette using the base directory of the GAMBLR repository as your working directory. See the Readme in the repository for more tips on getting set up.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Load the GAMBLR package and other packages required by these examples. </span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">GAMBLR</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va"><a href="http://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va"><a href="https://github.com/PoisonAlien/maftools">maftools</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/circlize">circlize</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va"><a href="http://r-datatable.com">data.table</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va"><a href="https://rmariadb.r-dbi.org">RMariaDB</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va"><a href="http://r-dbi.github.io/DBI">DBI</a></span><span class="op">)</span></code></pre></div>
<p>The key level-1 GAMBL outputs are making their way into a MySQL (a.k.a. MariaDB) database. The most intuitive way to query the data is using the dbplyr R package. The initial connection and a method to determine what each table contains is shown below.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">con</span> <span class="op">&lt;-</span> <span class="fu">DBI</span><span class="fu">::</span><span class="fu"><a href="https://dbi.r-dbi.org/reference/dbConnect.html">dbConnect</a></span><span class="op">(</span><span class="fu">RMariaDB</span><span class="fu">::</span><span class="fu"><a href="https://rmariadb.r-dbi.org/reference/dbConnect-MariaDBDriver-method.html">MariaDB</a></span><span class="op">(</span><span class="op">)</span>, dbname <span class="op">=</span> <span class="st">"gambl_test"</span><span class="op">)</span>
<span class="co"># use DBI function to list the tables</span>
<span class="va">all_table_names</span> <span class="op">=</span> <span class="fu">DBI</span><span class="fu">::</span><span class="fu"><a href="https://dbi.r-dbi.org/reference/dbListTables.html">dbListTables</a></span><span class="op">(</span><span class="va">con</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">all_table_names</span><span class="op">)</span>
<span class="co">#&gt;  [1] "analysis"            "bedpe_manta_hg19"    "biopsy_metadata"    </span>
<span class="co">#&gt;  [4] "derived_data"        "maf_slms3_hg19"      "maf_slms3_hg19_icgc"</span>
<span class="co">#&gt;  [7] "maf_slms3_hg38"      "outcome_metadata"    "sample_metadata"    </span>
<span class="co">#&gt; [10] "seg_battenberg_hg19"</span>
<span class="co"># peek at the contents of each table. </span>
<span class="kw">for</span><span class="op">(</span><span class="va">table_name</span> <span class="kw">in</span> <span class="va">all_table_names</span><span class="op">)</span><span class="op">{</span>
  <span class="va">table_db</span> <span class="op">&lt;-</span> <span class="fu">tbl</span><span class="op">(</span><span class="va">con</span>, <span class="va">table_name</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">table_name</span><span class="op">)</span>
  
<span class="op">}</span>
<span class="co">#&gt; [1] "analysis"</span>
<span class="co">#&gt; [1] "bedpe_manta_hg19"</span>
<span class="co">#&gt; [1] "biopsy_metadata"</span>
<span class="co">#&gt; [1] "derived_data"</span>
<span class="co">#&gt; [1] "maf_slms3_hg19"</span>
<span class="co">#&gt; [1] "maf_slms3_hg19_icgc"</span>
<span class="co">#&gt; [1] "maf_slms3_hg38"</span>
<span class="co">#&gt; [1] "outcome_metadata"</span>
<span class="co">#&gt; [1] "sample_metadata"</span>
<span class="co">#&gt; [1] "seg_battenberg_hg19"</span></code></pre></div>
<p>Get the metadata, automagically joining both tables and dropping normals and RNA-seq samples.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">my_metadata</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># reduce to some of the more useful metadata fields</span>

<span class="va">my_metadata</span> <span class="op">=</span> <span class="va">my_metadata</span> <span class="op">%&gt;%</span> <span class="fu">select</span><span class="op">(</span><span class="va">sample_id</span>,<span class="va">biopsy_id</span>,<span class="va">myc_ba</span>,<span class="va">cohort</span>,<span class="va">pathology</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">my_metadata</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt;                   sample_id            biopsy_id myc_ba       cohort pathology</span>
<span class="co">#&gt; 1 BLGSP-71-06-00001-01A-11D BLGSP-71-06-00001-01    POS BL_Pediatric        BL</span>
<span class="co">#&gt; 2 BLGSP-71-06-00004-01A-11D BLGSP-71-06-00004-01    POS BL_Pediatric        BL</span>
<span class="co">#&gt; 3 BLGSP-71-06-00013-01B-11D BLGSP-71-06-00013-01    POS BL_Pediatric        BL</span>
<span class="co">#&gt; 4 BLGSP-71-06-00080-01A-11D BLGSP-71-06-00080-01    POS BL_Pediatric        BL</span>
<span class="co">#&gt; 5 BLGSP-71-06-00019-01A-11D BLGSP-71-06-00019-01    POS BL_Pediatric        BL</span>
<span class="co">#&gt; 6 BLGSP-71-06-00008-01A-11D BLGSP-71-06-00008-01    NEG BL_Pediatric        BL</span>

<span class="co">#An important feature for reproducibility is that we all use the same exact subset of samples when performing various level-3 analyses for a study. This function effectively "locks in" a set of cases for a study based on some filters it applies automatically based on a study set identifier. For retrieving the metadata for all BLGSP cases with WGS data, you can use the following:</span>

<span class="va">blgsp_metadata</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span>case_set<span class="op">=</span><span class="st">"BLGSP-study"</span><span class="op">)</span>

<span class="co">#as you can see this spans several cohorts and doesn't even just include BL pathology (for complex reasons)</span>

<span class="va">blgsp_metadata</span> <span class="op">%&gt;%</span> <span class="fu">pull</span><span class="op">(</span><span class="va">cohort</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; .</span>
<span class="co">#&gt;        BL_Adult   BL_cell_lines BLGSP_Bcell_UNC         BL_ICGC    BL_Pediatric </span>
<span class="co">#&gt;              81              23               6              17             124</span>

<span class="co">#If you want the sample_id (i.e. the Tumor_Sample_Barcode) for all these cases, for example to subset a MAF file, you can extract them into a vector:</span>

<span class="va">blgsp_sample_ids</span> <span class="op">=</span> <span class="fu">pull</span><span class="op">(</span><span class="va">blgsp_metadata</span>,<span class="va">sample_id</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">blgsp_sample_ids</span><span class="op">)</span>
<span class="co">#&gt; [1] 251</span></code></pre></div>
<p>The largest tables in the database are those that contain mutations from MAF files. The tables are simply all the rows and columns from a given merged MAF file. There will likely be four MAF-derived tables in this database, one for each reference build for “GAMBL” and the complete “GAMBL + ICGC”. Although the names may change, currently the tables are implicitly GAMBL-only unless they contain “icgc” in their name. For example, maf_slms3_hg19_icgc contains all hg19 (grch37/hs37d5) mutations from GAMBL and icgc_dart/external cases.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hg19_maf</span> <span class="op">=</span> <span class="fu">tbl</span><span class="op">(</span><span class="va">con</span>,<span class="st">"maf_slms3_hg19_icgc"</span><span class="op">)</span>
<span class="va">mutation_counts</span> <span class="op">=</span> <span class="va">hg19_maf</span> <span class="op">%&gt;%</span> <span class="fu">group_by</span><span class="op">(</span><span class="va">Tumor_Sample_Barcode</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">tally</span><span class="op">(</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">mutation_counts</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; # Source:   lazy query [?? x 2]</span>
<span class="co">#&gt; # Database: mysql [rmorin@mysql01:NA/gambl_test]</span>
<span class="co">#&gt;   Tumor_Sample_Barcode       n</span>
<span class="co">#&gt;   &lt;chr&gt;                &lt;int64&gt;</span>
<span class="co">#&gt; 1 00-14595_tumorA         7139</span>
<span class="co">#&gt; 2 00-14595_tumorB        18212</span>
<span class="co">#&gt; 3 00-14595_tumorC        27633</span>
<span class="co">#&gt; 4 00-15201_tumorA         6423</span>
<span class="co">#&gt; 5 00-15201_tumorB         8255</span>
<span class="co">#&gt; 6 00-20702T               6335</span>

<span class="va">mutation_counts</span>  <span class="op">%&gt;%</span> <span class="fu">ggplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_histogram</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x<span class="op">=</span><span class="va">n</span><span class="op">)</span>,bins<span class="op">=</span><span class="fl">100</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.window.html">xlim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">25000</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Warning: Removed 142 rows containing non-finite values (stat_bin).</span>
<span class="co">#&gt; Warning: Removed 2 rows containing missing values (geom_bar).</span></code></pre></div>
<p><img src="examples_files/figure-html/mutation_totals-1.png" width="75%"></p>
<p>Here’s a histogram of the total number of mutations per genome across all GAMBL cases. Pretty hard to interpret this without metadata. There are two metadata tables. Often we need to join them both to get all the information we want.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sample_meta</span> <span class="op">=</span> <span class="fu">tbl</span><span class="op">(</span><span class="va">con</span>,<span class="st">"sample_metadata"</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">seq_type</span> <span class="op">==</span> <span class="st">"genome"</span> <span class="op">&amp;</span> <span class="va">tissue_status</span> <span class="op">==</span> <span class="st">"tumour"</span><span class="op">)</span>
<span class="co">#if we only care about genomes, we can drop/filter anything that isn't a tumour genome</span>
<span class="co">#The key for joining this table to the mutation information is to use sample_id. Think of this as equivalent to a library_id. It will differ depending on what assay was done to the sample. </span>

<span class="va">biopsy_meta</span> <span class="op">=</span> <span class="fu">tbl</span><span class="op">(</span><span class="va">con</span>,<span class="st">"biopsy_metadata"</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">select</span><span class="op">(</span><span class="op">-</span><span class="va">patient_id</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">select</span><span class="op">(</span><span class="op">-</span><span class="va">pathology</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">select</span><span class="op">(</span><span class="op">-</span><span class="va">time_point</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">select</span><span class="op">(</span><span class="op">-</span><span class="va">EBV_status_inf</span><span class="op">)</span>
<span class="co"># this table is keyed on biopsy_id. One biopsy can have more than one sample_id. This table must be joined to the sample table using biopsy_id. Because of some redundancy in columns between the two tables, I've dropped all redundant columns with the exception of biopsy_id.</span>

<span class="va">all_meta</span> <span class="op">=</span> <span class="fu">left_join</span><span class="op">(</span><span class="va">sample_meta</span>,<span class="va">biopsy_meta</span>,by<span class="op">=</span><span class="st">"biopsy_id"</span><span class="op">)</span>
<span class="co"># IMPORTANT: the dbplyr package uses mysql queries under the hood to lazily retrieve the datat you need for each table on-the-fly. To properly use the efficiency of indexing and joins in MySQL, don't convert your tables into data frames until you're done all the necessary joins. </span></code></pre></div>
<p>Obtain all SV breakpoints (called by Manta) within a specific region of the genome and visualize them.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">myc_locus_sv</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_manta_sv.html">get_manta_sv</a></span><span class="op">(</span>region<span class="op">=</span><span class="st">"8:128723128-128774067"</span>,pass<span class="op">=</span><span class="cn">FALSE</span>,with_chr_prefix<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> 

<span class="co"># we can override default that requires SV to meet the Manta "Pass" filtering criterion</span>
<span class="co"># here we are also asking for the chromosomes to be named with a chr prefix (for circlize compatability)</span>

<span class="va">bed1</span> <span class="op">=</span> <span class="va">myc_locus_sv</span> <span class="op">%&gt;%</span> <span class="fu">select</span><span class="op">(</span><span class="va">CHROM_A</span>,<span class="va">START_A</span>,<span class="va">END_A</span>,<span class="va">tumour_sample_id</span><span class="op">)</span>
<span class="va">bed2</span> <span class="op">=</span> <span class="va">myc_locus_sv</span> <span class="op">%&gt;%</span> <span class="fu">select</span><span class="op">(</span><span class="va">CHROM_B</span>,<span class="va">START_B</span>,<span class="va">END_B</span>,<span class="va">tumour_sample_id</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">bed1</span><span class="op">)</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"chrom"</span>,<span class="st">"start"</span>,<span class="st">"end"</span>,<span class="st">"sample_id"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">bed2</span><span class="op">)</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"chrom"</span>,<span class="st">"start"</span>,<span class="st">"end"</span>,<span class="st">"sample_id"</span><span class="op">)</span>

<span class="va">myc_locus_sv</span> <span class="op">=</span> <span class="va">myc_locus_sv</span> <span class="op">%&gt;%</span> <span class="fu">select</span><span class="op">(</span><span class="va">tumour_sample_id</span>,<span class="va">CHROM_A</span>,<span class="va">START_A</span>,<span class="va">CHROM_B</span>,<span class="va">START_B</span>,<span class="va">STRAND_A</span>,<span class="va">STRAND_B</span>,<span class="va">VAF_tumour</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">myc_locus_sv</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt;   tumour_sample_id CHROM_A   START_A CHROM_B   START_B STRAND_A STRAND_B</span>
<span class="co">#&gt; 1  00-15201_tumorB    chr8 128727469    chr8 128727743        +        +</span>
<span class="co">#&gt; 2  01-12047_tumorC    chr8 128748814   chr14 106994498        -        +</span>
<span class="co">#&gt; 3  01-14774_tumorA    chr8 128727469    chr8 128727743        +        +</span>
<span class="co">#&gt; 4        01-20774T    chr8  66878096    chr8 128735744        +        -</span>
<span class="co">#&gt; 5        01-20774T    chr8 128771911    chr8 128920622        +        -</span>
<span class="co">#&gt; 6  01-23117_tumorB    chr8 128749370   chr14 106323836        -        +</span>
<span class="co">#&gt;   VAF_tumour</span>
<span class="co">#&gt; 1      0.151</span>
<span class="co">#&gt; 2      0.429</span>
<span class="co">#&gt; 3      0.163</span>
<span class="co">#&gt; 4      0.358</span>
<span class="co">#&gt; 5      0.951</span>
<span class="co">#&gt; 6      0.345</span>

<span class="fu"><a href="https://rdrr.io/pkg/circlize/man/circos.initializeWithIdeogram.html">circos.initializeWithIdeogram</a></span><span class="op">(</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/circlize/man/circos.genomicLink.html">circos.genomicLink</a></span><span class="op">(</span><span class="va">bed1</span>, <span class="va">bed2</span>,col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/circlize/man/rand_color.html">rand_color</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">bed1</span><span class="op">)</span>,transparency<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="examples_files/figure-html/manta-1.png" width="80%"> This package also includes a convenience function to facilitate annotation of common oncogene fusions/SVs in B-cell lymphomas.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">unannotated_sv</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_manta_sv.html">get_manta_sv</a></span><span class="op">(</span><span class="op">)</span> 

<span class="co">#in this example, let's just look at the SVs annotated as likely driving BCL6 expression</span>
<span class="va">annotated_sv</span> <span class="op">=</span> <span class="fu"><a href="../reference/annotate_sv.html">annotate_sv</a></span><span class="op">(</span><span class="va">unannotated_sv</span>,with_chr_prefix <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">partner</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="co">#select(-entrez) %&gt;% </span>
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">gene</span><span class="op">==</span><span class="st">"BCL6"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; Warning in if (!grepl("chr", all.annotated$chrom1)) {: the condition has length</span>
<span class="co">#&gt; &gt; 1 and only the first element will be used</span>

<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">annotated_sv</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt;   chrom1    start1      end1 chrom2   start2     end2 name score strand1</span>
<span class="co">#&gt; 1   chr3 187462585 187462585   chr6 27107409 27107409    .    78       +</span>
<span class="co">#&gt; 2   chr3 187462600 187462600   chr6 27107409 27107409    .    90       -</span>
<span class="co">#&gt; 3   chr3 187462585 187462585   chr6 27107409 27107409    .    67       +</span>
<span class="co">#&gt; 4   chr3 187462600 187462600   chr6 27107409 27107409    .    93       -</span>
<span class="co">#&gt; 5   chr3 187462525 187462525  chr12 25207319 25207319    .   118       +</span>
<span class="co">#&gt; 6   chr3 187462548 187462548  chr12 25207342 25207342    .    97       -</span>
<span class="co">#&gt;   strand2 tumour_sample_id gene   partner         fusion</span>
<span class="co">#&gt; 1       +  01-14774_tumorA BCL6 HIST1H2BK HIST1H2BK-BCL6</span>
<span class="co">#&gt; 2       -  01-14774_tumorA BCL6 HIST1H2BK HIST1H2BK-BCL6</span>
<span class="co">#&gt; 3       +  01-14774_tumorB BCL6 HIST1H2BK HIST1H2BK-BCL6</span>
<span class="co">#&gt; 4       -  01-14774_tumorB BCL6 HIST1H2BK HIST1H2BK-BCL6</span>
<span class="co">#&gt; 5       +  02-24492_tumorA BCL6      LRMP      LRMP-BCL6</span>
<span class="co">#&gt; 6       -  02-24492_tumorA BCL6      LRMP      LRMP-BCL6</span>

<span class="co">#for labelling, get the unique set of partners</span>
<span class="va">to_label</span> <span class="op">=</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/duplicated.html">unique</a></span><span class="op">(</span><span class="va">annotated_sv</span><span class="op">$</span><span class="va">partner</span><span class="op">)</span>
<span class="va">partner_label</span> <span class="op">=</span> <span class="va">grch37_partners</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">gene</span> <span class="op">%in%</span> <span class="va">to_label</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">mutate</span><span class="op">(</span>chrom <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>,<span class="va">chrom</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span>

<span class="va">bed1</span> <span class="op">=</span> <span class="va">annotated_sv</span> <span class="op">%&gt;%</span> <span class="fu">select</span><span class="op">(</span><span class="va">chrom1</span>,<span class="va">start1</span>,<span class="va">end1</span>,<span class="va">tumour_sample_id</span>,<span class="va">fusion</span><span class="op">)</span>
<span class="va">bed2</span> <span class="op">=</span> <span class="va">annotated_sv</span> <span class="op">%&gt;%</span> <span class="fu">select</span><span class="op">(</span><span class="va">chrom2</span>,<span class="va">start2</span>,<span class="va">end2</span>,<span class="va">tumour_sample_id</span>,<span class="va">fusion</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">bed1</span><span class="op">)</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"chrom"</span>,<span class="st">"start"</span>,<span class="st">"end"</span>,<span class="st">"sample_id"</span>,<span class="st">"fusion"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">bed2</span><span class="op">)</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"chrom"</span>,<span class="st">"start"</span>,<span class="st">"end"</span>,<span class="st">"sample_id"</span>,<span class="st">"fusion"</span><span class="op">)</span>

<span class="co">#circos.initializeWithIdeogram()</span>
<span class="fu"><a href="https://rdrr.io/pkg/circlize/man/circos.clear.html">circos.clear</a></span><span class="op">(</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/circlize/man/circos.initializeWithIdeogram.html">circos.initializeWithIdeogram</a></span><span class="op">(</span>plotType <span class="op">=</span> <span class="cn">NULL</span>,chromosome.index <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span>,<span class="st">"X"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/circlize/man/circos.genomicLabels.html">circos.genomicLabels</a></span><span class="op">(</span><span class="va">partner_label</span>, labels.column <span class="op">=</span> <span class="fl">4</span>, side <span class="op">=</span> <span class="st">"outside"</span>, cex<span class="op">=</span><span class="fl">0.4</span>,col<span class="op">=</span><span class="st">'black'</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/circlize/man/circos.genomicIdeogram.html">circos.genomicIdeogram</a></span><span class="op">(</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/circlize/man/circos.genomicLink.html">circos.genomicLink</a></span><span class="op">(</span><span class="va">bed1</span>, <span class="va">bed2</span>,col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/circlize/man/rand_color.html">rand_color</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">bed1</span><span class="op">)</span>,transparency<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="examples_files/figure-html/oncogene_annotations-1.png" width="100%"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">unannotated_sv</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_manta_sv.html">get_manta_sv</a></span><span class="op">(</span><span class="op">)</span> 

<span class="co">#Now let's try the SVs annotated as likely driving MYC expression</span>
<span class="va">annotated_sv</span> <span class="op">=</span> <span class="fu"><a href="../reference/annotate_sv.html">annotate_sv</a></span><span class="op">(</span><span class="va">unannotated_sv</span>,with_chr_prefix <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">partner</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">gene</span><span class="op">==</span><span class="st">"MYC"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; Warning in if (!grepl("chr", all.annotated$chrom1)) {: the condition has length</span>
<span class="co">#&gt; &gt; 1 and only the first element will be used</span>

<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">annotated_sv</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt;   chrom1    start1      end1 chrom2    start2      end2 name score strand1</span>
<span class="co">#&gt; 1   chr8 134207510 134207517   chr9  36763818  36763825    .    46       +</span>
<span class="co">#&gt; 2   chr8 128748814 128748814  chr14 106994498 106994498    .    85       -</span>
<span class="co">#&gt; 3   chr8 128749370 128749370  chr14 106323836 106323836    .   122       -</span>
<span class="co">#&gt; 4   chr8 128749348 128749348  chr14 106323928 106323928    .    53       +</span>
<span class="co">#&gt; 5   chr8 128746623 128746625  chr14 106328385 106328387    .    60       -</span>
<span class="co">#&gt; 6   chr8 128922317 128922939  chr22  23114594  23114952    .    88       +</span>
<span class="co">#&gt;   strand2 tumour_sample_id gene partner   fusion</span>
<span class="co">#&gt; 1       -  00-26427_tumorA  MYC    PAX5 PAX5-MYC</span>
<span class="co">#&gt; 2       +  01-12047_tumorC  MYC     IGH  IGH-MYC</span>
<span class="co">#&gt; 3       +  01-23117_tumorB  MYC     IGH  IGH-MYC</span>
<span class="co">#&gt; 4       -  01-23117_tumorB  MYC     IGH  IGH-MYC</span>
<span class="co">#&gt; 5       +  02-24492_tumorB  MYC     IGH  IGH-MYC</span>
<span class="co">#&gt; 6       -        05-10981T  MYC     IGL  IGL-MYC</span>

<span class="co">#for labelling, get the unique set of partners</span>
<span class="va">to_label</span> <span class="op">=</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/duplicated.html">unique</a></span><span class="op">(</span><span class="va">annotated_sv</span><span class="op">$</span><span class="va">partner</span><span class="op">)</span>
<span class="va">partner_label</span> <span class="op">=</span> <span class="va">grch37_partners</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">gene</span> <span class="op">%in%</span> <span class="va">to_label</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">mutate</span><span class="op">(</span>chrom <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>,<span class="va">chrom</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">onco_label</span> <span class="op">=</span> <span class="va">grch37_oncogene</span> <span class="op">%&gt;%</span> <span class="fu">mutate</span><span class="op">(</span>chrom <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>,<span class="va">chrom</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span>

<span class="va">bed1</span> <span class="op">=</span> <span class="va">annotated_sv</span> <span class="op">%&gt;%</span> <span class="fu">select</span><span class="op">(</span><span class="va">chrom1</span>,<span class="va">start1</span>,<span class="va">end1</span>,<span class="va">tumour_sample_id</span>,<span class="va">fusion</span><span class="op">)</span>
<span class="va">bed2</span> <span class="op">=</span> <span class="va">annotated_sv</span> <span class="op">%&gt;%</span> <span class="fu">select</span><span class="op">(</span><span class="va">chrom2</span>,<span class="va">start2</span>,<span class="va">end2</span>,<span class="va">tumour_sample_id</span>,<span class="va">fusion</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">bed1</span><span class="op">)</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"chrom"</span>,<span class="st">"start"</span>,<span class="st">"end"</span>,<span class="st">"sample_id"</span>,<span class="st">"fusion"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">bed2</span><span class="op">)</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"chrom"</span>,<span class="st">"start"</span>,<span class="st">"end"</span>,<span class="st">"sample_id"</span>,<span class="st">"fusion"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/circlize/man/circos.initializeWithIdeogram.html">circos.initializeWithIdeogram</a></span><span class="op">(</span>plotType <span class="op">=</span> <span class="cn">NULL</span>,chromosome.index <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span>,<span class="st">"X"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="co">#circos.genomicLabels(partner_label, labels.column = 4, side = "outside", cex=0.4,col='black')</span>

<span class="co">#circos.genomicLabels(onco_label, labels.column = 4, side = "outside", cex=0.4,col='red')</span>

<span class="va">all_labels</span> <span class="op">=</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/rbindlist.html">rbind</a></span><span class="op">(</span><span class="va">onco_label</span>,<span class="va">partner_label</span><span class="op">)</span>
<span class="va">cols</span><span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"red"</span>,<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">onco_label</span><span class="op">$</span><span class="va">chrom</span><span class="op">)</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"black"</span>,<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">partner_label</span><span class="op">$</span><span class="va">chrom</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/circlize/man/circos.genomicLabels.html">circos.genomicLabels</a></span><span class="op">(</span><span class="va">all_labels</span>, labels.column <span class="op">=</span> <span class="fl">4</span>, side <span class="op">=</span> <span class="st">"outside"</span>, cex<span class="op">=</span><span class="fl">0.7</span>,col<span class="op">=</span><span class="va">cols</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/circlize/man/circos.genomicIdeogram.html">circos.genomicIdeogram</a></span><span class="op">(</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/circlize/man/circos.genomicLink.html">circos.genomicLink</a></span><span class="op">(</span><span class="va">bed1</span>, <span class="va">bed2</span>,col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/circlize/man/rand_color.html">rand_color</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">bed1</span><span class="op">)</span>,transparency<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="examples_files/figure-html/oncogene_annotations_myc-1.png" width="100%"></p>
<p>This package has convenience features such as the ability to run liftOver on a set of coordinates (e.g. in bedpe format)</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bedpe_hg19</span> <span class="op">=</span> <span class="va">unannotated_sv</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fl">20</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">bedpe_hg19</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt;   CHROM_A  START_A    END_A CHROM_B  START_B    END_B NAME SOMATIC_SCORE</span>
<span class="co">#&gt; 1       1  1556541  1556547       1  1556664  1556670    .            40</span>
<span class="co">#&gt; 2       1  6012725  6012732       1  6012825  6012832    .            48</span>
<span class="co">#&gt; 3       1  8464072  8464090       1  8464293  8464311    .            40</span>
<span class="co">#&gt; 4       1 10084099 10084251       1 10084266 10084411    .            48</span>
<span class="co">#&gt; 5       1 10526162 10526172       1 10526290 10526300    .            40</span>
<span class="co">#&gt; 6       1 15878430 15878436       1 15878608 15878614    .            40</span>
<span class="co">#&gt;   STRAND_A STRAND_B TYPE FILTER VAF_tumour VAF_normal DP_tumour DP_normal</span>
<span class="co">#&gt; 1        -        -  BND   PASS      0.145          0        55        73</span>
<span class="co">#&gt; 2        +        +  BND   PASS      0.156          0        32        85</span>
<span class="co">#&gt; 3        +        +  BND   PASS      0.112          0       215        84</span>
<span class="co">#&gt; 4        -        -  BND   PASS      0.186          0        43        50</span>
<span class="co">#&gt; 5        -        -  BND   PASS      0.157          0       121        92</span>
<span class="co">#&gt; 6        +        +  BND   PASS      0.119          0       126       101</span>
<span class="co">#&gt;   tumour_sample_id normal_sample_id pair_status</span>
<span class="co">#&gt; 1  00-14595_tumorA  00-14595_normal     matched</span>
<span class="co">#&gt; 2  00-14595_tumorA  00-14595_normal     matched</span>
<span class="co">#&gt; 3  00-14595_tumorA  00-14595_normal     matched</span>
<span class="co">#&gt; 4  00-14595_tumorA  00-14595_normal     matched</span>
<span class="co">#&gt; 5  00-14595_tumorA  00-14595_normal     matched</span>
<span class="co">#&gt; 6  00-14595_tumorA  00-14595_normal     matched</span>
<span class="va">bedpe_hg38</span> <span class="op">=</span> <span class="fu"><a href="../reference/liftover_bedpe.html">liftover_bedpe</a></span><span class="op">(</span>bedpe_df <span class="op">=</span> <span class="va">bedpe_hg19</span>,target_build <span class="op">=</span> <span class="st">"hg38"</span><span class="op">)</span>
<span class="co">#&gt; Warning in if (!grepl("chr", original_bedpe$CHROM_A)) {: the condition has</span>
<span class="co">#&gt; length &gt; 1 and only the first element will be used</span>
<span class="co">#&gt; Warning: 20 parsing failures.</span>
<span class="co">#&gt; row col  expected     actual         file</span>
<span class="co">#&gt;   1  -- 6 columns 12 columns literal data</span>
<span class="co">#&gt;   2  -- 6 columns 12 columns literal data</span>
<span class="co">#&gt;   3  -- 6 columns 12 columns literal data</span>
<span class="co">#&gt;   4  -- 6 columns 12 columns literal data</span>
<span class="co">#&gt;   5  -- 6 columns 12 columns literal data</span>
<span class="co">#&gt; ... ... ......... .......... ............</span>
<span class="co">#&gt; See problems(...) for more details.</span>

<span class="co">#&gt; Warning: 20 parsing failures.</span>
<span class="co">#&gt; row col  expected     actual         file</span>
<span class="co">#&gt;   1  -- 6 columns 12 columns literal data</span>
<span class="co">#&gt;   2  -- 6 columns 12 columns literal data</span>
<span class="co">#&gt;   3  -- 6 columns 12 columns literal data</span>
<span class="co">#&gt;   4  -- 6 columns 12 columns literal data</span>
<span class="co">#&gt;   5  -- 6 columns 12 columns literal data</span>
<span class="co">#&gt; ... ... ......... .......... ............</span>
<span class="co">#&gt; See problems(...) for more details.</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">bedpe_hg38</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 6 x 19</span>
<span class="co">#&gt;   CHROM_A START_A  END_A CHROM_B START_B  END_B NAME  SOMATIC_SCORE STRAND_A</span>
<span class="co">#&gt;   &lt;chr&gt;     &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;     &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;         &lt;dbl&gt; &lt;chr&gt;   </span>
<span class="co">#&gt; 1 chr1     1.62e6 1.62e6 chr1     1.62e6 1.62e6 .                40 -       </span>
<span class="co">#&gt; 2 chr1     5.95e6 5.95e6 chr1     5.95e6 5.95e6 .                48 +       </span>
<span class="co">#&gt; 3 chr1     8.40e6 8.40e6 chr1     8.40e6 8.40e6 .                40 +       </span>
<span class="co">#&gt; 4 chr1     1.00e7 1.00e7 chr1     1.00e7 1.00e7 .                48 -       </span>
<span class="co">#&gt; 5 chr1     1.05e7 1.05e7 chr1     1.05e7 1.05e7 .                40 -       </span>
<span class="co">#&gt; 6 chr1     1.56e7 1.56e7 chr1     1.56e7 1.56e7 .                40 +       </span>
<span class="co">#&gt; # … with 10 more variables: STRAND_B &lt;chr&gt;, TYPE &lt;chr&gt;, FILTER &lt;chr&gt;,</span>
<span class="co">#&gt; #   VAF_tumour &lt;dbl&gt;, VAF_normal &lt;dbl&gt;, DP_tumour &lt;int&gt;, DP_normal &lt;int&gt;,</span>
<span class="co">#&gt; #   tumour_sample_id &lt;chr&gt;, normal_sample_id &lt;chr&gt;, pair_status &lt;chr&gt;</span></code></pre></div>
<p>Working with all mutations from a sample.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># we can also directly query the database to get a MAF per patient for patient-centric visualizations. # this is not using the GAMBLR functions but shows an example of how unsupported queries can be accomplished. Beware queries that will return many thousands of variants. Thes will be slow and may fail if they're too greedy </span>
<span class="co"># note that here I'm not restricting to only coding variants</span>
<span class="va">example_dlbcl</span> <span class="op">=</span> <span class="va">hg19_maf</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">Tumor_Sample_Barcode</span> <span class="op">==</span> <span class="st">"HTMCP-01-06-00422-01A-01D"</span><span class="op">)</span>
<span class="va">example_dlbcl_df</span> <span class="op">=</span> <span class="va">example_dlbcl</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">example_dlbcl_maf</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/maftools/man/read.maf.html">read.maf</a></span><span class="op">(</span><span class="va">example_dlbcl_df</span><span class="op">)</span>
<span class="co">#&gt; -Validating</span>
<span class="co">#&gt; -Silent variants: 4552 </span>
<span class="co">#&gt; -Summarizing</span>
<span class="co">#&gt; -Processing clinical data</span>
<span class="co">#&gt; --Missing clinical data</span>
<span class="co">#&gt; -Finished in 0.142s elapsed (0.404s cpu)</span>
<span class="fu"><a href="https://rdrr.io/pkg/maftools/man/rainfallPlot.html">rainfallPlot</a></span><span class="op">(</span><span class="va">example_dlbcl_maf</span><span class="op">)</span>
<span class="co">#&gt; Processing HTMCP-01-06-00422-01A-01D..</span></code></pre></div>
<p><img src="examples_files/figure-html/rainfall_plots-1.png" width="100%"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">DBI</span><span class="fu">::</span><span class="fu"><a href="https://dbi.r-dbi.org/reference/dbDisconnect.html">dbDisconnect</a></span><span class="op">(</span><span class="va">con</span><span class="op">)</span></code></pre></div>
<p>What if you want to view the copy number landscape of a single patient and the VAF of somatic mutations in different copy number states? This can sometimes help you scrutinize whether the copy number fit looks reasonable.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#use the same sample as the previous example</span>
<span class="va">my_sample</span> <span class="op">=</span> <span class="st">"HTMCP-01-06-00422-01A-01D"</span>

<span class="fu"><a href="../reference/copy_number_vaf_plot.html">copy_number_vaf_plot</a></span><span class="op">(</span>this_sample<span class="op">=</span><span class="va">my_sample</span><span class="op">)</span>
<span class="co">#&gt; [1] "copy number colours"</span>
<span class="co">#&gt; Warning: Removed 52 rows containing missing values (geom_point).</span></code></pre></div>
<p><img src="examples_files/figure-html/copy_number_and_vaf-1.png" width="700"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co">#what if we want to focus on putative driver mutations? You can restrict this plot just to coding mutations and label genes of your choice.</span>

<span class="co">#use the built in lymphoma gene list and subset for BL or DLBCL</span>

<span class="va">my_genes</span><span class="op">=</span><span class="va">lymphoma_genes</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">BL</span><span class="op">==</span><span class="cn">TRUE</span> <span class="op">|</span> <span class="va">DLBCL</span> <span class="op">==</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">pull</span><span class="op">(</span><span class="va">Gene</span><span class="op">)</span>

<span class="fu"><a href="../reference/copy_number_vaf_plot.html">copy_number_vaf_plot</a></span><span class="op">(</span>this_sample<span class="op">=</span><span class="va">my_sample</span>,coding_only <span class="op">=</span> <span class="cn">TRUE</span>,genes_to_label <span class="op">=</span> <span class="va">my_genes</span><span class="op">)</span>
<span class="co">#&gt; [1] "copy number colours"</span>
<span class="co">#&gt; Warning: Removed 2 rows containing missing values (geom_point).</span></code></pre></div>
<p><img src="examples_files/figure-html/copy_number_and_vaf-2.png" width="700"></p>
<p>You may also be interested in visualizing all mutations across a given cohort (or every genome). This can only be done in relatively small regions. Here’s a way to visualize the non-coding mutation pattern across a region of interest.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># set up some coordinates to annotate in your plot (optional)</span>
<span class="va">mybed</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>start<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">128806578</span>,<span class="fl">128805652</span>,<span class="fl">128748315</span><span class="op">)</span>, end<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">128806992</span>,<span class="fl">128809822</span>,<span class="fl">128748880</span><span class="op">)</span>, name<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"TSS"</span>,<span class="st">"enhancer"</span>,<span class="st">"MYC-e1"</span><span class="op">)</span><span class="op">)</span>

<span class="co"># get the mutations within a region of interest</span>
<span class="co"># note that we can specify the query chromosome with or without a chr prefix and it will be handled elegantly</span>
<span class="va">my_mutations</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_ssm_by_region.html">get_ssm_by_region</a></span><span class="op">(</span>chromosome<span class="op">=</span><span class="st">"chr8"</span>,qstart<span class="op">=</span><span class="fl">128743606</span>,qend<span class="op">=</span><span class="fl">128820015</span><span class="op">)</span>

<span class="fu"><a href="../reference/ashm_rainbow_plot.html">ashm_rainbow_plot</a></span><span class="op">(</span>mutations_maf<span class="op">=</span><span class="va">my_mutations</span>,metadata<span class="op">=</span><span class="va">my_metadata</span>,bed<span class="op">=</span><span class="va">mybed</span><span class="op">)</span></code></pre></div>
<p><img src="examples_files/figure-html/rainbow_plot_ashm-1.png" width="100%"></p>
<p>A more advanced and fairly automatic plotting function for exploring multiple aSHM regions at once.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Handy function that provides a vector of colours for giving points for different pathology/subgroups reproducible and distinguishable colours</span>
<span class="va">lymphgen_colours</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_gambl_colours.html">get_gambl_colours</a></span><span class="op">(</span>classification<span class="op">=</span><span class="st">"lymphgen"</span><span class="op">)</span>
<span class="co"># This package comes with some custom (curated) data such as the regions recurrently affected by hypermutation in B-NHLs</span>
<span class="fu"><a href="../reference/ashm_multi_rainbow_plot.html">ashm_multi_rainbow_plot</a></span><span class="op">(</span>regions_to_display<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"BCL2-TSS"</span>,<span class="st">"MYC-TSS"</span>,<span class="st">"SGK1-TSS"</span>,<span class="st">"IGL"</span><span class="op">)</span>,custom_colours <span class="op">=</span> <span class="va">lymphgen_colours</span><span class="op">)</span>
<span class="co">#&gt; Joining, by = "sample_id"</span>
<span class="co">#&gt; Joining, by = "start"</span>
<span class="co">#&gt; Joining, by = "sample_id"</span></code></pre></div>
<p><img src="examples_files/figure-html/more_ashm_plotting-1.png" width="700"></p>
<p>Get all CN segments that overlap with a gene of interest and annotate them using the metadata.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">my_segments</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_cn_segments.html">get_cn_segments</a></span><span class="op">(</span>chromosome<span class="op">=</span><span class="st">"4"</span>,qstart<span class="op">=</span><span class="fl">83274467</span>,qend<span class="op">=</span><span class="fl">83295149</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">my_segments</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt;             ID chrom    start       end LOH_flag log.ratio</span>
<span class="co">#&gt; 1  1064-01-1TD     4    11870 190789536        0         0</span>
<span class="co">#&gt; 2 1182-01-03TD     4    69567 190785390        0         1</span>
<span class="co">#&gt; 3 1227-02-02TD     4 62724209 190790246        0         0</span>
<span class="co">#&gt; 4 1242-02-01BD     4    70761 190790246        0         0</span>
<span class="co">#&gt; 5 1296-01-03TD     4    98530 190762427        0         0</span>
<span class="co">#&gt; 6 1318-02-04BD     4    81303 190784332        0         0</span>

<span class="va">deleted_segments</span> <span class="op">=</span> <span class="va">my_segments</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">log.ratio</span><span class="op">&lt;</span><span class="fl">0</span><span class="op">)</span> <span class="co">#use some lower value if you want to be more stringent</span>

<span class="va">annotated_segments</span> <span class="op">=</span> <span class="fu">left_join</span><span class="op">(</span><span class="va">deleted_segments</span>,<span class="va">my_metadata</span>,by<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ID"</span> <span class="op">=</span> <span class="st">"sample_id"</span><span class="op">)</span><span class="op">)</span>

<span class="va">annotated_segments</span> <span class="op">%&gt;%</span> <span class="fu">pull</span><span class="op">(</span><span class="va">cohort</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; .</span>
<span class="co">#&gt;           BL_Adult       BL_Pediatric   CLL_GenomeCanada        DLBCL_ctDNA </span>
<span class="co">#&gt;                  2                  3                  4                  2 </span>
<span class="co">#&gt;     DLBCL_Gascoyne DLBCL_GenomeCanada        DLBCL_HTMCP         DLBCL_ICGC </span>
<span class="co">#&gt;                  5                  9                  3                  5 </span>
<span class="co">#&gt;  DLBCL_LSARP_Trios        DLBCL_Marra    FL_GenomeCanada            FL_ICGC </span>
<span class="co">#&gt;                 53                 10                  1                  4 </span>
<span class="co">#&gt;          FL_Kridel    MALY_Other_ICGC          MCL_Morin        MM_mmsanger </span>
<span class="co">#&gt;                  6                  3                  4                  4</span></code></pre></div>
<p>Retrieve all somatic mutations in a given gene and visualize them using MAFtools.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">all_ssms</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_ssm_by_gene.html">get_ssm_by_gene</a></span><span class="op">(</span>gene_symbol<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CCND3"</span><span class="op">)</span>,coding_only <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="va">all_ssms</span> <span class="op">=</span> <span class="va">all_ssms</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># make a MAFtools object and plot a lollipop plot</span>

<span class="va">maf_obj</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/maftools/man/read.maf.html">read.maf</a></span><span class="op">(</span><span class="va">all_ssms</span><span class="op">)</span>
<span class="co">#&gt; -Validating</span>
<span class="co">#&gt; -Silent variants: 2 </span>
<span class="co">#&gt; -Summarizing</span>
<span class="co">#&gt; -Processing clinical data</span>
<span class="co">#&gt; --Missing clinical data</span>
<span class="co">#&gt; -Finished in 0.098s elapsed (1.163s cpu)</span>
<span class="fu"><a href="https://rdrr.io/pkg/maftools/man/lollipopPlot.html">lollipopPlot</a></span><span class="op">(</span><span class="va">maf_obj</span>,gene<span class="op">=</span><span class="st">"CCND3"</span><span class="op">)</span>
<span class="co">#&gt; Assuming protein change information are stored under column HGVSp_Short. Use argument AACol to override if necessary.</span>
<span class="co">#&gt; 4 transcripts available. Use arguments refSeqID or proteinID to manually specify tx name.</span>
<span class="co">#&gt;     HGNC    refseq.ID   protein.ID aa.length</span>
<span class="co">#&gt; 1: CCND3 NM_001136125 NP_001129597       220</span>
<span class="co">#&gt; 2: CCND3    NM_001760    NP_001751       292</span>
<span class="co">#&gt; 3: CCND3 NM_001136017 NP_001129489       211</span>
<span class="co">#&gt; 4: CCND3 NM_001136126 NP_001129598        96</span>
<span class="co">#&gt; Using longer transcript NM_001760 for now.</span></code></pre></div>
<p><img src="examples_files/figure-html/ssms_and_maftools-1.png" width="100%"></p>
<p>We can also extract a MAF-like data frame for a set of samples using any combination of cohorts or pathology to subset the data.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># load the master merged MAF (coding only). It's usually more efficient to do this than to try to add filters to this query. Instead, just filter the data afterward</span>
<span class="va">maf_data</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_coding_ssm.html">get_coding_ssm</a></span><span class="op">(</span><span class="op">)</span> 
<span class="va">maf_metadata</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html">rename</a></span><span class="op">(</span><span class="st">"Tumor_Sample_Barcode"</span><span class="op">=</span><span class="st">"sample_id"</span><span class="op">)</span>
<span class="va">maf</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/maftools/man/read.maf.html">read.maf</a></span><span class="op">(</span><span class="va">maf_data</span>,clinicalData <span class="op">=</span> <span class="va">maf_metadata</span><span class="op">)</span>
<span class="co">#&gt; -Validating</span>
<span class="co">#&gt; --Removed 230 duplicated variants</span>
<span class="co">#&gt; -Silent variants: 48987 </span>
<span class="co">#&gt; -Summarizing</span>
<span class="co">#&gt; --Possible FLAGS among top ten genes:</span>
<span class="co">#&gt;   TTN</span>
<span class="co">#&gt; -Processing clinical data</span>
<span class="co">#&gt; -Finished in 20.5s elapsed (00:01:00 cpu)</span>
<span class="co"># subset metadata, for example, to only BLGSP samples</span>
<span class="va">blgsp_metadata</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span>case_set<span class="op">=</span><span class="st">"BLGSP-study"</span><span class="op">)</span>
<span class="co"># subset maf</span>
<span class="va">blgsp_maf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/maftools/man/subsetMaf.html">subsetMaf</a></span><span class="op">(</span><span class="va">maf</span>, tsb<span class="op">=</span><span class="va">blgsp_sample_ids</span><span class="op">)</span>
<span class="co"># perform analysis for subset maf</span>
<span class="fu"><a href="https://rdrr.io/pkg/maftools/man/oncoplot.html">oncoplot</a></span><span class="op">(</span><span class="va">blgsp_maf</span>,clinicalFeatures <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"sex"</span>,<span class="st">"cohort"</span><span class="op">)</span>,sortByAnnotation <span class="op">=</span> <span class="cn">TRUE</span>,genesToIgnore <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"TTN"</span>,<span class="st">"LILRB1"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="examples_files/figure-html/get_big_maf-1.png" width="768"></p>
<p>Commonly, we will want to generate visualizations that only show genes deemed relevant for a given malignancy (and above some frequency threshold). There is a lymphoma_genes data frame bundled with this package that allows us a quick way to get a list of genes that are recurrently mutated in that disease. This list is quite long for most because it includes anything mutated at a frequency of at least (roughly) 0.5% in GAMBL.</p>
<p>#```{r tabulate_genes} #BL_genes_all = lymphoma_genes %&gt;% filter(BL==TRUE) %&gt;% pull(Gene) #print(BL_genes_all)</p>
<p>#determine what frequency each gene is mutated in across the extracted data set</p>
<p>#n = maf_data %&gt;% select(Tumor_Sample_Barcode) %&gt;% unique() %&gt;% count()</p>
<p>#gene_sample_mut_freq = maf_data %&gt;% dplyr::filter(Hugo_Symbol %in% BL_genes_all) %&gt;% # dplyr::select(Hugo_Symbol, Tumor_Sample_Barcode) %&gt;% unique() %&gt;% # dplyr::group_by(Hugo_Symbol) %&gt;% dplyr::tally() %&gt;% # dplyr::mutate(frequency= 100 * n / 222)</p>
<p>#which genes are mutated in at least 10% of patients? #gene_sample_mut_freq %&gt;% filter(frequency &gt;=10)</p>
<p>#top_BL_genes = gene_sample_mut_freq %&gt;% dplyr::filter(frequency &gt;=10) %&gt;% dplyr::pull(Hugo_Symbol)</p>
<p>#maftools::oncoplot(maf,clinicalFeatures = c(“sex”,“cohort”),sortByAnnotation = TRUE,genes=top_BL_genes)</p>
<p>#```</p>
<p>#```{r igv} # WARNING: this code chunk will fail unless you have IGV running and listening on port 60506 # The simplest scenario is to run this command on a terminal (if using a Mac), assuming you are using R on gphost10 and you have a ssh config that routes gp10 to that host. # Stretch goal/TODO: get this working with a headless igv # ssh -X gp10 # then launch IGV (e.e. from a conda installation): # conda activate igv; igv &amp; #hg19/grch37 example #this_sv = annotated_sv %&gt;% filter(partner==“HIST1H2BK”) %&gt;% head(1) #bam_details = get_bams(sample=this_sv<span class="math inline">\(tumour_sample_id) #snapshot_home = "/home/rmorin/IGV_snapshots/" #snapshot = make_igv_snapshot(chrom=this_sv\)</span>chrom2, start=this_sv<span class="math inline">\(start2, # end=this_sv\)</span>end2, # sample_id=this_sv<span class="math inline">\(tumour_sample_id, # genome_build=bam_details\)</span>igv_build, # bams = #c(bam_details<span class="math inline">\(tumour_bams,bam_details\)</span>normal_genome_bams), # out_path=snapshot_home)</p>
<p>#library(imager)</p>
<p>#snapshot_image = imager::load.image(snapshot) #plot(snapshot_image,axes = FALSE) #```</p>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Ryan Morin.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
