<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>SSM Tutorial â€¢ GAMBLR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="SSM Tutorial">
<meta property="og:description" content="GAMBLR">
<meta property="og:image" content="https://github.com/morinlab/GAMBLR/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-12345678-0"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-12345678-0');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9500</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-light fa-memo-circle-info"></span>
     
    How To
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="https://morinlab.github.io/GAMBLR/#installation" class="external-link">Install</a>
    </li>
    <li>
      <a href="https://morinlab.github.io/GAMBLR/#running-gamblr-on-your-own-computer" class="external-link">Running GAMBLR On Your Own Computer</a>
    </li>
    <li>
      <a href="https://morinlab.github.io/GAMBLR/#contributing" class="external-link">Contribute</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-light fa-person-chalkboard"></span>
     
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/ssm_tutorial.html">SSM</a>
    </li>
    <li>
      <a href="../articles/cn_sv_tutorial.html">CN &amp; SV</a>
    </li>
    <li>
      <a href="../articles/fancy_plots.html">Fancy Plots</a>
    </li>
    <li>
      <a href="../articles/utilities_tutorial.html">Utilities</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="../reference/index.html">
    <span class="fa fa-file-code-o"></span>
     
    Functions &amp; Bundled Data
  </a>
</li>
<li>
  <a href="https://github.com/morinlab/GAMBLR">
    <span class="fa fa-github fa-lg"></span>
     
    github
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="ssm_tutorial_files/htmlwidgets-1.6.2/htmlwidgets.js"></script><script src="ssm_tutorial_files/d3-4.13.0/./d3/d3.v4.min.js"></script><link href="ssm_tutorial_files/g3viz-0.4.0/./g3viz/g3-lollipop.min.css" rel="stylesheet">
<script src="ssm_tutorial_files/g3viz-0.4.0/./g3viz/g3-lollipop.min.js"></script><script src="ssm_tutorial_files/g3Lollipop-binding-1.1.3/g3Lollipop.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>SSM Tutorial</h1>
            
      
      
      <div class="hidden name"><code>ssm_tutorial.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="simple-somatic-mutations">Simple Somatic Mutations<a class="anchor" aria-label="anchor" href="#simple-somatic-mutations"></a>
</h2>
<p>This vignette demonstrates how to work with Simple Somatic Mutations (SSM) results within GAMBLR, all the way from obtaining results (in various ways), linking with metadata, performing data analysis and annotations to producing publication ready figures.</p>
<div class="section level3">
<h3 id="metadata">Metadata<a class="anchor" aria-label="anchor" href="#metadata"></a>
</h3>
<p>First, lets start with retrieving metadata for all GAMBL samples with <code>seqtype</code> <em>genome</em>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#get gambl metadata for all samples</span></span>
<span><span class="va">this_metadata</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span>seq_type_filter <span class="op">=</span> <span class="st">"genome"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#subset metadata on a set of samples (samples classified as DLBCL for pathology)</span></span>
<span><span class="va">dlbcl_metadata</span> <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">this_metadata</span>, </span>
<span>                               <span class="va">pathology</span> <span class="op">==</span> <span class="st">"DLBCL"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="get-ssm-data">Get SSM data<a class="anchor" aria-label="anchor" href="#get-ssm-data"></a>
</h3>
<p>Based on the information available to the user, application, or downstream analysis, there are multiple ways to retrieve SSM data. For example, if the user knows the sample ID and is only interested in looking at SSM results for that particular sample, one can use <code>get_ssm_by_sample</code>. If multiple samples are to be queried, <code>get_ssm_by_samples</code> is recommended. The user can also use patient IDs for retrieving this data, in this case <code>get_ssm_by_patients</code> is available. In addition, the user can also restrict SSM calls to specific genomic regions with <code>get_ssm_by_regions</code> or <code>get_ssm_by_region</code>. Another possibility for returning coding SSM is to call <code>get_coding_ssm</code>, this function returns all coding SSMs from the GAMBL database in MAF-like format. If the user have an already subset metadata, the <code>these_samples_metadata</code> parameter can be used with most ssm functions to restrict the variants returned to the sample IDs in this data frame, handy! In this section we will be demonstrating how these functions can be used.</p>
<div class="section level4">
<h4 id="by-sample">By Sample<a class="anchor" aria-label="anchor" href="#by-sample"></a>
</h4>
<p>Return SSMs for one specific sample with <code>get_ssm_by_sample</code>. In the example bellow, the augmented MAF is loaded (grch37, slims-3 and clustered) and variants with a minimum read support of 3 are kept. Basic column is set to <em>TRUE</em>, this means the function will only return the standard 45 columns out of all 116 columns present in the original MAF. In addition, the parameter <code>these_genes</code> is available for easy subset of variants to specific genomic loci.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#set sample ID</span></span>
<span><span class="va">this_sample</span> <span class="op">=</span> <span class="st">"00-15201_tumorA"</span></span>
<span></span>
<span><span class="co">#get_ssm_by_sample</span></span>
<span><span class="va">this_sample_ssm</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_ssm_by_sample.html">get_ssm_by_sample</a></span><span class="op">(</span>this_sample_id <span class="op">=</span> <span class="va">this_sample</span>, </span>
<span>                                    this_seq_type <span class="op">=</span> <span class="st">"genome"</span>,</span>
<span>                                    projection <span class="op">=</span> <span class="st">"grch37"</span><span class="op">)</span></span></code></pre></div>
<p>To return SSMs for multiple samples one can use <code>get_ssm_by_samples</code>. Note, this function wraps <code>get_ssm_by_sample</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#set sample IDs</span></span>
<span><span class="va">these_samples</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"HTMCP-01-06-00485-01A-01D"</span>, </span>
<span>                  <span class="st">"14-35472_tumorA"</span>, </span>
<span>                  <span class="st">"14-35472_tumorB"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#get_ssm_by_samples</span></span>
<span><span class="va">these_samples_ssm</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_ssm_by_samples.html">get_ssm_by_samples</a></span><span class="op">(</span>these_sample_ids <span class="op">=</span> <span class="va">these_samples</span>,</span>
<span>                                       projection <span class="op">=</span> <span class="st">"grch37"</span>,</span>
<span>                                       seq_type <span class="op">=</span> <span class="st">"genome"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="by-patient">By Patient<a class="anchor" aria-label="anchor" href="#by-patient"></a>
</h4>
<p>Return SSMs for a set of patients with <code>get_ssm_by_patients</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#set patient IDs</span></span>
<span><span class="va">these_patients</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"00-14595"</span>, </span>
<span>                   <span class="st">"00-15201"</span>, </span>
<span>                   <span class="st">"01-12047"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#get_ssm_by_patients</span></span>
<span><span class="va">these_patients_ssm</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_ssm_by_patients.html">get_ssm_by_patients</a></span><span class="op">(</span>these_patient_ids <span class="op">=</span> <span class="va">these_patients</span>,</span>
<span>                                         seq_type <span class="op">=</span> <span class="st">"genome"</span>,</span>
<span>                                         projection <span class="op">=</span> <span class="st">"grch37"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="by-region">By Region<a class="anchor" aria-label="anchor" href="#by-region"></a>
</h4>
<p>Return SSMs based on a specific region with <code>get_ssm_by_region</code>. In this example we are using <code>gene_to_region</code> to get genomic coordinates for a specific gene that will be called by <code>get_ssm_by_region</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#get region</span></span>
<span><span class="va">myc_region</span> <span class="op">=</span> <span class="fu"><a href="../reference/gene_to_region.html">gene_to_region</a></span><span class="op">(</span>gene_symbol <span class="op">=</span> <span class="st">"MYC"</span>, </span>
<span>                            return_as <span class="op">=</span> <span class="st">"region"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#get all SSMs that are residing in the defined MYC region</span></span>
<span><span class="va">myc_ssm</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_ssm_by_region.html">get_ssm_by_region</a></span><span class="op">(</span>region <span class="op">=</span> <span class="va">myc_region</span>,</span>
<span>                            projection <span class="op">=</span> <span class="st">"grch37"</span>,</span>
<span>                            seq_type <span class="op">=</span> <span class="st">"genome"</span><span class="op">)</span></span></code></pre></div>
<p>For multiple regions, refer to <code>get_ssm_by_regions</code>. In this example we are returning SSMs for all ASHM regions (hyper mutated) across all samples.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#convert regions to bed format</span></span>
<span><span class="va">my_bed</span> <span class="op">=</span> <span class="fu"><a href="../reference/gene_to_region.html">gene_to_region</a></span><span class="op">(</span>gene_symbol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MYC"</span>, <span class="st">"CD58"</span><span class="op">)</span>, return_as <span class="op">=</span> <span class="st">"bed"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#get ssm for all ashm regions</span></span>
<span><span class="va">bed_ssm</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_ssm_by_regions.html">get_ssm_by_regions</a></span><span class="op">(</span>regions_bed <span class="op">=</span> <span class="va">my_bed</span>,</span>
<span>                              projection <span class="op">=</span> <span class="st">"grch37"</span>,</span>
<span>                              seq_type <span class="op">=</span> <span class="st">"genome"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "1:117057157-117113661" "8:128747680-128753674"</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="coding-ssm">Coding SSM<a class="anchor" aria-label="anchor" href="#coding-ssm"></a>
</h4>
<p>Lastly, another way to retrieve SSMs is to call <code>get_coding_ssm</code>. This function returns coding SSMs for any given sample. This function is a convenient option for anyone interested in plotting coding mutations. Convenient filtering parameters are included in this function for easy and straight-forward subset to specific pathologist, cohorts, and sample IDs. If these parameters are not called, coding SSMs will be returned for all samples. In the example bellow, we are using a metadata subset (DLBCL cases) for retrieving coding ssm.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#create metadata subset</span></span>
<span><span class="va">my_sample_metadata</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span>seq_type_filter <span class="op">=</span> <span class="st">"genome"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pathology</span> <span class="op">==</span> <span class="st">"DLBCL"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#get SSMs for all samples in metadata subset</span></span>
<span><span class="va">all_coding_ssm</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_coding_ssm.html">get_coding_ssm</a></span><span class="op">(</span>these_samples_metadata <span class="op">=</span> <span class="va">my_sample_metadata</span>,</span>
<span>                                projection <span class="op">=</span> <span class="st">"grch37"</span>, </span>
<span>                                seq_type <span class="op">=</span> <span class="st">"genome"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="annotating-ssms">Annotating SSMs<a class="anchor" aria-label="anchor" href="#annotating-ssms"></a>
</h3>
<p>In this section we are exploring functions that can be used to annotate SSMs that were retrieved in the previous section.</p>
<div class="section level4">
<h4 id="blacklist">Blacklist<a class="anchor" aria-label="anchor" href="#blacklist"></a>
</h4>
<p>Annotate and auto-drop a MAF data frame with existing blacklists to remove variants that would be dropped during the merge process with <code>annotate_blacklist_ssm</code>. This function returns a MAF format data frame with two new columns, indicating the number of occurrences of each variant in the two blacklists. Note that there are a collection of parameters to this function to improve flexibility for many applications, such as <code>return_blacklist</code> (returns the used blacklist to the vector given the function, or printed to terminal if blank). For returning variants that would be dropped, one can specify <code>invert = TRUE</code>, please use with caution, this is most likely the opposite of what you want from this function. Lastly, the minimum count from one of the blacklists to drop a variant is specified with <code>drop_threshold = 4</code>. This function also, conveniently lets you know how many variants that were dropped in the annotation process.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#annotate ssm calls (all coding SSMs)</span></span>
<span><span class="va">deblacklisted_ssm</span> <span class="op">=</span> <span class="fu"><a href="../reference/annotate_ssm_blacklist.html">annotate_ssm_blacklist</a></span><span class="op">(</span>mutations_df <span class="op">=</span> <span class="va">all_coding_ssm</span>, </span>
<span>                                           genome_build <span class="op">=</span> <span class="st">"grch37"</span>,   </span>
<span>                                           seq_type <span class="op">=</span> <span class="st">"genome"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="drivers">Drivers<a class="anchor" aria-label="anchor" href="#drivers"></a>
</h4>
<p>Another useful way to annotate SSMs is with <code>annotate_driver_ssm</code>. This function indicates what rows (mutations) are putative driver mutations. For this example, we are annotating the single-sample MAF file that got returned with <code>get_ssm_by_sample</code>. This function can also take additional optional parameters to restrict the annotations to specific lymphoma types and known driver genes.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#annotate sample-level maf with putative driver mutations</span></span>
<span><span class="va">driver_ssm</span> <span class="op">=</span> <span class="fu"><a href="../reference/annotate_driver_ssm.html">annotate_driver_ssm</a></span><span class="op">(</span>maf_df <span class="op">=</span> <span class="va">this_sample_ssm</span>, </span>
<span>                                 lymphoma_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DLBCL"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 14 45</span></span>
<span><span class="co">## [1] 14 45</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="hotspots">Hotspots<a class="anchor" aria-label="anchor" href="#hotspots"></a>
</h4>
<p>Another way to annotate an incoming MAF with SSMs is to call <code>annotate_hotspots</code>. This function annotates MAF-like data frame with a <em>hot_spot</em> column indicating recurrent mutations. The function takes the following parameters; <code>mutation_maf</code> - which is expecting a data frame in MAF-like format, <code>recurrence_min</code> - minimum number of recurrences for mutation to be included (default is 5), <code>analysis_base</code> - base name for hot spot output directory and <code>p_thresh</code> - P value threshold, default is 0.05. In this example, we are querying the MAF that was returned from the one-sample (<code>get_ssm_by_sample</code>).</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#annotate mutational hotspots</span></span>
<span><span class="va">this_sample_ssm_hot</span> <span class="op">=</span> <span class="fu"><a href="../reference/annotate_hotspots.html">annotate_hotspots</a></span><span class="op">(</span>mutation_maf <span class="op">=</span> <span class="va">this_sample_ssm</span>, </span>
<span>                                        recurrence_min <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                                        p_thresh <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                                        analysis_base <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"FL--DLBCL"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="review-hotspots">Review Hotspots<a class="anchor" aria-label="anchor" href="#review-hotspots"></a>
</h4>
<p>Returned hotspots can also be further reviewed with <code>review_hotspots</code>. This function expects an MAF that has already been annotated with <code>annotate_hotspots</code>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#review hotspots</span></span>
<span><span class="va">this_sample_ssm_hot_rev</span> <span class="op">=</span> <span class="fu"><a href="../reference/review_hotspots.html">review_hotspots</a></span><span class="op">(</span>annotated_maf <span class="op">=</span> <span class="va">this_sample_ssm_hot</span>, </span>
<span>                                          genes_of_interest <span class="op">=</span> <span class="st">"MYD88"</span>, </span>
<span>                                          genome_build <span class="op">=</span> <span class="st">"grch37"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="custom-track-ucsc">Custom Track (UCSC)<a class="anchor" aria-label="anchor" href="#custom-track-ucsc"></a>
</h4>
<p>Lastly, we can also transform the annotated MAF into a custom track file for visualization on th UCSC genome browser. This can be achieved with the function <code>maf_to_custum_track</code>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#transform the annotated maf into a custom track for UCSC genome browser interrogation</span></span>
<span><span class="va">ucsc_customtrack</span> <span class="op">=</span> <span class="fu"><a href="../reference/maf_to_custom_track.html">maf_to_custom_track</a></span><span class="op">(</span>maf_data <span class="op">=</span> <span class="va">this_sample_ssm_hot_rev</span>,</span>
<span>                                       as_bigbed <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                       as_biglolly <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                       output_file <span class="op">=</span> <span class="st">"this_sample_custom_track.bed"</span>,</span>
<span>                                       track_name <span class="op">=</span> <span class="st">"GAMBL Mutations"</span>,</span>
<span>                                       track_description <span class="op">=</span> <span class="st">"Mutations from GAMBL"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="ssm-visualization">SSM Visualization<a class="anchor" aria-label="anchor" href="#ssm-visualization"></a>
</h3>
<div class="section level4">
<h4 id="variant-counts">Variant Counts<a class="anchor" aria-label="anchor" href="#variant-counts"></a>
</h4>
<p>Interested in how many variants are present in a given sample? A quick way to overview such information would be to call <code>fancy_svbar</code> on a given sample ID. This function returns a bar plot visualizing SVs or SSMs. A variety of sub-setting options are also available. Such as chromosome selection and variant types to be included in plot. In the example bellow, we are looking at genome-wide SSMs for one selected sample.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#get ssm data for one sample</span></span>
<span><span class="va">this_sample_ssm</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_ssm_by_sample.html">get_ssm_by_sample</a></span><span class="op">(</span>this_sample_id <span class="op">=</span> <span class="st">"HTMCP-01-06-00422-01A-01D"</span>, </span>
<span>                                    this_seq_type <span class="op">=</span> <span class="st">"genome"</span>,</span>
<span>                                    projection <span class="op">=</span> <span class="st">"grch37"</span>,</span>
<span>                                    augmented <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/fancy_v_count.html">fancy_v_count</a></span><span class="op">(</span>maf_data <span class="op">=</span> <span class="va">this_sample_ssm</span>,</span>
<span>              ssm <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>              log10_y <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>              from_flatfile <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>              use_augmented_maf <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>              variant_select <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DEL"</span>, <span class="st">"DNP"</span>, <span class="st">"INS"</span>, <span class="st">"SNP"</span><span class="op">)</span>, </span>
<span>              snp_colours <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SNP"</span> <span class="op">=</span> <span class="st">"#2B9971"</span>, <span class="st">"DNP"</span> <span class="op">=</span> <span class="st">"#993F2B"</span><span class="op">)</span>,</span>
<span>              plot_subtitle <span class="op">=</span> <span class="st">"grch37"</span>,</span>
<span>              plot_title <span class="op">=</span> <span class="st">"Fig 1."</span>,</span>
<span>              chr_select <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="ssm_tutorial_files/figure-html/fancy_v_count-1.png" width="1093.75"></p>
</div>
<div class="section level4">
<h4 id="variant-counts-per-chromosome-">Variant Counts Per Chromosome.<a class="anchor" aria-label="anchor" href="#variant-counts-per-chromosome-"></a>
</h4>
<p>Sometimes it can be useful to overview variant distributions across the complete genomic landscape. To do so, one could call <code>fancy_v_chrcount</code> with sample ID. Optional arguments for sub-setting your data are available (see parameter descriptions for more information). In the following example weâ€™re plotting SSM counts for all autosomes. One can also call <code>add_qc_metric</code> to add a second y-axis annotating mean corrected coverage for the selected sample.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fancy_v_chrcount.html">fancy_v_chrcount</a></span><span class="op">(</span>maf_data <span class="op">=</span> <span class="va">this_sample_ssm</span>,  </span>
<span>                 plot_subtitle <span class="op">=</span>  <span class="st">"grch37"</span>,</span>
<span>                 plot_title <span class="op">=</span> <span class="st">"Fig 2."</span>,</span>
<span>                 y_interval <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                 projection <span class="op">=</span> <span class="st">"grch37"</span>,</span>
<span>                 min_vaf <span class="op">=</span> <span class="fl">4</span>,</span>
<span>                 chr_select <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                 ssm <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                 coding_only <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="ssm_tutorial_files/figure-html/fancy_v_chrcount-1.png" width="1093.75"></p>
</div>
<div class="section level4">
<h4 id="variant-size-distributions">Variant Size Distributions<a class="anchor" aria-label="anchor" href="#variant-size-distributions"></a>
</h4>
<p>A useful way to visualize variant size distributions is to call <code>fancy_v_sizedis</code> to generate a violin plot showing variant size distributions (sample-level) across selected contigs (per default chr1-22). Optional parameters to sub-set data to include only coding regions are also available for this function, as well as to plot SVs by setting <code>ssm = FALSE</code>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fancy_v_sizedis.html">fancy_v_sizedis</a></span><span class="op">(</span>maf_data <span class="op">=</span> <span class="va">this_sample_ssm</span>,</span>
<span>                ssm <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                plot_trim <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                log_10 <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                scale_value <span class="op">=</span> <span class="st">"width"</span>,</span>
<span>                plot_title <span class="op">=</span> <span class="st">"Fig 3."</span>,</span>
<span>                plot_subtitle <span class="op">=</span>  <span class="st">"grch37"</span>,</span>
<span>                chr_select <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="ssm_tutorial_files/figure-html/fancy_v_sizedis-1.png" width="1093.75"></p>
</div>
<div class="section level4">
<h4 id="single-nucleotide-variants-chromosome-distributions">Single Nucleotide Variants Chromosome Distributions<a class="anchor" aria-label="anchor" href="#single-nucleotide-variants-chromosome-distributions"></a>
</h4>
<p>Curious to see how SNVs are distributed across specific chromosomes? For this purpose <code>fancy_snv_chrdistplot</code> was developed. This function takes a sample ID and retrieves SNV information (using <code>assign_cn_to_ssm</code>) and plots variants in an ideogram dependent manner. Similarly to other <code>fanxy_x_plot</code> functions, a selection of parameters are available for data sub-setting and plot customization. For example, by calling the include_dnp, DNP counts will be added to each bar.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fancy_snv_chrdistplot.html">fancy_snv_chrdistplot</a></span><span class="op">(</span>maf_data <span class="op">=</span> <span class="va">this_sample_ssm</span>,</span>
<span>                      plot_subtitle <span class="op">=</span> <span class="st">"SNV Distribution Example"</span>,</span>
<span>                      plot_title <span class="op">=</span> <span class="st">"Fig 4."</span>,</span>
<span>                      chr_select <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                      include_dnp <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="ssm_tutorial_files/figure-html/fancy_snv_chrdistplot-1.png" width="1093.75"></p>
</div>
<div class="section level4">
<h4 id="forest-plot">Forest Plot<a class="anchor" aria-label="anchor" href="#forest-plot"></a>
</h4>
<p>The <code>prettyForestPlot</code> function is used to compare mutation frequencies for a set of genes between two groups. In this example, pathology is set to <strong>FL</strong> and <strong>DLBCL</strong> (the two groups to be compared). Next, to get coding mutations for the selected samples, <code>get_coding_ssm</code> is called with <code>limit_samples</code> parameter, giving it the sample id column from the recently subset metadata. This function returns two types of plot (box plot and forest plot), the user can either view them separately or arranged on the same grob (showed in this example).</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata_fl_dlbcl</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pairing_status</span> <span class="op">==</span> <span class="st">"matched"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">consensus_pathology</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"FL"</span>, <span class="st">"DLBCL"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">maf_fl_dlbcl_coding</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_coding_ssm.html">get_coding_ssm</a></span><span class="op">(</span>limit_samples <span class="op">=</span> <span class="va">metadata_fl_dlbcl</span><span class="op">$</span><span class="va">sample_id</span>,</span>
<span>                                     basic_columns <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                     seq_type <span class="op">=</span> <span class="st">"genome"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">forest_plot</span> <span class="op">=</span> <span class="fu"><a href="../reference/prettyForestPlot.html">prettyForestPlot</a></span><span class="op">(</span>maf <span class="op">=</span> <span class="va">maf_fl_dlbcl_coding</span>,</span>
<span>                               metadata <span class="op">=</span> <span class="va">metadata_fl_dlbcl</span>,</span>
<span>                               genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ATP6V1B2"</span>, <span class="st">"EZH2"</span>, <span class="st">"TNFRSF14"</span>, <span class="st">"RRAGC"</span><span class="op">)</span>,</span>
<span>                               comparison_column <span class="op">=</span> <span class="st">"consensus_pathology"</span>,</span>
<span>                               comparison_values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DLBCL"</span>, <span class="st">"FL"</span><span class="op">)</span>,</span>
<span>                               separate_hotspots <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                               comparison_name <span class="op">=</span> <span class="st">"FL vs DLBCL"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">forest_plot</span><span class="op">$</span><span class="va">arranged</span></span></code></pre></div>
<p><img src="ssm_tutorial_files/figure-html/pretyyForestPlot-1.png" width="1093.75"></p>
</div>
<div class="section level4">
<h4 id="rainbow-plots">Rainbow Plots<a class="anchor" aria-label="anchor" href="#rainbow-plots"></a>
</h4>
<p>The <code>ashm_rainbow_plot</code> allows the user to plot all mutations in a region, ordered and coloured by defined metadata columns. This function also has convenient parameters for highlighting specific regions within the larger defined region. For example, here we are looking at mutations withing MYC, the plot is showing 10K bases upstream and downstream of MYC, with MYC highlighted. Samples are plotted along the y-axis and genomic positions are in respect to the x-axis. On this example, samples are colour-coded based on their pathology status.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#set region for the plot to highlight.</span></span>
<span><span class="va">mybed</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>start <span class="op">=</span> <span class="fl">128747680</span>,</span>
<span>                   end <span class="op">=</span> <span class="fl">128753674</span>,</span>
<span>                   name <span class="op">=</span> <span class="st">"MYC"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#full region to be plotted.</span></span>
<span><span class="va">region</span> <span class="op">=</span> <span class="st">"chr8:128737680-128763674"</span></span>
<span></span>
<span><span class="co">#get gambl metadata for all FL and DLBCL samples.</span></span>
<span><span class="va">fl_dlbcl_metadata</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pathology</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"FL"</span>, <span class="st">"DLBCL"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#get SSm data for the specified region.</span></span>
<span><span class="va">my_mutations</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_ssm_by_region.html">get_ssm_by_region</a></span><span class="op">(</span>region <span class="op">=</span> <span class="va">region</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#construct plot</span></span>
<span><span class="fu"><a href="../reference/ashm_rainbow_plot.html">ashm_rainbow_plot</a></span><span class="op">(</span>mutations_maf <span class="op">=</span> <span class="va">my_mutations</span>,</span>
<span>                  drop_unmutated <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                  metadata <span class="op">=</span> <span class="va">fl_dlbcl_metadata</span>,</span>
<span>                  hide_ids <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                  bed <span class="op">=</span> <span class="va">mybed</span>,</span>
<span>                  region <span class="op">=</span> <span class="va">region</span>,</span>
<span>                  classification_column <span class="op">=</span> <span class="st">"pathology"</span>,</span>
<span>                  custom_colours <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"FL"</span> <span class="op">=</span> <span class="st">"#EA8368"</span>, <span class="st">"DLBCL"</span> <span class="op">=</span> <span class="st">"#479450"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="ssm_tutorial_files/figure-html/ashm_rainbow_plot-1.png" width="1500"></p>
<p>If you are in fact interested in plotting multi-panel overview of hypermutations in regions of interest across many samples. You should try out <code>ashm_multi_rainbow_plot</code>.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#construct a multi-panel ashm rainbow plot</span></span>
<span><span class="fu"><a href="../reference/ashm_multi_rainbow_plot.html">ashm_multi_rainbow_plot</a></span><span class="op">(</span>custom_colours <span class="op">=</span> <span class="fu"><a href="../reference/get_gambl_colours.html">get_gambl_colours</a></span><span class="op">(</span><span class="st">"pathology"</span><span class="op">)</span>,</span>
<span>                        metadata <span class="op">=</span> <span class="va">fl_dlbcl_metadata</span>,</span>
<span>                        classification_column <span class="op">=</span> <span class="st">"pathology"</span>,</span>
<span>                        seq_type <span class="op">=</span> <span class="st">"genome"</span>,</span>
<span>                        verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                        regions_to_display <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"BCL2-TSS"</span>, <span class="st">"MYC-TSS"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="ssm_tutorial_files/figure-html/ashm_multi-rainbow_plot-1.png" width="1500"></p>
</div>
<div class="section level4">
<h4 id="rainfall-plot">Rainfall Plot<a class="anchor" aria-label="anchor" href="#rainfall-plot"></a>
</h4>
<p>This plotting function is available for plotting one or multiple samples and takes either an already loaded MAF data frame or a MAF path. If none are supplied the function will get ssm data using one of the functions available in GAMBLR. There are also other convenient parameters for further customizing the returned plot. For example, we can label ASHM genes <code>(label_ashm_genes = TRUE)</code>, subset to specific chromosomes, zoom in to regions of interest. If plot is restricted to specific chromosomes or zoomed in to a region, we can also add SV labels, with the <code>labe_sv = TRUE</code>. In this example, we are looking at one sample (HTMCP-01-06-0042-01A-01D), all chromosomes with ASHM genes labeled.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#build plot</span></span>
<span><span class="fu"><a href="../reference/prettyRainfallPlot.html">prettyRainfallPlot</a></span><span class="op">(</span>this_sample_id <span class="op">=</span> <span class="st">"HTMCP-01-06-00422-01A-01D"</span>,</span>
<span>                   seq_type <span class="op">=</span> <span class="st">"genome"</span>,</span>
<span>                   projection <span class="op">=</span> <span class="st">"grch37"</span>, </span>
<span>                   label_ashm_genes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="ssm_tutorial_files/figure-html/prettyRainfallPlot-1.png" width="2700"></p>
</div>
<div class="section level4">
<h4 id="lollipop-plot">Lollipop Plot<a class="anchor" aria-label="anchor" href="#lollipop-plot"></a>
</h4>
<p>The function <code>pretty_lollipop_plot</code> is perfect for creating an appealing, flexible and interactive HTML-style lollipop plot from mutational data using a modified version of g3Lollipop function from g3viz package. This function only takes for parameters; <code>maf_df</code> which is the data frame containing the mutations to be plotted. <code>gene</code> this parameter controls what gene to plot mutations for. Lastly, <code>plot_title</code> lets you specify the name of the plot and with <code>plot_theme</code> you can specify the plot theme. Please note that in order for this function to run, you must first installed the forked version of <a href="https://github.com/morinlab/g3viz/" class="external-link">g3viz</a>.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#get metadata (Fl and DLBCL) with pairing status matched </span></span>
<span><span class="va">metadata</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pairing_status</span> <span class="op">==</span> <span class="st">"matched"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">consensus_pathology</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"FL"</span>, <span class="st">"DLBCL"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#get maf data for returned samples</span></span>
<span><span class="va">maf</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_coding_ssm.html">get_coding_ssm</a></span><span class="op">(</span>limit_samples <span class="op">=</span> <span class="va">metadata</span><span class="op">$</span><span class="va">sample_id</span>, </span>
<span>                     basic_columns <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                     seq_type <span class="op">=</span> <span class="st">"genome"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#construct pretty_lollipop_plot.</span></span>
<span><span class="fu"><a href="../reference/pretty_lollipop_plot.html">pretty_lollipop_plot</a></span><span class="op">(</span>maf_df <span class="op">=</span> <span class="va">maf</span>, </span>
<span>                     gene <span class="op">=</span> <span class="st">"MYC"</span>,</span>
<span>                     plot_title <span class="op">=</span> <span class="st">"Mutation data for MYC"</span>, </span>
<span>                     plot_theme <span class="op">=</span> <span class="st">"nature2"</span><span class="op">)</span></span></code></pre></div>
<div class="g3Lollipop html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-0f4d8b21e816e5f64287" style="width:1093.75px;height:675.988875154512px;"></div>
<script type="application/json" data-for="htmlwidget-0f4d8b21e816e5f64287">{"x":{"domainData":{"symbol":"MYC","uniprot":"P01106","length":439,"pfam":[{"hmm.acc":"PF01056","hmm.name":"Myc_N","start":1,"end":345,"type":"Family"},{"hmm.acc":"PF00010","hmm.name":"HLH","start":355,"end":407,"type":"Domain"},{"hmm.acc":"PF02344","hmm.name":"Myc-LZ","start":408,"end":438,"type":"Coiled-coil"}]},"domainDataFormat":{"length":"length","domainType":"pfam","details":{"start":"start","end":"end","name":"hmm.name"}},"snvData":[{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128748842,"End_Position":128748842,"Strand":"+","Variant_Classification":"Silent","Variant_Type":"SNP","Reference_Allele":"G","Tumor_Seq_Allele1":"G","Tumor_Seq_Allele2":"A","dbSNP_RS":"novel","Tumor_Sample_Barcode":"14-35026T","Matched_Norm_Sample_Barcode":"14-35026N","Match_Norm_Seq_Allele1":"G","Match_Norm_Seq_Allele2":"G","HGVSc":"c.3G>A","HGVSp":"p.=","HGVSp_Short":"p.L1=","Transcript_ID":"ENST00000377970","Exon_Number":"1/3","t_depth":79,"t_ref_count":32,"t_alt_count":47,"n_depth":29,"n_ref_count":29,"n_alt_count":0,"Mutation_Class":"Inframe","AA_Position":1},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128748851,"End_Position":128748851,"Strand":"+","Variant_Classification":"Silent","Variant_Type":"SNP","Reference_Allele":"T","Tumor_Seq_Allele1":"T","Tumor_Seq_Allele2":"C","dbSNP_RS":"rs777974417","Tumor_Sample_Barcode":"SP192993","Matched_Norm_Sample_Barcode":"SP192992","Match_Norm_Seq_Allele1":"T","Match_Norm_Seq_Allele2":"T","HGVSc":"c.12T>C","HGVSp":"p.=","HGVSp_Short":"p.F4=","Transcript_ID":"ENST00000377970","Exon_Number":"1/3","t_depth":38,"t_ref_count":17,"t_alt_count":21,"n_depth":41,"n_ref_count":41,"n_alt_count":0,"Mutation_Class":"Inframe","AA_Position":4},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128748852,"End_Position":128748852,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"T","dbSNP_RS":"novel","Tumor_Sample_Barcode":"02-28397_tumorA","Matched_Norm_Sample_Barcode":"02-28397_normal","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.13C>T","HGVSp":"p.Arg5Trp","HGVSp_Short":"p.R5W","Transcript_ID":"ENST00000377970","Exon_Number":"1/3","t_depth":64,"t_ref_count":52,"t_alt_count":12,"n_depth":41,"n_ref_count":41,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":5},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128748856,"End_Position":128748856,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"T","Tumor_Seq_Allele1":"T","Tumor_Seq_Allele2":"G","dbSNP_RS":"novel","Tumor_Sample_Barcode":"15-21654T","Matched_Norm_Sample_Barcode":"15-21654N","Match_Norm_Seq_Allele1":"T","Match_Norm_Seq_Allele2":"T","HGVSc":"c.17T>G","HGVSp":"p.Val6Gly","HGVSp_Short":"p.V6G","Transcript_ID":"ENST00000377970","Exon_Number":"1/3","t_depth":60,"t_ref_count":38,"t_alt_count":22,"n_depth":37,"n_ref_count":37,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":6},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128748858,"End_Position":128748858,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"G","Tumor_Seq_Allele1":"G","Tumor_Seq_Allele2":"A","dbSNP_RS":"","Tumor_Sample_Barcode":"SP59280","Matched_Norm_Sample_Barcode":"SP59278","Match_Norm_Seq_Allele1":"G","Match_Norm_Seq_Allele2":"G","HGVSc":"c.19G>A","HGVSp":"p.Val7Met","HGVSp_Short":"p.V7M","Transcript_ID":"ENST00000377970","Exon_Number":"1/3","t_depth":25,"t_ref_count":13,"t_alt_count":12,"n_depth":24,"n_ref_count":24,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":7},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128748863,"End_Position":128748863,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"A","Tumor_Seq_Allele1":"A","Tumor_Seq_Allele2":"C","dbSNP_RS":"novel","Tumor_Sample_Barcode":"SP192993","Matched_Norm_Sample_Barcode":"SP192992","Match_Norm_Seq_Allele1":"A","Match_Norm_Seq_Allele2":"A","HGVSc":"c.24A>C","HGVSp":"p.Glu8Asp","HGVSp_Short":"p.E8D","Transcript_ID":"ENST00000377970","Exon_Number":"1/3","t_depth":42,"t_ref_count":18,"t_alt_count":24,"n_depth":44,"n_ref_count":44,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":8},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128748870,"End_Position":128748870,"Strand":"+","Variant_Classification":"Splice_Site","Variant_Type":"SNP","Reference_Allele":"G","Tumor_Seq_Allele1":"G","Tumor_Seq_Allele2":"A","dbSNP_RS":"novel","Tumor_Sample_Barcode":"SP116697","Matched_Norm_Sample_Barcode":"SP116695","Match_Norm_Seq_Allele1":"G","Match_Norm_Seq_Allele2":"G","HGVSc":"c.30+1G>A","HGVSp":"","HGVSp_Short":"p.X10_splice","Transcript_ID":"ENST00000377970","Exon_Number":"","t_depth":47,"t_ref_count":31,"t_alt_count":16,"n_depth":31,"n_ref_count":31,"n_alt_count":0,"Mutation_Class":"Truncating","AA_Position":10},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128748870,"End_Position":128748870,"Strand":"+","Variant_Classification":"Splice_Site","Variant_Type":"SNP","Reference_Allele":"G","Tumor_Seq_Allele1":"G","Tumor_Seq_Allele2":"T","dbSNP_RS":"","Tumor_Sample_Barcode":"SP59376","Matched_Norm_Sample_Barcode":"SP59374","Match_Norm_Seq_Allele1":"G","Match_Norm_Seq_Allele2":"G","HGVSc":"c.30+1G>T","HGVSp":"","HGVSp_Short":"p.X10_splice","Transcript_ID":"ENST00000377970","Exon_Number":"","t_depth":58,"t_ref_count":43,"t_alt_count":15,"n_depth":18,"n_ref_count":18,"n_alt_count":0,"Mutation_Class":"Truncating","AA_Position":10},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750522,"End_Position":128750522,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"T","Tumor_Seq_Allele1":"T","Tumor_Seq_Allele2":"C","dbSNP_RS":"","Tumor_Sample_Barcode":"14-35026T","Matched_Norm_Sample_Barcode":"14-35026N","Match_Norm_Seq_Allele1":"T","Match_Norm_Seq_Allele2":"T","HGVSc":"c.59T>C","HGVSp":"p.Val20Ala","HGVSp_Short":"p.V20A","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":69,"t_ref_count":33,"t_alt_count":36,"n_depth":27,"n_ref_count":27,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":20},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750526,"End_Position":128750526,"Strand":"+","Variant_Classification":"Silent","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"T","dbSNP_RS":"rs772792914","Tumor_Sample_Barcode":"15-15757T","Matched_Norm_Sample_Barcode":"15-15757N","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.63C>T","HGVSp":"p.=","HGVSp_Short":"p.S21=","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":56,"t_ref_count":37,"t_alt_count":19,"n_depth":32,"n_ref_count":32,"n_alt_count":0,"Mutation_Class":"Inframe","AA_Position":21},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750526,"End_Position":128750526,"Strand":"+","Variant_Classification":"Silent","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"T","dbSNP_RS":"rs772792914","Tumor_Sample_Barcode":"HTMCP-01-06-00634-01A-01D","Matched_Norm_Sample_Barcode":"HTMCP-01-06-00634-10A-01D","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.63C>T","HGVSp":"p.=","HGVSp_Short":"p.S21=","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":66,"t_ref_count":45,"t_alt_count":21,"n_depth":48,"n_ref_count":48,"n_alt_count":0,"Mutation_Class":"Inframe","AA_Position":21},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750525,"End_Position":128750525,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"G","Tumor_Seq_Allele1":"G","Tumor_Seq_Allele2":"A","dbSNP_RS":"novel","Tumor_Sample_Barcode":"SP192870","Matched_Norm_Sample_Barcode":"SP192864","Match_Norm_Seq_Allele1":"G","Match_Norm_Seq_Allele2":"G","HGVSc":"c.62G>A","HGVSp":"p.Ser21Asn","HGVSp_Short":"p.S21N","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":62,"t_ref_count":44,"t_alt_count":18,"n_depth":51,"n_ref_count":51,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":21},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750534,"End_Position":128750534,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"A","Tumor_Seq_Allele1":"A","Tumor_Seq_Allele2":"G","dbSNP_RS":"rs148915481","Tumor_Sample_Barcode":"SP192870","Matched_Norm_Sample_Barcode":"SP192864","Match_Norm_Seq_Allele1":"A","Match_Norm_Seq_Allele2":"A","HGVSc":"c.71A>G","HGVSp":"p.Asn24Ser","HGVSp_Short":"p.N24S","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":62,"t_ref_count":27,"t_alt_count":35,"n_depth":53,"n_ref_count":53,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":24},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750556,"End_Position":128750556,"Strand":"+","Variant_Classification":"Silent","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"T","dbSNP_RS":"novel","Tumor_Sample_Barcode":"16-12281_tumorB","Matched_Norm_Sample_Barcode":"16-12281_normal","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.93C>T","HGVSp":"p.=","HGVSp_Short":"p.Y31=","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":79,"t_ref_count":56,"t_alt_count":23,"n_depth":41,"n_ref_count":41,"n_alt_count":0,"Mutation_Class":"Inframe","AA_Position":31},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750569,"End_Position":128750569,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"T","dbSNP_RS":"novel","Tumor_Sample_Barcode":"HTMCP-01-06-00242-01A-01D","Matched_Norm_Sample_Barcode":"HTMCP-01-06-00242-10A-01D","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.106C>T","HGVSp":"p.Pro36Ser","HGVSp_Short":"p.P36S","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":55,"t_ref_count":42,"t_alt_count":13,"n_depth":31,"n_ref_count":31,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":36},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750592,"End_Position":128750592,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"G","Tumor_Seq_Allele1":"G","Tumor_Seq_Allele2":"C","dbSNP_RS":"novel","Tumor_Sample_Barcode":"SP116635","Matched_Norm_Sample_Barcode":"SP116634","Match_Norm_Seq_Allele1":"G","Match_Norm_Seq_Allele2":"G","HGVSc":"c.129G>C","HGVSp":"p.Glu43Asp","HGVSp_Short":"p.E43D","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":13,"t_ref_count":5,"t_alt_count":8,"n_depth":20,"n_ref_count":20,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":43},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750611,"End_Position":128750611,"Strand":"+","Variant_Classification":"Nonsense_Mutation","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"T","dbSNP_RS":"novel","Tumor_Sample_Barcode":"HTMCP-01-06-00634-01A-01D","Matched_Norm_Sample_Barcode":"HTMCP-01-06-00634-10A-01D","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.148C>T","HGVSp":"p.Gln50Ter","HGVSp_Short":"p.Q50*","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":79,"t_ref_count":51,"t_alt_count":28,"n_depth":37,"n_ref_count":37,"n_alt_count":0,"Mutation_Class":"Truncating","AA_Position":50},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750613,"End_Position":128750613,"Strand":"+","Variant_Classification":"Silent","Variant_Type":"SNP","Reference_Allele":"G","Tumor_Seq_Allele1":"G","Tumor_Seq_Allele2":"A","dbSNP_RS":"","Tumor_Sample_Barcode":"18-14625_tumorA","Matched_Norm_Sample_Barcode":"18-14625_normal","Match_Norm_Seq_Allele1":"G","Match_Norm_Seq_Allele2":"G","HGVSc":"c.150G>A","HGVSp":"p.=","HGVSp_Short":"p.Q50=","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":42,"t_ref_count":25,"t_alt_count":17,"n_depth":72,"n_ref_count":72,"n_alt_count":0,"Mutation_Class":"Inframe","AA_Position":50},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750613,"End_Position":128750613,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"G","Tumor_Seq_Allele1":"G","Tumor_Seq_Allele2":"C","dbSNP_RS":"novel","Tumor_Sample_Barcode":"HTMCP-01-06-00634-01A-01D","Matched_Norm_Sample_Barcode":"HTMCP-01-06-00634-10A-01D","Match_Norm_Seq_Allele1":"G","Match_Norm_Seq_Allele2":"G","HGVSc":"c.150G>C","HGVSp":"p.Gln50His","HGVSp_Short":"p.Q50H","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":79,"t_ref_count":51,"t_alt_count":28,"n_depth":37,"n_ref_count":37,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":50},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750613,"End_Position":128750613,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"G","Tumor_Seq_Allele1":"G","Tumor_Seq_Allele2":"T","dbSNP_RS":"novel","Tumor_Sample_Barcode":"SP116627","Matched_Norm_Sample_Barcode":"SP116626","Match_Norm_Seq_Allele1":"G","Match_Norm_Seq_Allele2":"G","HGVSc":"c.150G>T","HGVSp":"p.Gln50His","HGVSp_Short":"p.Q50H","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":26,"t_ref_count":18,"t_alt_count":8,"n_depth":22,"n_ref_count":22,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":50},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750616,"End_Position":128750616,"Strand":"+","Variant_Classification":"Silent","Variant_Type":"SNP","Reference_Allele":"G","Tumor_Seq_Allele1":"G","Tumor_Seq_Allele2":"A","dbSNP_RS":"novel","Tumor_Sample_Barcode":"SP116688","Matched_Norm_Sample_Barcode":"SP116687","Match_Norm_Seq_Allele1":"G","Match_Norm_Seq_Allele2":"G","HGVSc":"c.153G>A","HGVSp":"p.=","HGVSp_Short":"p.Q51=","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":33,"t_ref_count":10,"t_alt_count":23,"n_depth":31,"n_ref_count":31,"n_alt_count":0,"Mutation_Class":"Inframe","AA_Position":51},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750621,"End_Position":128750621,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"G","Tumor_Seq_Allele1":"G","Tumor_Seq_Allele2":"T","dbSNP_RS":"novel","Tumor_Sample_Barcode":"03-23488_tumorA","Matched_Norm_Sample_Barcode":"03-23488_normal","Match_Norm_Seq_Allele1":"G","Match_Norm_Seq_Allele2":"G","HGVSc":"c.158G>T","HGVSp":"p.Ser53Ile","HGVSp_Short":"p.S53I","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":36,"t_ref_count":17,"t_alt_count":19,"n_depth":34,"n_ref_count":34,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":53},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750625,"End_Position":128750625,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"G","Tumor_Seq_Allele1":"G","Tumor_Seq_Allele2":"C","dbSNP_RS":"rs121918684","Tumor_Sample_Barcode":"03-23488_tumorA","Matched_Norm_Sample_Barcode":"03-23488_normal","Match_Norm_Seq_Allele1":"G","Match_Norm_Seq_Allele2":"G","HGVSc":"c.162G>C","HGVSp":"p.Glu54Asp","HGVSp_Short":"p.E54D","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":36,"t_ref_count":17,"t_alt_count":19,"n_depth":33,"n_ref_count":33,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":54},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750625,"End_Position":128750625,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"G","Tumor_Seq_Allele1":"G","Tumor_Seq_Allele2":"C","dbSNP_RS":"rs121918684","Tumor_Sample_Barcode":"SP192993","Matched_Norm_Sample_Barcode":"SP192992","Match_Norm_Seq_Allele1":"G","Match_Norm_Seq_Allele2":"G","HGVSc":"c.162G>C","HGVSp":"p.Glu54Asp","HGVSp_Short":"p.E54D","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":36,"t_ref_count":21,"t_alt_count":15,"n_depth":40,"n_ref_count":40,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":54},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750638,"End_Position":128750638,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"G","Tumor_Seq_Allele1":"G","Tumor_Seq_Allele2":"A","dbSNP_RS":"","Tumor_Sample_Barcode":"SP116659","Matched_Norm_Sample_Barcode":"SP116658","Match_Norm_Seq_Allele1":"G","Match_Norm_Seq_Allele2":"G","HGVSc":"c.175G>A","HGVSp":"p.Ala59Thr","HGVSp_Short":"p.A59T","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":39,"t_ref_count":26,"t_alt_count":13,"n_depth":28,"n_ref_count":28,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":59},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750673,"End_Position":128750673,"Strand":"+","Variant_Classification":"Silent","Variant_Type":"SNP","Reference_Allele":"G","Tumor_Seq_Allele1":"G","Tumor_Seq_Allele2":"A","dbSNP_RS":"novel","Tumor_Sample_Barcode":"SP116627","Matched_Norm_Sample_Barcode":"SP116626","Match_Norm_Seq_Allele1":"G","Match_Norm_Seq_Allele2":"G","HGVSc":"c.210G>A","HGVSp":"p.=","HGVSp_Short":"p.L70=","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":17,"t_ref_count":10,"t_alt_count":7,"n_depth":17,"n_ref_count":17,"n_alt_count":0,"Mutation_Class":"Inframe","AA_Position":70},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750677,"End_Position":128750677,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"T","dbSNP_RS":"rs28933407","Tumor_Sample_Barcode":"HTMCP-01-06-00206-01A-01D","Matched_Norm_Sample_Barcode":"HTMCP-01-06-00206-10A-01D","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.214C>T","HGVSp":"p.Pro72Ser","HGVSp_Short":"p.P72S","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":64,"t_ref_count":44,"t_alt_count":20,"n_depth":27,"n_ref_count":27,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":72},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750681,"End_Position":128750681,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"T","dbSNP_RS":"","Tumor_Sample_Barcode":"03-23488_tumorA","Matched_Norm_Sample_Barcode":"03-23488_normal","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.218C>T","HGVSp":"p.Thr73Ile","HGVSp_Short":"p.T73I","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":35,"t_ref_count":14,"t_alt_count":21,"n_depth":23,"n_ref_count":23,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":73},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750681,"End_Position":128750681,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"T","dbSNP_RS":"","Tumor_Sample_Barcode":"15-15757T","Matched_Norm_Sample_Barcode":"15-15757N","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.218C>T","HGVSp":"p.Thr73Ile","HGVSp_Short":"p.T73I","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":51,"t_ref_count":41,"t_alt_count":10,"n_depth":22,"n_ref_count":22,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":73},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750681,"End_Position":128750681,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"T","dbSNP_RS":"","Tumor_Sample_Barcode":"HTMCP-01-06-00307-01A-01D","Matched_Norm_Sample_Barcode":"HTMCP-01-06-00307-10A-01D","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.218C>T","HGVSp":"p.Thr73Ile","HGVSp_Short":"p.T73I","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":55,"t_ref_count":39,"t_alt_count":16,"n_depth":42,"n_ref_count":42,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":73},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750681,"End_Position":128750682,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"DNP","Reference_Allele":"CC","Tumor_Seq_Allele1":"CC","Tumor_Seq_Allele2":"TT","dbSNP_RS":"novel","Tumor_Sample_Barcode":"SP192870","Matched_Norm_Sample_Barcode":"SP192864","Match_Norm_Seq_Allele1":"CC","Match_Norm_Seq_Allele2":"CC","HGVSc":"c.218_219delCCinsTT","HGVSp":"p.Thr73Ile","HGVSp_Short":"p.T73I","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":32,"t_ref_count":15,"t_alt_count":17,"n_depth":46,"n_ref_count":46,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":73},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750681,"End_Position":128750681,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"A","dbSNP_RS":"novel","Tumor_Sample_Barcode":"15-16885T","Matched_Norm_Sample_Barcode":"15-16885N","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.218C>A","HGVSp":"p.Thr73Asn","HGVSp_Short":"p.T73N","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":47,"t_ref_count":24,"t_alt_count":23,"n_depth":21,"n_ref_count":21,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":73},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750681,"End_Position":128750681,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"A","dbSNP_RS":"novel","Tumor_Sample_Barcode":"HTMCP-01-01-00451-01A-01D","Matched_Norm_Sample_Barcode":"HTMCP-01-01-00451-11B-01D","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.218C>A","HGVSp":"p.Thr73Asn","HGVSp_Short":"p.T73N","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":82,"t_ref_count":51,"t_alt_count":31,"n_depth":32,"n_ref_count":32,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":73},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750681,"End_Position":128750681,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"A","dbSNP_RS":"novel","Tumor_Sample_Barcode":"SP116688","Matched_Norm_Sample_Barcode":"SP116687","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.218C>A","HGVSp":"p.Thr73Asn","HGVSp_Short":"p.T73N","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":24,"t_ref_count":10,"t_alt_count":14,"n_depth":23,"n_ref_count":23,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":73},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750681,"End_Position":128750681,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"A","dbSNP_RS":"novel","Tumor_Sample_Barcode":"SP124957","Matched_Norm_Sample_Barcode":"SP124956","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.218C>A","HGVSp":"p.Thr73Asn","HGVSp_Short":"p.T73N","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":19,"t_ref_count":13,"t_alt_count":6,"n_depth":22,"n_ref_count":22,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":73},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750683,"End_Position":128750683,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"T","dbSNP_RS":"","Tumor_Sample_Barcode":"SP59280","Matched_Norm_Sample_Barcode":"SP59278","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.220C>T","HGVSp":"p.Pro74Ser","HGVSp_Short":"p.P74S","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":13,"t_ref_count":7,"t_alt_count":6,"n_depth":15,"n_ref_count":15,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":74},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750683,"End_Position":128750683,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"A","dbSNP_RS":"novel","Tumor_Sample_Barcode":"365-11-01BD","Matched_Norm_Sample_Barcode":"365-01-01ND","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.220C>A","HGVSp":"p.Pro74Thr","HGVSp_Short":"p.P74T","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":29,"t_ref_count":16,"t_alt_count":13,"n_depth":33,"n_ref_count":33,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":74},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750686,"End_Position":128750686,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"G","dbSNP_RS":"","Tumor_Sample_Barcode":"181712-PL01","Matched_Norm_Sample_Barcode":"181677-BC01","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.223N>G","HGVSp":"p.Pro75Ala","HGVSp_Short":"p.P75A","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":34,"t_ref_count":22,"t_alt_count":12,"n_depth":50,"n_ref_count":50,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":75},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750686,"End_Position":128750686,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"G","dbSNP_RS":"","Tumor_Sample_Barcode":"FL1019T2","Matched_Norm_Sample_Barcode":"FL1019N","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.223C>G","HGVSp":"p.Pro75Ala","HGVSp_Short":"p.P75A","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":46,"t_ref_count":36,"t_alt_count":10,"n_depth":37,"n_ref_count":37,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":75},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750686,"End_Position":128750686,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"A","dbSNP_RS":"novel","Tumor_Sample_Barcode":"03-23488_tumorB","Matched_Norm_Sample_Barcode":"03-23488_normal","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.223C>A","HGVSp":"p.Pro75Thr","HGVSp_Short":"p.P75T","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":67,"t_ref_count":53,"t_alt_count":14,"n_depth":24,"n_ref_count":24,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":75},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750694,"End_Position":128750695,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"DNP","Reference_Allele":"CC","Tumor_Seq_Allele1":"CC","Tumor_Seq_Allele2":"GT","dbSNP_RS":"novel","Tumor_Sample_Barcode":"191621-PL01","Matched_Norm_Sample_Barcode":"190906-CS01","Match_Norm_Seq_Allele1":"CC","Match_Norm_Seq_Allele2":"CC","HGVSc":"c.231_232delNNinsGT","HGVSp":"p.Pro78Ser","HGVSp_Short":"p.P78S","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":36,"t_ref_count":17,"t_alt_count":19,"n_depth":39,"n_ref_count":39,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":78},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750768,"End_Position":128750768,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"A","Tumor_Seq_Allele1":"A","Tumor_Seq_Allele2":"C","dbSNP_RS":"novel","Tumor_Sample_Barcode":"03-23488_tumorA","Matched_Norm_Sample_Barcode":"03-23488_normal","Match_Norm_Seq_Allele1":"A","Match_Norm_Seq_Allele2":"A","HGVSc":"c.305A>C","HGVSp":"p.Asp102Ala","HGVSp_Short":"p.D102A","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":29,"t_ref_count":14,"t_alt_count":15,"n_depth":27,"n_ref_count":27,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":102},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750802,"End_Position":128750802,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"G","Tumor_Seq_Allele1":"G","Tumor_Seq_Allele2":"C","dbSNP_RS":"","Tumor_Sample_Barcode":"03-23488_tumorA","Matched_Norm_Sample_Barcode":"03-23488_normal","Match_Norm_Seq_Allele1":"G","Match_Norm_Seq_Allele2":"G","HGVSc":"c.339G>C","HGVSp":"p.Gln113His","HGVSp_Short":"p.Q113H","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":32,"t_ref_count":13,"t_alt_count":19,"n_depth":23,"n_ref_count":23,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":113},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750821,"End_Position":128750821,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"G","dbSNP_RS":"novel","Tumor_Sample_Barcode":"FL1008T2","Matched_Norm_Sample_Barcode":"FL1008N","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.358C>G","HGVSp":"p.Leu120Val","HGVSp_Short":"p.L120V","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":44,"t_ref_count":22,"t_alt_count":22,"n_depth":32,"n_ref_count":32,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":120},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750824,"End_Position":128750824,"Strand":"+","Variant_Classification":"Silent","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"T","dbSNP_RS":"novel","Tumor_Sample_Barcode":"SP116688","Matched_Norm_Sample_Barcode":"SP116687","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.361C>T","HGVSp":"p.=","HGVSp_Short":"p.L121=","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":39,"t_ref_count":10,"t_alt_count":29,"n_depth":28,"n_ref_count":28,"n_alt_count":0,"Mutation_Class":"Inframe","AA_Position":121},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750844,"End_Position":128750844,"Strand":"+","Variant_Classification":"Silent","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"T","dbSNP_RS":"novel","Tumor_Sample_Barcode":"SP116688","Matched_Norm_Sample_Barcode":"SP116687","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.381C>T","HGVSp":"p.=","HGVSp_Short":"p.N127=","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":41,"t_ref_count":12,"t_alt_count":29,"n_depth":30,"n_ref_count":30,"n_alt_count":0,"Mutation_Class":"Inframe","AA_Position":127},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750849,"End_Position":128750849,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"G","Tumor_Seq_Allele1":"G","Tumor_Seq_Allele2":"A","dbSNP_RS":"novel","Tumor_Sample_Barcode":"16-12281_tumorB","Matched_Norm_Sample_Barcode":"16-12281_normal","Match_Norm_Seq_Allele1":"G","Match_Norm_Seq_Allele2":"G","HGVSc":"c.386G>A","HGVSp":"p.Ser129Asn","HGVSp_Short":"p.S129N","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":84,"t_ref_count":62,"t_alt_count":22,"n_depth":32,"n_ref_count":32,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":129},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750849,"End_Position":128750849,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"G","Tumor_Seq_Allele1":"G","Tumor_Seq_Allele2":"C","dbSNP_RS":"","Tumor_Sample_Barcode":"08-15460_tumorB","Matched_Norm_Sample_Barcode":"08-15460N","Match_Norm_Seq_Allele1":"G","Match_Norm_Seq_Allele2":"G","HGVSc":"c.386G>C","HGVSp":"p.Ser129Thr","HGVSp_Short":"p.S129T","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":42,"t_ref_count":32,"t_alt_count":10,"n_depth":23,"n_ref_count":23,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":129},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750878,"End_Position":128750878,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"T","Tumor_Seq_Allele1":"T","Tumor_Seq_Allele2":"G","dbSNP_RS":"novel","Tumor_Sample_Barcode":"SP124957","Matched_Norm_Sample_Barcode":"SP124956","Match_Norm_Seq_Allele1":"T","Match_Norm_Seq_Allele2":"T","HGVSc":"c.415T>G","HGVSp":"p.Phe139Val","HGVSp_Short":"p.F139V","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":19,"t_ref_count":12,"t_alt_count":7,"n_depth":19,"n_ref_count":19,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":139},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750883,"End_Position":128750883,"Strand":"+","Variant_Classification":"Silent","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"T","dbSNP_RS":"novel","Tumor_Sample_Barcode":"16-12281_tumorB","Matched_Norm_Sample_Barcode":"16-12281_normal","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.420C>T","HGVSp":"p.=","HGVSp_Short":"p.I140=","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":84,"t_ref_count":60,"t_alt_count":24,"n_depth":34,"n_ref_count":34,"n_alt_count":0,"Mutation_Class":"Inframe","AA_Position":140},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750921,"End_Position":128750921,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"T","Tumor_Seq_Allele1":"T","Tumor_Seq_Allele2":"C","dbSNP_RS":"","Tumor_Sample_Barcode":"13-26835_tumorA","Matched_Norm_Sample_Barcode":"13-26835_normal","Match_Norm_Seq_Allele1":"T","Match_Norm_Seq_Allele2":"T","HGVSc":"c.458T>C","HGVSp":"p.Phe153Ser","HGVSp_Short":"p.F153S","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":46,"t_ref_count":31,"t_alt_count":15,"n_depth":38,"n_ref_count":38,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":153},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750938,"End_Position":128750938,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"T","dbSNP_RS":"novel","Tumor_Sample_Barcode":"00-14595_tumorC","Matched_Norm_Sample_Barcode":"00-14595_normal","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.475C>T","HGVSp":"p.Leu159Phe","HGVSp_Short":"p.L159F","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":24,"t_ref_count":15,"t_alt_count":9,"n_depth":45,"n_ref_count":45,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":159},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750942,"End_Position":128750942,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"T","Tumor_Seq_Allele1":"T","Tumor_Seq_Allele2":"A","dbSNP_RS":"novel","Tumor_Sample_Barcode":"FL3020T1","Matched_Norm_Sample_Barcode":"FL3020N","Match_Norm_Seq_Allele1":"T","Match_Norm_Seq_Allele2":"T","HGVSc":"c.479T>A","HGVSp":"p.Val160Asp","HGVSp_Short":"p.V160D","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":71,"t_ref_count":52,"t_alt_count":19,"n_depth":33,"n_ref_count":33,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":160},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750941,"End_Position":128750941,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"G","Tumor_Seq_Allele1":"G","Tumor_Seq_Allele2":"C","dbSNP_RS":"novel","Tumor_Sample_Barcode":"HTMCP-01-06-00634-01A-01D","Matched_Norm_Sample_Barcode":"HTMCP-01-06-00634-10A-01D","Match_Norm_Seq_Allele1":"G","Match_Norm_Seq_Allele2":"G","HGVSc":"c.478G>C","HGVSp":"p.Val160Leu","HGVSp_Short":"p.V160L","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":74,"t_ref_count":49,"t_alt_count":25,"n_depth":32,"n_ref_count":32,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":160},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750946,"End_Position":128750946,"Strand":"+","Variant_Classification":"Silent","Variant_Type":"SNP","Reference_Allele":"A","Tumor_Seq_Allele1":"A","Tumor_Seq_Allele2":"T","dbSNP_RS":"novel","Tumor_Sample_Barcode":"SP192993","Matched_Norm_Sample_Barcode":"SP192992","Match_Norm_Seq_Allele1":"A","Match_Norm_Seq_Allele2":"A","HGVSc":"c.483A>T","HGVSp":"p.=","HGVSp_Short":"p.S161=","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":47,"t_ref_count":28,"t_alt_count":19,"n_depth":47,"n_ref_count":47,"n_alt_count":0,"Mutation_Class":"Inframe","AA_Position":161},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750945,"End_Position":128750945,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"T","dbSNP_RS":"","Tumor_Sample_Barcode":"16-12281_tumorB","Matched_Norm_Sample_Barcode":"16-12281_normal","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.482C>T","HGVSp":"p.Ser161Leu","HGVSp_Short":"p.S161L","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":78,"t_ref_count":53,"t_alt_count":25,"n_depth":39,"n_ref_count":39,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":161},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750953,"End_Position":128750953,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"G","dbSNP_RS":"","Tumor_Sample_Barcode":"01-23117_tumorB","Matched_Norm_Sample_Barcode":"01-23117_normal","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.490C>G","HGVSp":"p.Leu164Val","HGVSp_Short":"p.L164V","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":43,"t_ref_count":21,"t_alt_count":22,"n_depth":36,"n_ref_count":36,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":164},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750964,"End_Position":128750964,"Strand":"+","Variant_Classification":"Silent","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"T","dbSNP_RS":"rs752022129","Tumor_Sample_Barcode":"SP59412","Matched_Norm_Sample_Barcode":"SP59410","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.501C>T","HGVSp":"p.=","HGVSp_Short":"p.Y167=","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":36,"t_ref_count":28,"t_alt_count":8,"n_depth":27,"n_ref_count":27,"n_alt_count":0,"Mutation_Class":"Inframe","AA_Position":167},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750962,"End_Position":128750962,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"T","Tumor_Seq_Allele1":"T","Tumor_Seq_Allele2":"A","dbSNP_RS":"novel","Tumor_Sample_Barcode":"SP124957","Matched_Norm_Sample_Barcode":"SP124956","Match_Norm_Seq_Allele1":"T","Match_Norm_Seq_Allele2":"T","HGVSc":"c.499T>A","HGVSp":"p.Tyr167Asn","HGVSp_Short":"p.Y167N","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":26,"t_ref_count":16,"t_alt_count":10,"n_depth":26,"n_ref_count":26,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":167},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750970,"End_Position":128750970,"Strand":"+","Variant_Classification":"Silent","Variant_Type":"SNP","Reference_Allele":"T","Tumor_Seq_Allele1":"T","Tumor_Seq_Allele2":"A","dbSNP_RS":"novel","Tumor_Sample_Barcode":"01-23117_tumorB","Matched_Norm_Sample_Barcode":"01-23117_normal","Match_Norm_Seq_Allele1":"T","Match_Norm_Seq_Allele2":"T","HGVSc":"c.507T>A","HGVSp":"p.=","HGVSp_Short":"p.A169=","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":45,"t_ref_count":22,"t_alt_count":23,"n_depth":39,"n_ref_count":39,"n_alt_count":0,"Mutation_Class":"Inframe","AA_Position":169},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128750997,"End_Position":128750997,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"A","dbSNP_RS":"novel","Tumor_Sample_Barcode":"HTMCP-01-06-00634-01A-01D","Matched_Norm_Sample_Barcode":"HTMCP-01-06-00634-10A-01D","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.534C>A","HGVSp":"p.Asn178Lys","HGVSp_Short":"p.N178K","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":80,"t_ref_count":51,"t_alt_count":29,"n_depth":32,"n_ref_count":32,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":178},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128751033,"End_Position":128751033,"Strand":"+","Variant_Classification":"Silent","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"T","dbSNP_RS":"novel","Tumor_Sample_Barcode":"HTMCP-01-06-00634-01A-01D","Matched_Norm_Sample_Barcode":"HTMCP-01-06-00634-10A-01D","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.570C>T","HGVSp":"p.=","HGVSp_Short":"p.S190=","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":62,"t_ref_count":40,"t_alt_count":22,"n_depth":33,"n_ref_count":33,"n_alt_count":0,"Mutation_Class":"Inframe","AA_Position":190},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128751036,"End_Position":128751036,"Strand":"+","Variant_Classification":"Silent","Variant_Type":"SNP","Reference_Allele":"G","Tumor_Seq_Allele1":"G","Tumor_Seq_Allele2":"A","dbSNP_RS":"","Tumor_Sample_Barcode":"SP116630","Matched_Norm_Sample_Barcode":"SP116628","Match_Norm_Seq_Allele1":"G","Match_Norm_Seq_Allele2":"G","HGVSc":"c.573G>A","HGVSp":"p.=","HGVSp_Short":"p.L191=","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":34,"t_ref_count":21,"t_alt_count":13,"n_depth":23,"n_ref_count":23,"n_alt_count":0,"Mutation_Class":"Inframe","AA_Position":191},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128751039,"End_Position":128751039,"Strand":"+","Variant_Classification":"Silent","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"T","dbSNP_RS":"rs369978120","Tumor_Sample_Barcode":"HTMCP-01-06-00634-01A-01D","Matched_Norm_Sample_Barcode":"HTMCP-01-06-00634-10A-01D","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.576C>T","HGVSp":"p.=","HGVSp_Short":"p.Y192=","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":63,"t_ref_count":40,"t_alt_count":23,"n_depth":33,"n_ref_count":33,"n_alt_count":0,"Mutation_Class":"Inframe","AA_Position":192},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128751087,"End_Position":128751087,"Strand":"+","Variant_Classification":"Silent","Variant_Type":"SNP","Reference_Allele":"G","Tumor_Seq_Allele1":"G","Tumor_Seq_Allele2":"A","dbSNP_RS":"novel","Tumor_Sample_Barcode":"SP192993","Matched_Norm_Sample_Barcode":"SP192992","Match_Norm_Seq_Allele1":"G","Match_Norm_Seq_Allele2":"G","HGVSc":"c.624G>A","HGVSp":"p.=","HGVSp_Short":"p.V208=","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":36,"t_ref_count":18,"t_alt_count":18,"n_depth":37,"n_ref_count":37,"n_alt_count":0,"Mutation_Class":"Inframe","AA_Position":208},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128751093,"End_Position":128751093,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"A","dbSNP_RS":"novel","Tumor_Sample_Barcode":"SP116688","Matched_Norm_Sample_Barcode":"SP116687","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.630C>A","HGVSp":"p.Phe210Leu","HGVSp_Short":"p.F210L","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":32,"t_ref_count":8,"t_alt_count":24,"n_depth":32,"n_ref_count":32,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":210},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128751116,"End_Position":128751116,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"G","Tumor_Seq_Allele1":"G","Tumor_Seq_Allele2":"T","dbSNP_RS":"novel","Tumor_Sample_Barcode":"SP116688","Matched_Norm_Sample_Barcode":"SP116687","Match_Norm_Seq_Allele1":"G","Match_Norm_Seq_Allele2":"G","HGVSc":"c.653G>T","HGVSp":"p.Ser218Ile","HGVSp_Short":"p.S218I","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":37,"t_ref_count":10,"t_alt_count":27,"n_depth":36,"n_ref_count":36,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":218},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128751195,"End_Position":128751195,"Strand":"+","Variant_Classification":"Silent","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"G","dbSNP_RS":"novel","Tumor_Sample_Barcode":"SP116688","Matched_Norm_Sample_Barcode":"SP116687","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.732C>G","HGVSp":"p.=","HGVSp_Short":"p.S244=","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":37,"t_ref_count":11,"t_alt_count":26,"n_depth":31,"n_ref_count":31,"n_alt_count":0,"Mutation_Class":"Inframe","AA_Position":244},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128751242,"End_Position":128751242,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"A","dbSNP_RS":"rs749498067","Tumor_Sample_Barcode":"SP116627","Matched_Norm_Sample_Barcode":"SP116626","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.779C>A","HGVSp":"p.Pro260Gln","HGVSp_Short":"p.P260Q","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":20,"t_ref_count":12,"t_alt_count":8,"n_depth":29,"n_ref_count":29,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":260},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128751244,"End_Position":128751244,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"T","dbSNP_RS":"novel","Tumor_Sample_Barcode":"SP116688","Matched_Norm_Sample_Barcode":"SP116687","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.781C>T","HGVSp":"p.Pro261Ser","HGVSp_Short":"p.P261S","Transcript_ID":"ENST00000377970","Exon_Number":"2/3","t_depth":27,"t_ref_count":13,"t_alt_count":14,"n_depth":28,"n_ref_count":28,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":261},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128752702,"End_Position":128752702,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"A","dbSNP_RS":"novel","Tumor_Sample_Barcode":"HTMCP-01-06-00634-01A-01D","Matched_Norm_Sample_Barcode":"HTMCP-01-06-00634-10A-01D","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.863C>A","HGVSp":"p.Pro288His","HGVSp_Short":"p.P288H","Transcript_ID":"ENST00000377970","Exon_Number":"3/3","t_depth":66,"t_ref_count":44,"t_alt_count":22,"n_depth":36,"n_ref_count":36,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":288},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128752715,"End_Position":128752715,"Strand":"+","Variant_Classification":"Silent","Variant_Type":"SNP","Reference_Allele":"A","Tumor_Seq_Allele1":"A","Tumor_Seq_Allele2":"G","dbSNP_RS":"novel","Tumor_Sample_Barcode":"75-61923_tumorA","Matched_Norm_Sample_Barcode":"75-61923_normal","Match_Norm_Seq_Allele1":"A","Match_Norm_Seq_Allele2":"A","HGVSc":"c.876A>G","HGVSp":"p.=","HGVSp_Short":"p.S292=","Transcript_ID":"ENST00000377970","Exon_Number":"3/3","t_depth":36,"t_ref_count":26,"t_alt_count":10,"n_depth":64,"n_ref_count":64,"n_alt_count":0,"Mutation_Class":"Inframe","AA_Position":292},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128752767,"End_Position":128752767,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"G","dbSNP_RS":"novel","Tumor_Sample_Barcode":"SP116688","Matched_Norm_Sample_Barcode":"SP116687","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.928C>G","HGVSp":"p.Leu310Val","HGVSp_Short":"p.L310V","Transcript_ID":"ENST00000377970","Exon_Number":"3/3","t_depth":36,"t_ref_count":12,"t_alt_count":24,"n_depth":41,"n_ref_count":41,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":310},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128752782,"End_Position":128752782,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"T","Tumor_Seq_Allele1":"T","Tumor_Seq_Allele2":"A","dbSNP_RS":"novel","Tumor_Sample_Barcode":"SP116688","Matched_Norm_Sample_Barcode":"SP116687","Match_Norm_Seq_Allele1":"T","Match_Norm_Seq_Allele2":"T","HGVSc":"c.943T>A","HGVSp":"p.Cys315Ser","HGVSp_Short":"p.C315S","Transcript_ID":"ENST00000377970","Exon_Number":"3/3","t_depth":36,"t_ref_count":14,"t_alt_count":22,"n_depth":38,"n_ref_count":38,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":315},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128752791,"End_Position":128752791,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"T","Tumor_Seq_Allele1":"T","Tumor_Seq_Allele2":"G","dbSNP_RS":"novel","Tumor_Sample_Barcode":"SP116688","Matched_Norm_Sample_Barcode":"SP116687","Match_Norm_Seq_Allele1":"T","Match_Norm_Seq_Allele2":"T","HGVSc":"c.952T>G","HGVSp":"p.Ser318Ala","HGVSp_Short":"p.S318A","Transcript_ID":"ENST00000377970","Exon_Number":"3/3","t_depth":35,"t_ref_count":12,"t_alt_count":23,"n_depth":36,"n_ref_count":36,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":318},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128752808,"End_Position":128752808,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"G","dbSNP_RS":"novel","Tumor_Sample_Barcode":"16-12281_tumorB","Matched_Norm_Sample_Barcode":"16-12281_normal","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.969C>G","HGVSp":"p.Asn323Lys","HGVSp_Short":"p.N323K","Transcript_ID":"ENST00000377970","Exon_Number":"3/3","t_depth":56,"t_ref_count":33,"t_alt_count":23,"n_depth":35,"n_ref_count":35,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":323},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128752954,"End_Position":128752954,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"G","Tumor_Seq_Allele1":"G","Tumor_Seq_Allele2":"A","dbSNP_RS":"novel","Tumor_Sample_Barcode":"HTMCP-01-06-00634-01A-01D","Matched_Norm_Sample_Barcode":"HTMCP-01-06-00634-10A-01D","Match_Norm_Seq_Allele1":"G","Match_Norm_Seq_Allele2":"G","HGVSc":"c.1115G>A","HGVSp":"p.Arg372Gln","HGVSp_Short":"p.R372Q","Transcript_ID":"ENST00000377970","Exon_Number":"3/3","t_depth":64,"t_ref_count":44,"t_alt_count":20,"n_depth":34,"n_ref_count":34,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":372},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128752960,"End_Position":128752960,"Strand":"+","Variant_Classification":"Missense_Mutation","Variant_Type":"SNP","Reference_Allele":"A","Tumor_Seq_Allele1":"A","Tumor_Seq_Allele2":"G","dbSNP_RS":"","Tumor_Sample_Barcode":"SP192870","Matched_Norm_Sample_Barcode":"SP192864","Match_Norm_Seq_Allele1":"A","Match_Norm_Seq_Allele2":"A","HGVSc":"c.1121A>G","HGVSp":"p.His374Arg","HGVSp_Short":"p.H374R","Transcript_ID":"ENST00000377970","Exon_Number":"3/3","t_depth":39,"t_ref_count":18,"t_alt_count":21,"n_depth":48,"n_ref_count":48,"n_alt_count":0,"Mutation_Class":"Missense","AA_Position":374},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128753183,"End_Position":128753183,"Strand":"+","Variant_Classification":"Silent","Variant_Type":"SNP","Reference_Allele":"G","Tumor_Seq_Allele1":"G","Tumor_Seq_Allele2":"A","dbSNP_RS":"novel","Tumor_Sample_Barcode":"SP116627","Matched_Norm_Sample_Barcode":"SP116626","Match_Norm_Seq_Allele1":"G","Match_Norm_Seq_Allele2":"G","HGVSc":"c.1344G>A","HGVSp":"p.=","HGVSp_Short":"p.Q448=","Transcript_ID":"ENST00000377970","Exon_Number":"3/3","t_depth":27,"t_ref_count":19,"t_alt_count":8,"n_depth":31,"n_ref_count":31,"n_alt_count":0,"Mutation_Class":"Inframe","AA_Position":448},{"Hugo_Symbol":"MYC","Entrez_Gene_Id":0,"Center":".","NCBI_Build":"GRCh37","Chromosome":"8","Start_Position":128753184,"End_Position":128753184,"Strand":"+","Variant_Classification":"Silent","Variant_Type":"SNP","Reference_Allele":"C","Tumor_Seq_Allele1":"C","Tumor_Seq_Allele2":"T","dbSNP_RS":"novel","Tumor_Sample_Barcode":"HTMCP-01-15-00366-01A-01E","Matched_Norm_Sample_Barcode":"HTMCP-01-15-00366-10A-01D","Match_Norm_Seq_Allele1":"C","Match_Norm_Seq_Allele2":"C","HGVSc":"c.1345C>T","HGVSp":"p.=","HGVSp_Short":"p.L449=","Transcript_ID":"ENST00000377970","Exon_Number":"3/3","t_depth":58,"t_ref_count":50,"t_alt_count":8,"n_depth":29,"n_ref_count":29,"n_alt_count":0,"Mutation_Class":"Inframe","AA_Position":449}],"snvDataFormat":{"x":"AA_Position","y":"HGVSp_Short","factor":"Mutation_Class"},"plotSettings":{"chartWidth":800,"chartType":"pie","chartMargin":{"left":40,"right":20,"top":15,"bottom":25},"chartBackground":"transparent","transitionTime":600,"yAxisLabel":"# of mutations","axisLabelFont":"normal 12px Arial","axisLabelColor":"#4f4f4f","axisLabelAlignment":"middle","axisLabelDy":"-2em","yAxisLineColor":"#c4c8ca","yAxisLineStyle":"dash","yAxisLineWidth":1,"yMaxRangeRatio":1.1,"legendMargin":{"left":10,"right":0,"top":5,"bottom":5},"legendInteractive":true,"legendTitle":"Mutation_Class","lollipopTrackHeight":360,"lollipopTrackBackground":"#F5F5DC","lollipopPopMinSize":2,"lollipopPopMaxSize":12,"lollipopPopInfoLimit":8,"lollipopPopInfoColor":"#EEE","lollipopPopInfoDy":"0.35em","lollipopLineColor":"rgb(42,42,42)","lollipopLineWidth":0.5,"lollipopCircleColor":"#b2b3b4","lollipopCircleWidth":1,"lollipopLabelRatio":1.4,"lollipopLabelMinFontSize":10,"lollipopColorScheme":"accent","titleText":"Mutation data for MYC","titleFont":"normal 16px Arial","titleColor":"#424242","titleAlignment":"middle","titleDy":"0.35em","highlightTextAngle":"90","annoHeight":30,"annoMargin":{"top":2,"bottom":0},"annoBackground":"#F5F5DC","annoBarFill":"grey","annoBarMargin":{"top":8,"bottom":8},"domainColorScheme":"dark2","domainMargin":{"top":0,"bottom":0},"domainTextFont":"italic 18px Serif","domainTextColor":"white","brush":true,"brushBackground":"transparent","brushOpacity":0.2,"brushBorderColor":"#808000","brushBorderWidth":1,"brushHandler":"#808000","legend":true,"tooltip":true,"zoom":true},"pngButton":true,"svgButton":true,"btnStyle":null,"outputFN":"my_lollipop_plot_MYC"},"evals":[],"jsHooks":[]}</script>
</div>
<div class="section level4">
<h4 id="mutation-frequency-bin-matrix">Mutation Frequency Bin Matrix<a class="anchor" aria-label="anchor" href="#mutation-frequency-bin-matrix"></a>
</h4>
<p><code>get_mutation_frequency_bin_matrix</code> counts the number of mutations in a sliding window across a region for all samples (in the provided metadata parameter) and clusters the data.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#load metadata.</span></span>
<span><span class="va">dlbcl_bl_meta</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pathology</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DLBCL"</span>,</span>
<span>                                   <span class="st">"BL"</span><span class="op">)</span><span class="op">)</span> <span class="co">#subset on specific pathology.</span></span>
<span></span>
<span><span class="co">#bring together all derived sample-level results from many GAMBL pipelines.</span></span>
<span><span class="va">dlbcl_bl_meta</span> <span class="op">=</span> <span class="fu"><a href="../reference/collate_results.html">collate_results</a></span><span class="op">(</span>join_with_full_metadata <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                these_samples_metadata <span class="op">=</span> <span class="va">dlbcl_bl_meta</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">genetic_subgroup</span> <span class="op">!=</span> <span class="st">"0"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## /projects/nhl_meta_analysis_scratch/gambl/results_local/shared/gambl_genome_results.tsv</span></span></code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#get ashm regions (grch37) and filter out a set of genes.</span></span>
<span><span class="va">some_regions</span> <span class="op">=</span> <span class="fu">GAMBLR.data</span><span class="fu">::</span><span class="va">somatic_hypermutation_locations_GRCh37_v_latest</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="va">gene</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"BCL11A"</span>, <span class="st">"CD44"</span>, <span class="st">"CD74"</span>, <span class="st">"IRF1"</span>, <span class="st">"IRF4"</span>, </span>
<span>                             <span class="st">"PTPN1"</span>, <span class="st">"KLHL6"</span>, <span class="st">"ZCCHC7"</span>, <span class="st">"ITPKB"</span>, <span class="st">"IL4R"</span>, </span>
<span>                             <span class="st">"NCOA3"</span>, <span class="st">"MYO1E"</span>, <span class="st">"ETV6"</span>, <span class="st">"FOXP1"</span>, <span class="st">"MEF2C"</span>, </span>
<span>                             <span class="st">"MS4A1"</span>, <span class="st">"SEPT9"</span>, <span class="st">"P2RX5"</span>, <span class="st">"BCL2L11"</span>, <span class="st">"POU2AF1"</span>,</span>
<span>                             <span class="st">"RFTN1"</span>, <span class="st">"C13orf18"</span>, <span class="st">"MIR155HG"</span>, <span class="st">"EBF1"</span>, <span class="st">"LRMP"</span>, </span>
<span>                             <span class="st">"BLK"</span>, <span class="st">"NFKBIZ"</span>, <span class="st">"BCL2"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">select</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#reorder factor levels of genetic subgroups.</span></span>
<span><span class="va">some_regions</span><span class="op">$</span><span class="va">gene</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">some_regions</span><span class="op">$</span><span class="va">gene</span>, </span>
<span>                           levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"BACH2"</span>, <span class="st">"TCL1A"</span>, <span class="st">"MYC"</span>, <span class="st">"PAX5"</span>,</span>
<span>                                      <span class="st">"ST6GAL1"</span>, <span class="st">"BCL6"</span>, <span class="st">"BCL7A"</span>, <span class="st">"BIRC3"</span>,</span>
<span>                                      <span class="st">"BTG1"</span>, <span class="st">"BTG2"</span>, <span class="st">"CD83"</span>, <span class="st">"CIITA"</span>, </span>
<span>                                      <span class="st">"CXCR4"</span>, <span class="st">"DTX1"</span>, <span class="st">"FANK1"</span>, <span class="st">"IRF8"</span>,</span>
<span>                                      <span class="st">"PIM1"</span>, <span class="st">"PIM2"</span>, <span class="st">"S1PR2"</span>, <span class="st">"SERPINA9"</span>,</span>
<span>                                      <span class="st">"TMSB4X"</span>, <span class="st">"ZFP36L1"</span>, <span class="st">"RHOH"</span>, <span class="st">"MEF2B"</span>,</span>
<span>                                      <span class="st">"GRHPR"</span>, <span class="st">"LPP"</span>, <span class="st">"SGK1"</span>, <span class="st">"IGLL5"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#reorder factor levels of genetic subgroups.</span></span>
<span><span class="va">dlbcl_bl_meta</span><span class="op">$</span><span class="va">genetic_subgroup</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">dlbcl_bl_meta</span><span class="op">$</span><span class="va">genetic_subgroup</span>, </span>
<span>                                        levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"M53-BL"</span>, <span class="st">"DGG-BL"</span>, </span>
<span>                                                   <span class="st">"IC-BL"</span>, <span class="st">"DLBCL-1"</span>,</span>
<span>                                                   <span class="st">"DLBCL-2"</span>, <span class="st">"DLBCL-3"</span>, </span>
<span>                                                   <span class="st">"ST2"</span>, <span class="st">"ST2-COMP"</span>,</span>
<span>                                                   <span class="st">"EZB"</span>, <span class="st">"EZB-COMP"</span>, </span>
<span>                                                   <span class="st">"BN2"</span>, <span class="st">"MCD"</span>,</span>
<span>                                                   <span class="st">"MCD-COMP"</span>, <span class="st">"N1-COMP"</span>, </span>
<span>                                                   <span class="st">"N1"</span>, <span class="st">"Other"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/get_mutation_frequency_bin_matrix.html">get_mutation_frequency_bin_matrix</a></span><span class="op">(</span>these_samples_metadata <span class="op">=</span> <span class="va">dlbcl_bl_meta</span>,</span>
<span>                                  regions_df <span class="op">=</span> <span class="va">some_regions</span>,</span>
<span>                                  min_bin_recurrence <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                                  region_fontsize <span class="op">=</span> <span class="fl">12</span>,</span>
<span>                                  window_size <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>                                  slide_by <span class="op">=</span> <span class="fl">500</span>,</span>
<span>                                  orientation <span class="op">=</span> <span class="st">"sample_rows"</span>,</span>
<span>                                  min_count_per_bin <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                                  sortByColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pathology"</span>, </span>
<span>                                                    <span class="st">"genetic_subgroup"</span>, </span>
<span>                                                    <span class="st">"manta_MYC_sv"</span><span class="op">)</span>,</span>
<span>                                  metadataColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"manta_MYC_sv"</span>, </span>
<span>                                                      <span class="st">"genetic_subgroup"</span>, </span>
<span>                                                      <span class="st">"pathology"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="ssm_tutorial_files/figure-html/get_mutation_frequency_bin_matrix-1.png" width="2700"></p>
</div>
<div class="section level4">
<h4 id="oncoplots">Oncoplots<a class="anchor" aria-label="anchor" href="#oncoplots"></a>
</h4>
<p>Commonly, we will want to generate visualizations that only show genes deemed relevant for a given malignancy (and above some frequency threshold). For this purpose we have <code>prettyOncoplot</code> available. In the example bellow, we are looking at all coding SSM (genome samples) and the relevant genes for two malignancies (BL and DLBCL). This plotting function has a large collection of parameters that are available for further customization and manipulation of the returned plot. Please refer to the documentation for a complete explanation of each parameter.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load the master merged MAF (coding only). It's usually more efficient to do this than to try to add filters to this query. Instead, just filter the data afterward</span></span>
<span><span class="va">maf_data</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_coding_ssm.html">get_coding_ssm</a></span><span class="op">(</span>seq_type <span class="op">=</span> <span class="st">"genome"</span><span class="op">)</span></span>
<span><span class="va">maf_metadata</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">maf</span> <span class="op">=</span> <span class="fu">read.maf</span><span class="op">(</span><span class="va">maf_data</span>, clinicalData <span class="op">=</span> <span class="va">maf_metadata</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## -Validating</span></span>
<span><span class="co">## --Removed 1078 duplicated variants</span></span>
<span><span class="co">## -Silent variants: 65357 </span></span>
<span><span class="co">## -Summarizing</span></span>
<span><span class="co">## --Possible FLAGS among top ten genes:</span></span>
<span><span class="co">##   TTN</span></span>
<span><span class="co">## -Processing clinical data</span></span>
<span><span class="co">## -Finished in 28.4s elapsed (57.4s cpu)</span></span></code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#define some genes of interest</span></span>
<span><span class="va">bl_genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"NFKBIZ"</span>, <span class="st">"ID3"</span>, <span class="st">"TP53"</span>, <span class="st">"ARID1A"</span>, <span class="st">"FBXO11"</span>, </span>
<span>             <span class="st">"GNA13"</span>, <span class="st">"TCF3"</span>, <span class="st">"TFAP4"</span>, <span class="st">"HNRNPU"</span>, <span class="st">"FOXO1"</span>,</span>
<span>             <span class="st">"CCND3"</span>, <span class="st">"SMARCA4"</span>, <span class="st">"DDX3X"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dlbcl_genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EZH2"</span>, <span class="st">"KMT2D"</span>, <span class="st">"MEF2B"</span>, <span class="st">"CREBBP"</span>, <span class="st">"MYD88"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">bl_genes</span>, <span class="va">dlbcl_genes</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#define gene groups</span></span>
<span><span class="va">gene_groups</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"BL"</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">bl_genes</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"DLBCL"</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">dlbcl_genes</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gene_groups</span><span class="op">)</span> <span class="op">=</span> <span class="va">genes</span></span>
<span></span>
<span><span class="co">#filter metadata</span></span>
<span><span class="va">maf_metadata</span> <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">maf_metadata</span>,<span class="op">!</span><span class="va">lymphgen</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"COMPOSITE"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#convert metadata column into factor</span></span>
<span><span class="va">maf_metadata</span><span class="op">$</span><span class="va">pathology</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">maf_metadata</span><span class="op">$</span><span class="va">pathology</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#define order of factors for selected metadata column</span></span>
<span><span class="va">maf_metadata</span><span class="op">$</span><span class="va">pathology</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">maf_metadata</span><span class="op">$</span><span class="va">pathology</span>, </span>
<span>                                levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DLBCL"</span>, <span class="st">"BL"</span>,</span>
<span>                                           <span class="st">"B-ALL"</span>, <span class="st">"CLL"</span>,</span>
<span>                                           <span class="st">"COMFL"</span>, <span class="st">"DLBCL-BL-like"</span>,</span>
<span>                                           <span class="st">"FL"</span>, <span class="st">"HGBL"</span>,</span>
<span>                                           <span class="st">"MCL"</span>, <span class="st">"PBL"</span>,</span>
<span>                                           <span class="st">"SCBC"</span>, <span class="st">"UNSPECIFIED"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">maf_metadata</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">maf_metadata</span>, <span class="va">maf_metadata</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">pathology</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#create prettyOncoplot</span></span>
<span><span class="fu"><a href="../reference/prettyOncoplot.html">prettyOncoplot</a></span><span class="op">(</span>maftools_obj <span class="op">=</span> <span class="va">maf</span>,</span>
<span>               genes <span class="op">=</span> <span class="va">genes</span>,</span>
<span>               these_samples_metadata <span class="op">=</span> <span class="va">maf_metadata</span>,</span>
<span>               splitGeneGroups <span class="op">=</span> <span class="va">gene_groups</span>,</span>
<span>               keepGeneOrder <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>               splitColumnName <span class="op">=</span> <span class="st">"pathology"</span>,</span>
<span>               metadataBarHeight <span class="op">=</span> <span class="fl">5</span>,</span>
<span>               metadataBarFontsize <span class="op">=</span> <span class="fl">8</span>,</span>
<span>               fontSizeGene <span class="op">=</span> <span class="fl">11</span>,</span>
<span>               metadataColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pathology"</span>, <span class="st">"lymphgen"</span>, <span class="st">"sex"</span>, <span class="st">"EBV_status_inf"</span>, <span class="st">"cohort"</span><span class="op">)</span>,</span>
<span>               sortByColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pathology"</span>, <span class="st">"lymphgen"</span>, <span class="st">"sex"</span>, <span class="st">"EBV_status_inf"</span>, <span class="st">"cohort"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "numcases: 1646"</span></span>
<span><span class="co">## [1] "numgenes: 18"</span></span></code></pre>
<p><img src="ssm_tutorial_files/figure-html/pretty_oncoplot_1-1.png" width="2700"></p>
<p>As previous stated, <code>prettyOncoplot</code> has many parameters that can be called to further customize the plot and aggregate the data that are to be plotted. In reality, if all features of this plotting functions were to be demonstrated, a dedicated vignette for this purpose might be the way to go. Even so, here is another example were we include non coding regions (3â€™ UTR, NFKBIZ).</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#get SSM for NFKBIZ UTR region</span></span>
<span><span class="va">nfkbiz_utr_region</span> <span class="op">=</span> <span class="st">"chr3:101,578,185-101,579,902"</span></span>
<span><span class="va">nfkbiz_ssm</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_ssm_by_region.html">get_ssm_by_region</a></span><span class="op">(</span>region <span class="op">=</span> <span class="va">nfkbiz_utr_region</span>, seq_type <span class="op">=</span> <span class="st">"genome"</span>, projection <span class="op">=</span> <span class="st">"grch37"</span>, basic_columns <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#rbind nfkbiz SSM with big mutation data frame</span></span>
<span><span class="va">maf_data</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">maf_data</span>, <span class="va">nfkbiz_ssm</span><span class="op">)</span></span>
<span><span class="va">maf</span> <span class="op">=</span> <span class="fu">read.maf</span><span class="op">(</span><span class="va">maf_data</span>, clinicalData <span class="op">=</span> <span class="va">maf_metadata</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## -Validating</span></span>
<span><span class="co">## --Removed 1079 duplicated variants</span></span>
<span><span class="co">## -Silent variants: 65404 </span></span>
<span><span class="co">## -Summarizing</span></span>
<span><span class="co">## --Possible FLAGS among top ten genes:</span></span>
<span><span class="co">##   TTN</span></span>
<span><span class="co">## -Processing clinical data</span></span>
<span><span class="co">## -Finished in 28.3s elapsed (54.3s cpu)</span></span></code></pre>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#create prettyOncoplot</span></span>
<span><span class="fu"><a href="../reference/prettyOncoplot.html">prettyOncoplot</a></span><span class="op">(</span>maftools_obj <span class="op">=</span> <span class="va">maf</span>,</span>
<span>               genes <span class="op">=</span> <span class="va">genes</span>,</span>
<span>               these_samples_metadata <span class="op">=</span> <span class="va">maf_metadata</span>,</span>
<span>               include_noncoding<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"NFKBIZ"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"3'UTR"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>               splitGeneGroups <span class="op">=</span> <span class="va">gene_groups</span>,</span>
<span>               keepGeneOrder <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>               splitColumnName <span class="op">=</span> <span class="st">"pathology"</span>,</span>
<span>               metadataBarHeight <span class="op">=</span> <span class="fl">5</span>,</span>
<span>               metadataBarFontsize <span class="op">=</span> <span class="fl">8</span>,</span>
<span>               fontSizeGene <span class="op">=</span> <span class="fl">11</span>,</span>
<span>               metadataColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pathology"</span>, <span class="st">"lymphgen"</span>, <span class="st">"sex"</span>, <span class="st">"EBV_status_inf"</span>, <span class="st">"cohort"</span><span class="op">)</span>,</span>
<span>               sortByColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pathology"</span>, <span class="st">"lymphgen"</span>, <span class="st">"sex"</span>, <span class="st">"EBV_status_inf"</span>, <span class="st">"cohort"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "numcases: 1646"</span></span>
<span><span class="co">## [1] "numgenes: 18"</span></span></code></pre>
<p><img src="ssm_tutorial_files/figure-html/pretty_oncoplot_2-1.png" width="2700"></p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Ryan Morin, Kostia Dreval, Adam Mattsson.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
