<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Utilities Tutorial • GAMBLR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Utilities Tutorial">
<meta property="og:description" content="GAMBLR">
<meta property="og:image" content="https://github.com/morinlab/GAMBLR/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-12345678-0"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-12345678-0');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9500</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-light fa-memo-circle-info"></span>
     
    How To
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="https://morinlab.github.io/GAMBLR/#installation" class="external-link">Install</a>
    </li>
    <li>
      <a href="https://morinlab.github.io/GAMBLR/#run-remote-on-a-local-machine" class="external-link">Run Remote On A Local Machine</a>
    </li>
    <li>
      <a href="https://morinlab.github.io/GAMBLR/#contributing" class="external-link">Contribute</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-light fa-person-chalkboard"></span>
     
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/ssm_tutorial.html">SSM</a>
    </li>
    <li>
      <a href="../articles/cn_sv_tutorial.html">CN &amp; SV</a>
    </li>
    <li>
      <a href="../articles/fancy_plots.html">Fancy Plots</a>
    </li>
    <li>
      <a href="../articles/utilities_tutorial.html">Utilities</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="../reference/index.html">
    <span class="fa fa-file-code-o"></span>
     
    Functions &amp; Bundled Data
  </a>
</li>
<li>
  <a href="https://github.com/morinlab/GAMBLR">
    <span class="fa fa-github fa-lg"></span>
     
    github
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Utilities Tutorial</h1>
            
      
      
      <div class="hidden name"><code>utilities_tutorial.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="utilities">Utilities<a class="anchor" aria-label="anchor" href="#utilities"></a>
</h2>
<p>In this vignette we are exploring functions that fall under the category, utilities. This is a collection of functions that can be used for a variety of applications, such as retrieving gene coordinates from regions (and the other way around), cleaning and sanitizing data, creating custom tracks for the UCSC Genome Browser, getting collated results, accessing the GAMBLR colour paletter, etc.</p>
<div class="section level3">
<h3 id="genes-regions-and-statistical-computing">Genes, Regions and Statistical Computing<a class="anchor" aria-label="anchor" href="#genes-regions-and-statistical-computing"></a>
</h3>
<p>This section demonstrates how to return information related to genes and other specified genomic regions</p>
<div class="section level4">
<h4 id="gene-to-region">Gene To Region<a class="anchor" aria-label="anchor" href="#gene-to-region"></a>
</h4>
<p>If you have a set of genes and you are interested in returning the genomic coordinates for these genes, it is recommended to call <code>gene_to_region</code>. This function takes four parameters; <strong>gene_symbol</strong>, a vector of on or more gene symbols in Hugo format. <strong>ensembl_id</strong>, a vector of one or more Ensembl IDs. <strong>genome_build</strong>, the desired projection for returned regions, acceptable values are hg38 and grch37. <strong>return_as</strong>, the format of returned regions, acceptable values are (1) <strong>region</strong> (a string formatted as chr:start-end), (2) <strong>bed</strong> and (3) <strong>df</strong>. Let’s have a look at a few different examples, using a few different parameter combinations.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#gene to region, for a single gene returned as "region"</span></span>
<span><span class="va">myc_region</span> <span class="op">=</span> <span class="fu"><a href="../reference/gene_to_region.html">gene_to_region</a></span><span class="op">(</span>gene_symbol <span class="op">=</span> <span class="st">"MYC"</span>, genome_build <span class="op">=</span> <span class="st">"grch37"</span>, return_as <span class="op">=</span> <span class="st">"region"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#print region</span></span>
<span><span class="va">myc_region</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "8:128747680-128753674"</span></span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#gene to region, for multiple genes (MYC and BCL2) in a vector of characters, returned in bed format.</span></span>
<span><span class="va">genes_bed</span> <span class="op">=</span> <span class="fu"><a href="../reference/gene_to_region.html">gene_to_region</a></span><span class="op">(</span>ensembl_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ENSG00000136997"</span>, <span class="st">"ENSG00000171791"</span><span class="op">)</span>, genome_build <span class="op">=</span> <span class="st">"hg38"</span>, return_as <span class="op">=</span> <span class="st">"bed"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#print regions</span></span>
<span><span class="va">genes_bed</span></span></code></pre></div>
<pre><code><span><span class="co">##   chromosome     start       end hugo_symbol</span></span>
<span><span class="co">## 1      chr18  63123345  63320128        BCL2</span></span>
<span><span class="co">## 2       chr8 127735433 127742951         MYC</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#gene to region, for multiple genes specified in an already subset list of gene names, return format is data frame</span></span>
<span><span class="co">#subset FL genes</span></span>
<span><span class="va">fl_genes</span> <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">lymphoma_genes</span>, <span class="va">FL</span> <span class="op">==</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">Gene</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#get regions for Fl genes</span></span>
<span><span class="va">fl_genes_df</span> <span class="op">=</span> <span class="fu"><a href="../reference/gene_to_region.html">gene_to_region</a></span><span class="op">(</span>gene_symbol <span class="op">=</span> <span class="va">fl_genes</span>, genome_build <span class="op">=</span> <span class="st">"grch37"</span>, return_as <span class="op">=</span> <span class="st">"df"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="region-to-gene">Region To Gene<a class="anchor" aria-label="anchor" href="#region-to-gene"></a>
</h4>
<p>Similarly, we can also return genes that are residing within a specified region. For this purpose <code>region_to_gene</code> was developed. This function operates in a similar way as the previous described function. Three parameters are available for this function; <strong>region</strong>, regions to intersect genes with, this can be either a data frame with regions sorted in the following columns; chromosome, start, end. Or it can be a character vector in “region” format, i.e chromosome:start-end. <strong>gene_format</strong>, parameter for specifying the format of returned genes, default is “hugo”, other acceptable inputs are “ensembl”. <strong>genome_build</strong>, the desired projection for returned regions, acceptable values are hg38 and grch37. Here are a few examples showcasing how the different parameters can be used, as well as examples on how to make the most out of this function.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#simple example using a previously defined region (MYC)</span></span>
<span><span class="va">genes_in_this_region</span> <span class="op">=</span> <span class="fu"><a href="../reference/region_to_gene.html">region_to_gene</a></span><span class="op">(</span>region <span class="op">=</span> <span class="va">myc_region</span>, gene_format <span class="op">=</span> <span class="st">"hugo"</span>, genome_build <span class="op">=</span> <span class="st">"grch37"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#print genes</span></span>
<span><span class="va">genes_in_this_region</span></span></code></pre></div>
<pre><code><span><span class="co">##   chromosome     start       end hugo_symbol region_start region_end</span></span>
<span><span class="co">## 1       chr8 128747680 128753674         MYC    128747680  128753674</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#wider example using larger region (q-arm, chromosome 1) and using another genome build (hg38) and gene format (ensembl)</span></span>
<span><span class="va">chr1q_genes</span> <span class="op">=</span> <span class="fu"><a href="../reference/region_to_gene.html">region_to_gene</a></span><span class="op">(</span>region <span class="op">=</span> <span class="st">"chr1:143200000-248956422"</span>, gene_format <span class="op">=</span> <span class="st">"ensembl"</span>, genome_build <span class="op">=</span> <span class="st">"hg38"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="regions-to-bins">Regions To Bins<a class="anchor" aria-label="anchor" href="#regions-to-bins"></a>
</h4>
<p>This function was developed to split a contiguous genomic region on a chromosome into non-overlapping bins. This function takes four parameter; <strong>chr</strong>, <strong>start</strong>, <strong>end</strong>, and <strong>bin_size</strong>. In this example we are splitting the q-arm on chromosome 8 into bins of 20000 base pairs.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#regions to bin, bin chromosome 8 into 20000 bins</span></span>
<span><span class="va">chr8q_bins</span> <span class="op">=</span> <span class="fu"><a href="../reference/region_to_bins.html">region_to_bins</a></span><span class="op">(</span>chromosome <span class="op">=</span> <span class="st">"8"</span>, start <span class="op">=</span> <span class="fl">48100000</span>, end <span class="op">=</span> <span class="fl">146364022</span>, bin_size <span class="op">=</span> <span class="fl">20000</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="count-ssm-by-region">Count SSM by Region<a class="anchor" aria-label="anchor" href="#count-ssm-by-region"></a>
</h4>
<p>Sometimes it can be helpful to tally all mutations (SSM) in a region for a set number of samples. To do so, one can call <code>count_ssm_by_region</code>. This function takes genomic coordinates either specified in region format (chr:start-end) with the <code>region</code> parameter, or the chromosome, start and end can be specified individually with the <code>chr</code>, <code>start</code>, and <code>end</code> parameters. In this example we are first creating a metadata subset, only including samples with FL as the pathology. Next, we are defining a region for MYC with help from <code>gene_to_region</code> with he return format specified as “region”. The return is the sum of all SSMs overlapping with MYC for all FL samples.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#define a region.</span></span>
<span><span class="va">my_region</span> <span class="op">=</span> <span class="fu"><a href="../reference/gene_to_region.html">gene_to_region</a></span><span class="op">(</span>gene_symbol <span class="op">=</span> <span class="st">"MYC"</span>, return_as <span class="op">=</span> <span class="st">"region"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#subset metadata.</span></span>
<span><span class="va">my_metadata</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pathology</span> <span class="op">==</span> <span class="st">"FL"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#count SSMs for the selected sample subset and defined region.</span></span>
<span><span class="va">fl_ssm_counts_myc</span> <span class="op">=</span> <span class="fu"><a href="../reference/count_ssm_by_region.html">count_ssm_by_region</a></span><span class="op">(</span>region <span class="op">=</span> <span class="va">my_region</span>, these_samples_metadata <span class="op">=</span> <span class="va">my_metadata</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#print number of SSMs in the defined region for the selected samples</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">fl_ssm_counts_myc</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 12</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="coding-ssm-status">Coding SSM Status<a class="anchor" aria-label="anchor" href="#coding-ssm-status"></a>
</h4>
<p>Another way to summarize SSM across samples for coding regions is with <code>get_coding_ssm_status</code>. This function takes a list of gene symbols and subsets the incoming MAF to specified genes. If no gene list is provided, the function will default to all lymphoma genes. The function can accept a wide range of incoming MAFs. For example, the user can call this function with <code>these_samples_metadata</code> (preferably a metadata table that has been subset to the sample IDs of interest). If this parameter is not called, the function will default to all samples available with <code>get_gambl_metadata</code>. The user can also provide a path to a MAF, or MAF-like file with <code>maf_path</code>, or an already loaded MAF can be used with the <code>maf_data</code> parameter. If both <code>maf_path</code> and <code>maf_data</code> is missing, the function will default to run <code>get_coding_ssm</code>. In this example, we are returning the mutation status for FL samples (using the already subset metadata table from the previous example) in respect to MYC.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#tabulate coding SSMs for FL samples.</span></span>
<span><span class="va">coding_tabulated_df</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_coding_ssm_status.html">get_coding_ssm_status</a></span><span class="op">(</span>these_samples_metadata <span class="op">=</span> <span class="va">my_metadata</span>, review_hotspots <span class="op">=</span> <span class="cn">FALSE</span>, include_hotspots <span class="op">=</span> <span class="cn">FALSE</span>, gene_symbols <span class="op">=</span> <span class="st">"MYC"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="calculate-proportion-of-genome-altered-by-cnv">Calculate Proportion of Genome Altered by CNV<a class="anchor" aria-label="anchor" href="#calculate-proportion-of-genome-altered-by-cnv"></a>
</h4>
<p>Interested to see the percentage of a genome altered by CNVs for specific sample, or sample subsets? For this purpose <code>calcualte_pga</code> was developed. This function takes into account the total length of sample’s CNV and relates it to the total genome length to return the proportion affected by CNV. The function can work with either individual or multi-sample seg files. The telomeres are always excluded from calculation, and centromeres/sex chromosomes can be optionally included or excluded.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#get CN segments for a specific sample.</span></span>
<span><span class="va">sample_seg</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_sample_cn_segments.html">get_sample_cn_segments</a></span><span class="op">(</span>this_sample_id <span class="op">=</span> <span class="st">"14-36022T"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span><span class="st">"sample"</span> <span class="op">=</span> <span class="st">"ID"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#calculate PGA for the same sample.</span></span>
<span><span class="fu"><a href="../reference/calculate_pga.html">calculate_pga</a></span><span class="op">(</span>this_seg <span class="op">=</span> <span class="va">sample_seg</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   sample_id       PGA</span></span>
<span><span class="co">## 1 14-36022T 0.8705917</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="adjust-ploidy">Adjust Ploidy<a class="anchor" aria-label="anchor" href="#adjust-ploidy"></a>
</h4>
<p>The <code>adjust_ploidy</code> function returns a seg file with log.ratios adjusted to the overall sample ploidy. This function adjusts the ploidy of the sample using the percent of genome altered (PGA). The PGA is calculated internally, but can also be optionally provided as data frame if calculated from other sources. Only the samples above the threshold-provided PGA (<code>pga_cutoff</code>) will have ploidy adjusted. The function can work with either individual or multi-sample seg file.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#ajdust ploidy</span></span>
<span><span class="fu"><a href="../reference/adjust_ploidy.html">adjust_ploidy</a></span><span class="op">(</span>this_seg <span class="op">=</span> <span class="va">sample_seg</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 187 × 6</span></span></span>
<span><span class="co">##    sample    chrom     start       end LOH_flag log.ratio</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> 14-36022T 1         <span style="text-decoration: underline;">10</span>001    <span style="text-decoration: underline;">762</span>600        0         0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> 14-36022T 1        <span style="text-decoration: underline;">762</span>601 121<span style="text-decoration: underline;">500</span>000        0         0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> 14-36022T 1     142<span style="text-decoration: underline;">600</span>000 248<span style="text-decoration: underline;">277</span>662        0         0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> 14-36022T 1     248<span style="text-decoration: underline;">277</span>663 248<span style="text-decoration: underline;">278</span>622        0         0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> 14-36022T 1     248<span style="text-decoration: underline;">278</span>623 249<span style="text-decoration: underline;">226</span>346        0        -<span style="color: #BB0000;">1</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> 14-36022T 1     249<span style="text-decoration: underline;">226</span>347 249<span style="text-decoration: underline;">250</span>620        0         0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> 14-36022T 2         <span style="text-decoration: underline;">10</span>001     <span style="text-decoration: underline;">11</span>319        0         0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> 14-36022T 2         <span style="text-decoration: underline;">11</span>320  90<span style="text-decoration: underline;">500</span>000        0         0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> 14-36022T 2      96<span style="text-decoration: underline;">800</span>000 186<span style="text-decoration: underline;">704</span>965        0         0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> 14-36022T 2     186<span style="text-decoration: underline;">704</span>966 186<span style="text-decoration: underline;">712</span>276        0         0</span></span>
<span><span class="co">## <span style="color: #949494;"># … with 177 more rows</span></span></span></code></pre>
</div>
<div class="section level4">
<h4 id="compare-cnvs-for-multiple-samples">Compare CNVs For Multiple Samples<a class="anchor" aria-label="anchor" href="#compare-cnvs-for-multiple-samples"></a>
</h4>
<p><code>cnvKompare</code> will compare CNV data between samples with multiple time points. It can also handle same-sample comparison between different CNV callers if the sample ID is specified in unique fashion. For groups with more than 2 samples, optionally pairwise comparisons can be performed. The comparison is made based on the internally calculated score, which reflects the percentage of each cytoband covered by CNV (rounded to the nearest 5%) and its absolute CN. Optionally, the heatmap of <code>cnvKompare</code> scores can be returned. In addition, the function will return all concordant and discordant cytobands. Finally, the time series plot of CNV log ratios will be returned for all lymphoma genes, with further functionality to subset it to a panel of genes of interest.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#get CNV comparison data.</span></span>
<span><span class="va">kompare_out</span> <span class="op">=</span> <span class="fu"><a href="../reference/cnvKompare.html">cnvKompare</a></span><span class="op">(</span>patient_id <span class="op">=</span> <span class="st">"13-26835"</span>, </span>
<span>                         projection <span class="op">=</span> <span class="st">"hg38"</span>,</span>
<span>                         genes_of_interest <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EZH2"</span>,</span>
<span>                                               <span class="st">"TP53"</span>,</span>
<span>                                               <span class="st">"MYC"</span>,</span>
<span>                                               <span class="st">"CREBBP"</span>,</span>
<span>                                               <span class="st">"GNA13"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># concordance as percent</span></span>
<span><span class="va">kompare_out</span><span class="op">$</span><span class="va">overall_concordance_pct</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 12.12</span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#visualizations</span></span>
<span><span class="va">kompare_out</span><span class="op">$</span><span class="va">Heatmap</span></span></code></pre></div>
<p><img src="utilities_tutorial_files/figure-html/cnvKompare-1.png" width="700"></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kompare_out</span><span class="op">$</span><span class="va">time_plot</span></span></code></pre></div>
<p><img src="utilities_tutorial_files/figure-html/cnvKompare-2.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="genome-to-exome">Genome to Exome<a class="anchor" aria-label="anchor" href="#genome-to-exome"></a>
</h4>
<p>If you are interested to subset a maf to only features that would be available in Whole Genome Exome (WEX) data, <code>genome_to_exome</code> is available for this purpose. This function takes the following parameters; <code>maf</code>, the incoming maf object. Can be maf-like data frame or maftools maf object. Minimum columns that should be present are Chromosome, Start_Position, and End_Position. <code>custom_bed</code>, optional argument specifying a path to custom bed file for covered regions. Must be bed-like and contain chrom, start, and end position information in first 3 columns. Other columns are disregarded if provided. <code>genome_build</code>, string indicating genome build of the maf file. Default is grch37, but can accept modifications of both grch37- and hg38-based builds. <code>padding</code>, numeric value that will be used to pad probes in WEX data from both ends. Default is 100. After padding, overlapping features are squished together. let’s have a look at an example using this function.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#get all ssm in the MYC aSHM region</span></span>
<span><span class="va">myc_ashm_maf</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_ssm_by_region.html">get_ssm_by_region</a></span><span class="op">(</span>region <span class="op">=</span> <span class="st">"8:128748352-128749427"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#get mutations with 100 bp padding (default)</span></span>
<span><span class="va">wex</span> <span class="op">=</span> <span class="fu"><a href="../reference/genome_to_exome.html">genome_to_exome</a></span><span class="op">(</span>maf <span class="op">=</span> <span class="va">myc_ashm_maf</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="miscellaneous">Miscellaneous<a class="anchor" aria-label="anchor" href="#miscellaneous"></a>
</h3>
<div class="section level4">
<h4 id="collate-results">Collate Results<a class="anchor" aria-label="anchor" href="#collate-results"></a>
</h4>
<p>Another useful function for collecting all derived sample-level results from a variety of GAMBL pipelines is to call <code>collate_results</code>. This function takes a variety of parameters that will be explained here. First, <strong>sample_table</strong> which is a data frame with sample IDs in the first column. This parameter is used to control for what samples you are retrieving collated results for. Another option for controlling the same variable is to use <strong>these_samples_metadata</strong>. This parameter takes a metadata table (ideally subset to the sample IDs of interest). If this parameter is called, it overwrites the <code>get_gambl_metadata</code> statement for when the user wants to join with metadata. Meaning if this parameter is set to <code>TRUE</code> and no <strong>these_sample_metadata</strong> is provided, it will link the collated results with all samples in the unfiltered meta data table. Other parameters include <code>write_to_file</code>, this parameter is Boolean parameter and if set to TRUE, it writes the results to disk (default is <code>FALSE</code>). Another parameter that is good to be aware of is <code>from_cache</code>, which controls how the function gets the results. Default is <code>TRUE</code>. In most user-cases, these two parameters should be run with their respective default values. Let’s have a look at a few examples using this function.</p>
<pre class="collate_results"><code><span><span class="co">#collate all genome samples, using cached results</span></span>
<span><span class="va">genome_collated</span> <span class="op">=</span> <span class="fu"><a href="../reference/collate_results.html">collate_results</a></span><span class="op">(</span>seq_type_filter <span class="op">=</span> <span class="st">"genome"</span>, from_cache <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#use an already subset metadata table for getting collated results (cached)</span></span>
<span><span class="va">fl_metadata</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span>seq_type_filter <span class="op">=</span> <span class="st">"genome"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pathology</span> <span class="op">==</span> <span class="st">"FL"</span><span class="op">)</span></span>
<span><span class="va">fl_collated</span> <span class="op">=</span> <span class="fu"><a href="../reference/collate_results.html">collate_results</a></span><span class="op">(</span>seq_type_filter <span class="op">=</span> <span class="st">"genome"</span>, join_with_full_metadata <span class="op">=</span> <span class="cn">TRUE</span>, these_samples_metadata <span class="op">=</span> <span class="va">fl_metadata</span>, from_cache <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#get collated results (cached) for all genome samples and join with full metadata</span></span>
<span><span class="va">everything_collated</span> <span class="op">=</span> <span class="fu"><a href="../reference/collate_results.html">collate_results</a></span><span class="op">(</span>seq_type_filter <span class="op">=</span> <span class="st">"genome"</span>, from_cache <span class="op">=</span> <span class="cn">TRUE</span>, join_with_full_metadata <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#another example demonstrating correct usage of the sample_table parameter.</span></span>
<span><span class="va">fl_samples</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span>seq_type_filter <span class="op">=</span> <span class="st">"genome"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pathology</span> <span class="op">==</span> <span class="st">"FL"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">sample_id</span>, <span class="va">patient_id</span>, <span class="va">biopsy_id</span><span class="op">)</span></span>
<span><span class="va">fl_collated</span> <span class="op">=</span> <span class="fu"><a href="../reference/collate_results.html">collate_results</a></span><span class="op">(</span>sample_table <span class="op">=</span> <span class="va">fl_samples</span>, seq_type_filter <span class="op">=</span> <span class="st">"genome"</span>, from_cache <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="data-exports">Data Exports<a class="anchor" aria-label="anchor" href="#data-exports"></a>
</h3>
<div class="section level4">
<h4 id="sv-to-bedpe">SV To bedpe<a class="anchor" aria-label="anchor" href="#sv-to-bedpe"></a>
</h4>
<p>This function takes an input SV data frame and writes it to a bedpe formatted file, that is compatible with the IGV genome browser. This function is rather straight forward, it expects a data frame with variant calls, it also takes a parameter called <code>file</code> which is the path to the to-be exported file. In addition, we also have <code>add_chr_prefix</code> that controls if the contigs should be prefixed with “chr”, default is TRUE.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#get some SVs</span></span>
<span><span class="va">my_svs</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_manta_sv_by_sample.html">get_manta_sv_by_sample</a></span><span class="op">(</span>this_sample_id <span class="op">=</span> <span class="st">"191621-PL01"</span>, these_samples_metadata <span class="op">=</span> <span class="fu"><a href="../reference/get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span>, force_lift <span class="op">=</span> <span class="cn">TRUE</span>, projection <span class="op">=</span> <span class="st">"hg38"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#write to bedpe file</span></span>
<span><span class="va">my_bedpe</span> <span class="op">=</span> <span class="fu"><a href="../reference/sv_to_bedpe_file.html">sv_to_bedpe_file</a></span><span class="op">(</span>sv_df <span class="op">=</span> <span class="va">my_svs</span>, add_chr_prefix <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="sv-to-custom-track">SV To Custom Track<a class="anchor" aria-label="anchor" href="#sv-to-custom-track"></a>
</h4>
<p>Similarly, SVs can also be output as custom track for visualization on the UCSC genome Browser with <code>sv_to_custom_track</code>. This function takes an incoming SV data frame and outputs a bed file, ready for visualization on the UCSC Genome Browser. First, specify the output file with <code>output_file</code>. Next, indicate if the incoming SVs are annotated with <code>is_annotated</code> (default is TRUE). Lastly, the user can also specify if the incoming SV data frame should be subset to specific mutation types (e.g deletions, duplications, insertions, etc.) with the <code>sv_name</code> parameter. Default is to include all SV sub types.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#get SVs</span></span>
<span><span class="va">all_sv</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_combined_sv.html">get_combined_sv</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#create a custom track with selected SVs</span></span>
<span><span class="fu"><a href="../reference/sv_to_custom_track.html">sv_to_custom_track</a></span><span class="op">(</span><span class="va">annotated_sv</span>, output_file <span class="op">=</span> <span class="st">"GAMBL_sv_custom_track_annotated.bed"</span>, is_annotated <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="maf-to-custom-track">MAF To Custom Track<a class="anchor" aria-label="anchor" href="#maf-to-custom-track"></a>
</h4>
<p>In addition, we also have convenience functions available for generating custom tracks compatible with UCSC Genome Browser. This function offers a lot of flexibility in regards to the parameters at hand. For example, there are Boolean parameters dictating if output should be <em>bigbed</em> or <em>biglolly</em>. Parameters for specifying the custom fields of the output file are also available. Here is a straightforward example for output a maf data frame into a bed file for visualization on the Genome Browser.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#get some mutations</span></span>
<span><span class="va">my_mutations</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_ssm_by_region.html">get_ssm_by_region</a></span><span class="op">(</span>region <span class="op">=</span> <span class="st">"chr8:128723128-128774067"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#output mutation matrix to a bedfile</span></span>
<span><span class="va">my_bed</span> <span class="op">=</span> <span class="fu"><a href="../reference/maf_to_custom_track.html">maf_to_custom_track</a></span><span class="op">(</span>maf_data <span class="op">=</span> <span class="va">my_mutations</span>, colour_column <span class="op">=</span> <span class="st">"Tumor_Sample_Barcode"</span>, output_file <span class="op">=</span> <span class="st">"my_mutations.bed"</span>, track_name <span class="op">=</span> <span class="st">"My Mutations"</span>, track_description <span class="op">=</span> <span class="st">"MYC mutations"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="visualizations">Visualizations<a class="anchor" aria-label="anchor" href="#visualizations"></a>
</h3>
<div class="section level4">
<h4 id="gambl-colours">GAMBL Colours<a class="anchor" aria-label="anchor" href="#gambl-colours"></a>
</h4>
<p>For your convenience, there is a dedicated GAMBL colour palette available. Useful for annotating figures in a standardized way. This function holds custom colour palettes for a variety of different applications and data subsets. To access all available palettes, one should start with <code>get_gambl_colours</code>. This function takes a few different parameters. The most important being, <code>clasification</code>. This parameter allows the user to return a palette for specific classifications (e.g pathology, lymphgen, mutation or copy_number, etc.). In addition, this function also have an <code>alpha</code> parameter for controlling the opacity of the returned colours. Another useful parameter for formatting the return, is <code>as_list</code>, and <code>as_dataframe</code>. If you want to return all available classification types, call this function with <code>return_available = TRUE</code>. Let’s start with exploring a few of the different palettes, then let’s arrange all colour palettes into a ggplot object for easy overview.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#print all available classification columns</span></span>
<span><span class="fu"><a href="../reference/get_gambl_colours.html">get_gambl_colours</a></span><span class="op">(</span>return_available <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "seq_type"          "type"              "hmrn"             </span></span>
<span><span class="co">##  [4] "EBV"               "BL"                "FL"               </span></span>
<span><span class="co">##  [7] "lymphgenerator"    "chapuy_classifier" "lacy_classifier"  </span></span>
<span><span class="co">## [10] "lymphgen"          "mutation"          "rainfall"         </span></span>
<span><span class="co">## [13] "pos_neg"           "copy_number"       "blood"            </span></span>
<span><span class="co">## [16] "sex"               "clinical"          "pathology"        </span></span>
<span><span class="co">## [19] "coo"               "cohort"            "indels"           </span></span>
<span><span class="co">## [22] "svs"               "genetic_subgroup"</span></span></code></pre>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#get the lymphgen palette</span></span>
<span><span class="fu"><a href="../reference/get_gambl_colours.html">get_gambl_colours</a></span><span class="op">(</span>classification <span class="op">=</span> <span class="st">"lymphgen"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   EZB-MYC       EZB  EZB-COMP       ST2  ST2-COMP       MCD  MCD-COMP       BN2 </span></span>
<span><span class="co">## "#52000F" "#721F0F" "#C7371A" "#C41230" "#EC3251" "#3B5FAC" "#6787CB" "#7F3293" </span></span>
<span><span class="co">##  BN2-COMP        N1   N1-COMP       A53     Other COMPOSITE </span></span>
<span><span class="co">## "#A949C1" "#55B55E" "#7FC787" "#5b6d8a" "#ACADAF" "#ACADAF"</span></span></code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#get copy number states palette</span></span>
<span><span class="fu"><a href="../reference/get_gambl_colours.html">get_gambl_colours</a></span><span class="op">(</span>classification <span class="op">=</span> <span class="st">"copy_number"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      nLOH        14        15        13        12        11        10         9 </span></span>
<span><span class="co">## "#E026D7" "#380015" "#380015" "#380015" "#380015" "#380015" "#380015" "#380015" </span></span>
<span><span class="co">##         8         7         6         5         4         3         2         1 </span></span>
<span><span class="co">## "#380015" "#380015" "#380015" "#67001F" "#B2182B" "#D6604D" "#ede4c7" "#92C5DE" </span></span>
<span><span class="co">##         0 </span></span>
<span><span class="co">## "#4393C3"</span></span></code></pre>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#get pathology palette</span></span>
<span><span class="fu"><a href="../reference/get_gambl_colours.html">get_gambl_colours</a></span><span class="op">(</span>classification <span class="op">=</span> <span class="st">"pathology"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          B-ALL            CLL            MCL             BL            mBL </span></span>
<span><span class="co">##      "#C1C64B"      "#889BE5"      "#F37A20"      "#926CAD"      "#34C7F4" </span></span>
<span><span class="co">##            tFL  DLBCL-BL-like         pre-HT           PMBL          PMBCL </span></span>
<span><span class="co">##      "#FF8595"      "#34C7F4"      "#754F5B"      "#227C9D"      "#227C9D" </span></span>
<span><span class="co">##             FL          no-HT          COMFL            COM        post-HT </span></span>
<span><span class="co">##      "#EA8368"      "#EA8368"      "#8BBC98"      "#8BBC98"      "#479450" </span></span>
<span><span class="co">##          DLBCL   denovo-DLBCL       HGBL-NOS           HGBL     HGBL-DH/TH </span></span>
<span><span class="co">##      "#479450"      "#479450"      "#294936"      "#294936"      "#7A1616" </span></span>
<span><span class="co">##            PBL  Plasmablastic            CNS        THRLBCL             MM </span></span>
<span><span class="co">##      "#E058C0"      "#E058C0"      "#E2EF60"      "#A5F2B3"      "#CC9A42" </span></span>
<span><span class="co">##           SCBC    UNSPECIFIED          OTHER            MZL           SMZL </span></span>
<span><span class="co">##      "#8c9c90"      "#cfba7c"      "#cfba7c"      "#065A7F"      "#065A7F" </span></span>
<span><span class="co">## Prolymphocytic </span></span>
<span><span class="co">##      "#7842f5"</span></span></code></pre>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#print selected palettes to a ggplot object</span></span>
<span><span class="co">#get all gambl colours in a data frame format</span></span>
<span><span class="va">all_gambl_colours</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_gambl_colours.html">get_gambl_colours</a></span><span class="op">(</span>as_dataframe <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">classes</span> <span class="op">=</span> <span class="va">all_gambl_colours</span><span class="op">$</span><span class="va">group</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#create a function for building the plots</span></span>
<span><span class="va">plot_this_colour</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">this_group</span><span class="op">)</span><span class="op">{</span></span>
<span>  </span>
<span>  <span class="va">p</span> <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">all_gambl_colours</span>, <span class="va">group</span> <span class="op">==</span> <span class="va">this_group</span><span class="op">)</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">name</span>, y <span class="op">=</span> <span class="fl">0</span>, fill <span class="op">=</span> <span class="va">colour</span>, label <span class="op">=</span> <span class="va">name</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_tile</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">0.9</span>, height <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_text</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"white"</span>, fontface <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_identity.html" class="external-link">scale_fill_identity</a></span><span class="op">(</span>guide <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">this_group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>lineheight <span class="op">=</span> <span class="fl">0.9</span>, hjust <span class="op">=</span> <span class="fl">0.5</span>, face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">p</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co">#wrap the newly created function</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">classes</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="fu">plot_this_colour</span><span class="op">(</span>this_group <span class="op">=</span> <span class="va">x</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [[1]]</span></span></code></pre>
<p><img src="utilities_tutorial_files/figure-html/get_gambl_colours-1.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[2]]</span></span></code></pre>
<p><img src="utilities_tutorial_files/figure-html/get_gambl_colours-2.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[3]]</span></span></code></pre>
<p><img src="utilities_tutorial_files/figure-html/get_gambl_colours-3.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[4]]</span></span></code></pre>
<p><img src="utilities_tutorial_files/figure-html/get_gambl_colours-4.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[5]]</span></span></code></pre>
<p><img src="utilities_tutorial_files/figure-html/get_gambl_colours-5.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[6]]</span></span></code></pre>
<p><img src="utilities_tutorial_files/figure-html/get_gambl_colours-6.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[7]]</span></span></code></pre>
<p><img src="utilities_tutorial_files/figure-html/get_gambl_colours-7.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[8]]</span></span></code></pre>
<p><img src="utilities_tutorial_files/figure-html/get_gambl_colours-8.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[9]]</span></span></code></pre>
<p><img src="utilities_tutorial_files/figure-html/get_gambl_colours-9.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[10]]</span></span></code></pre>
<p><img src="utilities_tutorial_files/figure-html/get_gambl_colours-10.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[11]]</span></span></code></pre>
<p><img src="utilities_tutorial_files/figure-html/get_gambl_colours-11.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[12]]</span></span></code></pre>
<p><img src="utilities_tutorial_files/figure-html/get_gambl_colours-12.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[13]]</span></span></code></pre>
<p><img src="utilities_tutorial_files/figure-html/get_gambl_colours-13.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[14]]</span></span></code></pre>
<p><img src="utilities_tutorial_files/figure-html/get_gambl_colours-14.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[15]]</span></span></code></pre>
<p><img src="utilities_tutorial_files/figure-html/get_gambl_colours-15.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[16]]</span></span></code></pre>
<p><img src="utilities_tutorial_files/figure-html/get_gambl_colours-16.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[17]]</span></span></code></pre>
<p><img src="utilities_tutorial_files/figure-html/get_gambl_colours-17.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[18]]</span></span></code></pre>
<p><img src="utilities_tutorial_files/figure-html/get_gambl_colours-18.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[19]]</span></span></code></pre>
<p><img src="utilities_tutorial_files/figure-html/get_gambl_colours-19.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[20]]</span></span></code></pre>
<p><img src="utilities_tutorial_files/figure-html/get_gambl_colours-20.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[21]]</span></span></code></pre>
<p><img src="utilities_tutorial_files/figure-html/get_gambl_colours-21.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[22]]</span></span></code></pre>
<p><img src="utilities_tutorial_files/figure-html/get_gambl_colours-22.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[23]]</span></span></code></pre>
<p><img src="utilities_tutorial_files/figure-html/get_gambl_colours-23.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="gene-clouds">Gene Clouds<a class="anchor" aria-label="anchor" href="#gene-clouds"></a>
</h4>
<p>Create a wordcloud from an incoming MAF with the <code>prettyGeneCloud</code> function. This function creates a figure with gene symbols from the provided maf, were the size of each printed gene symbol is proportional to the total n occurances in the maf. Required parameter is <code>maf_df</code>. Optional parameters are <code>these_genes</code>, <code>other_genes</code>, <code>these_genes_colour</code>, <code>other_genes_colour</code> and <code>colour_index</code>. If no genes are supplied when calling the function, this function will default to all lymphoma genes.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#get all coding SSM</span></span>
<span><span class="va">maf</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_coding_ssm.html">get_coding_ssm</a></span><span class="op">(</span>seq_type <span class="op">=</span> <span class="st">"genome"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#get gene symbols from MAF and subset to chr1</span></span>
<span><span class="va">maf_genes</span> <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">maf</span>, <span class="va">Hugo_Symbol</span> <span class="op">!=</span> <span class="st">"Unknown"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">Chromosome</span> <span class="op">==</span> <span class="st">"1"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">Hugo_Symbol</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#build wordcloud</span></span>
<span><span class="fu"><a href="../reference/prettyGeneCloud.html">prettyGeneCloud</a></span><span class="op">(</span>maf_df <span class="op">=</span> <span class="va">maf</span>, these_genes <span class="op">=</span> <span class="va">maf_genes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     Hugo_Symbol  n</span></span>
<span><span class="co">## 1       A3GALT2  4</span></span>
<span><span class="co">## 2       AADACL3  6</span></span>
<span><span class="co">## 3       AADACL4  7</span></span>
<span><span class="co">## 4         ABCA4 30</span></span>
<span><span class="co">## 5        ABCB10  5</span></span>
<span><span class="co">## 6         ABCD3  5</span></span>
<span><span class="co">## 7          ABL2  8</span></span>
<span><span class="co">## 8    AC096677.1  3</span></span>
<span><span class="co">## 9    AC119673.1  1</span></span>
<span><span class="co">## 10        ACADM  6</span></span>
<span><span class="co">## 11        ACAP3  7</span></span>
<span><span class="co">## 12        ACBD3  7</span></span>
<span><span class="co">## 13        ACBD6  1</span></span>
<span><span class="co">## 14       ACOT11  8</span></span>
<span><span class="co">## 15        ACOT7  5</span></span>
<span><span class="co">## 16         ACP6  4</span></span>
<span><span class="co">## 17        ACTA1  5</span></span>
<span><span class="co">## 18        ACTL8  3</span></span>
<span><span class="co">## 19        ACTN2 15</span></span>
<span><span class="co">## 20       ACTRT2  1</span></span>
<span><span class="co">## 21       ADAM15  8</span></span>
<span><span class="co">## 22       ADAM30  9</span></span>
<span><span class="co">## 23      ADAMTS4  5</span></span>
<span><span class="co">## 24     ADAMTSL4 22</span></span>
<span><span class="co">## 25 ADAMTSL4-AS1  6</span></span></code></pre>
<p><img src="utilities_tutorial_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 2,006 × 2</span></span></span>
<span><span class="co">##    Hugo_Symbol     n</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> AC119673.1      1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> ACBD6           1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> ACTRT2          1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> AGTRAP          1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> AHCYL1          1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> AKIRIN1         1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> AL109927.1      1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> AL136115.1      1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> AL138847.1      1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> AL139147.1      1</span></span>
<span><span class="co">## <span style="color: #949494;"># … with 1,996 more rows</span></span></span></code></pre>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Ryan Morin, Kostia Dreval, Adam Mattsson.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
