<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Estimate Purity. — estimate_purity • GAMBLR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Estimate Purity. — estimate_purity"><meta property="og:description" content="Annotate a MAF with segmented absolute copy number data and added additional columns (VAF, Ploidy and Final_purity)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-12345678-0"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-12345678-0');
</script></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9500</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-light fa-memo-circle-info"></span>
     
    How To
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../path_once_integrated_on_github">Install</a>
    </li>
    <li>
      <a href="../path_once_integrated_on_github">Run Remote On A Local Machine</a>
    </li>
    <li>
      <a href="../path_once_integrated_on_github">Contribute</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-light fa-person-chalkboard"></span>
     
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/ssm_tutorial.html">SSM</a>
    </li>
    <li>
      <a href="../articles/cn_sv_tutorial.html">CN &amp; SV</a>
    </li>
    <li>
      <a href="../articles/fancy_plots.html">Fancy Plots</a>
    </li>
    <li>
      <a href="../articles/utiliteis_tutorial.html">Utilities</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="../reference/index.html">
    <span class="fa fa-file-code-o"></span>
     
    Functions &amp; Bundled Data
  </a>
</li>
<li>
  <a href="https://github.com/morinlab/GAMBLR">
    <span class="fa fa-github fa-lg"></span>
     
    github
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Estimate Purity.</h1>
    
    <div class="hidden name"><code>estimate_purity.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Annotate a MAF with segmented absolute copy number data and added additional columns (VAF, Ploidy and Final_purity).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">estimate_purity</span><span class="op">(</span></span>
<span>  <span class="va">in_maf</span>,</span>
<span>  <span class="va">maf_df</span>,</span>
<span>  <span class="va">in_seg</span>,</span>
<span>  <span class="va">this_sample_id</span>,</span>
<span>  seg_file_source <span class="op">=</span> <span class="st">"battenberg"</span>,</span>
<span>  show_plots <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  assume_diploid <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  coding_only <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">genes</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>in_maf</dt>
<dd><p>Path to a local maf file.</p></dd>


<dt>maf_df</dt>
<dd><p>Optional. Instead of using the path to a maf file, use a local dataframe as the maf file.</p></dd>


<dt>in_seg</dt>
<dd><p>Path to a local corresponding seg file for the same sample ID as the input maf.</p></dd>


<dt>this_sample_id</dt>
<dd><p>Specify the sample_id or any other string you want embedded in the file name.</p></dd>


<dt>seg_file_source</dt>
<dd><p>Specify what copy number calling program the input seg file is from, as it handles ichorCNA differently than WisecondorX, battenberg, etc.</p></dd>


<dt>show_plots</dt>
<dd><p>Optional. Show two faceted plots that display the VAF and purity distributions for each copy number state in the sample. Default is FALSE.</p></dd>


<dt>assume_diploid</dt>
<dd><p>Optional. If no local seg file is provided, instead of defaulting to a GAMBL sample, this parameter annotates every mutation as copy neutral. Default is FALSE.</p></dd>


<dt>coding_only</dt>
<dd><p>Optional. set to TRUE to restrict to only coding variants. Default is FALSE.</p></dd>


<dt>genes</dt>
<dd><p>Genes of interest.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A list containing a data frame (MAF-like format) with the segmented absolute copy number data and three extra columns:
VAF is the variant allele frequency calculated from the t_ref_count and t_alt_count
Ploidy is the number of copies of an allele in the tumour cell
Final_purity is the finalized purity estimation per mutation after considering different copy number states and LOH events.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>This function takes a sample ID with the <code>this_sample_id</code> parameter and calls <a href="assign_cn_to_ssm.html">assign_cn_to_ssm</a> to get CN information.
The user can also use an already loaded maf file with <code>maf_df</code>. In addition, a path to the maf/seq file of interest can also be passed to this function with
<code>in_maf</code> and <code>in_seg</code>. To visualize VAF and purity distributions, set the <code>show_plots</code> to TRUE (default is FALSE).
For more information on how to run this function with the parameters at hand, refer to the parameter descriptions and function examples.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">#load a maf</span></span></span>
<span class="r-in"><span><span class="va">this_maf</span> <span class="op">=</span> <span class="fu"><a href="get_ssm_by_sample.html">get_ssm_by_sample</a></span><span class="op">(</span>this_sample_id <span class="op">=</span> <span class="st">"HTMCP-01-06-00422-01A-01D"</span>, </span></span>
<span class="r-in"><span>                             this_seq_type <span class="op">=</span> <span class="st">"genome"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#estimate purity based on an already loaded maf object</span></span></span>
<span class="r-in"><span><span class="fu">estimate_purity</span><span class="op">(</span>maf_df <span class="op">=</span> <span class="va">this_maf</span>, </span></span>
<span class="r-in"><span>                show_plots <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "WARNING: A seg file was not provided! Annotating all mutation calls as copy neutral"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $VAF_plot</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span>
<span class="r-plt img"><img src="estimate_purity-1.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Purity_plot</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span>
<span class="r-plt img"><img src="estimate_purity-2.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $sample_purity_estimation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.4471171</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $CN_final</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       Hugo_Symbol   Chrom_pos Chromosome Start_Position End_Position</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    1:      TTLL10   1:1114072          1        1114072      1114072</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    2:       ACAP3   1:1228897          1        1228897      1228897</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    3:      PRDM16   1:3152295          1        3152295      3152295</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    4:     Unknown   1:4047201          1        4047201      4047201</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    5:     Unknown   1:4413966          1        4413966      4413966</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ---                                                               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4608:     MIR2114 X:149391826          X      149391826    149391826</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4609:     Unknown X:150719323          X      150719323    150719323</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4610:       PASD1 X:150739245          X      150739245    150739245</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4611:      ZNF275 X:152614366          X      152614366    152614366</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4612:      ATP2B3 X:152792650          X      152792650    152792650</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       t_ref_count t_alt_count CN       VAF Ploidy    Purity CCF_mutation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    1:          61          33  2 0.3510638      1 0.7021277    0.6820669</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    2:          63          21  2 0.2500000      1 0.5000000    0.4857143</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    3:          50          19  2 0.2753623      1 0.5507246    0.5349896</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    4:          78          16  2 0.1702128      1 0.3404255    0.3306991</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    5:          60          15  2 0.2000000      1 0.4000000    0.3885714</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ---                                                                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4608:          35          21  2 0.3750000      1 0.7500000    0.7285714</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4609:          37          17  2 0.3148148      1 0.6296296    0.6116402</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4610:          40          21  2 0.3442623      1 0.6885246    0.6688525</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4611:          38          19  2 0.3333333      1 0.6666667    0.6476190</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4612:          33          35  2 0.5147059      1 1.0294118    1.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       CCF_sample</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    1:  1.5703442</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    2:  1.1182754</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    3:  1.2317236</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    4:  0.7613790</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    5:  0.8946203</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ---           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4608:  1.6774131</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4609:  1.4081986</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4610:  1.5399202</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4611:  1.4910339</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4612:  2.3023317</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#estimate purity based sole on a smaple ID + added seg data.</span></span></span>
<span class="r-in"><span><span class="fu">estimate_purity</span><span class="op">(</span>this_sample_id <span class="op">=</span> <span class="st">"HTMCP-01-06-00422-01A-01D"</span>, </span></span>
<span class="r-in"><span>                show_plots <span class="op">=</span> <span class="cn">TRUE</span>, coding_only <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> trying to find output from: battenberg</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> looking for flatfile: /projects/nhl_meta_analysis_scratch/gambl/results_local/gambl/battenberg_current/99-outputs/seg/genome--projection/HTMCP-01-06-00422-01A-01D--HTMCP-01-06-00422-10A-01D--matched.battenberg.grch37.seg</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $VAF_plot</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span>
<span class="r-plt img"><img src="estimate_purity-3.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Purity_plot</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span>
<span class="r-plt img"><img src="estimate_purity-4.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $sample_purity_estimation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.4483024</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $CN_final</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       Hugo_Symbol    Chrom_pos Chromosome Start_Position End_Position</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  1:          BTG2  1:203274877          1      203274877    203274877</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  2:          BTG2  1:203274883          1      203274883    203274883</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  3:         DUSP2   2:96809892          2       96809892     96809892</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  4:         USP38  4:144135996          4      144135996    144135996</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  5:         DDX60  4:169195215          4      169195215    169195215</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  6:           ZFR   5:32417803          5       32417803     32417803</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  7:      HIST1H1E   6:26157109          6       26157109     26157109</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  8:      HIST1H4H   6:26285627          6       26285627     26285627</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  9:          PIM1   6:37138549          6       37138549     37138549</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10:          MYO6   6:76602374          6       76602374     76602374</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11:          ADGB  6:147090183          6      147090183    147090183</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12:         MEOX2   7:15652211          7       15652211     15652211</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 13:        SAMD9L   7:92763210          7       92763210     92763210</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 14:         MUC12  7:100635711          7      100635711    100635711</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 15:      CYB561A3  11:61118471         11       61118471     61118471</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 16:      B4GALNT3    12:661271         12         661271       661272</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 17:         RERGL  12:18237497         12       18237497     18237497</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 18:       SLCO1A2  12:21453503         12       21453503     21453503</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 19:        HECTD4 12:112746674         12      112746674    112746674</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 20:     KIAA0226L  13:46961263         13       46961263     46961263</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 21:       ZFP36L1  14:69257090         14       69257090     69257090</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 22:       ZFP36L1  14:69259603         14       69259603     69259603</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 23:         IGHG2 14:106109960         14      106109960    106109960</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 24:         IGHG2 14:106110428         14      106110428    106110428</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 25:         IGHG2 14:106110883         14      106110883    106110883</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 26:         IGHG1 14:106208502         14      106208502    106208502</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 27:       IGHV1-8 14:106539389         14      106539389    106539389</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 28:        SCNN1G  16:23200859         16       23200859     23200859</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 29:          SNPH   20:1285578         20        1285578      1285578</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 30:         LZTS3   20:3147779         20        3147779      3147779</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 31:         TRPM1  15:31319180         15       31319180     31319180</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 32:      C17orf77  17:72588641         17       72588641     72588641</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 33: RP11-1055B8.7  17:79430805         17       79430805     79430805</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 34:          CLPP   19:6361633         19        6361633      6361633</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 35:         ACTR5  20:37378710         20       37378710     37378710</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 36:           RAX  18:56936659         18       56936659     56936659</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 37:           BOC  3:112993366          3      112993366    112993366</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 38:         USP13  3:179499554          3      179499554    179499554</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       Hugo_Symbol    Chrom_pos Chromosome Start_Position End_Position</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     t_ref_count t_alt_count CN       VAF Ploidy    Purity CCF_mutation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  1:          78          26  2 0.2500000      1 0.5000000    0.8020833</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  2:          79          26  2 0.2476190      1 0.4952381    0.7944444</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  3:          67          15  2 0.1829268      1 0.3658537    0.5868902</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  4:          63          15  2 0.1923077      1 0.3846154    0.6169872</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  5:          60          19  2 0.2405063      1 0.4810127    0.7716245</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  6:          54          21  2 0.2800000      1 0.5600000    0.8983333</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  7:          63          15  2 0.1923077      1 0.3846154    0.6169872</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  8:          64          12  2 0.1578947      1 0.3157895    0.5065789</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  9:          60          17  2 0.2207792      1 0.4415584    0.7083333</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10:          67          15  2 0.1829268      1 0.3658537    0.5868902</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11:          59          18  2 0.2337662      1 0.4675325    0.7500000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12:          51          19  2 0.2714286      1 0.5428571    0.8708333</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 13:          41          16  2 0.2807018      1 0.5614035    0.9005848</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 14:          71          19  2 0.2111111      1 0.4222222    0.6773148</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 15:          46          18  2 0.2812500      1 0.5625000    0.9023438</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 16:          59          17  2 0.2236842      1 0.4473684    0.7176535</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 17:          54          15  2 0.2173913      1 0.4347826    0.6974638</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 18:          55          16  2 0.2253521      1 0.4507042    0.7230047</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 19:          67          11  2 0.1410256      1 0.2820513    0.4524573</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 20:          53          20  2 0.2739726      1 0.5479452    0.8789954</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 21:          60          16  2 0.2105263      1 0.4210526    0.6754386</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 22:          71          17  2 0.1931818      1 0.3863636    0.6197917</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 23:          53          24  2 0.3116883      1 0.6233766    1.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 24:          62          21  2 0.2530120      1 0.5060241    0.8117470</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 25:          63          17  2 0.2125000      1 0.4250000    0.6817708</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 26:          61          15  2 0.1973684      1 0.3947368    0.6332237</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 27:          60          13  2 0.1780822      1 0.3561644    0.5713470</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 28:          68          18  2 0.2093023      1 0.4186047    0.6715116</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 29:          56          16  2 0.2222222      1 0.4444444    0.7129630</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 30:          65          24  2 0.2696629      1 0.5393258    0.8651685</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 31:          53          11  2 0.1718750      1 0.3437500    0.5514323</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 32:          72          19  2 0.2087912      1 0.4175824    0.6698718</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 33:          68          21  2 0.2359551      1 0.4719101    0.7570225</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 34:          80          29  2 0.2660550      1 0.5321101    0.8535933</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 35:          44          19  1 0.3015873      1 0.3015873    0.4837963</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 36:         104          31  4 0.2296296      2 0.4592593    0.7367284</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 37:          70          28  3 0.2857143      2 0.4285714    0.6875000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 38:          71          16  3 0.1839080      1 0.5517241    0.8850575</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     t_ref_count t_alt_count CN       VAF Ploidy    Purity CCF_mutation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     CCF_sample</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  1:  1.1153186</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  2:  1.1046965</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  3:  0.8160867</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  4:  0.8579373</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  5:  1.0729647</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  6:  1.2491568</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  7:  0.8579373</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  8:  0.7044117</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  9:  0.9849566</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10:  0.8160867</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11:  1.0428953</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12:  1.2109173</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 13:  1.2522875</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 14:  0.9418246</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 15:  1.2547334</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 16:  0.9979166</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 17:  0.9698422</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 18:  1.0053576</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 19:  0.6291541</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 20:  1.2222669</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 21:  0.9392156</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 22:  0.8618371</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 23:  1.3905270</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 24:  1.1287561</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 25:  0.9480208</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 26:  0.8805146</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 27:  0.7944735</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 28:  0.9337551</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 29:  0.9913943</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 30:  1.2030402</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 31:  0.7667815</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 32:  0.9314748</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 33:  1.0526602</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 34:  1.1869445</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 35:  0.6727318</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 36:  1.0244407</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 37:  0.9559873</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 38:  1.2306963</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     CCF_sample</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ryan Morin, Kostia Dreval, Adam Mattsson.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

