<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Total n variants count plot. — fancy_v_count • GAMBLR</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Total n variants count plot. — fancy_v_count"><meta property="og:description" content="Generate a bar plot visualizing total variant (SSM or SVs) count for selected contigs."><meta property="og:image" content="https://github.com/morinlab/GAMBLR/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-12345678-0"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-12345678-0');
</script></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9500</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-light fa-memo-circle-info"></span>
     
    How To
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="https://morinlab.github.io/GAMBLR/#installation" class="external-link">Install</a>
    </li>
    <li>
      <a href="https://morinlab.github.io/GAMBLR/#run-remote-on-a-local-machine" class="external-link">Run Remote On A Local Machine</a>
    </li>
    <li>
      <a href="https://morinlab.github.io/GAMBLR/#contributing" class="external-link">Contribute</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-light fa-person-chalkboard"></span>
     
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/ssm_tutorial.html">SSM</a>
    </li>
    <li>
      <a href="../articles/cn_sv_tutorial.html">CN &amp; SV</a>
    </li>
    <li>
      <a href="../articles/fancy_plots.html">Fancy Plots</a>
    </li>
    <li>
      <a href="../articles/utilities_tutorial.html">Utilities</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="../reference/index.html">
    <span class="fa fa-file-code-o"></span>
     
    Functions &amp; Bundled Data
  </a>
</li>
<li>
  <a href="https://github.com/morinlab/GAMBLR">
    <span class="fa fa-github fa-lg"></span>
     
    github
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Total n variants count plot.</h1>
    
    <div class="hidden name"><code>fancy_v_count.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Generate a bar plot visualizing total variant (SSM or SVs) count for selected contigs.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">fancy_v_count</span><span class="op">(</span></span>
<span>  <span class="va">this_sample_id</span>,</span>
<span>  <span class="va">maf_data</span>,</span>
<span>  maf_path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ssm <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  projection <span class="op">=</span> <span class="st">"grch37"</span>,</span>
<span>  this_seq_type <span class="op">=</span> <span class="st">"genome"</span>,</span>
<span>  min_vaf <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  variant_type_col <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  chromosome_col <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  plot_title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">this_sample_id</span><span class="op">)</span>,</span>
<span>  plot_subtitle <span class="op">=</span> <span class="st">"Variant Count For Selected Contigs"</span>,</span>
<span>  chr_select <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  variant_select <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DEL"</span>, <span class="st">"INS"</span>, <span class="st">"DUP"</span><span class="op">)</span>,</span>
<span>  snp_colours <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>SNP <span class="op">=</span> <span class="st">"#2B9971"</span>, DNP <span class="op">=</span> <span class="st">"#993F2B"</span>, TNP <span class="op">=</span> <span class="st">"#A62656"</span><span class="op">)</span>,</span>
<span>  hide_legend <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  coding_only <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  log10_y <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  from_flatfile <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  use_augmented_maf <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>this_sample_id</dt>
<dd><p>Sample to be plotted.</p></dd>


<dt>maf_data</dt>
<dd><p>Optional parameter with maf like df already loaded into R.</p></dd>


<dt>maf_path</dt>
<dd><p>Optional parameter with path to external maf like file.</p></dd>


<dt>ssm</dt>
<dd><p>Set to FALSE to get plotting data from get_combined_sv (SVs). Default value is TRUE (plots SSM retrieved from annotate_cn_by_ssm$maf).</p></dd>


<dt>projection</dt>
<dd><p>Genome build for returned variants (only applicable for ssm = FALSE).</p></dd>


<dt>this_seq_type</dt>
<dd><p>Seq type for returned CN segments. One of "genome" (default) or "capture".</p></dd>


<dt>min_vaf</dt>
<dd><p>The minimum tumour VAF for a SV to be returned. Recommended: 0 (only applicable for ssm = FALSE).</p></dd>


<dt>variant_type_col</dt>
<dd><p>Index of column holding Variant Type (to be used with either maf_data or maf_path).</p></dd>


<dt>chromosome_col</dt>
<dd><p>Index of column holding Chromosome (to be used with either maf_data or maf_path).</p></dd>


<dt>plot_title</dt>
<dd><p>Title of plot (default to sample ID).</p></dd>


<dt>plot_subtitle</dt>
<dd><p>Subtitle for created plot.</p></dd>


<dt>chr_select</dt>
<dd><p>vector of chromosomes to be included in plot, defaults to autosomes.</p></dd>


<dt>variant_select</dt>
<dd><p>Subtypes of SVs to be included in plot, default is DEL, INS and DUP.</p></dd>


<dt>snp_colours</dt>
<dd><p>Optional vector with colours for SNPs (DNP and TNP).</p></dd>


<dt>hide_legend</dt>
<dd><p>Set to True to remove legend from plot, default is FALSE.</p></dd>


<dt>coding_only</dt>
<dd><p>Optional. Set to TRUE to restrict to plotting only coding mutations.</p></dd>


<dt>log10_y</dt>
<dd><p>Set to TRUE to force y axis to be in log10.</p></dd>


<dt>from_flatfile</dt>
<dd><p>If set to true the function will use flat files instead of the database.</p></dd>


<dt>use_augmented_maf</dt>
<dd><p>Boolean statement if to use augmented maf, default is TRUE.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A plot as a ggplot object (grob).</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>This function creates a barplot showing the total number of variants for a selected sample.
Convenience parameters for restricting the returned plot are available. For example, with <code>ssm</code> (Boolean)
you can toggle if the plot will be in respect to SSM (<code>ssm = TRUE</code>) or if you wish to count SVs (<code>ssm = FALSE</code>).
In addition, this plot can also accept a variety of incoming data types. Either, you supply the function with a sample ID
(<code>this_sample_id</code>) and the function will retrieve data using <a href="assign_cn_to_ssm.html">assign_cn_to_ssm</a> or <a href="get_combined_sv.html">get_combined_sv</a> (depending on how the <code>ssm</code> parameter is used).
This function also supports a maf or maf-like data frame directly, this is done with <code>maf_data</code> or <code>maf_path</code>. If data is supplied with either of these parameters,
the user can specify what column holds the variant type information as well as chromosome information (<code>variant_type_col</code> and <code>chromosome_col</code>).
Restricting the plot to coding mutations is done with <code>coding_only = TRUE</code>. Flat-file and augmented maf options can be toggled with <code>from_flatfile</code>
and <code>use_augmented_maf</code>. Both are TRUE by default and should rarely be set to FALSE. Lastly, this plotting function also have convenient parameters for
customizing the returned plot, e.g <code>plot_title</code>, <code>y_interval</code>, <code>hide_legend</code>, and<code>plot_subtitle</code> and <code>snp_colours</code>. lastly, it is also possible
to control what variants are to be counted with <code>variant_select</code>. Default is deletions, insertions and duplications, c("DEL", "DUP", "INS"). Not that
the variant types specified in this parameter must match with whatever is present in the corresponding <code>variant_type_col</code>.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">#count all variants for one sample (default parameters)</span></span></span>
<span class="r-in"><span><span class="fu">fancy_v_count</span><span class="op">(</span>this_sample_id <span class="op">=</span> <span class="st">"HTMCP-01-06-00422-01A-01D"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> trying to find output from: battenberg</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> looking for flatfile: /projects/nhl_meta_analysis_scratch/gambl/results_local/gambl/battenberg_current/99-outputs/seg/genome--projection/HTMCP-01-06-00422-01A-01D--HTMCP-01-06-00422-10A-01D--matched.battenberg.grch37.seg</span>
<span class="r-plt img"><img src="fancy_v_count-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#count and plot all variants on chromosome 1</span></span></span>
<span class="r-in"><span><span class="fu">fancy_v_count</span><span class="op">(</span>this_sample_id <span class="op">=</span> <span class="st">"HTMCP-01-06-00422-01A-01D"</span>,</span></span>
<span class="r-in"><span>              chr_select <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr1"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> trying to find output from: battenberg</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> looking for flatfile: /projects/nhl_meta_analysis_scratch/gambl/results_local/gambl/battenberg_current/99-outputs/seg/genome--projection/HTMCP-01-06-00422-01A-01D--HTMCP-01-06-00422-10A-01D--matched.battenberg.grch37.seg</span>
<span class="r-plt img"><img src="fancy_v_count-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ryan Morin, Kostia Dreval, Adam Mattsson.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

