<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Variant size distribution plot — fancy_v_sizedis • GAMBLR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Variant size distribution plot — fancy_v_sizedis"><meta property="og:description" content="Generate a violin plot showing variant (SSM or SVs) size distributions for selected contigs."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9500</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/cn_sv_tutorial.html">CN &amp; SV Tutorial - GAMBLR</a>
    </li>
    <li>
      <a href="../articles/fancy_plots.html">GAMBLR: Fancy Plotting Functions</a>
    </li>
    <li>
      <a href="../articles/ssm_tutorial.html">SSM Tutorial - GAMBLR</a>
    </li>
    <li>
      <a href="../articles/utilities_tutorial.html">Utilities Tutorial - GAMBLR</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Variant size distribution plot</h1>
    
    <div class="hidden name"><code>fancy_v_sizedis.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Generate a violin plot showing variant (SSM or SVs) size distributions for selected contigs.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">fancy_v_sizedis</span><span class="op">(</span></span>
<span>  <span class="va">this_sample_id</span>,</span>
<span>  <span class="va">maf_data</span>,</span>
<span>  maf_path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ssm <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  projection <span class="op">=</span> <span class="st">"grch37"</span>,</span>
<span>  min_vaf <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  variant_type_col <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  chromosome_col <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  start_col <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  end_col <span class="op">=</span> <span class="fl">7</span>,</span>
<span>  plot_title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">this_sample_id</span><span class="op">)</span>,</span>
<span>  plot_subtitle <span class="op">=</span> <span class="st">"Variant Size Distribution"</span>,</span>
<span>  scale_value <span class="op">=</span> <span class="st">"width"</span>,</span>
<span>  log_10 <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  plot_trim <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  chr_select <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  coding_only <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  from_flatfile <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  use_augmented_maf <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>this_sample_id</dt>
<dd><p>Sample to be plotted.</p></dd>


<dt>maf_data</dt>
<dd><p>Optional parameter with maf like df already loaded into R.</p></dd>


<dt>maf_path</dt>
<dd><p>Optional parameter with path to external maf like file.</p></dd>


<dt>ssm</dt>
<dd><p>Set to FALSE to get plotting data from get_combined_sv (SVs). Default value is TRUE (plots SSM retrieved from annotate_cn_by_ssm$maf).</p></dd>


<dt>projection</dt>
<dd><p>Genome build for returned variants (only applicable for ssm = FALSE).</p></dd>


<dt>min_vaf</dt>
<dd><p>The minimum tumour VAF for a SV to be returned. Recommended: 0 (only applicable for ssm = FALSE).</p></dd>


<dt>variant_type_col</dt>
<dd><p>Index of column holding Variant Type (to be used with either maf_data or maf_path).</p></dd>


<dt>chromosome_col</dt>
<dd><p>Index of column holding Chromosome (to be used with either maf_data or maf_path).</p></dd>


<dt>start_col</dt>
<dd><p>Index of column with variant start coordinates (to be used with either maf_data or maf_path).</p></dd>


<dt>end_col</dt>
<dd><p>Index of column with variant end coordinates (to be used with either maf_data or maf_path).</p></dd>


<dt>plot_title</dt>
<dd><p>Title of plot (default to sample ID).</p></dd>


<dt>plot_subtitle</dt>
<dd><p>Subtitle for created plot.</p></dd>


<dt>scale_value</dt>
<dd><p>Scale type for violin plot, accepted values are "area", "width", and "count", default is "count.</p></dd>


<dt>log_10</dt>
<dd><p>Boolean statement for y-axis, default is TRUE.</p></dd>


<dt>plot_trim</dt>
<dd><p>If TRUE, trim the tails of the violins to the range of the data. If FALSE (default), don't trim the tails.</p></dd>


<dt>chr_select</dt>
<dd><p>vector of chromosomes to be included in plot, defaults to autosomes.</p></dd>


<dt>coding_only</dt>
<dd><p>Optional. Set to TRUE to restrict to plotting only coding mutations.</p></dd>


<dt>from_flatfile</dt>
<dd><p>If set to true the function will use flat files instead of the database.</p></dd>


<dt>use_augmented_maf</dt>
<dd><p>Boolean statement if to use augmented maf, default is FALSE.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A plot as a ggplot object (grob).</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Function for plotting variant size distributions. This function takes either a sample ID given to the <code>this_sample</code> parameter.
In addition, the function can also accept an already loaded MAF or MAF-like object given to the <code>maf_data</code> parameter.
As a third option, the function can also read a maf from disk (provide path to maf with <code>maf_path</code>).
A collection of convenient filtering and data subsetting parameters are also available for this function.
For restricting your data (if plotting data retrieved with <code>this_sample_id</code>), the user can choose to only plot coding mutations with setting <code>coding_only</code> to TRUE.
This plot can also deal with SVs as well as SSM data. To control this, please use the <code>ssm</code> parameter. If set to TRUE and if <code>this_sample</code> is called,
the function gets data with annotate_cn_by_ssm and if set to FALSE, the function calls <code>get_combined_sv</code> to get SV calls for plotting.
If the user calls either <code>maf_data</code> or <code>maf_path</code>, there are a collection of parameters available for specifying
the relevant columns in the given data frame (<code>variant_type_col</code>, <code>chhromosome_col</code>, <code>start_col</code>, <code>end_col</code>). It is also possible to
restrict the returned plot to any given chromosome. This is done with the <code>chr_select</code> parameter (default is all autosomes).
In addition, plot aesthetics can also be controlled with <code>plot_title</code>, <code>plot_subtitle</code>, <code>scale_value</code>, <code>log10</code>, and <code>trim</code>.
For more info on how to run with these parameters, refer to the parameter descriptions.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">#plot SSM size distributions:</span></span></span>
<span class="r-in"><span><span class="fu">fancy_v_sizedis</span><span class="op">(</span>this_sample_id <span class="op">=</span> <span class="st">"HTMCP-01-06-00422-01A-01D"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> trying to find output from: battenberg</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> looking for flatfile: /projects/nhl_meta_analysis_scratch/gambl/results_local/gambl/battenberg_current/99-outputs/seg/genome--projection/HTMCP-01-06-00422-01A-01D--HTMCP-01-06-00422-10A-01D--matched.battenberg.grch37.seg</span>
<span class="r-plt img"><img src="fancy_v_sizedis-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ryan Morin.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

