<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Plot Alignment Metrics — fancy_alignment_plot • GAMBLR</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Plot Alignment Metrics — fancy_alignment_plot"><meta property="og:description" content="Visualize (stacked barplot) genomic read-subsets (metrics) across a selection of samples."><meta property="og:image" content="https://github.com/morinlab/GAMBLR/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-12345678-0"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-12345678-0');
</script></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9500</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-light fa-memo-circle-info"></span>
     
    How To
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="https://morinlab.github.io/GAMBLR/#installation" class="external-link">Install</a>
    </li>
    <li>
      <a href="https://morinlab.github.io/GAMBLR/#running-gamblr-on-your-own-computer" class="external-link">Running GAMBLR On Your Own Computer</a>
    </li>
    <li>
      <a href="https://morinlab.github.io/GAMBLR/#contributing" class="external-link">Contribute</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-light fa-person-chalkboard"></span>
     
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/ssm_tutorial.html">SSM</a>
    </li>
    <li>
      <a href="../articles/cn_sv_tutorial.html">CN &amp; SV</a>
    </li>
    <li>
      <a href="../articles/fancy_plots.html">Fancy Plots</a>
    </li>
    <li>
      <a href="../articles/utilities_tutorial.html">Utilities</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="../reference/index.html">
    <span class="fa fa-file-code-o"></span>
     
    Functions &amp; Bundled Data
  </a>
</li>
<li>
  <a href="https://github.com/morinlab/GAMBLR">
    <span class="fa fa-github fa-lg"></span>
     
    github
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Plot Alignment Metrics</h1>
    
    <div class="hidden name"><code>fancy_alignment_plot.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Visualize (stacked barplot) genomic read-subsets (metrics) across a selection of samples.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">fancy_alignment_plot</span><span class="op">(</span></span>
<span>  <span class="va">these_sample_ids</span>,</span>
<span>  <span class="va">metadata</span>,</span>
<span>  <span class="va">these_samples_metadata</span>,</span>
<span>  <span class="va">comparison_group</span>,</span>
<span>  seq_type <span class="op">=</span> <span class="st">"genome"</span>,</span>
<span>  add_mean <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  add_corrected_coverage <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">keep_cohort</span>,</span>
<span>  <span class="va">keep_pathology</span>,</span>
<span>  this_color_palette <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>TotalReads <span class="op">=</span> <span class="st">"#3D405B"</span>, TotalUniquelyMapped <span class="op">=</span> <span class="st">"#81B29A"</span>,</span>
<span>    TotalDuplicatedreads <span class="op">=</span> <span class="st">"#E07A5F"</span><span class="op">)</span>,</span>
<span>  plot_sub <span class="op">=</span> <span class="st">""</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>these_sample_ids</dt>
<dd><p>Data frame with sample IDs (to be plotted) in the first column.</p></dd>


<dt>metadata</dt>
<dd><p>Optional argument, used to derive sample IDs if sample_table is Null.</p></dd>


<dt>these_samples_metadata</dt>
<dd><p>GAMBL metadata subset to the cases you want to process.</p></dd>


<dt>comparison_group</dt>
<dd><p>Optional argument for plotting mean alignment metrics. Default is plotting the mean for samples provided. This parameter takes a vector of sample IDs.</p></dd>


<dt>seq_type</dt>
<dd><p>Subset qc metrics to a specific seq_type, default is genome.</p></dd>


<dt>add_mean</dt>
<dd><p>Set to TRUE to superimpose mean values of plotted variables. Default is TRUE.</p></dd>


<dt>add_corrected_coverage</dt>
<dd><p>Set to TRUE to add corrected coverage for selected samples.</p></dd>


<dt>keep_cohort</dt>
<dd><p>If no df with sample IDs is supplied (these_sample_ids = NULL) the function calls get_gambl_metadata and subsets on selected cohorts.</p></dd>


<dt>keep_pathology</dt>
<dd><p>If no df with sample IDs is supplied (these_sample_ids = NULL) the function calls get_gambl_metadata and subsets on selected pathology.</p></dd>


<dt>this_color_palette</dt>
<dd><p>Optional parameter that holds the selected colours for the plotted bars.</p></dd>


<dt>plot_sub</dt>
<dd><p>Optional parameter, add a subtitle to the alignment metric plot.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A plot as a ggplot object (grob).</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>This function is available for plotting relevant alignment metrics (read-subsets) for a selection of samples. Per default, this plot returns the following read-metrics;
total n reads, total n uniquely mapped reads, total n duplicated reads. This plot can also be superimposed with read metrics from additional samples,
allowing for easy comparisons between different sample populations. To run this function, simply specify the sample IDs you are interested in with <code>these_sample_ids</code>.
This parameter expects a data frame with sample IDs in the first column. Optionally, the user can also provide an already subset (with the sample IDS of interest)
metadata table with <code>these_samples_metadata</code>. For adding a comparison group to the returned plot, simply give another cohort/set of samples to the <code>comparison_group</code> parameter.
Similarly to <code>these_sample_ids</code>, this parameter also expects a data frame with sample IDs in the first column. In addition, this plot can also add additional read-metrics such as
mean values for all plotted metrics and corrected coverage. To enable these features, simply set <code>add_mean</code> and <code>add_corrected_coverage</code> to TRUE (default).</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">#Example 1 - using these_sample_ids parameter</span></span></span>
<span class="r-in"><span><span class="co">#subset on FL cases with QC metrics available and plot</span></span></span>
<span class="r-in"><span><span class="va">metadata</span> <span class="op">=</span> <span class="fu"><a href="get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">kridel_fl</span> <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">metadata</span>, <span class="va">pathology</span> <span class="op">==</span> <span class="st">"FL"</span>,</span></span>
<span class="r-in"><span>               <span class="va">cohort</span> <span class="op">==</span> <span class="st">"FL_Kridel"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">kridel_fl_samples</span> <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">kridel_fl</span>, <span class="va">sample_id</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">fancy_alignment_plot</span><span class="op">(</span>these_sample_ids <span class="op">=</span> <span class="va">kridel_fl_samples</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> /projects/nhl_meta_analysis_scratch/gambl/results_local/shared/gambl_genome_results.tsv</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> QC Metric successfully retreived for 56 samples out of a total of 56 samples in input sample table.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>The melt generic in data.table has been passed a tbl_df and will attempt to redirect to the relevant reshape2 method; please note that reshape2 is deprecated, and this redirection is now deprecated as well. To continue using melt methods from reshape2 while both libraries are attached, e.g. melt.list, you can prepend the namespace like reshape2::melt(.). In the next version, this warning will become an error.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>The melt generic in data.table has been passed a tbl_df and will attempt to redirect to the relevant reshape2 method; please note that reshape2 is deprecated, and this redirection is now deprecated as well. To continue using melt methods from reshape2 while both libraries are attached, e.g. melt.list, you can prepend the namespace like reshape2::melt(.). In the next version, this warning will become an error.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Width not defined</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Set with `position_dodge(width = ...)`</span>
<span class="r-plt img"><img src="fancy_alignment_plot-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#Example 2 - using already filtered metadata (these_samples_metadata)</span></span></span>
<span class="r-in"><span><span class="fu">fancy_alignment_plot</span><span class="op">(</span>these_samples_metadata <span class="op">=</span> <span class="va">kridel_fl</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> /projects/nhl_meta_analysis_scratch/gambl/results_local/shared/gambl_genome_results.tsv</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> QC Metric successfully retreived for 56 samples out of a total of 56 samples in input sample table.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>The melt generic in data.table has been passed a tbl_df and will attempt to redirect to the relevant reshape2 method; please note that reshape2 is deprecated, and this redirection is now deprecated as well. To continue using melt methods from reshape2 while both libraries are attached, e.g. melt.list, you can prepend the namespace like reshape2::melt(.). In the next version, this warning will become an error.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>The melt generic in data.table has been passed a tbl_df and will attempt to redirect to the relevant reshape2 method; please note that reshape2 is deprecated, and this redirection is now deprecated as well. To continue using melt methods from reshape2 while both libraries are attached, e.g. melt.list, you can prepend the namespace like reshape2::melt(.). In the next version, this warning will become an error.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Width not defined</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Set with `position_dodge(width = ...)`</span>
<span class="r-plt img"><img src="fancy_alignment_plot-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#Example 3 - using in-house metadata filtering options</span></span></span>
<span class="r-in"><span><span class="fu">fancy_alignment_plot</span><span class="op">(</span>keep_cohort <span class="op">=</span> <span class="st">"FL_Kridel"</span>,</span></span>
<span class="r-in"><span>                     keep_pathology <span class="op">=</span> <span class="st">"FL"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> /projects/nhl_meta_analysis_scratch/gambl/results_local/shared/gambl_genome_results.tsv</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> QC Metric successfully retreived for 56 samples out of a total of 56 samples in input sample table.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>The melt generic in data.table has been passed a tbl_df and will attempt to redirect to the relevant reshape2 method; please note that reshape2 is deprecated, and this redirection is now deprecated as well. To continue using melt methods from reshape2 while both libraries are attached, e.g. melt.list, you can prepend the namespace like reshape2::melt(.). In the next version, this warning will become an error.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>The melt generic in data.table has been passed a tbl_df and will attempt to redirect to the relevant reshape2 method; please note that reshape2 is deprecated, and this redirection is now deprecated as well. To continue using melt methods from reshape2 while both libraries are attached, e.g. melt.list, you can prepend the namespace like reshape2::melt(.). In the next version, this warning will become an error.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Width not defined</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Set with `position_dodge(width = ...)`</span>
<span class="r-plt img"><img src="fancy_alignment_plot-3.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ryan Morin, Kostia Dreval, Adam Mattsson.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

