<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>genome-wide ideogram annotated with SSM and CN information — fancy_ideogram • GAMBLR</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="genome-wide ideogram annotated with SSM and CN information — fancy_ideogram"><meta property="og:description" content="Generate sample-level ideogram with copy number information, ssm and gene annotations, etc."><meta property="og:image" content="https://github.com/morinlab/GAMBLR/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-12345678-0"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-12345678-0');
</script></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9500</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-light fa-memo-circle-info"></span>
     
    How To
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="https://morinlab.github.io/GAMBLR/#installation" class="external-link">Install</a>
    </li>
    <li>
      <a href="https://morinlab.github.io/GAMBLR/#run-remote-on-a-local-machine" class="external-link">Run Remote On A Local Machine</a>
    </li>
    <li>
      <a href="https://morinlab.github.io/GAMBLR/#contributing" class="external-link">Contribute</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-light fa-person-chalkboard"></span>
     
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/ssm_tutorial.html">SSM</a>
    </li>
    <li>
      <a href="../articles/cn_sv_tutorial.html">CN &amp; SV</a>
    </li>
    <li>
      <a href="../articles/fancy_plots.html">Fancy Plots</a>
    </li>
    <li>
      <a href="../articles/utilities_tutorial.html">Utilities</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="../reference/index.html">
    <span class="fa fa-file-code-o"></span>
     
    Functions &amp; Bundled Data
  </a>
</li>
<li>
  <a href="https://github.com/morinlab/GAMBLR">
    <span class="fa fa-github fa-lg"></span>
     
    github
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>genome-wide ideogram annotated with SSM and CN information</h1>
    
    <div class="hidden name"><code>fancy_ideogram.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Generate sample-level ideogram with copy number information, ssm and gene annotations, etc.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">fancy_ideogram</span><span class="op">(</span></span>
<span>  <span class="va">this_sample_id</span>,</span>
<span>  <span class="va">gene_annotation</span>,</span>
<span>  <span class="va">seq_data</span>,</span>
<span>  seq_path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">maf_data</span>,</span>
<span>  maf_path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  variant_type_col_maf <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  chromosome_col_maf <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  start_col_maf <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  end_col_maf <span class="op">=</span> <span class="fl">7</span>,</span>
<span>  chrom_col_seq <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  start_col_seq <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  end_col_seq <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  cn_col_seq <span class="op">=</span> <span class="fl">7</span>,</span>
<span>  plot_title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">this_sample_id</span><span class="op">)</span>,</span>
<span>  plot_subtitle <span class="op">=</span> <span class="st">"Genome-wide Ideogram (grch37)."</span>,</span>
<span>  <span class="va">intersect_regions</span>,</span>
<span>  include_ssm <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  ssm_count <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  coding_only <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  from_flatfile <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  use_augmented_maf <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  this_seq_type <span class="op">=</span> <span class="st">"genome"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>this_sample_id</dt>
<dd><p>Sample to be plotted (for multiple samples, see fancy_multisample_ideogram.</p></dd>


<dt>gene_annotation</dt>
<dd><p>Annotate ideogram with a set of genes. These genes can either be specified as a vector of characters or a data frame.</p></dd>


<dt>seq_data</dt>
<dd><p>Optional parameter with copy number df already loaded into R.</p></dd>


<dt>seq_path</dt>
<dd><p>Optional parameter with path to external cn file.</p></dd>


<dt>maf_data</dt>
<dd><p>Optional parameter with maf like df already loaded into R.</p></dd>


<dt>maf_path</dt>
<dd><p>Optional parameter with path to external maf like file.</p></dd>


<dt>variant_type_col_maf</dt>
<dd><p>Index of column holding Variant Type (to be used with either maf_data or maf_path).</p></dd>


<dt>chromosome_col_maf</dt>
<dd><p>Index of column holding Chromosome (to be used with either maf_data or maf_path).</p></dd>


<dt>start_col_maf</dt>
<dd><p>Index of column with variant start coordinates (to be used with either maf_data or maf_path).</p></dd>


<dt>end_col_maf</dt>
<dd><p>Index of column with variant end coordinates (to be used with either maf_data or maf_path).</p></dd>


<dt>chrom_col_seq</dt>
<dd><p>Index of column with chromosome annotations (to be used with either maf_data or maf_path).</p></dd>


<dt>start_col_seq</dt>
<dd><p>Index of column with copy number start coordinates (to be used with either maf_data or maf_path).</p></dd>


<dt>end_col_seq</dt>
<dd><p>Index of column with copy number end coordinates (to be used with either maf_data or maf_path).</p></dd>


<dt>cn_col_seq</dt>
<dd><p>Index of column holding copy number information (to be used with either maf_data or maf_path).</p></dd>


<dt>plot_title</dt>
<dd><p>Title of plot (default to sample ID).</p></dd>


<dt>plot_subtitle</dt>
<dd><p>Optional argument for plot subtitle.</p></dd>


<dt>intersect_regions</dt>
<dd><p>Optional parameter for subset variant calls to specific regions. Should be either a vector of characters (chr:start-end) or data frame with regions.</p></dd>


<dt>include_ssm</dt>
<dd><p>Set to TRUE to plot SSMs (dels and ins).</p></dd>


<dt>ssm_count</dt>
<dd><p>Optional parameter to summarize n variants per chromosome, inlcude_ssm must be set to TRUE.</p></dd>


<dt>coding_only</dt>
<dd><p>Optional. Set to TRUE to restrict to plotting only coding mutations.</p></dd>


<dt>from_flatfile</dt>
<dd><p>If set to true the function will use flat files instead of the database.</p></dd>


<dt>use_augmented_maf</dt>
<dd><p>Boolean statement if to use augmented maf, default is FALSE.</p></dd>


<dt>this_seq_type</dt>
<dd><p>Seq type for returned CN segments. One of "genome" (default) or "capture".</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A plot as a ggplot object (grob).</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>This function generates genome-wide ideograms, visualizing SSM data as well as CN segments.
It is also possible to superimpose the plot with gene annotations. Offering a comprehensive overview of all SSM and CN segments of different aneuploidy.
The plotting of SSM can be toggled with setting <code>include_ssm</code> to TRUE. If so, it is also possible to count the number of SSMs per chromosome with <code>ssm_count = TRUE</code>.
To get data for plotting, there are a few different options available; like all <code>fanncy_x_plots</code> a sample ID can be provided to the <code>this_sample</code>
parameter. If done so, the function will retrieve data (SSm and CN segments) by wrapping the appropriate functions.
This data can also be provided with <code>seq_data</code>, <code>seg_path</code>, <code>maf_data</code> and <code>maf_path</code>.
For more info on how to run with these parameters, refer to the parameter descriptions.
In order to annotate the ideogram with genes, simply give the <code>gene_annotations</code> parameter a set of genes as a vector of characters or a data frame with gene names in the first column.
Another useful parameter for restricting the plotted regions is to call the function with <code>intersect_regions</code>.
This parameter takes a vector of characters or a data frame with regions that the plotted calls are restricted to.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">#build plot</span></span></span>
<span class="r-in"><span><span class="fu">fancy_ideogram</span><span class="op">(</span>this_sample_id <span class="op">=</span> <span class="st">"HTMCP-01-06-00422-01A-01D"</span>,</span></span>
<span class="r-in"><span>               gene_annotation <span class="op">=</span> <span class="st">"MYC"</span>,</span></span>
<span class="r-in"><span>               plot_title <span class="op">=</span> <span class="st">"Sample-level Ideogram Example"</span>,</span></span>
<span class="r-in"><span>               plot_subtitle <span class="op">=</span> <span class="st">"grch37"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 1 region(s) returned for 1 gene(s)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>NAs introduced by coercion</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>invalid factor level, NA generated</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>‘&gt;’ not meaningful for factors</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> trying to find output from: battenberg</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> looking for flatfile: /projects/nhl_meta_analysis_scratch/gambl/results_local/gambl/battenberg_current/99-outputs/seg/genome--projection/HTMCP-01-06-00422-01A-01D--HTMCP-01-06-00422-10A-01D--matched.battenberg.grch37.seg</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>NAs introduced by coercion</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>NAs introduced by coercion</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Ignoring unknown aesthetics: label</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Ignoring unknown parameters: check_overlap</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Width not defined. Set with `position_dodge(width = ?)`</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Width not defined. Set with `position_dodge(width = ?)`</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Width not defined. Set with `position_dodge(width = ?)`</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Width not defined. Set with `position_dodge(width = ?)`</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Width not defined. Set with `position_dodge(width = ?)`</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Width not defined. Set with `position_dodge(width = ?)`</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Width not defined. Set with `position_dodge(width = ?)`</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Width not defined. Set with `position_dodge(width = ?)`</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Width not defined. Set with `position_dodge(width = ?)`</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Width not defined. Set with `position_dodge(width = ?)`</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Width not defined. Set with `position_dodge(width = ?)`</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Width not defined. Set with `position_dodge(width = ?)`</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Width not defined. Set with `position_dodge(width = ?)`</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Removed 8 rows containing missing values (geom_segment).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Removed 8 rows containing missing values (geom_segment).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Removed 8 rows containing missing values (geom_segment).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Removed 4 rows containing missing values (geom_segment).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Removed 4 rows containing missing values (geom_segment).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Removed 4 rows containing missing values (geom_segment).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Removed 1 rows containing missing values (geom_text).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Removed 1 rows containing missing values (geom_text).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Removed 6 rows containing missing values (geom_segment).</span>
<span class="r-plt img"><img src="fancy_ideogram-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ryan Morin, Kostia Dreval, Adam Mattsson.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

