<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Get Coding SSM Status. — get_coding_ssm_status • GAMBLR</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Get Coding SSM Status. — get_coding_ssm_status"><meta property="og:description" content="Tabulate mutation status (SSM) for a set of genes."><meta property="og:image" content="https://github.com/morinlab/GAMBLR/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-12345678-0"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-12345678-0');
</script></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9500</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-light fa-memo-circle-info"></span>
     
    How To
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="https://morinlab.github.io/GAMBLR/#installation" class="external-link">Install</a>
    </li>
    <li>
      <a href="https://morinlab.github.io/GAMBLR/#running-gamblr-on-your-own-computer" class="external-link">Running GAMBLR On Your Own Computer</a>
    </li>
    <li>
      <a href="https://morinlab.github.io/GAMBLR/#contributing" class="external-link">Contribute</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-light fa-person-chalkboard"></span>
     
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/ssm_tutorial.html">SSM</a>
    </li>
    <li>
      <a href="../articles/cn_sv_tutorial.html">CN &amp; SV</a>
    </li>
    <li>
      <a href="../articles/fancy_plots.html">Fancy Plots</a>
    </li>
    <li>
      <a href="../articles/utilities_tutorial.html">Utilities</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="../reference/index.html">
    <span class="fa fa-file-code-o"></span>
     
    Functions &amp; Bundled Data
  </a>
</li>
<li>
  <a href="https://github.com/morinlab/GAMBLR">
    <span class="fa fa-github fa-lg"></span>
     
    github
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Get Coding SSM Status.</h1>
    
    <div class="hidden name"><code>get_coding_ssm_status.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Tabulate mutation status (SSM) for a set of genes.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_coding_ssm_status</span><span class="op">(</span></span>
<span>  <span class="va">gene_symbols</span>,</span>
<span>  <span class="va">these_samples_metadata</span>,</span>
<span>  from_flatfile <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  augmented <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  min_read_support <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  maf_path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">maf_data</span>,</span>
<span>  include_hotspots <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  keep_multihit_hotspot <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  recurrence_min <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  seq_type <span class="op">=</span> <span class="st">"genome"</span>,</span>
<span>  projection <span class="op">=</span> <span class="st">"grch37"</span>,</span>
<span>  review_hotspots <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  genes_of_interest <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"FOXO1"</span>, <span class="st">"MYD88"</span>, <span class="st">"CREBBP"</span><span class="op">)</span>,</span>
<span>  genome_build <span class="op">=</span> <span class="st">"hg19"</span>,</span>
<span>  include_silent <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>gene_symbols</dt>
<dd><p>A vector of gene symbols for which the mutation status will be tabulated. If not provided, lymphoma genes will be returned by default.</p></dd>


<dt>these_samples_metadata</dt>
<dd><p>The metadata for samples of interest to be included in the returned matrix. Only the column "sample_id" is required. If not provided, the matrix is tabulated for all available samples as default.</p></dd>


<dt>from_flatfile</dt>
<dd><p>Optional argument whether to use database or flat file to retrieve mutations. Default is TRUE.</p></dd>


<dt>augmented</dt>
<dd><p>default: TRUE. Set to FALSE if you instead want the original MAF from each sample for multi-sample patients instead of the augmented MAF.</p></dd>


<dt>min_read_support</dt>
<dd><p>Only returns variants with at least this many reads in t_alt_count (for cleaning up augmented MAFs).</p></dd>


<dt>maf_path</dt>
<dd><p>If the status of coding SSM should be tabulated from a custom maf file, provide path to the maf in this argument. The default is set to NULL.</p></dd>


<dt>maf_data</dt>
<dd><p>Either a maf loaded from disk or from the database using a get_ssm function.</p></dd>


<dt>include_hotspots</dt>
<dd><p>Logical parameter indicating whether hotspots object should also be tabulated. Default is TRUE.</p></dd>


<dt>keep_multihit_hotspot</dt>
<dd><p>Logical parameter indicating whether to keep the gene annotation as mutated when the gene has both hot spot and non-hotspot mutation. Default is FALSE. If set to TRUE, will report the number of non-hotspot mutations instead of tabulating for just mutation presence.</p></dd>


<dt>recurrence_min</dt>
<dd><p>Integer value indicating minimal recurrence level.</p></dd>


<dt>seq_type</dt>
<dd><p>The seq_type you want back, default is genome.</p></dd>


<dt>projection</dt>
<dd><p>Specify projection (grch37 or hg38) of mutations. Default is grch37.</p></dd>


<dt>review_hotspots</dt>
<dd><p>Logical parameter indicating whether hotspots object should be reviewed to include functionally relevant mutations or rare lymphoma-related genes. Default is TRUE.</p></dd>


<dt>genes_of_interest</dt>
<dd><p>A vector of genes for hotspot review. Currently only FOXO1, MYD88, and CREBBP are supported.</p></dd>


<dt>genome_build</dt>
<dd><p>Reference genome build for the coordinates in the MAF file. The default is hg19 genome build.</p></dd>


<dt>include_silent</dt>
<dd><p>Logical parameter indicating whether to include silent mutations into coding mutations. Default is TRUE.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A data frame with tabulated mutation status.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>This function takes a vector of gene symbols and subsets the incoming MAF to specified genes. If no genes are provided, the function will default to all lymphoma genes.
The function can accept a wide range of incoming MAFs. For example, the user can call this function with <code>these_samples_metadata</code> (preferably a metadata table that has been subset to the sample IDs of interest).
If this parameter is not called, the function will default to all samples available with <a href="get_gambl_metadata.html">get_gambl_metadata</a>. The user can also provide a path to a MAF, or MAF-like file with <code>maf_path</code>,
or an already loaded MAF can be used with the <code>maf_data</code> parameter. If both <code>maf_path</code> and <code>maf_data</code> is missing, the function will default to run <code>get_coding_ssm</code>.
This function also has a lot of filtering and convenience parameters giving the user full control of the return. For more information, refer to the parameter descriptions and examples.
Is this function not what you are looking for? Try one of the following, similar, functions; <a href="get_coding_ssm.html">get_coding_ssm</a>, <a href="get_ssm_by_patients.html">get_ssm_by_patients</a>, <a href="get_ssm_by_sample.html">get_ssm_by_sample</a>,
<a href="get_ssm_by_samples.html">get_ssm_by_samples</a>, <a href="get_ssm_by_region.html">get_ssm_by_region</a>, <a href="get_ssm_by_regions.html">get_ssm_by_regions</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">coding_tabulated_df</span> <span class="op">=</span> <span class="fu">get_coding_ssm_status</span><span class="op">(</span>maf_data <span class="op">=</span> <span class="va">grande_maf</span>,</span></span>
<span class="r-in"><span>                                            gene_symbols <span class="op">=</span> <span class="st">"EGFR"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "missing:  /Users/rmorin/git/gambl/data/metadata/gambl_all_outcomes.tsv"</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Have you cloned the GAMBL repo and added the path to this directory under the local section of your config?</span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error:</span> '/Users/rmorin/git/gambl/data/metadata/gambl_all_outcomes.tsv' does not exist.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#all lymphoma genes from bundled NHL gene list</span></span></span>
<span class="r-in"><span><span class="va">coding_tabulated_df</span> <span class="op">=</span> <span class="fu">get_coding_ssm_status</span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> defaulting to all lymphoma genes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "missing:  /Users/rmorin/git/gambl/data/metadata/gambl_all_outcomes.tsv"</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Have you cloned the GAMBL repo and added the path to this directory under the local section of your config?</span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error:</span> '/Users/rmorin/git/gambl/data/metadata/gambl_all_outcomes.tsv' does not exist.</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ryan Morin, Kostia Dreval, Adam Mattsson.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

