<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Get Coding SSM. — get_coding_ssm • GAMBLR</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Get Coding SSM. — get_coding_ssm"><meta property="og:description" content="Retrieve all coding SSMs from the GAMBL database in MAF-like format."><meta property="og:image" content="https://github.com/morinlab/GAMBLR/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-12345678-0"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-12345678-0');
</script></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9500</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-light fa-memo-circle-info"></span>
     
    How To
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="https://morinlab.github.io/GAMBLR/#installation" class="external-link">Install</a>
    </li>
    <li>
      <a href="https://morinlab.github.io/GAMBLR/#running-gamblr-on-your-own-computer" class="external-link">Running GAMBLR On Your Own Computer</a>
    </li>
    <li>
      <a href="https://morinlab.github.io/GAMBLR/#contributing" class="external-link">Contribute</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-light fa-person-chalkboard"></span>
     
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/ssm_tutorial.html">SSM</a>
    </li>
    <li>
      <a href="../articles/cn_sv_tutorial.html">CN &amp; SV</a>
    </li>
    <li>
      <a href="../articles/fancy_plots.html">Fancy Plots</a>
    </li>
    <li>
      <a href="../articles/utilities_tutorial.html">Utilities</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="../reference/index.html">
    <span class="fa fa-file-code-o"></span>
     
    Functions &amp; Bundled Data
  </a>
</li>
<li>
  <a href="https://github.com/morinlab/GAMBLR">
    <span class="fa fa-github fa-lg"></span>
     
    github
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Get Coding SSM.</h1>
    
    <div class="hidden name"><code>get_coding_ssm.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Retrieve all coding SSMs from the GAMBL database in MAF-like format.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_coding_ssm</span><span class="op">(</span></span>
<span>  <span class="va">limit_cohort</span>,</span>
<span>  <span class="va">exclude_cohort</span>,</span>
<span>  <span class="va">limit_pathology</span>,</span>
<span>  <span class="va">limit_samples</span>,</span>
<span>  <span class="va">these_samples_metadata</span>,</span>
<span>  <span class="va">force_unmatched_samples</span>,</span>
<span>  projection <span class="op">=</span> <span class="st">"grch37"</span>,</span>
<span>  <span class="va">seq_type</span>,</span>
<span>  basic_columns <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  maf_cols <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  from_flatfile <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  augmented <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  min_read_support <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gambl"</span>, <span class="st">"icgc_dart"</span><span class="op">)</span>,</span>
<span>  include_silent <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  engine <span class="op">=</span> <span class="st">"fread_maf"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>limit_cohort</dt>
<dd><p>Supply this to restrict mutations to one or more cohorts in a vector.</p></dd>


<dt>exclude_cohort</dt>
<dd><p>Supply this to exclude mutations from one or more cohorts in a vector.</p></dd>


<dt>limit_pathology</dt>
<dd><p>Supply this to restrict mutations to one pathology.</p></dd>


<dt>limit_samples</dt>
<dd><p>Supply this to restrict mutations to a vector of sample_id (instead of subsetting using the provided metadata)</p></dd>


<dt>these_samples_metadata</dt>
<dd><p>Supply a metadata table to auto-subset the data to samples in that table before returning</p></dd>


<dt>force_unmatched_samples</dt>
<dd><p>Optional argument for forcing unmatched samples, using <a href="get_ssm_by_samples.html">get_ssm_by_samples</a>.</p></dd>


<dt>projection</dt>
<dd><p>Reference genome build for the coordinates in the MAF file. The default is hg19 genome build.</p></dd>


<dt>seq_type</dt>
<dd><p>The seq_type you want back, default is genome.</p></dd>


<dt>basic_columns</dt>
<dd><p>Set to FALSE to override the default behavior of returning only the first 45 columns of MAF data.</p></dd>


<dt>maf_cols</dt>
<dd><p>if basic_columns is set to FALSE, the user can specify what columns to be returned within the MAF. This parameter can either be a vector of indexes (integer) or a vector of characters (matching columns in MAF).</p></dd>


<dt>from_flatfile</dt>
<dd><p>Set to TRUE to obtain mutations from a local flatfile instead of the database. This can be more efficient and is currently the only option for users who do not have ICGC data access.</p></dd>


<dt>augmented</dt>
<dd><p>default: TRUE. Set to FALSE if you instead want the original MAF from each sample for multi-sample patients instead of the augmented MAF</p></dd>


<dt>min_read_support</dt>
<dd><p>Only returns variants with at least this many reads in t_alt_count (for cleaning up augmented MAFs)</p></dd>


<dt>groups</dt>
<dd><p>Unix groups for the samples to be included. Default is both gambl and icgc_dart samples.</p></dd>


<dt>include_silent</dt>
<dd><p>Logical parameter indicating whether to include silent mutations into coding mutations. Default is TRUE.</p></dd>


<dt>engine</dt>
<dd><p>Specify one of readr or fread_maf (default) to change how the large files are loaded prior to subsetting. You may have better performance with one or the other but for me fread_maf is faster and uses a lot less RAM.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A data frame containing all the MAF data columns (one row per mutation).</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Effectively retrieve coding SSM calls. Multiple filtering parameters are available for this function.
For more information on how to implement the filtering parameters, refer to the parameter descriptions as well as examples in the vignettes.
Is this function not what you are looking for? Try one of the following, similar, functions; <a href="get_coding_ssm_status.html">get_coding_ssm_status</a>,
<a href="get_ssm_by_patients.html">get_ssm_by_patients</a>, <a href="get_ssm_by_sample.html">get_ssm_by_sample</a>, <a href="get_ssm_by_samples.html">get_ssm_by_samples</a>, <a href="get_ssm_by_region.html">get_ssm_by_region</a>, <a href="get_ssm_by_regions.html">get_ssm_by_regions</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">#basic usage</span></span></span>
<span class="r-in"><span><span class="va">maf_data</span> <span class="op">=</span> <span class="fu">get_coding_ssm</span><span class="op">(</span>seq_type <span class="op">=</span> <span class="st">"genome"</span>, limit_cohort <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"BL_ICGC"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> reading from: /projects/nhl_meta_analysis_scratch/gambl/results_local/all_the_things/slms_3-1.0_vcf2maf-1.3/genome--projection/deblacklisted/augmented_maf/all_slms-3--grch37.CDS.maf</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> mutations from 1652 samples</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> after linking with metadata, we have mutations from 17 samples</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">maf_data</span> <span class="op">=</span> <span class="fu">get_coding_ssm</span><span class="op">(</span>seq_type <span class="op">=</span> <span class="st">"genome"</span>, limit_samples <span class="op">=</span> <span class="st">"HTMCP-01-06-00485-01A-01D"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> reading from: /projects/nhl_meta_analysis_scratch/gambl/results_local/all_the_things/slms_3-1.0_vcf2maf-1.3/genome--projection/deblacklisted/augmented_maf/all_slms-3--grch37.CDS.maf</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> mutations from 1652 samples</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> after linking with metadata, we have mutations from 1 samples</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ryan Morin, Kostia Dreval, Adam Mattsson.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

