<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Get SSM By Regions. — get_ssm_by_regions • GAMBLR</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Get SSM By Regions. — get_ssm_by_regions"><meta property="og:description" content="Efficiently retrieve all mutations across a range of genomic regions."><meta property="og:image" content="https://github.com/morinlab/GAMBLR/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-12345678-0"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-12345678-0');
</script></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9500</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-light fa-memo-circle-info"></span>
     
    How To
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="https://morinlab.github.io/GAMBLR/#installation" class="external-link">Install</a>
    </li>
    <li>
      <a href="https://morinlab.github.io/GAMBLR/#running-gamblr-on-your-own-computer" class="external-link">Running GAMBLR On Your Own Computer</a>
    </li>
    <li>
      <a href="https://morinlab.github.io/GAMBLR/#contributing" class="external-link">Contribute</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-light fa-person-chalkboard"></span>
     
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/ssm_tutorial.html">SSM</a>
    </li>
    <li>
      <a href="../articles/cn_sv_tutorial.html">CN &amp; SV</a>
    </li>
    <li>
      <a href="../articles/fancy_plots.html">Fancy Plots</a>
    </li>
    <li>
      <a href="../articles/utilities_tutorial.html">Utilities</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="../reference/index.html">
    <span class="fa fa-file-code-o"></span>
     
    Functions &amp; Bundled Data
  </a>
</li>
<li>
  <a href="https://github.com/morinlab/GAMBLR">
    <span class="fa fa-github fa-lg"></span>
     
    github
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Get SSM By Regions.</h1>
    
    <div class="hidden name"><code>get_ssm_by_regions.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Efficiently retrieve all mutations across a range of genomic regions.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_ssm_by_regions</span><span class="op">(</span></span>
<span>  <span class="va">regions_list</span>,</span>
<span>  <span class="va">regions_bed</span>,</span>
<span>  streamlined <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  maf_data <span class="op">=</span> <span class="va">maf_data</span>,</span>
<span>  use_name_column <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  from_indexed_flatfile <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"slms-3"</span>,</span>
<span>  augmented <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  seq_type <span class="op">=</span> <span class="st">"genome"</span>,</span>
<span>  projection <span class="op">=</span> <span class="st">"grch37"</span>,</span>
<span>  min_read_support <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  basic_columns <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>regions_list</dt>
<dd><p>Either provide a vector of regions in the chr:start-end format OR.</p></dd>


<dt>regions_bed</dt>
<dd><p>Better yet, provide a bed file with the coordinates you want to retrieve.</p></dd>


<dt>streamlined</dt>
<dd><p>Return a basic rather than full MAF format, default is TRUE.</p></dd>


<dt>maf_data</dt>
<dd><p>Use an already loaded MAF data frame.</p></dd>


<dt>use_name_column</dt>
<dd><p>If your bed-format data frame has a name column (must be named "name") these can be used to name your regions.</p></dd>


<dt>from_indexed_flatfile</dt>
<dd><p>Set to TRUE to avoid using the database and instead rely on flatfiles (only works for streamlined data, not full MAF details).</p></dd>


<dt>mode</dt>
<dd><p>Only works with indexed flatfiles. Accepts 2 options of "slms-3" and "strelka2" to indicate which variant caller to use. Default is "slms-3".</p></dd>


<dt>augmented</dt>
<dd><p>default: TRUE. Set to FALSE if you instead want the original MAF from each sample for multi-sample patients instead of the augmented MAF</p></dd>


<dt>seq_type</dt>
<dd><p>The seq_type you want back, default is genome.</p></dd>


<dt>projection</dt>
<dd><p>Obtain variants projected to this reference (one of grch37 or hg38).</p></dd>


<dt>min_read_support</dt>
<dd><p>Only returns variants with at least this many reads in t_alt_count (for cleaning up augmented MAFs).</p></dd>


<dt>basic_columns</dt>
<dd><p>Boolean parameter set to FALSE per default. Set to TRUE to return fewer columns.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>Returns a data frame of variants in MAF-like format.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>This function internally calls <a href="get_ssm_by_region.html">get_ssm_by_region</a> to retrieve SSM calls for the specified regions.
See parameter descriptions for <a href="get_ssm_by_region.html">get_ssm_by_region</a> for more information on how the different parameters can be called.
Is this function not what you are looking for? Try one of the following, similar, functions; <a href="get_coding_ssm.html">get_coding_ssm</a>,
<a href="get_coding_ssm_status.html">get_coding_ssm_status</a>, <a href="get_ssm_by_patients.html">get_ssm_by_patients</a>, <a href="get_ssm_by_sample.html">get_ssm_by_sample</a>,
<a href="get_ssm_by_samples.html">get_ssm_by_samples</a>, <a href="get_ssm_by_region.html">get_ssm_by_region</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">#basic usage, adding custom names from bundled ashm data frame</span></span></span>
<span class="r-in"><span><span class="va">regions_bed</span> <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span><span class="va">grch37_ashm_regions</span>, name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">gene</span>, <span class="va">region</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in dplyr::mutate(grch37_ashm_regions, name = paste(gene, region,     sep = "_")):</span> object 'grch37_ashm_regions' not found</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">ashm_basic_details</span> <span class="op">=</span> <span class="fu">get_ssm_by_regions</span><span class="op">(</span>regions_bed <span class="op">=</span> <span class="va">regions_bed</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in apply(regions_bed, 1, bed2region):</span> object 'regions_bed' not found</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">full_details_maf</span> <span class="op">=</span> <span class="fu">get_ssm_by_regions</span><span class="op">(</span>regions_bed <span class="op">=</span> <span class="va">regions_bed</span>,</span></span>
<span class="r-in"><span>                                      basic_columns<span class="op">=</span><span class="cn">T</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in apply(regions_bed, 1, bed2region):</span> object 'regions_bed' not found</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ryan Morin, Kostia Dreval, Adam Mattsson.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

