<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Get SSM By Patients. — get_ssm_by_patients • GAMBLR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Get SSM By Patients. — get_ssm_by_patients"><meta property="og:description" content="Get MAF-format data frame for more than one patient."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-12345678-0"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-12345678-0');
</script></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9500</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-light fa-memo-circle-info"></span>
     
    How To
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../path_once_integrated_on_github">Install</a>
    </li>
    <li>
      <a href="../path_once_integrated_on_github">Run Remote On A Local Machine</a>
    </li>
    <li>
      <a href="../path_once_integrated_on_github">Contribute</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-light fa-person-chalkboard"></span>
     
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/ssm_tutorial.html">SSM</a>
    </li>
    <li>
      <a href="../articles/cn_sv_tutorial.html">CN &amp; SV</a>
    </li>
    <li>
      <a href="../articles/fancy_plots.html">Fancy Plots</a>
    </li>
    <li>
      <a href="../articles/utiliteis_tutorial.html">Utilities</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="../reference/index.html">
    <span class="fa fa-file-code-o"></span>
     
    Functions &amp; Bundled Data
  </a>
</li>
<li>
  <a href="https://github.com/morinlab/GAMBLR">
    <span class="fa fa-github fa-lg"></span>
     
    github
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Get SSM By Patients.</h1>
    
    <div class="hidden name"><code>get_ssm_by_patients.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Get MAF-format data frame for more than one patient.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_ssm_by_patients</span><span class="op">(</span></span>
<span>  <span class="va">these_patient_ids</span>,</span>
<span>  <span class="va">these_samples_metadata</span>,</span>
<span>  tool_name <span class="op">=</span> <span class="st">"slms-3"</span>,</span>
<span>  projection <span class="op">=</span> <span class="st">"grch37"</span>,</span>
<span>  seq_type <span class="op">=</span> <span class="st">"genome"</span>,</span>
<span>  flavour <span class="op">=</span> <span class="st">"clustered"</span>,</span>
<span>  min_read_support <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  basic_columns <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  maf_cols <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  subset_from_merge <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  augmented <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  engine <span class="op">=</span> <span class="st">"fread_maf"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>these_patient_ids</dt>
<dd><p>A vector of patient IDs that you want results for. The user can also use a metadata table that has been subset to the patient IDs of interest (<code>these_samples_metadata</code>).</p></dd>


<dt>these_samples_metadata</dt>
<dd><p>A metadata subset to contain the rows corresponding to the patients of interest. If the vector of patient IDs is missing (<code>these_patient_ids</code>), this function will default to all patient IDs in the metadata table given to this parameter.</p></dd>


<dt>tool_name</dt>
<dd><p>Only supports slms-3 currently.</p></dd>


<dt>projection</dt>
<dd><p>Obtain variants projected to this reference (one of grch37 or hg38).</p></dd>


<dt>seq_type</dt>
<dd><p>The seq type you want results for. Default is "genome".</p></dd>


<dt>flavour</dt>
<dd><p>Currently this function only supports one flavour option but this feature is meant for eventual compatibility with additional variant calling parameters/versions.</p></dd>


<dt>min_read_support</dt>
<dd><p>Only returns variants with at least this many reads in t_alt_count (for cleaning up augmented MAFs).</p></dd>


<dt>basic_columns</dt>
<dd><p>Return first 43 columns of MAF rather than full details. Default is TRUE.</p></dd>


<dt>maf_cols</dt>
<dd><p>if basic_columns is set to FALSE, the user can specify what columns to be returned within the MAF. This parameter can either be a vector of indexes (integer) or a vector of characters (matching columns in MAF).</p></dd>


<dt>subset_from_merge</dt>
<dd><p>Instead of merging individual MAFs, the data will be subset from a pre-merged MAF of samples with the specified <code>seq_type</code>.</p></dd>


<dt>augmented</dt>
<dd><p>default: TRUE. Set to FALSE if you instead want the original MAF from each sample for multi-sample patients instead.</p></dd>


<dt>engine</dt>
<dd><p>Specify one of readr or fread_maf (default) to change how the large files are loaded prior to subsetting. You may have better performance with one or the other but for me fread_maf is faster and uses a lot less RAM.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A data frame with SSM calls for the selected patients in MAF format.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>This function returns variants from a set of patients avoiding duplicated mutations from multiple samples from that patient (i.e. unique superset of variants).
This is done either by combining the contents of individual MAF files or subset from a merged MAF (wraps <a href="get_ssm_by_samples.html">get_ssm_by_samples</a>).
In most situations, this should never need to be run with <code>subset_from_merge = TRUE</code>. Instead use one of <a href="get_coding_ssm.html">get_coding_ssm</a> or <a href="get_ssm_by_region.html">get_ssm_by_region</a>.
This function expects either a vector of patient IDs (<code>thse_patients_ids</code>) or an already subset metadata table (<code>these_samples_metadata</code>).
Is this function not what you are looking for? Try one of the following, similar, functions; <a href="get_coding_ssm.html">get_coding_ssm</a>, <a href="get_coding_ssm_status.html">get_coding_ssm_status</a>,
<a href="get_ssm_by_sample.html">get_ssm_by_sample</a>, <a href="get_ssm_by_samples.html">get_ssm_by_samples</a>, <a href="get_ssm_by_region.html">get_ssm_by_region</a>, <a href="get_ssm_by_regions.html">get_ssm_by_regions</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">parallel</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#example 1, using a vector of patient IDs.</span></span></span>
<span class="r-in"><span><span class="va">patients</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"00-14595"</span>, <span class="st">"00-15201"</span>, <span class="st">"01-12047"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">patients_maf</span> <span class="op">=</span> <span class="fu">get_ssm_by_patients</span><span class="op">(</span>these_patient_ids <span class="op">=</span> <span class="va">patients</span>,</span></span>
<span class="r-in"><span>                                   seq_type <span class="op">=</span> <span class="st">"genome"</span>,</span></span>
<span class="r-in"><span>                                   subset_from_merge <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> WARNING: on-the-fly merges can be extremely slow and consume a lot of memory. Use at your own risk. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> WARNING: on-the-fly merges can be extremely slow and consume a lot of memory if many samples are involved. Use at your own risk. </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#example 2, using a metadata table, subset to the patient IDs of interest.</span></span></span>
<span class="r-in"><span><span class="va">patient_meta</span> <span class="op">=</span> <span class="fu"><a href="get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span>seq_type_filter <span class="op">=</span> <span class="st">"genome"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">patient_meta</span> <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">patient_meta</span>, <span class="va">patient_id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">patients</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">patients_maf_2</span> <span class="op">=</span> <span class="fu">get_ssm_by_patients</span><span class="op">(</span>these_samples_metadata <span class="op">=</span> <span class="va">patient_meta</span>,</span></span>
<span class="r-in"><span>                                     subset_from_merge <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> WARNING: on-the-fly merges can be extremely slow and consume a lot of memory. Use at your own risk. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> WARNING: on-the-fly merges can be extremely slow and consume a lot of memory if many samples are involved. Use at your own risk. </span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ryan Morin, Kostia Dreval, Adam Mattsson.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

