<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Mutation Frequency Heatmap — get_mutation_frequency_bin_matrix • GAMBLR</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Mutation Frequency Heatmap — get_mutation_frequency_bin_matrix"><meta property="og:description" content="Count hypermutated bins and generate heatmap/cluster the data."><meta property="og:image" content="https://github.com/morinlab/GAMBLR/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-12345678-0"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-12345678-0');
</script></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9500</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-light fa-memo-circle-info"></span>
     
    How To
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="https://morinlab.github.io/GAMBLR/#installation" class="external-link">Install</a>
    </li>
    <li>
      <a href="https://morinlab.github.io/GAMBLR/#running-gamblr-on-your-own-computer" class="external-link">Running GAMBLR On Your Own Computer</a>
    </li>
    <li>
      <a href="https://morinlab.github.io/GAMBLR/#contributing" class="external-link">Contribute</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-light fa-person-chalkboard"></span>
     
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/ssm_tutorial.html">SSM</a>
    </li>
    <li>
      <a href="../articles/cn_sv_tutorial.html">CN &amp; SV</a>
    </li>
    <li>
      <a href="../articles/fancy_plots.html">Fancy Plots</a>
    </li>
    <li>
      <a href="../articles/utilities_tutorial.html">Utilities</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="../reference/index.html">
    <span class="fa fa-file-code-o"></span>
     
    Functions &amp; Bundled Data
  </a>
</li>
<li>
  <a href="https://github.com/morinlab/GAMBLR">
    <span class="fa fa-github fa-lg"></span>
     
    github
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Mutation Frequency Heatmap</h1>
    
    <div class="hidden name"><code>get_mutation_frequency_bin_matrix.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Count hypermutated bins and generate heatmap/cluster the data.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_mutation_frequency_bin_matrix</span><span class="op">(</span></span>
<span>  <span class="va">regions</span>,</span>
<span>  <span class="va">regions_df</span>,</span>
<span>  <span class="va">these_samples_metadata</span>,</span>
<span>  seq_type <span class="op">=</span> <span class="st">"genome"</span>,</span>
<span>  region_padding <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  metadataColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pathology"</span><span class="op">)</span>,</span>
<span>  sortByColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pathology"</span><span class="op">)</span>,</span>
<span>  expressionColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  orientation <span class="op">=</span> <span class="st">"sample_rows"</span>,</span>
<span>  skip_regions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MYC"</span>, <span class="st">"BCL2"</span>, <span class="st">"IGLL5"</span><span class="op">)</span>,</span>
<span>  customColour <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  slide_by <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  window_size <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  min_count_per_bin <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  min_bin_recurrence <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  min_bin_patient <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  region_fontsize <span class="op">=</span> <span class="fl">8</span>,</span>
<span>  cluster_rows_heatmap <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  cluster_cols_heatmap <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  show_gene_colours <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  legend_row <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  legend_col <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  legend_direction <span class="op">=</span> <span class="st">"horizontal"</span>,</span>
<span>  legendFontSize <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  from_indexed_flatfile <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"slms-3"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>regions</dt>
<dd><p>Vector of regions in the format "chr:start-end".</p></dd>


<dt>regions_df</dt>
<dd><p>Data frame of regions with four columns (chrom,start,end,gene_name).</p></dd>


<dt>these_samples_metadata</dt>
<dd><p>GAMBL metadata subset to the cases you want to process (or full metadata).</p></dd>


<dt>seq_type</dt>
<dd><p>The seq_type you want back, default is genome.</p></dd>


<dt>region_padding</dt>
<dd><p>How many bases will be added on the left and right of the regions to ensure any small regions are sufficiently covered by bins. Default is  1000.</p></dd>


<dt>metadataColumns</dt>
<dd><p>What metadata will be shown in the visualization.</p></dd>


<dt>sortByColumns</dt>
<dd><p>Which of the metadata to sort on for the heatmap.</p></dd>


<dt>expressionColumns</dt>
<dd><p>Optional variable for retrieving expression values for a specific gene(s).</p></dd>


<dt>orientation</dt>
<dd><p>Specify the sample orientation, default is sample_rows.</p></dd>


<dt>skip_regions</dt>
<dd><p>Regions to be filtered out from the regions data frame. Only applies if <code>regions_df</code> is not provided. Default is MYC, BCL2 and IGLL5.</p></dd>


<dt>customColour</dt>
<dd><p>Optional named list of named vectors for specifying all colours for metadata. Can be generated with map_metadata_to_colours. Default is NULL.</p></dd>


<dt>slide_by</dt>
<dd><p>How far to shift before starting the next window.</p></dd>


<dt>window_size</dt>
<dd><p>The width of your sliding window.</p></dd>


<dt>min_count_per_bin</dt>
<dd><p>Minimum counts per bin, default is 3.</p></dd>


<dt>min_bin_recurrence</dt>
<dd><p>How many samples a bin must be mutated in to retain in the visualization.</p></dd>


<dt>min_bin_patient</dt>
<dd><p>How many bins must a patient mutated in to retain in the visualization.</p></dd>


<dt>region_fontsize</dt>
<dd><p>Font size of regions in plot, default is 8ppt.</p></dd>


<dt>cluster_rows_heatmap</dt>
<dd><p>Optional parameter to enable/disable clustering of each dimension of the heatmap. Default is FALSE.</p></dd>


<dt>cluster_cols_heatmap</dt>
<dd><p>Optional parameter to enable/disable clustering of each dimension of the heatmap. Default is FALSE.</p></dd>


<dt>show_gene_colours</dt>
<dd><p>Optional logical argument indicating whether regions should have associated colours plotted as annotation track of heatmap.</p></dd>


<dt>legend_row</dt>
<dd><p>Fiddle with these to widen or narrow your legend.</p></dd>


<dt>legend_col</dt>
<dd><p>Fiddle with these to widen or narrow your legend.</p></dd>


<dt>legend_direction</dt>
<dd><p>Accepts one of "horizontal" (default) or "vertical" to indicate in which direction the legend will be drawn.</p></dd>


<dt>legendFontSize</dt>
<dd><p>Font size of legend in plot, default is 10ppt.</p></dd>


<dt>from_indexed_flatfile</dt>
<dd><p>Set to TRUE to avoid using the database and instead rely on flat files (only works for streamlined data, not full MAF details).</p></dd>


<dt>mode</dt>
<dd><p>Only works with indexed flat files. Accepts 2 options of "slms-3" and "strelka2" to indicate which variant caller to use. Default is "slms-3".</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>Nothing</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>This function takes a metadata table with <code>these_samples_metadata</code> parameter and internally calls <a href="calc_mutation_frequency_sliding_windows.html">calc_mutation_frequency_sliding_windows</a> (that internally calls <a href="get_ssm_by_regions.html">get_ssm_by_regions</a>)
to retrieve mutations for plotting. This plotting function has a variety of useful parameters, providing many customizable plotting options. For more details on how these parameters can be used,
and extended usage examples, refer to the SSM tutorial vignette section 1.4.9.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">#load metadata.</span></span></span>
<span class="r-in"><span><span class="va">metadata</span> <span class="op">=</span> <span class="fu"><a href="get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "missing:  /Users/rmorin/git/gambl/data/metadata/gambl_all_outcomes.tsv"</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Have you cloned the GAMBL repo and added the path to this directory under the local section of your config?</span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error:</span> '/Users/rmorin/git/gambl/data/metadata/gambl_all_outcomes.tsv' does not exist.</span>
<span class="r-in"><span><span class="va">dlbcl_bl_meta</span> <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">metadata</span>, <span class="va">pathology</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DLBCL"</span>, <span class="st">"BL"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in dplyr::filter(metadata, pathology %in% c("DLBCL", "BL")):</span> object 'metadata' not found</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#bring together all derived sample-level results from many GAMBL pipelines.</span></span></span>
<span class="r-in"><span><span class="va">dlbcl_bl_meta</span> <span class="op">=</span> <span class="fu"><a href="collate_results.html">collate_results</a></span><span class="op">(</span>join_with_full_metadata <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>                                these_samples_metadata <span class="op">=</span> <span class="va">dlbcl_bl_meta</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "missing:  /Users/rmorin/git/gambl/data/metadata/gambl_all_outcomes.tsv"</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Have you cloned the GAMBL repo and added the path to this directory under the local section of your config?</span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error:</span> '/Users/rmorin/git/gambl/data/metadata/gambl_all_outcomes.tsv' does not exist.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#get ashm regions</span></span></span>
<span class="r-in"><span><span class="va">some_regions</span> <span class="op">=</span> <span class="va">grch37_ashm_regions</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in eval(expr, envir, enclos):</span> object 'grch37_ashm_regions' not found</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">get_mutation_frequency_bin_matrix</span><span class="op">(</span>these_samples_metadata <span class="op">=</span> <span class="va">dlbcl_bl_meta</span>,</span></span>
<span class="r-in"><span>                                  regions_df <span class="op">=</span> <span class="va">some_regions</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in apply(regions_df, 1, function(x) {    paste0(x[1], ":", as.numeric(x[2]) - region_padding, "-",         as.numeric(x[3]) + region_padding)}):</span> object 'some_regions' not found</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ryan Morin, Kostia Dreval, Adam Mattsson.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

