<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Get Manta SVs — get_manta_sv • GAMBLR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Get Manta SVs — get_manta_sv"><meta property="og:description" content="Retrieve Manta SVs and filter."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-12345678-0"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-12345678-0');
</script></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9500</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-light fa-memo-circle-info"></span>
     
    How To
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../path_once_integrated_on_github">Install</a>
    </li>
    <li>
      <a href="../path_once_integrated_on_github">Run Remote On A Local Machine</a>
    </li>
    <li>
      <a href="../path_once_integrated_on_github">Contribute</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-light fa-person-chalkboard"></span>
     
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/ssm_tutorial.html">SSM</a>
    </li>
    <li>
      <a href="../articles/cn_sv_tutorial.html">CN &amp; SV</a>
    </li>
    <li>
      <a href="../articles/fancy_plots.html">Fancy Plots</a>
    </li>
    <li>
      <a href="../articles/utiliteis_tutorial.html">Utilities</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="../reference/index.html">
    <span class="fa fa-file-code-o"></span>
     
    Functions &amp; Bundled Data
  </a>
</li>
<li>
  <a href="https://github.com/morinlab/GAMBLR">
    <span class="fa fa-github fa-lg"></span>
     
    github
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Get Manta SVs</h1>
    
    <div class="hidden name"><code>get_manta_sv.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Retrieve Manta SVs and filter.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_manta_sv</span><span class="op">(</span></span>
<span>  <span class="va">these_sample_ids</span>,</span>
<span>  <span class="va">these_samples_metadata</span>,</span>
<span>  projection <span class="op">=</span> <span class="st">"grch37"</span>,</span>
<span>  <span class="va">chromosome</span>,</span>
<span>  <span class="va">qstart</span>,</span>
<span>  <span class="va">qend</span>,</span>
<span>  <span class="va">region</span>,</span>
<span>  min_vaf <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  min_score <span class="op">=</span> <span class="fl">40</span>,</span>
<span>  pass <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">pairing_status</span>,</span>
<span>  from_flatfile <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>these_sample_ids</dt>
<dd><p>A vector of multiple sample_id (or a single sample ID as a string) that you want results for.</p></dd>


<dt>these_samples_metadata</dt>
<dd><p>A metadata table to auto-subset the data to samples in that table before returning.</p></dd>


<dt>projection</dt>
<dd><p>The projection genome build.</p></dd>


<dt>chromosome</dt>
<dd><p>Optional, the chromosome you are restricting to.</p></dd>


<dt>qstart</dt>
<dd><p>Optional, query start coordinate of the range you are restricting to.</p></dd>


<dt>qend</dt>
<dd><p>Optional, query end coordinate of the range you are restricting to.</p></dd>


<dt>region</dt>
<dd><p>Optional, region formatted like chrX:1234-5678 instead of specifying chromosome, start and end separately.</p></dd>


<dt>min_vaf</dt>
<dd><p>The minimum tumour VAF for a SV to be returned. Default is 0.1.</p></dd>


<dt>min_score</dt>
<dd><p>The lowest Manta somatic score for a SV to be returned. Default is 40.</p></dd>


<dt>pass</dt>
<dd><p>If set to TRUE, only return SVs that are annotated with PASS in the FILTER column. Set to FALSE to keep all variants, regardless if they PASS the filters. Default is TRUE.</p></dd>


<dt>pairing_status</dt>
<dd><p>Use to restrict results (if desired) to matched or unmatched results (default is to return all).</p></dd>


<dt>from_flatfile</dt>
<dd><p>Set to TRUE by default, FALSE is no longer supported (database).</p></dd>


<dt>verbose</dt>
<dd><p>Set to FALSE to prevent the path of the requested bedpe file to be printed.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A data frame in a bedpe-like format with additional columns that allow filtering of high-confidence SVs.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Return Manta SVs with aditional VCF information to allow for filtering of high-confidence variants.
To return SV calls for multiple samples, give <code>these_sample_ids</code> a vector of sample IDs, if only one sample is desired,
give this parameter one sample ID, as a string (or a vector of characters). The user can also call the <code>these_samples_metadata</code>
parameter to make use of an already subset metadata table. In this case, the returned calls will be restricted to the sample_ids
within that data frame. This function relies on a set of specific functions to be successful in returning SV calls for any
available sample in gambl. First, this function calls <a href="get_combined_sv.html">get_combined_sv</a> and performs an <code>anit_join</code> with the full metadata to
identify what samples are currently missing from the return of <a href="get_combined_sv.html">get_combined_sv</a>. This function then calls <a href="get_manta_sv_by_samples.html">get_manta_sv_by_samples</a>
(wrapper function for <a href="get_manta_sv_by_sample.html">get_manta_sv_by_sample</a>) on the subset of the missing samples. The merged calls are subject to any
filtering that is specified within this function. This function can also restrict the returned calls to any genomic regions
specified within <code>chromosome</code>, <code>qstart</code>, <code>qend</code>, or the complete region specified under <code>region</code> (in chr:start-end format).
Useful filtering parameters are also available, use <code>min_vaf</code> to set the minimum tumour VAF for a SV to be returned and <code>min_score</code>
to set the lowest Manta somatic score for a SV to be returned. <code>pair_status</code> can be used to only return variants that are
annotated with PASS in the filtering column (VCF).
Is this function not what you are looking for? Try one of the following, similar, functions;
<a href="get_combined_sv.html">get_combined_sv</a>, <a href="get_manta_sv_by_sample.html">get_manta_sv_by_sample</a>, <a href="get_manta_sv_by_samples.html">get_manta_sv_by_samples</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">#lazily get every SV in the table with default quality filters</span></span></span>
<span class="r-in"><span><span class="va">all_sv</span> <span class="op">=</span> <span class="fu">get_manta_sv</span><span class="op">(</span>verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> WARNING! No SV calls found in flat-file for: 171116-PL02</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> WARNING! No SV calls found in flat-file for: 171447-PL01</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> WARNING! No SV calls found in flat-file for: 171451-PL01</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#get all SVs for a single sample</span></span></span>
<span class="r-in"><span><span class="va">some_sv</span> <span class="op">=</span> <span class="fu">get_manta_sv</span><span class="op">(</span>these_sample_ids <span class="op">=</span> <span class="st">"94-15772_tumorA"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#get the SVs in a region around MYC</span></span></span>
<span class="r-in"><span><span class="va">myc_locus_sv</span> <span class="op">=</span> <span class="fu">get_manta_sv</span><span class="op">(</span>region <span class="op">=</span> <span class="st">"8:128723128-128774067"</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> WARNING! No SV calls found in flat-file for: 171116-PL02</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> WARNING! No SV calls found in flat-file for: 171447-PL01</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> WARNING! No SV calls found in flat-file for: 171451-PL01</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#get SVs for multiple samples, using these_samples_id</span></span></span>
<span class="r-in"><span><span class="va">my_metadata</span> <span class="op">=</span> <span class="fu"><a href="get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">these_samples</span> <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">my_metadata</span>, <span class="va">sample_id</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">my_samples_df</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">these_samples</span>, <span class="fl">10</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">my_samples</span> <span class="op">=</span> <span class="fu">pull</span><span class="op">(</span><span class="va">my_samples_df</span>, <span class="va">sample_id</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in pull(my_samples_df, sample_id):</span> could not find function "pull"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">my_svs_2</span> <span class="op">=</span> <span class="fu">get_manta_sv</span><span class="op">(</span>these_sample_ids <span class="op">=</span> <span class="va">my_samples</span>,</span></span>
<span class="r-in"><span>                        projection <span class="op">=</span> <span class="st">"hg38"</span>,</span></span>
<span class="r-in"><span>                        verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in dplyr::filter(., sample_id %in% these_sample_ids):</span> Problem while computing `..1 = sample_id %in% these_sample_ids`.</span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span style="font-weight: bold;">Caused by error in `h()`:</span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">!</span> error in evaluating the argument 'table' in selecting a method for function '%in%': object 'my_samples' not found</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#get SVs for multiple samples using a metadata table and with no VAF/score filtering</span></span></span>
<span class="r-in"><span><span class="va">my_metadata</span> <span class="op">=</span> <span class="fu"><a href="get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> </span></span>
<span class="r-in"><span><span class="va">this_metadata</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">my_metadata</span>, <span class="fl">10</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in get_gambl_metadata() %&gt;% this_metadata = head(my_metadata, 10):</span> invalid (NULL) left side of assignment</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">my_svs</span> <span class="op">=</span> <span class="fu">get_manta_sv</span><span class="op">(</span>these_samples_metadata <span class="op">=</span> <span class="va">this_metadata</span>,</span></span>
<span class="r-in"><span>                      verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>                      min_vaf <span class="op">=</span> <span class="fl">0</span>,</span></span>
<span class="r-in"><span>                      min_score <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in get_manta_sv(these_samples_metadata = this_metadata, verbose = FALSE,     min_vaf = 0, min_score = 0):</span> object 'this_metadata' not found</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ryan Morin, Kostia Dreval, Adam Mattsson.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

