<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Grande et al. MAF. — grande_maf • GAMBLR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Grande et al. MAF. — grande_maf"><meta property="og:description" content="A MAF (data frame) drawn from the Grande et al. dataset."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9500</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/cn_sv_tutorial.html">CN &amp; SV Tutorial - GAMBLR</a>
    </li>
    <li>
      <a href="../articles/fancy_plots.html">GAMBLR: Fancy Plotting Functions</a>
    </li>
    <li>
      <a href="../articles/ssm_tutorial.html">SSM Tutorial - GAMBLR</a>
    </li>
    <li>
      <a href="../articles/utilities_tutorial.html">Utilities Tutorial - GAMBLR</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Grande et al. MAF.</h1>
    
    <div class="hidden name"><code>grande_maf.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>A MAF (data frame) drawn from the Grande et al. dataset.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="va">grande_maf</span></span></code></pre></div>
    </div>

    <div id="format">
    <h2>Format</h2>
    
<div class="section">
<h3 id="grande-maf"><code>grande_maf</code><a class="anchor" aria-label="anchor" href="#grande-maf"></a></h3>


<p>A MAF in data frame format. 12251 rows and 125 columns.</p><dl><dt>Hugo_Symbol</dt>
<dd><p>HUGO symbol for the gene (HUGO symbols are always in all caps). "Unknown" is used for regions that do not correspond to a gene</p></dd>

<dt>Entrez_Gene_ID</dt>
<dd><p>Entrez gene ID (an integer). "0" is used for regions that do not correspond to a gene region or Ensembl ID</p></dd>

<dt>Centre</dt>
<dd><p>One or more genome sequencing center reporting the variant</p></dd>

<dt>NCBI_Build</dt>
<dd><p>The reference genome used for the alignment</p></dd>

<dt>Chromosome</dt>
<dd><p>The affected chromosome</p></dd>

<dt>Start_Position</dt>
<dd><p>Lowest numeric position of the reported variant on the genomic reference sequence. Mutation start coordinate</p></dd>

<dt>End_Position</dt>
<dd><p>Highest numeric genomic position of the reported variant on the genomic reference sequence. Mutation end coordinate</p></dd>

<dt>Strand</dt>
<dd><p>Genomic strand of the reported allele. Currently, all variants will report the positive strand: '+'</p></dd>

<dt>Variant_Classification</dt>
<dd><p>Translational effect of variant allele</p></dd>

<dt>Variant_Type</dt>
<dd><p>Type of mutation. TNP (tri-nucleotide polymorphism) is analogous to DNP (di-nucleotide polymorphism) but for three consecutive nucleotides. ONP (oligo-nucleotide polymorphism) is analogous to TNP but for consecutive runs of four or more (SNP, DNP, TNP, ONP, INS, DEL, or Consolidated)</p></dd>

<dt>Reference_Allele</dt>
<dd><p>The plus strand reference allele at this position. Includes the deleted sequence for a deletion or "-" for an insertion</p></dd>

<dt>Tumor_Seq_Allele1</dt>
<dd><p>Primary data genotype for tumor sequencing (discovery) allele 1. A "-" symbol for a deletion represents a variant. A "-" symbol for an insertion represents wild-type allele. Novel inserted sequence for insertion does not include flanking reference bases</p></dd>

<dt>Tumor_Seq_Allele2</dt>
<dd><p>Tumor sequencing (discovery) allele 2</p></dd>

<dt>dbSNP_RS</dt>
<dd><p>The rs-IDs from the dbSNP database, "novel" if not found in any database used, or null if there is no dbSNP record, but it is found in other databases</p></dd>

<dt>dbSNP_Val_Status</dt>
<dd><p>The dbSNP validation status is reported as a semicolon-separated list of statuses. The union of all rs-IDs is taken when there are multiple</p></dd>

<dt>Tumor_Sample_Barcode</dt>
<dd><p>Aliquot barcode for the tumor sample</p></dd>

<dt>Matched_Norm_Sample_Barcode</dt>
<dd><p>Aliquot barcode for the matched normal sample</p></dd>

<dt>Match_Norm_Seq_Allele1</dt>
<dd><p>Primary data genotype. Matched normal sequencing allele 1. A "-" symbol for a deletion represents a variant. A "-" symbol for an insertion represents wild-type allele. Novel inserted sequence for insertion does not include flanking reference bases (cleared in somatic MAF)</p></dd>

<dt>Match_Norm_Seq_Allele2</dt>
<dd><p>Matched normal sequencing allele 2</p></dd>

<dt>Tumor_Validation_Allele1</dt>
<dd><p>Secondary data from orthogonal technology. Tumor genotyping (validation) for allele 1. A "-" symbol for a deletion represents a variant. A "-" symbol for an insertion represents wild-type allele. Novel inserted sequence for insertion does not include flanking reference bases</p></dd>

<dt>Tumor_Validation_Allele2</dt>
<dd><p>Secondary data from orthogonal technology. Tumor genotyping (validation) for allele 2</p></dd>

<dt>Verification_Status</dt>
<dd><p>Second pass results from independent attempt using same methods as primary data source. Generally reserved for 3730 Sanger Sequencing</p></dd>

<dt>Validation_Status</dt>
<dd><p>Second pass results from orthogonal technology</p></dd>

<dt>Mutation_Status</dt>
<dd><p>An assessment of the mutation as somatic, germline, LOH, post transcriptional modification, unknown, or none. The values allowed in this field are constrained by the value in the Validation_Status field</p></dd>

<dt>Sequencing_Phase</dt>
<dd><p>TCGA sequencing phase (if applicable). Phase should change under any circumstance that the targets under consideration change</p></dd>

<dt>Sequence_Source</dt>
<dd><p>Molecular assay type used to produce the analytes used for sequencing. Allowed values are a subset of the SRA 1.5 library_strategy field values. This subset matches those used at CGHub</p></dd>

<dt>Validation_Method</dt>
<dd><p>The assay platforms used for the validation call</p></dd>

<dt>Score</dt>
<dd><p>Boolean variable</p></dd>

<dt>BAM_File</dt>
<dd><p>Boolean column stating if BAM file exists or not</p></dd>

<dt>Sequencer</dt>
<dd><p>Instrument used to produce primary sequence data</p></dd>

<dt>Tumor_Sample_UUID</dt>
<dd><p>GDC aliquot UUID for tumor sample</p></dd>

<dt>Matched_Norm_Sample_UUID</dt>
<dd><p>GDC aliquot UUID for matched normal sample</p></dd>

<dt>HGVSc</dt>
<dd><p>The coding sequence of the variant in HGVS recommended format</p></dd>

<dt>HGVSp</dt>
<dd><p>The protein sequence of the variant in HGVS recommended format. "p.=" signifies no change in the protein</p></dd>

<dt>HGVS_Short</dt>
<dd><p>Same as the HGVSp column, but using 1-letter amino-acid codes</p></dd>

<dt>Transcript_ID</dt>
<dd><p>Ensembl ID of the transcript affected by the varian</p></dd>

<dt>Exon_Number</dt>
<dd><p>The exon number (out of total number)</p></dd>

<dt>t_depth</dt>
<dd><p>Read depth across this locus in tumor BAM</p></dd>

<dt>t_ref_count</dt>
<dd><p>Read depth supporting the reference allele in tumor BAM</p></dd>

<dt>t_alt_count</dt>
<dd><p>Read depth supporting the variant allele in tumor BAM</p></dd>

<dt>n_depth</dt>
<dd><p>Read depth across this locus in normal BAM</p></dd>

<dt>n_ref_count</dt>
<dd><p>Read depth supporting the reference allele in normal BAM (cleared in somatic MAF)</p></dd>

<dt>n_alt_count</dt>
<dd><p>Read depth supporting the variant allele in normal BAM (cleared in somatic MAF)</p></dd>

<dt>all_effects</dt>
<dd><p>A semicolon delimited list of all possible variant effects, sorted by priority</p></dd>

<dt>Allele</dt>
<dd><p>The variant allele used to calculate the consequence</p></dd>

<dt>Gene</dt>
<dd><p>Stable Ensembl ID of affected gene</p></dd>

<dt>Feature</dt>
<dd><p>Stable Ensembl ID of feature (transcript, regulatory, motif)</p></dd>

<dt>Feature_type</dt>
<dd><p>Type of feature. Currently one of Transcript, RegulatoryFeature, MotifFeature (or blank)</p></dd>

<dt>Consequence</dt>
<dd><p>Consequence type of this variant; sequence ontology terms</p></dd>

<dt>cDNA_position</dt>
<dd><p>Relative position of base pair in the cDNA sequence as a fraction. A "-" symbol is displayed as the numerator if the variant does not appear in cDNA</p></dd>

<dt>CDS_position</dt>
<dd><p>Relative position of base pair in coding sequence. A "-" symbol is displayed as the numerator if the variant does not appear in coding sequence</p></dd>

<dt>Protein_position</dt>
<dd><p>Relative position of affected amino acid in protein. A "-" symbol is displayed as the numerator if the variant does not appear in coding sequence</p></dd>

<dt>Amino_acids</dt>
<dd><p>Only given if the variation affects the protein-coding sequence</p></dd>

<dt>Codons</dt>
<dd><p>The alternative codons with the variant base in upper case</p></dd>

<dt>Existing_variation</dt>
<dd><p>Known identifier of existing variation</p></dd>

<dt>ALLELE_NUM</dt>
<dd><p>Allele number from input; 0 is reference, 1 is first alternate etc.</p></dd>

<dt>DISTANCE</dt>
<dd><p>Shortest distance from the variant to transcript</p></dd>

<dt>STRAND_VEP</dt>
<dd></dd>

<dt>SYMBOL</dt>
<dd><p>The gene symbol</p></dd>

<dt>SYMBOL_SOURCE</dt>
<dd><p>The source of the gene symbol</p></dd>

<dt>HGNC_ID</dt>
<dd><p>Gene identifier from the HUGO Gene Nomenclature Committee if applicable</p></dd>

<dt>BIOTYPE</dt>
<dd><p>Biotype of transcript</p></dd>

<dt>CANONICAL</dt>
<dd><p>A flag (YES) indicating that the VEP-based canonical transcript, the longest translation, was used for this gene. If not, the value is null</p></dd>

<dt>CCDS</dt>
<dd><p>The CCDS identifier for this transcript, where applicable</p></dd>

<dt>ENSP</dt>
<dd><p>The Ensembl protein identifier of the affected transcript</p></dd>

<dt>SWISSPROT</dt>
<dd><p>UniProtKB/Swiss-Prot accession</p></dd>

<dt>TREMBL</dt>
<dd><p>UniProtKB/TrEMBL identifier of protein product</p></dd>

<dt>UNIPARC</dt>
<dd><p>UniParc identifier of protein product</p></dd>

<dt>RefSeq</dt>
<dd><p>RefSeq identifier for this transcript</p></dd>

<dt>SIFT</dt>
<dd><p>The SIFT prediction and/or score, with both given as prediction (score)</p></dd>

<dt>PolyPhen</dt>
<dd><p>The PolyPhen prediction and/or score</p></dd>

<dt>EXON</dt>
<dd><p>The exon number (out of total number)</p></dd>

<dt>INTRON</dt>
<dd><p>The intron number (out of total number)</p></dd>

<dt>DOMAINS</dt>
<dd><p>The source and identifier of any overlapping protein domains</p></dd>

<dt>GMAF</dt>
<dd><p>Non-reference allele and frequency of existing variant in 1000 Genomes</p></dd>

<dt>GMAF_Allele</dt>
<dd><p>Non-reference allele and frequency of existing variant in 1000 Genomes</p></dd>

<dt>GMAF_AF</dt>
<dd><p>Non-reference allele and frequency of existing variant in 1000 Genomes</p></dd>

<dt>AFR_MAF</dt>
<dd><p>Non-reference allele and frequency of existing variant in 1000 Genomes combined African population</p></dd>

<dt>AMR_MAF</dt>
<dd><p>Non-reference allele and frequency of existing variant in 1000 Genomes combined American population</p></dd>

<dt>ASN_MAF</dt>
<dd><p>Non-reference allele and frequency of existing variant in 1000 Genomes combined Asian population</p></dd>

<dt>EAS_MAF</dt>
<dd><p>Non-reference allele and frequency of existing variant in 1000 Genomes combined East Asian population</p></dd>

<dt>EUR_MAF</dt>
<dd><p>Non-reference allele and frequency of existing variant in 1000 Genomes combined European population</p></dd>

<dt>SAS_MAF</dt>
<dd><p>Non-reference allele and frequency of existing variant in 1000 Genomes combined South Asian population</p></dd>

<dt>AA_MAF</dt>
<dd><p>Non-reference allele and frequency of existing variant in NHLBI-ESP African American population</p></dd>

<dt>EA_MAF</dt>
<dd><p>Non-reference allele and frequency of existing variant in NHLBI-ESP European American population</p></dd>

<dt>CLIN_SIG</dt>
<dd><p>Clinical significance of variant from dbSNP as annotated in ClinVar</p></dd>

<dt>SOMATIC</dt>
<dd><p>Somatic status of each ID reported under Existing_variation (0, 1, or null)</p></dd>

<dt>PUBMED</dt>
<dd><p>Pubmed ID(s) of publications that cite existing variant</p></dd>

<dt>MOTIF_NAME</dt>
<dd><p>The source and identifier of a transcription factor binding profile aligned at this position</p></dd>

<dt>MOTIF_POS</dt>
<dd><p>The relative position of the variation in the aligned TFBP</p></dd>

<dt>HIGH_INF_POS</dt>
<dd><p>A flag indicating if the variant falls in a high information position of a transcription factor binding profile (TFBP) (Y, N, or null)</p></dd>

<dt>MOTIF_SCORE_CHANGE</dt>
<dd><p>The difference in motif score of the reference and variant sequences for the TFBP</p></dd>

<dt>IMAPCT</dt>
<dd><p>The impact modifier for the consequence type</p></dd>

<dt>PICK</dt>
<dd><p>Indicates if this block of consequence data was picked by VEP's pick feature (1 or null)</p></dd>

<dt>VARIANT_CLASS</dt>
<dd><p>Sequence Ontology variant class</p></dd>

<dt>TSL</dt>
<dd><p>Transcript support level, which is based on independent RNA analyses</p></dd>

<dt>HGVS_OFFSET</dt>
<dd><p>Indicates by how many bases the HGVS notations for this variant have been shifted</p></dd>

<dt>PHENO</dt>
<dd><p>Indicates if existing variant is associated with a phenotype, disease or trait (0, 1, or null)</p></dd>

<dt>MINIMISED</dt>
<dd><p>Alleles in this variant have been converted to minimal representation before consequence calculation (1 or null)</p></dd>

<dt>ExAC_AF</dt>
<dd><p>Global Allele Frequency from ExAC</p></dd>

<dt>ExAC_AF_AFR</dt>
<dd><p>African/African American Allele Frequency from ExAC</p></dd>

<dt>ExAC_AF_AMR</dt>
<dd><p>American Allele Frequency from ExAC</p></dd>

<dt>ExAC_AF_EAS</dt>
<dd><p>East Asian Allele Frequency from ExAC</p></dd>

<dt>ExAC_AF_FIN</dt>
<dd><p>Finnish Allele Frequency from ExAC</p></dd>

<dt>ExAC_AF_NFE</dt>
<dd><p>Non-Finnish European Allele Frequency from ExAC</p></dd>

<dt>ExAC_AF_OTH</dt>
<dd><p>Other Allele Frequency from ExAC</p></dd>

<dt>ExAC_AF_SAS</dt>
<dd><p>South Asian Allele Frequency from ExAC</p></dd>

<dt>GENE_PHENO</dt>
<dd><p>Indicates if gene that the variant maps to is associated with a phenotype, disease or trait (0, 1, or null)</p></dd>

<dt>FILTER</dt>
<dd><p>Copied from input VCF. This includes filters implemented directly by the variant caller and other external software used in the DNA-Seq pipeline. See below for additional details.</p></dd>

<dt>flanking_bps</dt>
<dd><p>The flanking basepairs</p></dd>

<dt>variant_id</dt>
<dd><p>Variant ID</p></dd>

<dt>variant_qual</dt>
<dd><p>Variant quality</p></dd>

<dt>ExAC_AF_Adj</dt>
<dd><p>Adjusted Global Allele Frequency from ExAC</p></dd>

<dt>ExAC_AC_AN_Adj</dt>
<dd><p>Adjusted Global Allele Frequency from ExAC</p></dd>

<dt>ExAC_AC_AN</dt>
<dd><p>Global Allele Frequency from ExAC</p></dd>

<dt>ExAC_AC_AN_AFR</dt>
<dd><p>African/African American Allele Frequency from ExAC</p></dd>

<dt>ExAC_AC_AN_AMR</dt>
<dd><p>American Allele Frequency from ExAC</p></dd>

<dt>ExAC_AC_AN_EAS</dt>
<dd><p>East Asian Allele Frequency from ExAC</p></dd>

<dt>ExAC_AC_AN_FIN</dt>
<dd><p>Finnish Allele Frequency from ExAC</p></dd>

<dt>ExAC_AC_AN_NFE</dt>
<dd><p>Non-Finnish European Allele Frequency from ExAC</p></dd>

<dt>ExAC_AC_AN_OTH</dt>
<dd><p>Other Allele Frequency from ExAC</p></dd>

<dt>ExAC_AC_AN_SAS</dt>
<dd><p>South Asian Allele Frequency from ExAC</p></dd>

<dt>ExAC_FILTER</dt>
<dd><p>Filter</p></dd>


</dl></div>

    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ryan Morin.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

