<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ GAMBLR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-12345678-0"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-12345678-0');
</script></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9500</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-light fa-memo-circle-info"></span>
     
    How To
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="https://morinlab.github.io/GAMBLR/#installation" class="external-link">Install</a>
    </li>
    <li>
      <a href="https://morinlab.github.io/GAMBLR/#run-remote-on-a-local-machine" class="external-link">Run Remote On A Local Machine</a>
    </li>
    <li>
      <a href="https://morinlab.github.io/GAMBLR/#contributing" class="external-link">Contribute</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-light fa-person-chalkboard"></span>
     
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/ssm_tutorial.html">SSM</a>
    </li>
    <li>
      <a href="../articles/cn_sv_tutorial.html">CN &amp; SV</a>
    </li>
    <li>
      <a href="../articles/fancy_plots.html">Fancy Plots</a>
    </li>
    <li>
      <a href="../articles/utiliteis_tutorial.html">Utilities</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="../reference/index.html">
    <span class="fa fa-file-code-o"></span>
     
    Functions &amp; Bundled Data
  </a>
</li>
<li>
  <a href="https://github.com/morinlab/GAMBLR">
    <span class="fa fa-github fa-lg"></span>
     
    github
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="annotations">Annotations <a href="#annotations" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions used to annotate GAMBL data.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="annotate_driver_ssm.html">annotate_driver_ssm()</a></code> </p>
        </td>
        <td><p>Annotate Driver SSM.</p></td>
      </tr><tr><td>
          <p><code><a href="annotate_hotspots.html">annotate_hotspots()</a></code> </p>
        </td>
        <td><p>Annotate Hotspots.</p></td>
      </tr><tr><td>
          <p><code><a href="annotate_igh_breakpoints.html">annotate_igh_breakpoints()</a></code> </p>
        </td>
        <td><p>Annotate IGH Breakpoints.</p></td>
      </tr><tr><td>
          <p><code><a href="annotate_recurrent_cnv.html">annotate_recurrent_cnv()</a></code> </p>
        </td>
        <td><p>Annotate Recurrent CNVs.</p></td>
      </tr><tr><td>
          <p><code><a href="annotate_ssm_blacklist.html">annotate_ssm_blacklist()</a></code> </p>
        </td>
        <td><p>Annotate SSM with Blacklists</p></td>
      </tr><tr><td>
          <p><code><a href="annotate_sv.html">annotate_sv()</a></code> </p>
        </td>
        <td><p>Annotate SVs.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="bundled-data">Bundled Data <a href="#bundled-data" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>This is the bundled data used by GAMBLR.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="chromosome_arms_grch37.html">chromosome_arms_grch37</a></code> </p>
        </td>
        <td><p>Chromosome Arms grch37.</p></td>
      </tr><tr><td>
          <p><code><a href="chromosome_arms_hg38.html">chromosome_arms_hg38</a></code> </p>
        </td>
        <td><p>Chromosome Arms hg38.</p></td>
      </tr><tr><td>
          <p><code><a href="dhitsig_genes_with_weights.html">dhitsig_genes_with_weights</a></code> </p>
        </td>
        <td><p>Double Hit Signature Genes With Weights.</p></td>
      </tr><tr><td>
          <p><code><a href="gene_blacklist.html">gene_blacklist</a></code> </p>
        </td>
        <td><p>Genes Blacklist.</p></td>
      </tr><tr><td>
          <p><code><a href="grande_maf.html">grande_maf</a></code> </p>
        </td>
        <td><p>Grande et al. MAF.</p></td>
      </tr><tr><td>
          <p><code><a href="grch37_all_gene_coordinates.html">grch37_all_gene_coordinates</a></code> </p>
        </td>
        <td><p>Gene Coordinates for grch37 (all).</p></td>
      </tr><tr><td>
          <p><code><a href="grch37_ashm_regions.html">grch37_ashm_regions</a></code> </p>
        </td>
        <td><p>grch37 ASHM Regions.</p></td>
      </tr><tr><td>
          <p><code><a href="grch37_gene_coordinates.html">grch37_gene_coordinates</a></code> </p>
        </td>
        <td><p>grch37 Gene Coordinates.</p></td>
      </tr><tr><td>
          <p><code><a href="grch37_lymphoma_genes_bed.html">grch37_lymphoma_genes_bed</a></code> </p>
        </td>
        <td><p>Lymphoma Genes (grch37).</p></td>
      </tr><tr><td>
          <p><code><a href="grch37_oncogene.html">grch37_oncogene</a></code> </p>
        </td>
        <td><p>Oncogenes in grch37 genome build.</p></td>
      </tr><tr><td>
          <p><code><a href="grch37_partners.html">grch37_partners</a></code> </p>
        </td>
        <td><p>grch37 Partner Genes.</p></td>
      </tr><tr><td>
          <p><code><a href="hg38_ashm_regions.html">hg38_ashm_regions</a></code> </p>
        </td>
        <td><p>hg38 ASHM Regions.</p></td>
      </tr><tr><td>
          <p><code><a href="hg38_gene_coordinates.html">hg38_gene_coordinates</a></code> </p>
        </td>
        <td><p>hg38 Gene Coordinates.</p></td>
      </tr><tr><td>
          <p><code><a href="hg38_lymphoma_genes_bed.html">hg38_lymphoma_genes_bed</a></code> </p>
        </td>
        <td><p>Lymphoma Genes (hg38).</p></td>
      </tr><tr><td>
          <p><code><a href="hg38_oncogene.html">hg38_oncogene</a></code> </p>
        </td>
        <td><p>Oncogenes in hg38 genome build.</p></td>
      </tr><tr><td>
          <p><code><a href="hg38_partners.html">hg38_partners</a></code> </p>
        </td>
        <td><p>hg38 Partner Genes.</p></td>
      </tr><tr><td>
          <p><code><a href="hotspot_regions_grch37.html">hotspot_regions_grch37</a></code> </p>
        </td>
        <td><p>grch37 Hotspot Regions.</p></td>
      </tr><tr><td>
          <p><code><a href="hotspot_regions_hg38.html">hotspot_regions_hg38</a></code> </p>
        </td>
        <td><p>hg38 Hotspot Regions.</p></td>
      </tr><tr><td>
          <p><code><a href="lymphoma_genes.html">lymphoma_genes</a></code> </p>
        </td>
        <td><p>Lymphoma Genes</p></td>
      </tr><tr><td>
          <p><code><a href="lymphoma_genes_comprehensive.html">lymphoma_genes_comprehensive</a></code> </p>
        </td>
        <td><p>Lymphoma Genes Comprehensive.</p></td>
      </tr><tr><td>
          <p><code><a href="reddy_genes.html">reddy_genes</a></code> </p>
        </td>
        <td><p>Reddy Genes.</p></td>
      </tr><tr><td>
          <p><code><a href="target_regions_grch37.html">target_regions_grch37</a></code> </p>
        </td>
        <td><p>Target Regions grch37.</p></td>
      </tr><tr><td>
          <p><code><a href="target_regions_hg38.html">target_regions_hg38</a></code> </p>
        </td>
        <td><p>Target Regions hg38.</p></td>
      </tr><tr><td>
          <p><code><a href="wright_genes_with_weights.html">wright_genes_with_weights</a></code> </p>
        </td>
        <td><p>Wright Genes With Weights.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="cbioportal">cBioPortal <a href="#cbioportal" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions tailored to setting up a new study or updating an existing cBioPortal instance.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="cbioportal_create.html">cbioportal_create()</a></code> </p>
        </td>
        <td><p>Create cBioPortal Study.</p></td>
      </tr><tr><td>
          <p><code><a href="custom_caselist.html">custom_caselist()</a></code> </p>
        </td>
        <td><p>Custom cBioPortal case list.</p></td>
      </tr><tr><td>
          <p><code><a href="finalize_study.html">finalize_study()</a></code> </p>
        </td>
        <td><p>Finalize Study (cBioPortal).</p></td>
      </tr><tr><td>
          <p><code><a href="get_study_info.html">get_study_info()</a></code> </p>
        </td>
        <td><p>Get Study Info.</p></td>
      </tr><tr><td>
          <p><code><a href="setup_expreession_data.html">setup_expreession_data()</a></code> </p>
        </td>
        <td><p>Setup Expression Data (cBioPortal).</p></td>
      </tr><tr><td>
          <p><code><a href="setup_fusions.html">setup_fusions()</a></code> </p>
        </td>
        <td><p>Setup Fusions (cBioPortal).</p></td>
      </tr><tr><td>
          <p><code><a href="setup_study.html">setup_study()</a></code> </p>
        </td>
        <td><p>Setup Study (cBioPortal).</p></td>
      </tr><tr><td>
          <p><code><a href="study_check.html">study_check()</a></code> </p>
        </td>
        <td><p>Study Check (cBioPortal).</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="checks">Checks <a href="#checks" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>GAMBLR functions that fall under the check category.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="check_gambl_metadata.html">check_gambl_metadata()</a></code> </p>
        </td>
        <td><p>Check GAMBL Metadata.</p></td>
      </tr><tr><td>
          <p><code><a href="check_gamblr_config.html">check_gamblr_config()</a></code> </p>
        </td>
        <td><p>Check GAMBLR Config.</p></td>
      </tr><tr><td>
          <p><code><a href="check_host.html">check_host()</a></code> </p>
        </td>
        <td><p>Check Host.</p></td>
      </tr><tr><td>
          <p><code><a href="check_remote_configuration.html">check_remote_configuration()</a></code> </p>
        </td>
        <td><p>Check Remote Configuration.</p></td>
      </tr><tr><td>
          <p><code><a href="fread_maf.html">fread_maf()</a></code> </p>
        </td>
        <td><p>Fread MAF.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="get-data">Get Data <a href="#get-data" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>A collection of functions used to retrieve data of different types.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="get_ashm_count_matrix.html">get_ashm_count_matrix()</a></code> </p>
        </td>
        <td><p>Get ASHM Count Matrix.</p></td>
      </tr><tr><td>
          <p><code><a href="get_bams.html">get_bams()</a></code> </p>
        </td>
        <td><p>Get BAMs.</p></td>
      </tr><tr><td>
          <p><code><a href="get_cn_segments.html">get_cn_segments()</a></code> </p>
        </td>
        <td><p>Get CN Segments.</p></td>
      </tr><tr><td>
          <p><code><a href="get_cn_states.html">get_cn_states()</a></code> </p>
        </td>
        <td><p>Get CN States.</p></td>
      </tr><tr><td>
          <p><code><a href="get_coding_ssm.html">get_coding_ssm()</a></code> </p>
        </td>
        <td><p>Get Coding SSM.</p></td>
      </tr><tr><td>
          <p><code><a href="get_coding_ssm_status.html">get_coding_ssm_status()</a></code> </p>
        </td>
        <td><p>Get Coding SSM Status.</p></td>
      </tr><tr><td>
          <p><code><a href="get_combined_sv.html">get_combined_sv()</a></code> </p>
        </td>
        <td><p>Get Combined SV.</p></td>
      </tr><tr><td>
          <p><code><a href="get_excluded_samples.html">get_excluded_samples()</a></code> </p>
        </td>
        <td><p>Get Excluded Samples.</p></td>
      </tr><tr><td>
          <p><code><a href="get_gambl_colours.html">get_gambl_colours()</a></code> </p>
        </td>
        <td><p>Get GAMBL Colours.</p></td>
      </tr><tr><td>
          <p><code><a href="get_gambl_metadata.html">get_gambl_metadata()</a></code> </p>
        </td>
        <td><p>Get GAMBL metadata.</p></td>
      </tr><tr><td>
          <p><code><a href="get_gene_cn_and_expression.html">get_gene_cn_and_expression()</a></code> </p>
        </td>
        <td><p>Get Gene CN and Expression.</p></td>
      </tr><tr><td>
          <p><code><a href="get_gene_expression.html">get_gene_expression()</a></code> </p>
        </td>
        <td><p>Get Gene Expression.</p></td>
      </tr><tr><td>
          <p><code><a href="get_lymphgen.html">get_lymphgen()</a></code> </p>
        </td>
        <td><p>Get Lymphgen.</p></td>
      </tr><tr><td>
          <p><code><a href="get_manta_sv.html">get_manta_sv()</a></code> </p>
        </td>
        <td><p>Get Manta SVs</p></td>
      </tr><tr><td>
          <p><code><a href="get_manta_sv_by_sample.html">get_manta_sv_by_sample()</a></code> </p>
        </td>
        <td><p>Get Manta SV By Sample.</p></td>
      </tr><tr><td>
          <p><code><a href="get_manta_sv_by_samples.html">get_manta_sv_by_samples()</a></code> </p>
        </td>
        <td><p>Get Manta SV By Samples.</p></td>
      </tr><tr><td>
          <p><code><a href="get_mutation_frequency_bin_matrix.html">get_mutation_frequency_bin_matrix()</a></code> </p>
        </td>
        <td><p>Mutation Frequency Heatmap</p></td>
      </tr><tr><td>
          <p><code><a href="get_sample_cn_segments.html">get_sample_cn_segments()</a></code> </p>
        </td>
        <td><p>GetSample CN Segments.</p></td>
      </tr><tr><td>
          <p><code><a href="get_sample_wildcards.html">get_sample_wildcards()</a></code> </p>
        </td>
        <td><p>Get Sample Wildcards</p></td>
      </tr><tr><td>
          <p><code><a href="get_samples_md5_hash.html">get_samples_md5_hash()</a></code> </p>
        </td>
        <td><p>Generate md5 Hash For Samples</p></td>
      </tr><tr><td>
          <p><code><a href="get_ssh_session.html">get_ssh_session()</a></code> </p>
        </td>
        <td><p>Get SSH Session.</p></td>
      </tr><tr><td>
          <p><code><a href="get_ssm_by_patients.html">get_ssm_by_patients()</a></code> </p>
        </td>
        <td><p>Get SSM By Patients.</p></td>
      </tr><tr><td>
          <p><code><a href="get_ssm_by_sample.html">get_ssm_by_sample()</a></code> </p>
        </td>
        <td><p>Get SSM By Sample.</p></td>
      </tr><tr><td>
          <p><code><a href="get_ssm_by_samples.html">get_ssm_by_samples()</a></code> </p>
        </td>
        <td><p>Get SSM By Samples.</p></td>
      </tr><tr><td>
          <p><code><a href="get_ssm_by_region.html">get_ssm_by_region()</a></code> </p>
        </td>
        <td><p>Get SSM By Region.</p></td>
      </tr><tr><td>
          <p><code><a href="get_ssm_by_regions.html">get_ssm_by_regions()</a></code> </p>
        </td>
        <td><p>Get SSM By Regions.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="plotting">Plotting <a href="#plotting" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions used for visualizing GAMBL data.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="ashm_multi_rainbow_plot.html">ashm_multi_rainbow_plot()</a></code> </p>
        </td>
        <td><p>ASHM Multi-panel Rainbow Plot</p></td>
      </tr><tr><td>
          <p><code><a href="ashm_rainbow_plot.html">ashm_rainbow_plot()</a></code> </p>
        </td>
        <td><p>ASHM Rainbow Plot</p></td>
      </tr><tr><td>
          <p><code><a href="calc_mutation_frequency_sliding_windows.html">calc_mutation_frequency_sliding_windows()</a></code> </p>
        </td>
        <td><p>Calculate Mutation Frequency By Sliding Window.</p></td>
      </tr><tr><td>
          <p><code><a href="cnvKompare.html">cnvKompare()</a></code> </p>
        </td>
        <td><p>Compare segmented data for multiple samples.</p></td>
      </tr><tr><td>
          <p><code><a href="comp_report.html">comp_report()</a></code> </p>
        </td>
        <td><p>sample-level SV/SSM/CN reports in PDF</p></td>
      </tr><tr><td>
          <p><code><a href="copy_number_vaf_plot.html">copy_number_vaf_plot()</a></code> </p>
        </td>
        <td><p>CN VAF Plot</p></td>
      </tr><tr><td>
          <p><code><a href="fancy_alignment_plot.html">fancy_alignment_plot()</a></code> </p>
        </td>
        <td><p>Plot Alignment Metrics</p></td>
      </tr><tr><td>
          <p><code><a href="fancy_circos_plot.html">fancy_circos_plot()</a></code> </p>
        </td>
        <td><p>SSM and SV Circos Plot.</p></td>
      </tr><tr><td>
          <p><code><a href="fancy_cnbar.html">fancy_cnbar()</a></code> </p>
        </td>
        <td><p>Copy Number states barplot</p></td>
      </tr><tr><td>
          <p><code><a href="fancy_ideogram.html">fancy_ideogram()</a></code> </p>
        </td>
        <td><p>genome-wide ideogram annotated with SSM and CN information</p></td>
      </tr><tr><td>
          <p><code><a href="fancy_multisamp_ideogram.html">fancy_multisamp_ideogram()</a></code> </p>
        </td>
        <td><p>Genome-wide ideogram (CN segments) for multiple samples.</p></td>
      </tr><tr><td>
          <p><code><a href="fancy_propcov_plot.html">fancy_propcov_plot()</a></code> </p>
        </td>
        <td><p>Proportional Coverage Plot.</p></td>
      </tr><tr><td>
          <p><code><a href="fancy_proportions_plot.html">fancy_proportions_plot()</a></code> </p>
        </td>
        <td><p>Proportional Metrics Plot.</p></td>
      </tr><tr><td>
          <p><code><a href="fancy_qc_plot.html">fancy_qc_plot()</a></code> </p>
        </td>
        <td><p>Plot Quality Control Metrics.</p></td>
      </tr><tr><td>
          <p><code><a href="fancy_snv_chrdistplot.html">fancy_snv_chrdistplot()</a></code> </p>
        </td>
        <td><p>n SNVs per chromosome plot</p></td>
      </tr><tr><td>
          <p><code><a href="fancy_sv_sizedens.html">fancy_sv_sizedens()</a></code> </p>
        </td>
        <td><p>Structural Variants Size Plot.</p></td>
      </tr><tr><td>
          <p><code><a href="fancy_v_chrcount.html">fancy_v_chrcount()</a></code> </p>
        </td>
        <td><p>n variants per chromosome plot.</p></td>
      </tr><tr><td>
          <p><code><a href="fancy_v_count.html">fancy_v_count()</a></code> </p>
        </td>
        <td><p>Total n variants count plot.</p></td>
      </tr><tr><td>
          <p><code><a href="fancy_v_sizedis.html">fancy_v_sizedis()</a></code> </p>
        </td>
        <td><p>Variant size distribution plot</p></td>
      </tr><tr><td>
          <p><code><a href="focal_cn_plot.html">focal_cn_plot()</a></code> </p>
        </td>
        <td><p>Copy Number Segments Plot</p></td>
      </tr><tr><td>
          <p><code><a href="FtestCNV.html">FtestCNV()</a></code> </p>
        </td>
        <td><p>Fishers Exact Test (CNV).</p></td>
      </tr><tr><td>
          <p><code><a href="plot_sample_circos.html">plot_sample_circos()</a></code> </p>
        </td>
        <td><p>Sample-level Circos Plot</p></td>
      </tr><tr><td>
          <p><code><a href="prettyChromoplot.html">prettyChromoplot()</a></code> </p>
        </td>
        <td><p>Chromosome Plot</p></td>
      </tr><tr><td>
          <p><code><a href="prettyCoOncoplot.html">prettyCoOncoplot()</a></code> </p>
        </td>
        <td><p>Side-by-side Oncoplots</p></td>
      </tr><tr><td>
          <p><code><a href="prettyForestPlot.html">prettyForestPlot()</a></code> </p>
        </td>
        <td><p>Forest Plot.</p></td>
      </tr><tr><td>
          <p><code><a href="prettyGeneCloud.html">prettyGeneCloud()</a></code> </p>
        </td>
        <td><p>Gene Cloud Plot</p></td>
      </tr><tr><td>
          <p><code><a href="prettyOncoplot.html">prettyOncoplot()</a></code> </p>
        </td>
        <td><p>Oncooplot</p></td>
      </tr><tr><td>
          <p><code><a href="prettyRainfallPlot.html">prettyRainfallPlot()</a></code> </p>
        </td>
        <td><p>Rainfall Plot</p></td>
      </tr><tr><td>
          <p><code><a href="pretty_lollipop_plot.html">pretty_lollipop_plot()</a></code> </p>
        </td>
        <td><p>Lollipop Plot</p></td>
      </tr><tr><td>
          <p><code><a href="splendidHeatmap.html">splendidHeatmap()</a></code> </p>
        </td>
        <td><p>Heatmap</p></td>
      </tr><tr><td>
          <p><code><a href="theme_Morons.html">theme_Morons()</a></code> </p>
        </td>
        <td><p>Morons Theme</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="statistical-testing-and-computing">Statistical testing and Computing <a href="#statistical-testing-and-computing" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions used for executing statistical and other calculations on incoming GAMBL data.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="FtestCNV.html">FtestCNV()</a></code> </p>
        </td>
        <td><p>Fishers Exact Test (CNV).</p></td>
      </tr><tr><td>
          <p><code><a href="calc_mutation_frequency_sliding_windows.html">calc_mutation_frequency_sliding_windows()</a></code> </p>
        </td>
        <td><p>Calculate Mutation Frequency By Sliding Window.</p></td>
      </tr><tr><td>
          <p><code><a href="calculate_pga.html">calculate_pga()</a></code> </p>
        </td>
        <td><p>Calculate proportion of genome altered by CNV.</p></td>
      </tr><tr><td>
          <p><code><a href="cnvKompare.html">cnvKompare()</a></code> </p>
        </td>
        <td><p>Compare segmented data for multiple samples.</p></td>
      </tr><tr><td>
          <p><code><a href="count_ssm_by_region.html">count_ssm_by_region()</a></code> </p>
        </td>
        <td><p>Count SSM In A Region</p></td>
      </tr><tr><td>
          <p><code><a href="estimate_purity.html">estimate_purity()</a></code> </p>
        </td>
        <td><p>Estimate Purity.</p></td>
      </tr><tr><td>
          <p><code><a href="prettyForestPlot.html">prettyForestPlot()</a></code> </p>
        </td>
        <td><p>Forest Plot.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="utilities">Utilities <a href="#utilities" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>A collection of functions that fall under the utility category.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="adjust_ploidy.html">adjust_ploidy()</a></code> </p>
        </td>
        <td><p>Adjust ploidy for samples with CNV data.</p></td>
      </tr><tr><td>
          <p><code><a href="assemble_file_details.html">assemble_file_details()</a></code> </p>
        </td>
        <td><p>Assemble File Details.</p></td>
      </tr><tr><td>
          <p><code><a href="assign_cn_to_ssm.html">assign_cn_to_ssm()</a></code> </p>
        </td>
        <td><p>Assign CN to SSM.</p></td>
      </tr><tr><td>
          <p><code><a href="cleanup_maf.html">cleanup_maf()</a></code> </p>
        </td>
        <td><p>Cleanup MAF.</p></td>
      </tr><tr><td>
          <p><code><a href="consolidate_lymphgen.html">consolidate_lymphgen()</a></code> </p>
        </td>
        <td><p>Consolidate Lymphgen.</p></td>
      </tr><tr><td>
          <p><code><a href="collate_lymphgen.html">collate_lymphgen()</a></code> </p>
        </td>
        <td><p>Collate Lymphgen.</p></td>
      </tr><tr><td>
          <p><code><a href="collate_results.html">collate_results()</a></code> </p>
        </td>
        <td><p>Collate Results</p></td>
      </tr><tr><td>
          <p><code><a href="fetch_output_files.html">fetch_output_files()</a></code> </p>
        </td>
        <td><p>Fetch Output Files.</p></td>
      </tr><tr><td>
          <p><code><a href="find_expected_outputs.html">find_expected_outputs()</a></code> </p>
        </td>
        <td><p>Find Expected Outputs.</p></td>
      </tr><tr><td>
          <p><code><a href="find_files_extract_wildcards.html">find_files_extract_wildcards()</a></code> </p>
        </td>
        <td><p>Find Files Extract Wildcards</p></td>
      </tr><tr><td>
          <p><code><a href="gene_to_region.html">gene_to_region()</a></code> </p>
        </td>
        <td><p>Gene To Region.</p></td>
      </tr><tr><td>
          <p><code><a href="genome_to_exome.html">genome_to_exome()</a></code> </p>
        </td>
        <td><p>Genome To Exome.</p></td>
      </tr><tr><td>
          <p><code><a href="intersect_maf.html">intersect_maf()</a></code> </p>
        </td>
        <td><p>Intersect MAF.</p></td>
      </tr><tr><td>
          <p><code><a href="liftover_bedpe.html">liftover_bedpe()</a></code> </p>
        </td>
        <td><p>Liftover Bedpe.</p></td>
      </tr><tr><td>
          <p><code><a href="maf_to_custom_track.html">maf_to_custom_track()</a></code> </p>
        </td>
        <td><p>Maf To Custom Track.</p></td>
      </tr><tr><td>
          <p><code><a href="make_igv_snapshot.html">make_igv_snapshot()</a></code> </p>
        </td>
        <td><p>Make IGV Snapshot</p></td>
      </tr><tr><td>
          <p><code><a href="populate_each_tool_result.html">populate_each_tool_result()</a></code> </p>
        </td>
        <td><p>Populate Each Tool Results.</p></td>
      </tr><tr><td>
          <p><code><a href="populate_tool_results.html">populate_tool_results()</a></code> </p>
        </td>
        <td><p>Populate Tool Results.</p></td>
      </tr><tr><td>
          <p><code><a href="region_to_bins.html">region_to_bins()</a></code> </p>
        </td>
        <td><p>Region To Bins.</p></td>
      </tr><tr><td>
          <p><code><a href="region_to_gene.html">region_to_gene()</a></code> </p>
        </td>
        <td><p>Region To Gene.</p></td>
      </tr><tr><td>
          <p><code><a href="review_hotspots.html">review_hotspots()</a></code> </p>
        </td>
        <td><p>Review Hotspots.</p></td>
      </tr><tr><td>
          <p><code><a href="sanitize_maf_data.html">sanitize_maf_data()</a></code> </p>
        </td>
        <td><p>Convert mutation data to a shareable format.</p></td>
      </tr><tr><td>
          <p><code><a href="sv_to_bedpe_file.html">sv_to_bedpe_file()</a></code> </p>
        </td>
        <td><p>SV To BED File.</p></td>
      </tr><tr><td>
          <p><code><a href="sv_to_custom_track.html">sv_to_custom_track()</a></code> </p>
        </td>
        <td><p>SV To Custom Track.</p></td>
      </tr><tr><td>
          <p><code><a href="supplement_maf.html">supplement_maf()</a></code> </p>
        </td>
        <td><p>Supplement MAF.</p></td>
      </tr><tr><td>
          <p><code><a href="tidy_gene_expression.html">tidy_gene_expression()</a></code> </p>
        </td>
        <td><p>Tidy gene Expression.</p></td>
      </tr><tr><td>
          <p><code><a href="tidy_lymphgen.html">tidy_lymphgen()</a></code> </p>
        </td>
        <td><p>Tidy Lymphgen.</p></td>
      </tr><tr><td>
          <p><code><a href="write_sample_set_hash.html">write_sample_set_hash()</a></code> </p>
        </td>
        <td><p>Write Sample Set Hash</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ryan Morin, Kostia Dreval, Adam Mattsson.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

