<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ GAMBLR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9500</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/GAMBLR_presentation.html">Doubling Down on GAMBL with GAMBLR</a>
    </li>
    <li>
      <a href="../articles/examples.html">GAMBLR - working with GAMBL data painlessly in R</a>
    </li>
    <li>
      <a href="../articles/lymphoma.html">Lymphoma colours</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="add_icgc_metadata.html">add_icgc_metadata()</a></code> </p>
        </td>
        <td><p>Layer on ICGC metadata from a supplemental table to fill in missing COO</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="annotate_hotspots.html">annotate_hotspots()</a></code> </p>
        </td>
        <td><p>Annotate MAF-like data frome with a hot_spot column indicating recurrent mutations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="annotate_sv.html">annotate_sv()</a></code> </p>
        </td>
        <td><p>Annotate a data frame of SV breakpoints after retrieval from the database</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="append_to_table.html">append_to_table()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ashm_multi_rainbow_plot.html">ashm_multi_rainbow_plot()</a></code> </p>
        </td>
        <td><p>Generate a colourful multi-panel overview of hypermutation in regions of interest across many samples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ashm_rainbow_plot.html">ashm_rainbow_plot()</a></code> </p>
        </td>
        <td><p>Make a rainbow plot of all mutations in a region, ordered and coloured by metadata</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assemble_file_details.html">assemble_file_details()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assign_cn_to_ssm.html">assign_cn_to_ssm()</a></code> </p>
        </td>
        <td><p>Annotate mutations with their copy number information</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="collate_ashm_results.html">collate_ashm_results()</a></code> </p>
        </td>
        <td><p>Determine the hypermutation status of a few genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="collate_curated_sv_results.html">collate_curated_sv_results()</a></code> </p>
        </td>
        <td><p>Collate all SV calls from the genome data and summarize for main oncogenes of interest per sample</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="collate_derived_results.html">collate_derived_results()</a></code> </p>
        </td>
        <td><p>Extract derived results stored in the database (these are usually slower to derive on the fly)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="collate_nfkbiz_results.html">collate_nfkbiz_results()</a></code> </p>
        </td>
        <td><p>Determine which cases have NFKBIZ UTR mutations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="collate_results.html">collate_results()</a></code> </p>
        </td>
        <td><p>Bring together all derived sample-level results from many GAMBL pipelines</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="collate_sbs_results.html">collate_sbs_results()</a></code> </p>
        </td>
        <td><p>Bring in the results from mutational signature analysis</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="collate_sv_results.html">collate_sv_results()</a></code> </p>
        </td>
        <td><p>Determine and summarize which cases have specific oncogene SVs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="copy_number_vaf_plot.html">copy_number_vaf_plot()</a></code> </p>
        </td>
        <td><p>Create a genome-wide copy number plot for one sample and (optionally) display mutation VAF</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fetch_output_files.html">fetch_output_files()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="finalize_study.html">finalize_study()</a></code> </p>
        </td>
        <td><p>Finish setting up a new cBioPortal instance or updating an existing portal data set</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="find_files_extract_wildcards.html">find_files_extract_wildcards()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fread_maf.html">fread_maf()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_ashm_count_matrix.html">get_ashm_count_matrix()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_bams.html">get_bams()</a></code> </p>
        </td>
        <td><p>Get full paths for bam files for a sample or patient</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_cn_segments.html">get_cn_segments()</a></code> </p>
        </td>
        <td><p>Retrieve all copy number segments from the GAMBL database that overlap with a single genomic coordinate range</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_cn_states.html">get_cn_states()</a></code> </p>
        </td>
        <td><p>Get a copy number matrix for all samples based on segmented data in database</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_coding_ssm.html">get_coding_ssm()</a></code> </p>
        </td>
        <td><p>Retrieve all coding SSMs from the GAMBL database in MAF-like format</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_gambl_colours.html">get_gambl_colours()</a></code> </p>
        </td>
        <td><p>Get some colour schemes for annotating figures</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_gambl_metadata.html">get_gambl_metadata()</a></code> </p>
        </td>
        <td><p>Get GAMBL metadata</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_gambl_outcomes.html">get_gambl_outcomes()</a></code> </p>
        </td>
        <td><p>Get the patient-centric clinical metadata</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_gene_cn_and_expression.html">get_gene_cn_and_expression()</a></code> </p>
        </td>
        <td><p>Get the copy number and expression for a single gene</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_gene_expression.html">get_gene_expression()</a></code> </p>
        </td>
        <td><p>Get the expression for one or more genes for all GAMBL samples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_manta_sv.html">get_manta_sv()</a></code> </p>
        </td>
        <td><p>Retrieve Manta SVs from the database and filter</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_ssm_by_gene.html">get_ssm_by_gene()</a></code> </p>
        </td>
        <td><p>Get all somatic mutations for a given gene or list of genes and optionally restrict to coding variants</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_ssm_by_region.html">get_ssm_by_region()</a></code> </p>
        </td>
        <td><p>Retrieve all SSMs from the GAMBL database within a single genomic coordinate range</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_ssm_by_regions.html">get_ssm_by_regions()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="liftover_bedpe.html">liftover_bedpe()</a></code> </p>
        </td>
        <td><p>Use liftOver to convert a bedpe file between the two main genome builds (grch37/hg38)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="maf_to_custom_track.html">maf_to_custom_track()</a></code> </p>
        </td>
        <td><p>Convert a maf-formatted data frame into a bed custom track file for UCSC</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="make_igv_snapshot.html">make_igv_snapshot()</a></code> </p>
        </td>
        <td><p>Load bams and generate an IGV screenshot for one or more regions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_multi_timepoint.html">plot_multi_timepoint()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="populate_each_tool_result.html">populate_each_tool_result()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="populate_tool_results.html">populate_tool_results()</a></code> </p>
        </td>
        <td><p>Populate the database with the per-sample summarized results of various tools</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="prettyChromoplot.html">prettyChromoplot()</a></code> </p>
        </td>
        <td><p>Use GISTIC2.0 scores output to reproduce maftools::chromoplot with more flexibility</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="prettyOncoplot.html">prettyOncoplot()</a></code> </p>
        </td>
        <td><p>Make an oncoplot that is pretty using ComplexHeatmap. The metadata is expected to follow the structure and column naming used in GAMBL.
If you provide your own non-GAMBL samples and metadata, you must include at least the following columns with these names.
The first one should match the Tumor_Sample_Barcode in the MAF object or onco_matrix you provide.
sample_id, pathology</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="process_all_manta_bedpe.html">process_all_manta_bedpe()</a></code> </p>
        </td>
        <td><p>This is a helper function that is not meant to be used routinely</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read_merge_manta_with_liftover.html">read_merge_manta_with_liftover()</a></code> </p>
        </td>
        <td><p>This is a helper function that is not meant to be used routinely</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="referesh_metadata_tables.html">referesh_metadata_tables()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refresh_full_table.html">refresh_full_table()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sanitize_maf_data.html">sanitize_maf_data()</a></code> </p>
        </td>
        <td><p>Write an oncomatrix from a MAF File for further plotting. This is meant to be run by individuals who have access to data sets to
"sanitize" a subset of data for subsequent use by them or others who don't have permission to access the raw data.
Example: User J has full permissions for ICGC data and has read permissions on a MAF file. User B needs to make some oncoplots
and/or perform some statistical analysis on the frequency and assortment of mutations in that data set but doesn't need all the details.
User J can run this function on a maf file and provide the path of the output to user B.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="setup_fusions.html">setup_fusions()</a></code> </p>
        </td>
        <td><p>Annotate SVs and create the input for fusions to be displayed in cBioPortal instance</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="setup_study.html">setup_study()</a></code> </p>
        </td>
        <td><p>Initialize a new cBioPortal instance or update existing portal data set</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sv_to_custom_track.html">sv_to_custom_track()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tidy_gene_expression.html">tidy_gene_expression()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Ryan Morin.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


