<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Rainfall Plot — prettyRainfallPlot • GAMBLR</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Rainfall Plot — prettyRainfallPlot"><meta property="og:description" content="Plot a rainfall plot for one sample. This function takes in the MAF data frame, or path to a custom MAF file.
If non are specified, the SSM will be obtained through GAMBLR directly (with get_ssm_by_region)."><meta property="og:image" content="https://github.com/morinlab/GAMBLR/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-12345678-0"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-12345678-0');
</script></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9500</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-light fa-memo-circle-info"></span>
     
    How To
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="https://morinlab.github.io/GAMBLR/#installation" class="external-link">Install</a>
    </li>
    <li>
      <a href="https://morinlab.github.io/GAMBLR/#running-gamblr-on-your-own-computer" class="external-link">Running GAMBLR On Your Own Computer</a>
    </li>
    <li>
      <a href="https://morinlab.github.io/GAMBLR/#contributing" class="external-link">Contribute</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-light fa-person-chalkboard"></span>
     
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/ssm_tutorial.html">SSM</a>
    </li>
    <li>
      <a href="../articles/cn_sv_tutorial.html">CN &amp; SV</a>
    </li>
    <li>
      <a href="../articles/fancy_plots.html">Fancy Plots</a>
    </li>
    <li>
      <a href="../articles/utilities_tutorial.html">Utilities</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="../reference/index.html">
    <span class="fa fa-file-code-o"></span>
     
    Functions &amp; Bundled Data
  </a>
</li>
<li>
  <a href="https://github.com/morinlab/GAMBLR">
    <span class="fa fa-github fa-lg"></span>
     
    github
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Rainfall Plot</h1>
    
    <div class="hidden name"><code>prettyRainfallPlot.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Plot a rainfall plot for one sample. This function takes in the MAF data frame, or path to a custom MAF file.
If non are specified, the SSM will be obtained through GAMBLR directly (with <a href="get_ssm_by_region.html">get_ssm_by_region</a>).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">prettyRainfallPlot</span><span class="op">(</span></span>
<span>  <span class="va">this_sample_id</span>,</span>
<span>  label_ashm_genes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  projection <span class="op">=</span> <span class="st">"grch37"</span>,</span>
<span>  <span class="va">chromosome</span>,</span>
<span>  <span class="va">this_maf</span>,</span>
<span>  <span class="va">maf_path</span>,</span>
<span>  <span class="va">zoom_in_region</span>,</span>
<span>  <span class="va">seq_type</span>,</span>
<span>  label_sv <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>this_sample_id</dt>
<dd><p>Sample id for the sample to display. This is argument is not required if you want a multi-sample plot but is otherwise needed.</p></dd>


<dt>label_ashm_genes</dt>
<dd><p>Boolean argument indicating whether the aSHM regions will be labeled or not.</p></dd>


<dt>projection</dt>
<dd><p>Specify projection (grch37 or hg38) of mutations. Default is grch37.</p></dd>


<dt>chromosome</dt>
<dd><p>Provide one or more chromosomes to plot. The chr prefix can be inconsistent with projection and will be handled.</p></dd>


<dt>this_maf</dt>
<dd><p>Specify custom MAF data frame of mutations.</p></dd>


<dt>maf_path</dt>
<dd><p>Specify path to MAF file if it is not already loaded into data frame.</p></dd>


<dt>zoom_in_region</dt>
<dd><p>Provide a specific region in the format "chromosome:start-end" to zoom in to a specific region.</p></dd>


<dt>seq_type</dt>
<dd><p>Specify one of "genome" or "capture" when relying on the function to obtain mutations from a region (i.e. if you haven't provided a MAF or single sample_id)</p></dd>


<dt>label_sv</dt>
<dd><p>Boolean argument to specify whether label SVs or not. Only supported if a specific chromosome or zoom in region are specified.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>a ggplot2 plot. Print it using print() or save it using ggsave()</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Create a sample-level rainfall plot visualizing single nucleotide substitutions mutations for selected chromosomes.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu">prettyRainfallPlot</span><span class="op">(</span>this_sample_id <span class="op">=</span> <span class="st">"Raji"</span>,</span></span>
<span class="r-in"><span>                   seq_type <span class="op">=</span> <span class="st">"genome"</span>,</span></span>
<span class="r-in"><span>                   zoom_in_region <span class="op">=</span> <span class="st">"8:125252796-135253201"</span>,</span></span>
<span class="r-in"><span>                   label_sv <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> MAF df or path to custom MAF file was not provided, getting SSM using GAMBLR ...</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Column name 'GENE_PHENO' not found in column name header (case sensitive), skipping.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Column name 'FILTER' not found in column name header (case sensitive), skipping.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Column name 'flanking_bps' not found in column name header (case sensitive), skipping.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Column name 'vcf_id' not found in column name header (case sensitive), skipping.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Column name 'vcf_qual' not found in column name header (case sensitive), skipping.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Column name 'gnomAD_AF' not found in column name header (case sensitive), skipping.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Column name 'gnomAD_AFR_AF' not found in column name header (case sensitive), skipping.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Column name 'gnomAD_AMR_AF' not found in column name header (case sensitive), skipping.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Column name 'gnomAD_SAS_AF' not found in column name header (case sensitive), skipping.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Column name 'vcf_pos' not found in column name header (case sensitive), skipping.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Column name 'gnomADg_AF' not found in column name header (case sensitive), skipping.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Column name 'blacklist_count' not found in column name header (case sensitive), skipping.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Column name 'FILTER' (colClasses[[1]][67]) not found</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Column name 'flanking_bps' (colClasses[[1]][68]) not found</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Column name 'vcf_id' (colClasses[[1]][69]) not found</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Column name 'vcf_qual' (colClasses[[1]][70]) not found</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Column name 'gnomADg_AF' (colClasses[[1]][71]) not found</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Column name 'vcf_pos' (colClasses[[2]][11]) not found</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Column name 'gnomAD_AF' (colClasses[[3]][20]) not found</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Column name 'gnomAD_AFR_AF' (colClasses[[3]][21]) not found</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Column name 'gnomAD_AMR_AF' (colClasses[[3]][22]) not found</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Column name 'gnomAD_SAS_AF' (colClasses[[3]][28]) not found</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Column name 'GENE_PHENO' (colClasses[[4]][5]) not found</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Column name 'blacklist_count' (colClasses[[4]][6]) not found</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Attempt to override column 92 &lt;&lt;MOTIF_SCORE_CHANGE&gt;&gt; of inherent type 'float64' down to 'bool8' ignored. Only overrides to a higher type are currently supported. If this was intended, please coerce to the lower type afterwards.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Getting combined manta + GRIDSS SVs using GAMBLR ...</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There were 4 warnings in `dplyr::filter()`.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The first warning was:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> In argument: `if (...) NULL`.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> In group 1: `fusion = "IGH-MYC"`, `chromosomeN = chrom2`.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> Caused by warning in `if (grepl("1", chromosomeN)) ...`:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">!</span> the condition has length &gt; 1 and only the first element will be used</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Run `dplyr::last_dplyr_warnings()` to see the 3 remaining warnings.</span>
<span class="r-plt img"><img src="prettyRainfallPlot-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#multi-sample rainfall plot for one gene region</span></span></span>
<span class="r-in"><span><span class="fu">prettyRainfallPlot</span><span class="op">(</span>zoom_in_region <span class="op">=</span> <span class="st">"chr3:5,221,286-5,269,723"</span>,</span></span>
<span class="r-in"><span>                   seq_type <span class="op">=</span> <span class="st">"genome"</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>No sample_id was provided. Using all mutations in the MAF within your region!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Warning: after subsetting to a regions you requested to plot, there are no aSHM features to overlap on the final graph.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Will use all mutations for genome in this region: 3Will use all mutations for genome in this region: 5221286Will use all mutations for genome in this region: 5269723</span>
<span class="r-plt img"><img src="prettyRainfallPlot-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ryan Morin, Kostia Dreval, Adam Mattsson.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

