<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Side-by-side Oncoplots — prettyCoOncoplot • GAMBLR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Side-by-side Oncoplots — prettyCoOncoplot"><meta property="og:description" content="prettyCoOncoplot returns ggplot-compatible figure of 2 prettyOncoplot side-by-side."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-12345678-0"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-12345678-0');
</script></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9500</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-light fa-memo-circle-info"></span>
     
    How To
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../path_once_integrated_on_github">Install</a>
    </li>
    <li>
      <a href="../path_once_integrated_on_github">Run Remote On A Local Machine</a>
    </li>
    <li>
      <a href="../path_once_integrated_on_github">Contribute</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-light fa-person-chalkboard"></span>
     
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/ssm_tutorial.html">SSM</a>
    </li>
    <li>
      <a href="../articles/cn_sv_tutorial.html">CN &amp; SV</a>
    </li>
    <li>
      <a href="../articles/fancy_plots.html">Fancy Plots</a>
    </li>
    <li>
      <a href="../articles/utiliteis_tutorial.html">Utilities</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="../reference/index.html">
    <span class="fa fa-file-code-o"></span>
     
    Functions &amp; Bundled Data
  </a>
</li>
<li>
  <a href="https://github.com/morinlab/GAMBLR">
    <span class="fa fa-github fa-lg"></span>
     
    github
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Side-by-side Oncoplots</h1>
    
    <div class="hidden name"><code>prettyCoOncoplot.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>prettyCoOncoplot</code> returns ggplot-compatible figure of 2 <a href="prettyOncoplot.html">prettyOncoplot</a> side-by-side.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">prettyCoOncoplot</span><span class="op">(</span></span>
<span>  <span class="va">maf</span>,</span>
<span>  <span class="va">metadata</span>,</span>
<span>  <span class="va">comparison_column</span>,</span>
<span>  <span class="va">comparison_values</span>,</span>
<span>  <span class="va">label1</span>,</span>
<span>  <span class="va">label2</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>maf</dt>
<dd><p>Required argument. A maftools object containing the mutations you want to plot on both oncoplots.</p></dd>


<dt>metadata</dt>
<dd><p>Required argument. A data.frame with metadata for both oncoplots.</p></dd>


<dt>comparison_column</dt>
<dd><p>Required: the name of the metadata column containing the comparison values.</p></dd>


<dt>comparison_values</dt>
<dd><p>Optional: If the comparison column contains more than two values or is not a factor, specify a character vector of length two in the order you would like the factor levels to be set, reference group first.</p></dd>


<dt>label1</dt>
<dd><p>Optional argument. Label to be shown as a title for the oncoplot #1.</p></dd>


<dt>label2</dt>
<dd><p>Optional argument. Label to be shown as a title for the oncoplot #2.</p></dd>


<dt>...</dt>
<dd><p><code>prettyOncoplot</code> arguments, see that function for more info on avaialble parameters.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A ggplot object with 2 oncoplots side-by-side.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>This function will generate a graphic displaying 2 oncoplots side-by-side. Optionally user can
annotate each oncoplot with it's own title that will be displayed at the top. All the arguments
recognized by <a href="prettyOncoplot.html">prettyOncoplot</a> are supported and can be specified when calling this function.
For both oncoplots the same specified parameters will be applied (e.g. genes to display, split columns,
font size, top annotation etc). If the provided argument is not recognized by prettyOncoplot,
it will be discarded. If you want a specific order of oncoplots on the left and right, please
ensure the argument <code>comparison_column</code> is a factor with first level being the group
you want to be plotted on the left side. For developers: new arguments added to prettyOncoplot in the future
are expected to be out-of-the-box compatible with this function nd would not need code modifications.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">#get data for plotting</span></span></span>
<span class="r-in"><span><span class="va">ssm</span> <span class="op">=</span> <span class="fu"><a href="get_coding_ssm.html">get_coding_ssm</a></span><span class="op">(</span>limit_cohort <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"BL_Adult"</span>, <span class="st">"BL_Pediatric"</span><span class="op">)</span>, seq_type <span class="op">=</span> <span class="st">"genome"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> reading from: /projects/nhl_meta_analysis_scratch/gambl/results_local/all_the_things/slms_3-1.0_vcf2maf-1.3/genome--projection/deblacklisted/augmented_maf/all_slms-3--grch37.CDS.maf</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> mutations from 1652 samples</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> after linking with metadata, we have mutations from 219 samples</span>
<span class="r-in"><span><span class="va">ssm</span> <span class="op">=</span> <span class="fu">maftools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/maftools/man/read.maf.html" class="external-link">read.maf</a></span><span class="op">(</span><span class="va">ssm</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -Validating</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -Silent variants: 10604 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -Summarizing</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -Processing clinical data</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> --Missing clinical data</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -Finished in 4.532s elapsed (9.577s cpu) </span>
<span class="r-in"><span><span class="va">meta</span> <span class="op">=</span> <span class="fu"><a href="get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">meta</span> <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">meta</span>, <span class="va">cohort</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"BL_Adult"</span>, <span class="st">"BL_Pediatric"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#build plot</span></span></span>
<span class="r-in"><span><span class="fu">prettyCoOncoplot</span><span class="op">(</span>maf <span class="op">=</span> <span class="va">ssm</span>,</span></span>
<span class="r-in"><span>                 metadata <span class="op">=</span> <span class="va">meta</span>,</span></span>
<span class="r-in"><span>                 comparison_column <span class="op">=</span> <span class="st">"cohort"</span>,</span></span>
<span class="r-in"><span>                 include_noncoding <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>                 minMutationPercent <span class="op">=</span> <span class="fl">0</span>,</span></span>
<span class="r-in"><span>                 genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MYC"</span>,</span></span>
<span class="r-in"><span>                           <span class="st">"TET2"</span>,</span></span>
<span class="r-in"><span>                           <span class="st">"TP53"</span>,</span></span>
<span class="r-in"><span>                           <span class="st">"DDX3X"</span>,</span></span>
<span class="r-in"><span>                           <span class="st">"ID3"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                 metadataColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pathology"</span>,</span></span>
<span class="r-in"><span>                                     <span class="st">"EBV_status_inf"</span>,</span></span>
<span class="r-in"><span>                                     <span class="st">"pairing_status"</span>,</span></span>
<span class="r-in"><span>                                     <span class="st">"cohort"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                 splitColumnName <span class="op">=</span> <span class="st">"EBV_status_inf"</span>,</span></span>
<span class="r-in"><span>                 metadataBarHeight <span class="op">=</span> <span class="fl">10</span>,</span></span>
<span class="r-in"><span>                 fontSizeGene <span class="op">=</span> <span class="fl">12</span>,</span></span>
<span class="r-in"><span>                 metadataBarFontsize <span class="op">=</span> <span class="fl">10</span>, </span></span>
<span class="r-in"><span>                 legend_row <span class="op">=</span> <span class="fl">2</span>,</span></span>
<span class="r-in"><span>                 label1 <span class="op">=</span> <span class="st">"Adult"</span>,</span></span>
<span class="r-in"><span>                 label2 <span class="op">=</span> <span class="st">"Pediatric"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -Processing clinical data</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> --Possible FLAGS among top ten genes:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   TTN</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -Processing clinical data</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "numcases: 124"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "numgenes: 5"</span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in gpar(fontsize = legendFontSize):</span> could not find function "gpar"</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ryan Morin, Kostia Dreval, Adam Mattsson.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

