<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Oncooplot — prettyOncoplot • GAMBLR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Oncooplot — prettyOncoplot"><meta property="og:description" content="Create a highly customizable oncoplot."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-12345678-0"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-12345678-0');
</script></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9500</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-light fa-memo-circle-info"></span>
     
    How To
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../path_once_integrated_on_github">Install</a>
    </li>
    <li>
      <a href="../path_once_integrated_on_github">Run Remote On A Local Machine</a>
    </li>
    <li>
      <a href="../path_once_integrated_on_github">Contribute</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-light fa-person-chalkboard"></span>
     
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/ssm_tutorial.html">SSM</a>
    </li>
    <li>
      <a href="../articles/cn_sv_tutorial.html">CN &amp; SV</a>
    </li>
    <li>
      <a href="../articles/fancy_plots.html">Fancy Plots</a>
    </li>
    <li>
      <a href="../articles/utiliteis_tutorial.html">Utilities</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="../reference/index.html">
    <span class="fa fa-file-code-o"></span>
     
    Functions &amp; Bundled Data
  </a>
</li>
<li>
  <a href="https://github.com/morinlab/GAMBLR">
    <span class="fa fa-github fa-lg"></span>
     
    github
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Oncooplot</h1>
    
    <div class="hidden name"><code>prettyOncoplot.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Create a highly customizable oncoplot.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">prettyOncoplot</span><span class="op">(</span></span>
<span>  <span class="va">maftools_obj</span>,</span>
<span>  <span class="va">onco_matrix_path</span>,</span>
<span>  <span class="va">genes</span>,</span>
<span>  include_noncoding <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  keepGeneOrder <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  keepSampleOrder <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  highlightHotspots <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">these_samples_metadata</span>,</span>
<span>  <span class="va">metadataColumns</span>,</span>
<span>  <span class="va">numericMetadataColumns</span>,</span>
<span>  expressionColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  <span class="va">numericMetadataMax</span>,</span>
<span>  <span class="va">sortByColumns</span>,</span>
<span>  arrange_descending <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  removeNonMutated <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">minMutationPercent</span>,</span>
<span>  mutAlpha <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  recycleOncomatrix <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">splitColumnName</span>,</span>
<span>  <span class="va">splitGeneGroups</span>,</span>
<span>  showTumorSampleBarcode <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">groupNames</span>,</span>
<span>  <span class="va">hide_annotations</span>,</span>
<span>  annotate_specific_genes <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  this_forest_object <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  custom_colours <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  hideTopBarplot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  tally_all_mutations <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  tally_all_mutations_max <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  hideSideBarplot <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  box_col <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  annoAlpha <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  legend_direction <span class="op">=</span> <span class="st">"horizontal"</span>,</span>
<span>  ylim <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  legend_position <span class="op">=</span> <span class="st">"bottom"</span>,</span>
<span>  legend_row <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  legend_col <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  metadataBarHeight <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>  metadataBarFontsize <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  legendFontSize <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  fontSizeGene <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  annotation_row <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  annotation_col <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>maftools_obj</dt>
<dd><p>A maftools object containing the mutations you want to plot.</p></dd>


<dt>onco_matrix_path</dt>
<dd><p>Provide a path to an onco_matrix file instead of a MAF object if the former is unavailable (this limits functionality a bit).</p></dd>


<dt>genes</dt>
<dd><p>An optional vector of genes to restrict your plot to.</p></dd>


<dt>include_noncoding</dt>
<dd><p>List of non-coding regions to be included, default is NULL. Specify like this: include_noncoding=list("NFKBIZ" = c("3'UTR"), "HNRNPH1" = "Splice_Region")</p></dd>


<dt>keepGeneOrder</dt>
<dd><p>Set to TRUE if you want to preserve the gene order specified.</p></dd>


<dt>keepSampleOrder</dt>
<dd><p>Set to TRUE if you want to preserve the sample order specified.</p></dd>


<dt>highlightHotspots</dt>
<dd><p>Set to TRUE to highlight hot spots. Default is FALSE.</p></dd>


<dt>these_samples_metadata</dt>
<dd><p>Data frame containing metadata for your samples.</p></dd>


<dt>metadataColumns</dt>
<dd><p>A vector containing the categorical column names you want to plot below.</p></dd>


<dt>numericMetadataColumns</dt>
<dd><p>A vector containing the numeric columns you want to plot below.</p></dd>


<dt>expressionColumns</dt>
<dd><p>Optional variable for retreiving expression values for a specific gene(s).</p></dd>


<dt>numericMetadataMax</dt>
<dd><p>A numeric vector of cutoffs to apply to numeric columns above.</p></dd>


<dt>sortByColumns</dt>
<dd><p>A vector containing the column names you want to sort columns (patients) on.</p></dd>


<dt>arrange_descending</dt>
<dd><p>A Boolean parameter. Set to TRUE to sort metadata in descending fashion. Default is FALSE.</p></dd>


<dt>removeNonMutated</dt>
<dd><p>Set to TRUE to drop unmutated cases.</p></dd>


<dt>minMutationPercent</dt>
<dd><p>Only genes mutated in more than minMutationPercent % patients will be included.</p></dd>


<dt>mutAlpha</dt>
<dd><p>Optional alpha to apply to mutation colours.</p></dd>


<dt>recycleOncomatrix</dt>
<dd><p>Set to TRUE most of the time to reuse the oncomatrix saved by maftools.</p></dd>


<dt>splitColumnName</dt>
<dd><p>Optional argument to indicate which metadata column to split on. Default is set to pathology.</p></dd>


<dt>splitGeneGroups</dt>
<dd><p>Split genes into groups for better seperation (between different gene-groups) in prettyOncoplot.</p></dd>


<dt>showTumorSampleBarcode</dt>
<dd><p>Optional argument for showing tumor barcode. Default is FALSE.</p></dd>


<dt>groupNames</dt>
<dd><p>optional vector of group names to be displayed above heatmap. Should be the same length as the number of groups that will be shown. Default is NULL (no labels).</p></dd>


<dt>hide_annotations</dt>
<dd><p>Hide annotations for specifc ashms. argument takes a list with annotations.</p></dd>


<dt>annotate_specific_genes</dt>
<dd><p>Optional argument, specifying whether the features should be labelled according to their significance in one of the pathologies. Default is FALSE (no annotation).</p></dd>


<dt>this_forest_object</dt>
<dd><p>If annotate_specific_genes is specified, this arguments takes the output of GAMBLR::prettyForestPlot directly to determine the annotations.</p></dd>


<dt>custom_colours</dt>
<dd><p>Provide named vector (or named list of vectors) containing custom annotation colours if you do not want to use standartized pallette.</p></dd>


<dt>hideTopBarplot</dt>
<dd><p>Optional argument for removing top bar plot. Default value is TRUE.</p></dd>


<dt>tally_all_mutations</dt>
<dd><p>Optional argument. Set to TRUE to tally all mutations. Default is FALSE.</p></dd>


<dt>tally_all_mutations_max</dt>
<dd><p>Optional argument. Default is 1000.</p></dd>


<dt>hideSideBarplot</dt>
<dd><p>Optional argument for removing side bar plot. Default value is FALSE.</p></dd>


<dt>box_col</dt>
<dd><p>Colour of boxes for outlining mutations (can be problematic with larger oncoprints).</p></dd>


<dt>annoAlpha</dt>
<dd><p>Optional alpha to apply to annotation colours.</p></dd>


<dt>legend_direction</dt>
<dd><p>Direction of lgend, defualt is "horizontal".</p></dd>


<dt>ylim</dt>
<dd><p>Limit for y-axis.</p></dd>


<dt>legend_position</dt>
<dd><p>Position of legend, default is "bottom".</p></dd>


<dt>legend_row</dt>
<dd><p>Fiddle with these to widen or narrow your legend.</p></dd>


<dt>legend_col</dt>
<dd><p>Fiddle with these to widen or narrow your legend.</p></dd>


<dt>metadataBarHeight</dt>
<dd><p>Optional argument to adjust the height of bar with annotations. The default is 1.5.</p></dd>


<dt>metadataBarFontsize</dt>
<dd><p>Optional argument to control for the font size of metadata annotations. The default is 5.</p></dd>


<dt>legendFontSize</dt>
<dd><p>Font size for legend, default is 10.</p></dd>


<dt>fontSizeGene</dt>
<dd><p>Font size for gene labels (default 6).</p></dd>


<dt>annotation_row</dt>
<dd><p>Row for annotations, default is 2.</p></dd>


<dt>annotation_col</dt>
<dd><p>Column for annotations, default is 1.</p></dd>


<dt>verbose</dt>
<dd><p>Set to TRUE to enable verbose mode (debugging messages.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>Nothing</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Make an oncoplot that is pretty using ComplexHeatmap. The metadata is expected to follow the structure and column naming used in GAMBL.
If you provide your own non-GAMBL samples and metadata, you must include at least the following columns with these names.
The first one should match the Tumor_Sample_Barcode in the MAF object or onco_matrix you provide.
sample_id, pathology</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">grid</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#get some data</span></span></span>
<span class="r-in"><span><span class="va">maf_data</span> <span class="op">=</span> <span class="fu"><a href="get_coding_ssm.html">get_coding_ssm</a></span><span class="op">(</span>seq_type <span class="op">=</span> <span class="st">"genome"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> reading from: /projects/nhl_meta_analysis_scratch/gambl/results_local/all_the_things/slms_3-1.0_vcf2maf-1.3/genome--projection/deblacklisted/augmented_maf/all_slms-3--grch37.CDS.maf</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> mutations from 1652 samples</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> after linking with metadata, we have mutations from 1646 samples</span>
<span class="r-in"><span><span class="va">maf_metadata</span> <span class="op">=</span> <span class="fu"><a href="get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">maf</span> <span class="op">=</span> <span class="fu">maftools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/maftools/man/read.maf.html" class="external-link">read.maf</a></span><span class="op">(</span><span class="va">maf_data</span>, clinicalData <span class="op">=</span> <span class="va">maf_metadata</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -Validating</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> --Removed 1078 duplicated variants</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -Silent variants: 65357 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -Summarizing</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> --Possible FLAGS among top ten genes:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   TTN</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -Processing clinical data</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -Finished in 28.4s elapsed (56.8s cpu) </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#define some genes of interest</span></span></span>
<span class="r-in"><span><span class="va">bl_genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"NFKBIZ"</span>, <span class="st">"ID3"</span>, <span class="st">"TP53"</span>, <span class="st">"ARID1A"</span>, <span class="st">"FBXO11"</span>, </span></span>
<span class="r-in"><span>             <span class="st">"GNA13"</span>, <span class="st">"TCF3"</span>, <span class="st">"TFAP4"</span>, <span class="st">"HNRNPU"</span>, <span class="st">"FOXO1"</span>,</span></span>
<span class="r-in"><span>             <span class="st">"CCND3"</span>, <span class="st">"SMARCA4"</span>, <span class="st">"DDX3X"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">dlbcl_genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EZH2"</span>, <span class="st">"KMT2D"</span>, <span class="st">"MEF2B"</span>, <span class="st">"CREBBP"</span>, <span class="st">"MYD88"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">bl_genes</span>, <span class="va">dlbcl_genes</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#define gene groups</span></span></span>
<span class="r-in"><span><span class="va">gene_groups</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"BL"</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">bl_genes</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"DLBCL"</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">dlbcl_genes</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gene_groups</span><span class="op">)</span> <span class="op">=</span> <span class="va">genes</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#filter metadata</span></span></span>
<span class="r-in"><span><span class="va">maf_metadata</span> <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">maf_metadata</span>,<span class="op">!</span><span class="va">lymphgen</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"COMPOSITE"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#convert metadata column into factor</span></span></span>
<span class="r-in"><span><span class="va">maf_metadata</span><span class="op">$</span><span class="va">pathology</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">maf_metadata</span><span class="op">$</span><span class="va">pathology</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#define order of factors for selected metadata column</span></span></span>
<span class="r-in"><span><span class="va">maf_metadata</span><span class="op">$</span><span class="va">pathology</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">maf_metadata</span><span class="op">$</span><span class="va">pathology</span>, </span></span>
<span class="r-in"><span>                                levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DLBCL"</span>, <span class="st">"BL"</span>,</span></span>
<span class="r-in"><span>                                           <span class="st">"B-ALL"</span>, <span class="st">"CLL"</span>,</span></span>
<span class="r-in"><span>                                           <span class="st">"COMFL"</span>, <span class="st">"DLBCL-BL-like"</span>,</span></span>
<span class="r-in"><span>                                           <span class="st">"FL"</span>, <span class="st">"HGBL"</span>,</span></span>
<span class="r-in"><span>                                           <span class="st">"MCL"</span>, <span class="st">"PBL"</span>,</span></span>
<span class="r-in"><span>                                           <span class="st">"SCBC"</span>, <span class="st">"UNSPECIFIED"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">maf_metadata</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">maf_metadata</span>, <span class="va">maf_metadata</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">pathology</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#create prettyOncoplot</span></span></span>
<span class="r-in"><span><span class="fu">prettyOncoplot</span><span class="op">(</span>maftools_obj <span class="op">=</span> <span class="va">maf</span>,</span></span>
<span class="r-in"><span>               genes <span class="op">=</span> <span class="va">genes</span>,</span></span>
<span class="r-in"><span>               these_samples_metadata <span class="op">=</span> <span class="va">maf_metadata</span>,</span></span>
<span class="r-in"><span>               splitGeneGroups <span class="op">=</span> <span class="va">gene_groups</span>,</span></span>
<span class="r-in"><span>               keepGeneOrder <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>               splitColumnName <span class="op">=</span> <span class="st">"pathology"</span>,</span></span>
<span class="r-in"><span>               metadataBarHeight <span class="op">=</span> <span class="fl">5</span>,</span></span>
<span class="r-in"><span>               metadataBarFontsize <span class="op">=</span> <span class="fl">8</span>, </span></span>
<span class="r-in"><span>               legend_row <span class="op">=</span> <span class="fl">2</span>,</span></span>
<span class="r-in"><span>               fontSizeGene <span class="op">=</span> <span class="fl">11</span>,</span></span>
<span class="r-in"><span>               metadataColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pathology"</span>, <span class="st">"lymphgen"</span>, <span class="st">"sex"</span>, <span class="st">"EBV_status_inf"</span>, <span class="st">"cohort"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>               sortByColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pathology"</span>, <span class="st">"lymphgen"</span>, <span class="st">"sex"</span>, <span class="st">"EBV_status_inf"</span>, <span class="st">"cohort"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "numcases: 1646"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "numgenes: 18"</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> All mutation types: Frame_Shift_Del, Missense_Mutation, Multi_Hit,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Splice_Site, Nonsense_Mutation, Frame_Shift_Ins, In_Frame_Del,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Translation_Start_Site, Nonstop_Mutation, In_Frame_Ins.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> `alter_fun` is assumed vectorizable. If it does not generate correct</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> plot, please set `alter_fun_is_vectorized = FALSE` in `oncoPrint()`.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Following `at` are removed: RNA, 3'UTR, Splice_Region, Silent,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> hot_spot, because no color was defined for them.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Following `at` are removed: RNA, 3'UTR, Splice_Region, Silent,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> hot_spot, because no color was defined for them.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Following `at` are removed: RNA, 3'UTR, Splice_Region, Silent,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> hot_spot, because no color was defined for them.</span>
<span class="r-plt img"><img src="prettyOncoplot-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ryan Morin, Kostia Dreval, Adam Mattsson.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

